diff --git a/src/Build/Construction/Solution/SolutionFile.cs b/src/Build/Construction/Solution/SolutionFile.cs
index a5a66d780e8..f12b942b0b8 100644
--- a/src/Build/Construction/Solution/SolutionFile.cs
+++ b/src/Build/Construction/Solution/SolutionFile.cs
@@ -84,13 +84,6 @@ public sealed class SolutionFile
 
         private const char CommentStartChar = '#';
 
-        /// <summary>
-        /// Cache the return value of this method, as each invocation returns a new instance.
-        /// The method has to defend against code changing the values within the array, but we don't.
-        /// By caching it we avoid a per-call allocation.
-        /// </summary>
-        private static readonly char[] s_invalidPathChars = Path.GetInvalidPathChars();
-
         #endregion
 
         #region Member data
@@ -120,6 +113,13 @@ public sealed class SolutionFile
         private Version _currentVisualStudioVersion;
         private int _currentLineNumber;
 
+        /// <summary>
+        /// Cache the return value of this method, as each invocation returns a new instance.
+        /// The method has to defend against code changing the values within the array, but we don't.
+        /// By caching it we avoid a per-call allocation.
+        /// </summary>
+        private static readonly char[] s_invalidPathChars = Path.GetInvalidPathChars();
+
         // TODO: Unify to NativeMethodsShared.OSUsesCaseSensitive paths when possible.
         private static readonly StringComparer s_pathComparer = RuntimeInformation.IsOSPlatform(OSPlatform.Linux)
             ? StringComparer.Ordinal
