diff --git a/src/Build/BackEnd/Components/Logging/LoggingContext.cs b/src/Build/BackEnd/Components/Logging/LoggingContext.cs
index f0572894854..4f0713a6e7e 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingContext.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingContext.cs
@@ -214,13 +214,16 @@ internal void LogWarning(string subcategoryResourceName, BuildEventFileInfo file
 
             // Log an error if the warning we were about to log is listed as a WarningAsError
             // https://github.com/dotnet/msbuild/issues/5511
-            string warningCode;
-            string message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword(out warningCode, out _, messageResourceName, messageArgs);
-
-            if(_loggingService.WarningsAsErrors.Contains(warningCode))
+            if(_loggingService.WarningsAsErrors != null)
             {
-                LogError(file, messageResourceName, messageArgs);
-                return;
+                string warningCode;
+                string message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword(out warningCode, out _, messageResourceName, messageArgs);
+
+                if (_loggingService.WarningsAsErrors.Contains(warningCode))
+                {
+                    LogError(file, messageResourceName, messageArgs);
+                    return;
+                }
             }
 
             _loggingService.LogWarning(_eventContext, subcategoryResourceName, file, messageResourceName, messageArgs);
