diff --git a/RunApiPort.bat b/RunApiPort.bat
deleted file mode 100644
index 4c147cc6cf0..00000000000
--- a/RunApiPort.bat
+++ /dev/null
@@ -1,7 +0,0 @@
-@echo off
-setlocal
-
-msbuild %~dp0src\MSBuild.sln /p:Configuration=Port-Progress
-
-set ANALYSIS_PATH=%~dp0bin\Windows_NT\Port-Progress
-%~dp0ApiPort\ApiPort.exe analyze -f %ANALYSIS_PATH%\Microsoft.Build.dll -f %ANALYSIS_PATH%\Microsoft.Build.Framework.dll -f %ANALYSIS_PATH%\Microsoft.Build.Tasks.Core.dll -f %ANALYSIS_PATH%\Microsoft.Build.Utilities.Core.dll -f %ANALYSIS_PATH%\MSBuild.exe --target "ASP.NET 5, Version=1.0"
\ No newline at end of file
