diff --git a/src/Framework/ILogger.cs b/src/Framework/ILogger.cs
index 86309b35644..5f3a31f5464 100644
--- a/src/Framework/ILogger.cs
+++ b/src/Framework/ILogger.cs
@@ -1,8 +1,6 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Linq;
-using System;
 using System.Runtime.InteropServices;
 
 namespace Microsoft.Build.Framework
diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index b813ff2e0ac..bcbba83416f 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -51,8 +51,6 @@
 using File = Microsoft.IO.File;
 using FileInfo = Microsoft.IO.FileInfo;
 using Path = Microsoft.IO.Path;
-using BuildXL.Processes.Tracing;
-using Microsoft.Build.Utilities;
 #endif
 
 #nullable disable
