diff --git a/src/Build/BackEnd/Client/MSBuildClient.cs b/src/Build/BackEnd/Client/MSBuildClient.cs
index 7876ddda5d2..e10c700422b 100644
--- a/src/Build/BackEnd/Client/MSBuildClient.cs
+++ b/src/Build/BackEnd/Client/MSBuildClient.cs
@@ -296,7 +296,7 @@ private bool ServerIsBusyWithWaitAndRetry(int milliseconds)
         {
             bool isBusy = ServerWasBusy();
             Stopwatch sw = Stopwatch.StartNew();
-            while (isBusy && sw.Elapsed < TimeSpan.FromMilliseconds(milliseconds))
+            while (isBusy && sw.ElapsedMilliseconds < milliseconds)
             {
                 CommunicationsUtilities.Trace("Wait for server to be not busy - will retry soon...");
                 Thread.Sleep(100);
