diff --git a/src/Framework/TaskExecutionContext.cs b/src/Framework/TaskExecutionContext.cs
index 98715d9e659..f6bef2cc2b4 100644
--- a/src/Framework/TaskExecutionContext.cs
+++ b/src/Framework/TaskExecutionContext.cs
@@ -43,7 +43,15 @@ public string GetFullPath(string path)
                 return path;
             }
 
-            return Path.Combine(StartupDirectory, path);
+            try
+            {
+                // TODO: Does GetFullPath access the file system? If so, find a way to remove internal ../ and ./ without it.
+                // Use URI, perhaps?
+                return Path.GetFullPath(Path.Combine(StartupDirectory, path));
+            }
+            catch { }
+
+            return path;
         }
     }
 
