diff --git a/eng/SourceBuild.props b/eng/SourceBuild.props
index 7218b37ccf0..d0989ae9c8b 100644
--- a/eng/SourceBuild.props
+++ b/eng/SourceBuild.props
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <GitHubRepositoryName>msbuild</GitHubRepositoryName>
+    <SourceBuildManagedOnly>true</SourceBuildManagedOnly>
   </PropertyGroup>
 
   <Target Name="ConfigureInnerBuildArgs" BeforeTargets="GetSourceBuildCommandConfiguration">
