diff --git a/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs b/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs
index 2bb340531df..9ffff6d2077 100644
--- a/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs
@@ -33,8 +33,8 @@ public void Dispose()
         }
 
         /// <summary>
-        /// If we pass in an item spec that is over the max path but it can be normalized down to something under the max path, we should still work and not
-        /// throw a path too long exception
+        /// If we pass in an item spec can be normalized down to something under the max path, it should still work and not
+        /// throw a path too long exception or file not found exception
         /// </summary>
         [Fact]
         public void ProjectItemSpecTooLong()
