diff --git a/src/UnitTests.Shared/DummyMappedDriveTestEnv.cs b/src/UnitTests.Shared/DummyMappedDriveTestEnv.cs
index 88510d1c5d6..90a43afcba1 100644
--- a/src/UnitTests.Shared/DummyMappedDriveTestEnv.cs
+++ b/src/UnitTests.Shared/DummyMappedDriveTestEnv.cs
@@ -2,11 +2,12 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 #nullable enable
+using System;
 using Microsoft.Build.Framework;
 
 namespace Microsoft.Build.UnitTests.Shared
 {
-    public class DummyMappedDriveTestEnv
+    public class DummyMappedDriveTestEnv : IDisposable
     {
         public DummyMappedDrive? _mappedDrive;
 
@@ -29,5 +30,13 @@ public string UpdatePathToMappedDrive(string path)
             }
             return path;
         }
+
+        public void Dispose()
+        {
+            if (_mappedDrive != null)
+            {
+                _mappedDrive.Dispose();
+            }
+        }
     }
 }
