diff --git a/.gitattributes b/.gitattributes
index 4e6853e13ff..af9d9e931ae 100644
--- a/.gitattributes
+++ b/.gitattributes
@@ -20,7 +20,7 @@ src/Tasks/*.targets linguist-detectable=true
 src/Tasks/*.tasks linguist-detectable=true
 
 # Don't include Arcade-owned path
-eng/common/** linguist-vendored
+eng/common/** linguist-vendored linguist-generated
 
 # Display XLF files collapsed by default in PR diffs
 *.xlf linguist-generated=true
