diff --git a/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs b/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs
index e851f29e8ee..798db221d79 100644
--- a/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs
+++ b/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs
@@ -2676,67 +2676,18 @@ public void UpdatingIndividualItemsProceedsInOrder()
 <afterFourth Include='@(i)' />
 ";
             IList<ProjectItem> items = ObjectModelHelpers.GetItemsFromFragment(contents, allItems: true);
-            Dictionary<string, string> expectedAfterFirsta = new Dictionary<string, string>
-            {
-                {"m1", "second" }
-            };
-            Dictionary<string, string> expectedAfterFirstb = new Dictionary<string, string>
-            {
-                {"m1", "third" }
-            };
-            Dictionary<string, string> expectedAfterFirstc = new Dictionary<string, string>
-            {
-                {"m1", "fourth" }
-            };
-            Dictionary<string, string> expectedAfterSeconda = new Dictionary<string, string>
-            {
-                {"m1", "sixth" }
-            };
-            Dictionary<string, string> expectedAfterSecondb = new Dictionary<string, string>
-            {
-                {"m1", "sixth" }
-            };
-            Dictionary<string, string> expectedAfterSecondc = new Dictionary<string, string>
-            {
-                {"m1", "sixth" }
-            };
-            Dictionary<string, string> expectedAfterThirda = new Dictionary<string, string>
-            {
-                {"m1", "sixth" }
-            };
-            Dictionary<string, string> expectedAfterThirdb = new Dictionary<string, string>
-            {
-                {"m1", "seventh" }
-            };
-            Dictionary<string, string> expectedAfterThirdc = new Dictionary<string, string>
-            {
-                {"m1", "sixth" }
-            };
-            Dictionary<string, string> expectedAfterFourtha = new Dictionary<string, string>
-            {
-                {"m1", "sixth" }
-            };
-            Dictionary<string, string> expectedAfterFourthb = new Dictionary<string, string>
-            {
-                {"m1", "seventh" }
-            };
-            Dictionary<string, string> expectedAfterFourthc = new Dictionary<string, string>
-            {
-                {"m1", "eighth" }
-            };
-
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterFirsta, items[3]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterFirstb, items[4]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterFirstc, items[5]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterSeconda, items[6]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterSecondb, items[7]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterSecondc, items[8]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterThirda, items[9]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterThirdb, items[10]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterThirdc, items[11]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterFourtha, items[12]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterFourthb, items[13]);
-            ObjectModelHelpers.AssertItemHasMetadata(expectedAfterFourthc, items[14]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "second", items[3]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "third", items[4]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "fourth", items[5]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "sixth", items[6]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "sixth", items[7]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "sixth", items[8]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "sixth", items[9]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "seventh", items[10]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "sixth", items[11]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "sixth", items[12]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "seventh", items[13]);
+            ObjectModelHelpers.AssertItemHasMetadata("m1", "eighth", items[14]);
         }
 
         [Fact]
diff --git a/src/Shared/UnitTests/ObjectModelHelpers.cs b/src/Shared/UnitTests/ObjectModelHelpers.cs
index 6c62d8044f4..1dc5b0e2b55 100644
--- a/src/Shared/UnitTests/ObjectModelHelpers.cs
+++ b/src/Shared/UnitTests/ObjectModelHelpers.cs
@@ -424,6 +424,12 @@ internal static void AssertItemHasMetadata(Dictionary<string, string> expected,
             AssertItemHasMetadata(expected, new ProjectItemTestItemAdapter(item));
         }
 
+        internal static void AssertItemHasMetadata(string key, string value, ProjectItem item)
+        {
+            item.DirectMetadataCount.ShouldBe(1, () => $"Expected 1 metadata, ({key}), got {item.DirectMetadataCount}");
+            item.GetMetadataValue(key).ShouldBe(value);
+        }
+
         internal static void AssertItemHasMetadata(Dictionary<string, string> expected, TestItem item)
         {
             expected ??= new Dictionary<string, string>();
