diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 06c53027f78..5f652d67c1a 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -782,6 +782,25 @@ public static ExitType Execute(
                     {
                         ReplayBinaryLog(projectFile, loggers, distributedLoggerRecords, cpuCount);
                     }
+                    else if ("a".Equals("A", StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (FileUtilities.IsSolutionFilename(projectFile))
+                        {
+                            ProjectInstance.LoadSolutionForBuild
+                        }
+                        else
+                        {
+
+                        }
+                        Project p = Project.FromFile(projectFile, new Definition.ProjectOptions()
+                        {
+                            GlobalProperties = globalProperties,
+                            ToolsVersion = toolsVersion,
+                        });
+
+                        Console.WriteLine($"\"propertyName\": \"{p.GetPropertyValue("myProperty")}\"");
+                        Console.WriteLine($"\"propertyName\": \"{p.GetPropertyValue("otherProperty")}\"");
+                    }
                     else // regular build
                     {
                         // if everything checks out, and sufficient information is available to start building
