diff --git a/src/Tasks/Unzip.cs b/src/Tasks/Unzip.cs
index ba53e939b10..4e00a677831 100644
--- a/src/Tasks/Unzip.cs
+++ b/src/Tasks/Unzip.cs
@@ -67,7 +67,7 @@ public sealed class Unzip : TaskExtension, ICancelableTask
         public string Include { get; set; }
 
         /// <summary>
-        /// Gets or sets a glob expression that will be used to determine which files to exclude from being unzipped from the archive.
+        /// Gets or sets an MSBuild glob expression that will be used to determine which files to exclude from being unzipped from the archive.
         /// </summary>
         public string Exclude { get; set; }
 
