diff --git a/src/Shared/NativeMethodsShared.cs b/src/Shared/NativeMethodsShared.cs
index 7d66b7731ce..7acf0c04523 100644
--- a/src/Shared/NativeMethodsShared.cs
+++ b/src/Shared/NativeMethodsShared.cs
@@ -508,13 +508,20 @@ public SystemInformationData()
             }
         }
 
+        public static int GetLogicalCoreCount()
+        {
+            return IsWindows
+                ? GetLogicalCoreCountOnWindows()
+                : Environment.ProcessorCount;
+        }
+
         /// <summary>
         /// Get the exact physical core count on Windows
         /// Useful for getting the exact core count in 32 bits processes,
-        /// as Environment.ProcessorCount has a 32-core limit in that case. 
+        /// as NativeMethodsShared.GetLogicalCoreCount() has a 32-core limit in that case. 
         /// https://github.com/dotnet/runtime/blob/221ad5b728f93489655df290c1ea52956ad8f51c/src/libraries/System.Runtime.Extensions/src/System/Environment.Windows.cs#L171-L210
         /// </summary>
-        public unsafe static int GetLogicalCoreCount()
+        private unsafe static int GetLogicalCoreCountOnWindows()
         {
             uint len = 0;
             const int ERROR_INSUFFICIENT_BUFFER = 122;
