diff --git a/src/Build/Construction/Solution/SolutionFile.cs b/src/Build/Construction/Solution/SolutionFile.cs
index fb2c9f9d757..dd9497fe1be 100644
--- a/src/Build/Construction/Solution/SolutionFile.cs
+++ b/src/Build/Construction/Solution/SolutionFile.cs
@@ -305,13 +305,6 @@ internal void ParseUsingNewParser()
                     SolutionModel solutionModel = serializer.OpenAsync(FullPath, CancellationToken.None).Result;
                     ReadSolutionModel(solutionModel);
                 }
-                catch (AggregateException aggregateException)
-                {
-                    ProjectFileErrorUtilities.ThrowInvalidProjectFile(
-                            new BuildEventFileInfo(FullPath),
-                            $"InvalidProjectFile",
-                            string.Join(" ", aggregateException.InnerExceptions.Select(inner => inner.Message)));
-                }
                 catch (Exception ex)
                 {
                     ProjectFileErrorUtilities.ThrowInvalidProjectFile(
