diff --git a/src/BuildCheck.UnitTests/EndToEndTests.cs b/src/BuildCheck.UnitTests/EndToEndTests.cs
index e72f5c6d674..fa350de4eea 100644
--- a/src/BuildCheck.UnitTests/EndToEndTests.cs
+++ b/src/BuildCheck.UnitTests/EndToEndTests.cs
@@ -281,7 +281,14 @@ public void CopyToOutputTest(bool skipUnchangedDuringCopy)
         var output2 = RunCopyToOutputTest(false, skipUnchangedDuringCopy);
 
         // CopyToOutputDirectory="Always"
-        output2.File1AccessUtc.ShouldBeGreaterThan(output1.File1AccessUtc);
+        if (skipUnchangedDuringCopy)
+        {
+            output2.File1AccessUtc.ShouldBeEquivalentTo(output1.File1AccessUtc);
+        }
+        else
+        {
+            output2.File1AccessUtc.ShouldBeGreaterThan(output1.File1AccessUtc);
+        }
         // CopyToOutputDirectory="IfDifferent"
         output2.File2AccessUtc.ShouldBeEquivalentTo(output1.File2AccessUtc);
 
