diff --git a/build/build.ps1 b/build/build.ps1
index 4ea481150ae..af021e72110 100644
--- a/build/build.ps1
+++ b/build/build.ps1
@@ -20,6 +20,8 @@ Param(
 Set-StrictMode -Version 2.0
 $ErrorActionPreference = "Stop"
 
+Set-PSDebug -Trace 2
+
 function Print-Usage() {
     Write-Host "Common settings:"
     Write-Host "  -configuration <value>  Build configuration Debug, Release"
diff --git a/build/cibuild.cmd b/build/cibuild.cmd
index cd509176676..32f3ec03b66 100644
--- a/build/cibuild.cmd
+++ b/build/cibuild.cmd
@@ -1,3 +1,3 @@
-@echo off
+@echo on
 powershell -NoLogo -NoProfile -ExecutionPolicy ByPass "& '%~dp0build.ps1'" -build -pack -sign -ci -prepareMachine %*
 exit /b %ErrorLevel%
\ No newline at end of file
