diff --git a/eng/Versions.props b/eng/Versions.props
index c46fb09fd9a..dc0c87f6e08 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -28,7 +28,7 @@
   <PropertyGroup>
     <!-- DotNetCliVersion MUST match the dotnet version in global.json.
          Otherwise, this version of dotnet will not be installed and the build will error out. -->
-    <DotNetCliVersion>5.0.102</DotNetCliVersion>
+    <DotNetCliVersion>5.0.401</DotNetCliVersion>
     <MicrosoftNetCompilersToolsetVersion>3.9.0-2.20574.26</MicrosoftNetCompilersToolsetVersion>
     <NuGetBuildTasksVersion>5.9.1-rc.8</NuGetBuildTasksVersion>
   </PropertyGroup>
