diff --git a/src/Framework/Sdk/SdkResultItem.cs b/src/Framework/Sdk/SdkResultItem.cs
index 0ca56e3605f..b9a5b361204 100644
--- a/src/Framework/Sdk/SdkResultItem.cs
+++ b/src/Framework/Sdk/SdkResultItem.cs
@@ -38,8 +38,7 @@ public override bool Equals(object obj)
             {
                 foreach (var kvp in Metadata)
                 {
-                    var itemValue = "";
-                    if (!item.Metadata.TryGetValue(kvp.Key, out itemValue))
+                    if (!item.Metadata.TryGetValue(kvp.Key, out var itemValue))
                     {
                         return false;
                     }
