diff --git a/src/Shared/TaskHostConfiguration.cs b/src/Shared/TaskHostConfiguration.cs
index bcff607b032..758d8b05226 100644
--- a/src/Shared/TaskHostConfiguration.cs
+++ b/src/Shared/TaskHostConfiguration.cs
@@ -419,7 +419,7 @@ public void Translate(ITranslator translator)
 #if FEATURE_APPDOMAIN
             if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10) || !Traits.Instance.EscapeHatches.IsBinaryFormatterSerializationAllowed)
             {
-                byte[] _appDomainConfigBytes = null;
+                byte[] appDomainConfigBytes = null;
 
                 // Set the configuration bytes just before serialization in case the SetConfigurationBytes was invoked during lifetime of this instance.
                 if (translator.Mode == TranslationDirection.WriteToStream)
