diff --git a/eng/BootStrapMsBuild.targets b/eng/BootStrapMsBuild.targets
index dcff8617638..8a0dba8101b 100644
--- a/eng/BootStrapMsBuild.targets
+++ b/eng/BootStrapMsBuild.targets
@@ -238,4 +238,6 @@
 
   </Target>
 
+   <Import Project=".\TestAssets.targets" />
+
 </Project>
diff --git a/eng/TestAssets.targets b/eng/TestAssets.targets
new file mode 100644
index 00000000000..ba55fe7b907
--- /dev/null
+++ b/eng/TestAssets.targets
@@ -0,0 +1,34 @@
+<Project>
+  <!-- This target creates packages needed for e2e testing. Inputs and outputs are defined to enable incremental builds.
+    It must be run after getting bootstrapped msbuild versions because copy operation blocks the dlls.-->
+  <PropertyGroup Label="TestAssests">
+    <ArtifactsNonShippingPackagesDir>$(ArtifactsBinDir)Microsoft.Build.BuildCheck.UnitTests\CustomChecks</ArtifactsNonShippingPackagesDir>
+  </PropertyGroup>
+  
+  <ItemGroup Label="TestAssets">
+    <ProjectsToPack Include="..\Build\Microsoft.Build.csproj"/>
+    <ProjectsToPack Include="..\Framework\Microsoft.Build.Framework.csproj"/>
+    <ProjectsToPack Include="..\StringTools\StringTools.csproj"/>
+  </ItemGroup>
+  
+  <Target Name="GetSourceFilesForPacking">
+    <ItemGroup>
+      <SourceFilesForPacking Include="%(ProjectsToPack.RootDir)%(ProjectsToPack.Directory)**\*.cs" />
+    </ItemGroup>
+  </Target>
+  
+  <Target Name="MSBuildPackagesForTests"
+          AfterTargets="BootstrapFull"
+          DependsOnTargets="GetSourceFilesForPacking"
+          Inputs="@(ProjectsToPack);@(SourceFilesForPacking)"
+          Outputs="$(ArtifactsNonShippingPackagesDir)\Microsoft.Build.$(Version).nupkg;$(ArtifactsNonShippingPackagesDir)\Microsoft.Build.Framework.$(Version).nupkg;$(ArtifactsNonShippingPackagesDir)\Microsoft.NET.StringTools.$(Version).nupkg">
+    
+    <MSBuild Projects="@(ProjectsToPack)"
+             Targets="Pack"
+             Properties="PackageOutputPath=$(ArtifactsNonShippingPackagesDir);PackageVersion=$(PackageVersion)"
+             BuildInParallel="true">
+      <Output TaskParameter="TargetOutputs" ItemName="PackageOutputs" />
+    </MSBuild>
+    
+  </Target>
+</Project>
diff --git a/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props b/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props
index d79475be5f0..989cfa9c609 100644
--- a/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props
+++ b/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props
@@ -11,7 +11,7 @@
 
   <!-- In the real world scenario, the dependencies are added as Nuget PackageReference, modified for test purposes only. -->
   <ItemGroup>
-    <ProjectReference Include="..\..\..\Build\Microsoft.Build.csproj" IncludeInPackage="true" />
+    <ProjectReference Include="..\..\..\Build\Microsoft.Build.csproj" />
   </ItemGroup>
 
 </Project>
