diff --git a/src/Framework/Sdk/SdkResultItem.cs b/src/Framework/Sdk/SdkResultItem.cs
index 7c66cce7999..f9427aacb53 100644
--- a/src/Framework/Sdk/SdkResultItem.cs
+++ b/src/Framework/Sdk/SdkResultItem.cs
@@ -56,7 +56,7 @@ public override int GetHashCode()
             {
                 foreach (var kvp in Metadata)
                 {
-                    hashCode ^= $"{kvp.Key.GetHashCode()}: {(kvp.Value ?? "V").GetHashCode()}".GetHashCode();
+                    hashCode ^= $"{kvp.Key}: {kvp.Value ?? "V"}".GetHashCode();
                 }
             }
 
