diff --git a/src/Build/BackEnd/Components/RequestBuilder/BatchingEngine.cs b/src/Build/BackEnd/Components/RequestBuilder/BatchingEngine.cs
index a985b059b81..f250ea8c88c 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/BatchingEngine.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/BatchingEngine.cs
@@ -396,7 +396,7 @@ private static Dictionary<string, string> GetItemMetadataValues(
                     {
                         // This returns String.Empty for both metadata that is undefined and metadata that has 
                         // an empty value; they are treated the same.
-                        itemMetadataValues[metadataQualifiedName] = ((IItem)item).GetMetadataValueEscaped(metadataName);
+                        itemMetadataValues[metadataQualifiedName] = ((IItem)item).GetMetadataValueEscaped(metadataName) ?? string.Empty;
                     }
                     catch (InvalidOperationException e)
                     {
diff --git a/src/Build/Instance/ProjectItemInstance.cs b/src/Build/Instance/ProjectItemInstance.cs
index 60bc9672d82..4fafe708aa7 100644
--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -555,7 +555,7 @@ string IMetadataTable.GetEscapedValueIfPresent(string itemType, string name)
         {
             if (itemType == null || String.Equals(itemType, _itemType, StringComparison.OrdinalIgnoreCase))
             {
-                string value = _taskItem.GetMetadataEscaped(name);
+                string value = _taskItem.GetMetadataEscaped(name) ?? string.Empty;
 
                 if (value.Length > 0 || HasMetadata(name))
                 {
@@ -1253,7 +1253,7 @@ ProjectMetadataInstance IItem<ProjectMetadataInstance>.SetMetadata(ProjectMetada
             /// </summary>
             public string GetMetadata(string metadataName)
             {
-                return EscapingUtilities.UnescapeAll(GetMetadataEscaped(metadataName));
+                return EscapingUtilities.UnescapeAll(GetMetadataEscaped(metadataName) ?? string.Empty);
             }
 
             /// <summary>
diff --git a/src/Utilities/TaskItem.cs b/src/Utilities/TaskItem.cs
index 34d7b498183..f01aa3ba584 100644
--- a/src/Utilities/TaskItem.cs
+++ b/src/Utilities/TaskItem.cs
@@ -280,7 +280,7 @@ public void SetMetadata(
         /// <returns>The metadata value.</returns>
         public string GetMetadata(string metadataName)
         {
-            string metadataValue = (this as ITaskItem2).GetMetadataValueEscaped(metadataName);
+            string metadataValue = (this as ITaskItem2).GetMetadataValueEscaped(metadataName) ?? string.Empty;
             return EscapingUtilities.UnescapeAll(metadataValue);
         }
 
