diff --git a/src/Framework/Features.cs b/src/Framework/Features.cs
index d823dca2924..c28036dc113 100644
--- a/src/Framework/Features.cs
+++ b/src/Framework/Features.cs
@@ -28,6 +28,11 @@ public enum FeatureStatus
         /// The feature is not available.
         /// </summary>
         NotAvailable,
+
+        /// <summary>
+        /// The feature is in preview.
+        /// </summary>
+        Preview,
     }
 
     /// <summary>
