diff --git a/src/Shared/UnitTests/ObjectModelHelpers.cs b/src/Shared/UnitTests/ObjectModelHelpers.cs
index 2cfa7198b9d..a8f02e7d2cc 100644
--- a/src/Shared/UnitTests/ObjectModelHelpers.cs
+++ b/src/Shared/UnitTests/ObjectModelHelpers.cs
@@ -1819,7 +1819,7 @@ internal static void DeleteFiles(params string[] paths)
         {
             // When we delete the file directory which has the sub folder/file firstly, it will not be deleted since not empty.
             // So sort paths descendingly by file directory length, it will delete sub folder/file at first.
-            var pathsSortedByDepth = paths.OrderByDescending(x => Path.GetDirectoryName(x).Length);
+            var pathsSortedByDepth = paths.OrderByDescending(x => Path.GetDirectoryName(Path.GetFullPath(x)).Length);
 
             foreach (string path in pathsSortedByDepth)
             {
