diff --git a/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs b/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs
index 0b6c7769f80..298f96a392b 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs
@@ -156,15 +156,15 @@ public async Task<BuildResult> BuildTargets(ProjectLoggingContext loggingContext
                     TargetBuiltReason buildReason = TargetBuiltReason.None;
                     if (entry.Request.Targets.Contains(targetName))
                     {
-                        buildReason = TargetBuiltReason.EntryTarget;
+                        buildReason = TargetBuiltReason.EntryTargets;
                     }
                     else if (configuration.ProjectInitialTargets.Contains(targetName))
                     {
-                        buildReason = TargetBuiltReason.InitialTarget;
+                        buildReason = TargetBuiltReason.InitialTargets;
                     } 
                     else if (configuration.ProjectDefaultTargets.Contains(targetName))
                     {
-                        buildReason = TargetBuiltReason.DefaultTarget;
+                        buildReason = TargetBuiltReason.DefaultTargets;
                     }
 
                     targets.Add(new TargetSpecification(targetName, targetExists ? targetInstance.Location : _projectInstance.ProjectFileLocation, buildReason));
diff --git a/src/Framework/TargetBuiltReason.cs b/src/Framework/TargetBuiltReason.cs
index 6726d5a1d59..856ff8fa886 100644
--- a/src/Framework/TargetBuiltReason.cs
+++ b/src/Framework/TargetBuiltReason.cs
@@ -33,17 +33,16 @@ public enum TargetBuiltReason
         /// <summary>
         /// The target was defined as an initial target of the project.
         /// </summary>
-        InitialTarget,
-
+        InitialTargets,
 
         /// <summary>
-        /// The target was the default target of the project
+        /// The target was one of the default targets of the project.
         /// </summary>
-        DefaultTarget,
+        DefaultTargets,
 
         /// <summary>
-        /// The target was the target explicitly called to be built.
+        /// The target was one of the targets explicitly called to be built.
         /// </summary>
-        EntryTarget
+        EntryTargets
     }
 }
