diff --git a/src/Framework/GeneratedFileUsedEventArgs.cs b/src/Framework/GeneratedFileUsedEventArgs.cs
index 1b9f38fb594..653ae71bd82 100644
--- a/src/Framework/GeneratedFileUsedEventArgs.cs
+++ b/src/Framework/GeneratedFileUsedEventArgs.cs
@@ -6,8 +6,9 @@
 namespace Microsoft.Build.Framework
 {
     /// <summary>
-    /// Arguments for the generated file used event.
+    /// Arguments for the generated file used event
     /// </summary>
+    [Serializable]
     public class GeneratedFileUsedEventArgs : BuildMessageEventArgs
     {
         public GeneratedFileUsedEventArgs()
