diff --git a/src/Build/Instance/TaskRegistry.cs b/src/Build/Instance/TaskRegistry.cs
index 718efd46482..b60a82db47b 100644
--- a/src/Build/Instance/TaskRegistry.cs
+++ b/src/Build/Instance/TaskRegistry.cs
@@ -149,8 +149,6 @@ internal sealed class TaskRegistry : ITranslatable
         /// </summary>
         internal ProjectRootElementCacheBase RootElementCache { get; set; }
 
-        internal bool PrioritizeArchitectureInUsingTasks = true;
-
         /// <summary>
         /// Creates a task registry that does not fall back to any other task registry.
         /// Default constructor does no work because the tables are initialized lazily when a task is registered
