diff --git a/src/BuildCheck.UnitTests/BuildCheckManagerProviderTests.cs b/src/BuildCheck.UnitTests/BuildCheckManagerProviderTests.cs
index 307afe28c86..f632b2054ba 100644
--- a/src/BuildCheck.UnitTests/BuildCheckManagerProviderTests.cs
+++ b/src/BuildCheck.UnitTests/BuildCheckManagerProviderTests.cs
@@ -54,11 +54,14 @@ public void ProcessAnalyzerAcquisitionTest(bool isAnalyzerRuleExist, string expe
         private void MockField(string fieldName, object mockedValue)
         {
             var mockedField = _testedInstance.GetType().GetField(fieldName, BindingFlags.NonPublic | BindingFlags.Instance);
-            mockedField.SetValue(_testedInstance, mockedValue);
+            if (mockedField != null)
+            {
+                mockedField.SetValue(_testedInstance, mockedValue);
+            }
         }
     }
 
-    internal class BuildCheckAcquisitionModuleMock : IBuildCheckAcquisitionModule
+    internal sealed class BuildCheckAcquisitionModuleMock : IBuildCheckAcquisitionModule
     {
         private readonly bool _isAnalyzerRuleExistForTest = true;
 
@@ -70,7 +73,7 @@ internal class BuildCheckAcquisitionModuleMock : IBuildCheckAcquisitionModule
             : null;
     }
 
-    internal class BuildAnalyzerRuleMock : BuildAnalyzer
+    internal sealed class BuildAnalyzerRuleMock : BuildAnalyzer
     {
         public static BuildAnalyzerRule SupportedRule = new BuildAnalyzerRule(
             "X01234",
diff --git a/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj b/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj
index 566b6eff6d4..a2e65562ae1 100644
--- a/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj
+++ b/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj
@@ -1,8 +1,5 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
-  <Import Project="..\Shared\FileSystemSources.proj" />
-  <Import Project="..\Shared\DebuggingSources.proj" />
-
   <PropertyGroup>
     <!-- Don't build a netfx version of this unit with dotnet, as the netfx bootstrap layout created with 'dotnet build' is incomplete. -->
     <TargetFrameworks>$(LatestDotNetCoreForMSBuild)</TargetFrameworks>
@@ -29,33 +26,6 @@
     <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
   </ItemGroup>
 
-  <ItemGroup>
-    <Compile Include="..\Shared\ErrorUtilities.cs">
-      <Link>Shared\ErrorUtilities.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\EscapingUtilities.cs">
-      <Link>Shared\EscapingUtilities.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\BuildEnvironmentHelper.cs">
-      <Link>Shared\BuildEnvironmentHelper.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\ProcessExtensions.cs">
-      <Link>Shared\ProcessExtensions.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\ResourceUtilities.cs">
-      <Link>Shared\ResourceUtilities.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\ExceptionHandling.cs">
-      <Link>Shared\ExceptionHandling.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\FileUtilitiesRegex.cs">
-      <Link>Shared\FileUtilitiesRegex.cs</Link>
-    </Compile>
-    <Compile Include="..\Shared\UnitTests\AssemblyResources.cs">
-      <Link>Shared\AssemblyResources.cs</Link>
-    </Compile>
-  </ItemGroup>
-
   <ItemGroup>
     <None Include="..\Shared\UnitTests\App.config">
       <Link>App.config</Link>
