diff --git a/src/BuildCheck.UnitTests/EndToEndTests.cs b/src/BuildCheck.UnitTests/EndToEndTests.cs
index f950a1bbc39..52b9731359e 100644
--- a/src/BuildCheck.UnitTests/EndToEndTests.cs
+++ b/src/BuildCheck.UnitTests/EndToEndTests.cs
@@ -77,7 +77,7 @@ public void PropertiesUsageAnalyzerTest()
         TransientTestFolder logFolder = env.CreateFolder(createFolder: true);
         TransientTestFile projectFile = env.CreateFile(logFolder, "myProj.proj", contents);
 
-        string output = RunnerUtilities.ExecBootstrapedMSBuild($"{projectFile.Path} -analyze /v:detailed", out bool success);
+        string output = RunnerUtilities.ExecBootstrapedMSBuild($"{projectFile.Path} -check /v:detailed", out bool success);
         _env.Output.WriteLine(output);
         _env.Output.WriteLine("=========================");
         success.ShouldBeTrue(output);
