diff --git a/src/MSBuild/LiveLogger/LiveLogger.cs b/src/MSBuild/LiveLogger/LiveLogger.cs
index 5bcb21138d1..0d7e59bc445 100644
--- a/src/MSBuild/LiveLogger/LiveLogger.cs
+++ b/src/MSBuild/LiveLogger/LiveLogger.cs
@@ -81,6 +81,11 @@ public override string ToString()
     /// </summary>
     private DateTime _buildStartTime;
 
+    /// <summary>
+    /// The working directory when the build starts, to trim relative output paths.
+    /// </summary>
+    private readonly string _initialWorkingDirectory = Environment.CurrentDirectory;
+
     /// <summary>
     /// True if the build has encountered at least one error.
     /// </summary>
@@ -468,6 +473,12 @@ private void MessageRaised(object sender, BuildMessageEventArgs e)
                     _projects.TryGetValue(new ProjectContext(buildEventContext), out Project? project))
                 {
                     ReadOnlyMemory<char> outputPath = e.Message.AsMemory().Slice(index + 4);
+
+                    if (outputPath.Span.Slice(0, _initialWorkingDirectory.Length).SequenceEqual(_initialWorkingDirectory.AsSpan()))
+                    {
+                        outputPath = outputPath.Slice(_initialWorkingDirectory.Length + 1);
+                    }
+
                     project.OutputPath = outputPath;
                 }
             }
