diff --git a/eng/Versions.props b/eng/Versions.props
index 11f2f2abd92..48d6f10ec56 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -3,7 +3,7 @@
 <Project>
   <PropertyGroup>
     <VersionPrefix>17.11.0</VersionPrefix>
-    <PackageValidationBaselineVersion>17.10.6</PackageValidationBaselineVersion>
+    <PackageValidationBaselineVersion>17.10.4</PackageValidationBaselineVersion>
     <AssemblyVersion>15.1.0.0</AssemblyVersion>
     <PreReleaseVersionLabel>preview</PreReleaseVersionLabel>
     <DotNetUseShippingVersions>true</DotNetUseShippingVersions>
