diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index edcecedae33..ff04f0e5845 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -980,18 +980,6 @@ public void EndBuild()
                             _overallBuildSuccess = false;
                         }
 
-                        if (loggingService is LoggingService logService && !BuildEnvironmentState.s_runningTests)
-                        {
-                            if (_overallBuildSuccess)
-                            {
-                                Debug.Assert(logService.HasLoggedErrors() == false, "Build succeeded but logged errors.");
-                            }
-                            else
-                            {
-                                Debug.Assert(logService.HasLoggedErrors() == true, "Build failed but did not log errors.");
-                            }
-                        }
-
                         loggingService.LogBuildFinished(_overallBuildSuccess);
 
                         if (KnownTelemetry.BuildTelemetry != null)
diff --git a/src/Build/BackEnd/Components/Logging/LoggingService.cs b/src/Build/BackEnd/Components/Logging/LoggingService.cs
index b95d13c74eb..b10bd378e32 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -775,11 +775,6 @@ public static ILoggingService CreateLoggingService(LoggerMode mode, int node)
             return new LoggingService(mode, node);
         }
 
-        public bool HasLoggedErrors()
-        {
-            return _buildSubmissionIdsThatHaveLoggedErrors.Any();
-        }
-
         /// <summary>
         /// NotThreadSafe, this method should only be called from the component host thread
         /// Called by the build component host when a component is first initialized.
