diff --git a/src/Build/Xml/XmlReaderExtension.cs b/src/Build/Xml/XmlReaderExtension.cs
index 7f2e04dc251..753672a1f49 100644
--- a/src/Build/Xml/XmlReaderExtension.cs
+++ b/src/Build/Xml/XmlReaderExtension.cs
@@ -5,6 +5,7 @@
 using System.Text;
 using System.Xml;
 using Microsoft.Build.Shared;
+using Microsoft.Build.Utilities;
 
 namespace Microsoft.Build.Internal
 {
@@ -104,7 +105,7 @@ private static XmlReader GetXmlReader(string file, StreamReader input, bool load
             string uri = new UriBuilder(Uri.UriSchemeFile, string.Empty) { Path = file }.ToString();
 
             XmlReader reader = null;
-            if (loadAsReadOnly && !_disableReadOnlyLoad)
+            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_0) && loadAsReadOnly && !_disableReadOnlyLoad)
             {
                 // Create an XML reader with IgnoreComments and IgnoreWhitespace set if we know that we won't be asked
                 // to write the DOM back to a file. This is a performance optimization.
