diff --git a/src/Framework/Sdk/SdkResultItem.cs b/src/Framework/Sdk/SdkResultItem.cs
index 93b867f5d14..b4a04a97ed8 100644
--- a/src/Framework/Sdk/SdkResultItem.cs
+++ b/src/Framework/Sdk/SdkResultItem.cs
@@ -20,6 +20,7 @@ public class SdkResultItem
 
         public SdkResultItem()
         {
+            ItemSpec = "";
             Metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
         }
 
