diff --git a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
index ac46566736c..6d5f972d458 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
@@ -877,7 +877,8 @@ internal void ReleaseAllCores()
 
         #region IBuildEngine10 Members
 
-        private class BuildEngineInterfaceImpl : BuildEngineInterface
+        [Serializable]
+        private sealed class BuildEngineInterfaceImpl : BuildEngineInterface
         {
             private TaskHost _taskHost;
 
diff --git a/src/Framework/BuildEngineInterface.cs b/src/Framework/BuildEngineInterface.cs
index cb55d8d49c7..100b91f46ce 100644
--- a/src/Framework/BuildEngineInterface.cs
+++ b/src/Framework/BuildEngineInterface.cs
@@ -12,6 +12,7 @@ namespace Microsoft.Build.Framework
     /// Make all members virtual but not abstract, ensuring that implementations can override them and external implementations
     /// won't break when the class is extended with new members. This base implementation should be throwing <see cref="NotImplementedException"/>.
     /// </remarks>
+    [Serializable]
     public abstract class BuildEngineInterface
     {
         /// <summary>
diff --git a/src/MSBuild/OutOfProcTaskHostNode.cs b/src/MSBuild/OutOfProcTaskHostNode.cs
index 25413b9b419..0b31a90e0fa 100644
--- a/src/MSBuild/OutOfProcTaskHostNode.cs
+++ b/src/MSBuild/OutOfProcTaskHostNode.cs
@@ -495,7 +495,8 @@ public void ReleaseCores(int coresToRelease)
 
         #region IBuildEngine10 Members
 
-        private class BuildEngineInterfaceImpl : BuildEngineInterface
+        [Serializable]
+        private sealed class BuildEngineInterfaceImpl : BuildEngineInterface
         {
             /// <summary>
             /// No logging verbosity optimization in OOP nodes.
