diff --git a/src/MSBuild/Microsoft.Build.CommonTypes.xsd b/src/MSBuild/Microsoft.Build.CommonTypes.xsd
index 2733d18df99..5b609e81317 100644
--- a/src/MSBuild/Microsoft.Build.CommonTypes.xsd
+++ b/src/MSBuild/Microsoft.Build.CommonTypes.xsd
@@ -5114,6 +5114,21 @@ elementFormDefault="qualified">
         </xs:complexType>
     </xs:element>
 
+    <xs:element name="Unzip" substitutionGroup="msb:Task">
+        <xs:complexType>
+            <xs:complexContent>
+                <xs:extension base="msb:TaskType">
+                    <xs:attribute name="DestinationFiles" />
+                    <xs:attribute name="DestinationFolder" use="required" />
+                    <xs:attribute name="OverwriteReadOnlyFiles" type="msb:boolean" />
+                    <xs:attribute name="SkipUnchangedFiles" type="msb:boolean" />
+                    <xs:attribute name="SourceFiles" use="required" />
+                    <xs:attribute name="UnzippedFiles" />
+                </xs:extension>
+            </xs:complexContent>
+        </xs:complexType>
+    </xs:element>
+
     <xs:element name="UpdateAppxManifestForBundle" substitutionGroup="msb:Task">
         <xs:complexType>
             <xs:complexContent>
@@ -5230,4 +5245,15 @@ elementFormDefault="qualified">
         </xs:complexType>
     </xs:element>
 
+    <xs:element name="ZipDirectory" substitutionGroup="msb:Task">
+        <xs:complexType>
+            <xs:complexContent>
+                <xs:extension base="msb:TaskType">
+                    <xs:attribute name="DestinationFile" use="required" />
+                    <xs:attribute name="SourceDirectory" type="msb:boolean" />
+                </xs:extension>
+            </xs:complexContent>
+        </xs:complexType>
+    </xs:element>
+
 </xs:schema>
