diff --git a/src/Build/BackEnd/Components/Logging/LoggingService.cs b/src/Build/BackEnd/Components/Logging/LoggingService.cs
index 385afe29480..9cd5a1c8bd7 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -1492,7 +1492,7 @@ private void RouteBuildEvent(object loggingEvent)
                 _warningsAsMessagesByProject?.Remove(key);
             }
 
-            if (buildEventArgs is BuildEventArgs loggingEventBuildArgs)
+            if (loggingEvent is BuildEventArgs loggingEventBuildArgs)
             {
                 RouteBuildEvent(loggingEventBuildArgs);
             }
