diff --git a/src/Framework/NativeMethods.cs b/src/Framework/NativeMethods.cs
index 8ec9b26fcee..549f2f002af 100644
--- a/src/Framework/NativeMethods.cs
+++ b/src/Framework/NativeMethods.cs
@@ -478,21 +478,15 @@ public SystemInformationData()
 
     public static int GetLogicalCoreCount()
     {
-        int numberOfCpus = Environment.ProcessorCount;
 #if !MONO
         // .NET Core on Windows returns a core count limited to the current NUMA node
         //     https://github.com/dotnet/runtime/issues/29686
         if (IsWindows)
         {
-            var result = GetLogicalCoreCountOnWindows();
-
-            if (result != 0)
-            {
-                numberOfCpus = result;
-            }
+            return GetLogicalCoreCountOnWindows();
         }
 #endif
-        return numberOfCpus;
+        return Environment.ProcessorCount;
     }
 
     /// <summary>
