diff --git a/src/Build/BackEnd/Components/Logging/LoggingService.cs b/src/Build/BackEnd/Components/Logging/LoggingService.cs
index 5508bfbbfb3..89d9008f90d 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -541,7 +541,10 @@ public ICollection<string> GetWarningsToBeLoggedAsErrorsByProject(BuildEventCont
 
             int key = GetWarningsAsErrorOrMessageKey(context);
 
-            HashSet<string> warningsAsErrorsExcludingMessages = new HashSet<string>(_warningsAsErrorsByProject[key]);
+            if (!_warningsAsErrorsByProject.TryGetValue(key, out ISet<string> warningsAsErrorsExcludingMessages))
+            {
+                return null;
+            }
 
             if (_warningsAsMessagesByProject != null)
             {
