diff --git a/src/Tasks/GetAssembliesMetadata.cs b/src/Tasks/GetAssembliesMetadata.cs
index 63f8d6425a2..127a47f1605 100644
--- a/src/Tasks/GetAssembliesMetadata.cs
+++ b/src/Tasks/GetAssembliesMetadata.cs
@@ -47,12 +47,15 @@ public override bool Execute()
             var assembliesMetadata = new List<ITaskItem>();
             foreach (string assemblyPath in AssemblyPaths)
             {
-                AssemblyInformation assemblyInformation = new(assemblyPath);
-                AssemblyAttributes attributes = assemblyInformation.GetAssemblyMetadata();
-
-                if (attributes != null)
+                if (Directory.Exists(assemblyPath))
                 {
-                    assembliesMetadata.Add(CreateItemWithMetadata(attributes));
+                    AssemblyInformation assemblyInformation = new(assemblyPath);
+                    AssemblyAttributes attributes = assemblyInformation.GetAssemblyMetadata();
+
+                    if (attributes != null)
+                    {
+                        assembliesMetadata.Add(CreateItemWithMetadata(attributes));
+                    }
                 }
             }
 
