diff --git a/src/StringTools/InternableString.cs b/src/StringTools/InternableString.cs
index 723b71aff3b..3d55b37c0c2 100644
--- a/src/StringTools/InternableString.cs
+++ b/src/StringTools/InternableString.cs
@@ -331,6 +331,7 @@ public override unsafe int GetHashCode()
         /// <param name="length">Number of characters at <paramref name="charPtr"/>.</param>
         /// <param name="hash">The running hash code.</param>
         /// <param name="hashedOddNumberOfCharacters">True if the incoming <paramref name="hash"/> was calculated from an odd number of characters.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static unsafe void GetHashCodeHelper(char* charPtr, int length, ref uint hash, ref bool hashedOddNumberOfCharacters)
         {
             if (hashedOddNumberOfCharacters && length > 0)
