diff --git a/src/Tasks/Microsoft.Common.CurrentVersion.targets b/src/Tasks/Microsoft.Common.CurrentVersion.targets
index 188351d7f89..5fdfbfe9db4 100644
--- a/src/Tasks/Microsoft.Common.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.Common.CurrentVersion.targets
@@ -5739,27 +5739,6 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
   </Target>
 
-  <!--
-    ***********************************************************************************************
-    ***********************************************************************************************
-                                                                Test Section
-    ***********************************************************************************************
-    ***********************************************************************************************
-  -->
-  <!--
-    ============================================================
-    This stub `VSTest` target allows for targets implementing Test execution functionality
-    to run after it.
-    
-    For example:
-    <Target Name="RunVSTest" AfterTargets="VSTest">
-     (implementation)
-    </Target>
-    ============================================================
-  -->
-
-  <Target Name="VSTest"></Target>
-
   <!--
     ***********************************************************************************************
     ***********************************************************************************************
@@ -6728,7 +6707,8 @@ Copyright (C) Microsoft Corporation. All rights reserved.
   </PropertyGroup>
 
   <Import Project="$(MsTestToolsTargets)" Condition="Exists('$(MsTestToolsTargets)')" />
-
+  
+  <Import Project="$(MSBuildToolsPath)\Microsoft.Common.Test.targets" />
   <!-- App packaging support -->
 
   <!--
@@ -6761,5 +6741,5 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
   <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.targets\ImportAfter\*" Condition="'$(ImportByWildcardAfterMicrosoftCommonTargets)' == 'true' and exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.targets\ImportAfter')"/>
   <Import Project="$(MSBuildUserExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.targets\ImportAfter\*" Condition="'$(ImportUserLocationsByWildcardAfterMicrosoftCommonTargets)' == 'true' and exists('$(MSBuildUserExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.targets\ImportAfter')"/>
-
+  
 </Project>
diff --git a/src/Tasks/Microsoft.Common.Test.targets b/src/Tasks/Microsoft.Common.Test.targets
new file mode 100644
index 00000000000..9922c1a477a
--- /dev/null
+++ b/src/Tasks/Microsoft.Common.Test.targets
@@ -0,0 +1,28 @@
+<!--
+***********************************************************************************************
+Microsoft.Common.Test.targets
+
+WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and have
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+***********************************************************************************************
+-->
+<!--
+    ============================================================
+    This stub `Test` target allows for targets implementing Test execution functionality
+    to run after it.
+    
+    For example:
+    <Target Name="RunVSTest" AfterTargets="Test">
+     (implementation)
+    </Target>
+    ============================================================
+  -->
+<Project>
+    <PropertyGroup>
+        <UseMSBuildTestInfrastructure>false</UseMSBuildTestInfrastructure>
+    </PropertyGroup>
+    <Target Name="Test" Condition="$(UseMSBuildTestInfrastructure) == 'true'"></Target>
+</Project>
\ No newline at end of file
