diff --git a/Directory.Build.props b/Directory.Build.props
index 6f80aaa5077..93ea74b4b6b 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -63,8 +63,7 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <AssemblyInformationCachePaths Condition="Exists('$(MSBuildBinPath)\SdkPrecomputedAssemblyReferences.cache')">$(MSBuildBinPath)\SDKPrecomputedAssemblyReferences.cache</AssemblyInformationCachePaths>
-    <AssemblyInformationCachePaths Condition="Exists('$(NetCoreRoot)$(NetCoreSdkVersion)\SdkPrecomputedAssemblyReferences.cache')">$(NetCoreRoot)$(NetCoreSdkVersion)\SDKPrecomputedAssemblyReferences.cache</AssemblyInformationCachePaths>
+    <AssemblyInformationCachePaths Condition="Exists('$(NetCoreRoot)sdk\$(NetCoreSdkVersion)\SdkPrecomputedAssemblyReferences.cache')">$(NetCoreRoot)$(NetCoreSdkVersion)\SDKPrecomputedAssemblyReferences.cache</AssemblyInformationCachePaths>
   </PropertyGroup>
 
   <PropertyGroup>
