diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 5f2d936f3ae..fc209378218 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2951,7 +2951,7 @@ internal static string AggregateParameters(string anyPrefixingParameter, string[
         {
             for (int i = 0; i < parametersToAggregate.Length; i++)
             {
-                parametersToAggregate[i] = parametersToAggregate[i].Trim(';');
+                parametersToAggregate[i] = parametersToAggregate[i].Trim(MSBuildConstants.SemicolonChar);
             }
 
             // Join the logger parameters into one string seperated by semicolons
