diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index e0affa31b16..4aceb179fae 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1567,22 +1567,19 @@ internal static bool BuildProject(
                                             nodeResultKvp.Key.ProjectInstance.GlobalProperties.All(propertyKvp => entryPoint.GlobalProperties.TryGetValue(propertyKvp.Key, out string entryValue) &&
                                                                                                                                         entryValue.Equals(propertyKvp.Value)))
                                             .Value;
+                                        success = result.OverallResult == BuildResultCode.Success;
+                                    }
+                                    else
+                                    {
+                                        success = graphResult.OverallResult == BuildResultCode.Success;
                                     }
                                 }
                                 else
                                 {
                                     result = ExecuteBuild(buildManager, buildRequest);
+                                    success = result.OverallResult == BuildResultCode.Success;
                                 }
                             }
-
-                            if (graphResult != null && !saveProjectResult)
-                            {
-                                success = graphResult.OverallResult == BuildResultCode.Success;
-                            }
-                            else if (result != null && result.Exception == null)
-                            {
-                                success = result.OverallResult == BuildResultCode.Success;
-                            }
                         }
                         finally
                         {
