diff --git a/Directory.Build.targets b/Directory.Build.targets
index 873c576a2d9..33347b713e1 100644
--- a/Directory.Build.targets
+++ b/Directory.Build.targets
@@ -11,8 +11,8 @@
   </ItemGroup>
 
   <!-- Global Analyzer Config -->
-  <ItemGroup Condition="'$(ProjectIsDeprecated)' != 'true' AND '$(IsTestProject)' != 'true'">
-    <!-- Include Common.globalconfig for non-deprecated and non-test projects-->
+  <ItemGroup Condition="'$(ProjectIsDeprecated)' != 'true'">
+    <!-- Include Common.globalconfig for non-deprecated projects-->
     <EditorConfigFiles Include="$(MSBuildThisFileDirectory)eng/Common.globalconfig" />
   </ItemGroup>
   
diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index 9ed50b502fd..fd878420d57 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -1,6 +1,10 @@
 # Top level entry required to mark this as a global AnalyzerConfig file
 is_global = true
 
+# Set precedence one lower than default to allow subsets of projects to override values
+# from this file with more specific configurations
+global_level = 99
+
 # Microsoft.Analyzers.ManagedCodeAnalysis
 # Description: Microsoft.Analyzers.ManagedCodeAnalysis
 
