diff --git a/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs b/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs
index 9858c1dd2b4..65181d9e742 100644
--- a/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs
+++ b/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs
@@ -284,7 +284,7 @@ private void RunReadLoop(Stream localStream, ManualResetEvent localPacketPumpShu
                                 {
 #if NET
                                     ValueTask<int> bytesReadTask = localStream.ReadAsync(packetData.AsMemory(packetBytesRead, packetLength - packetBytesRead));
-                                    int bytesRead = readTask.IsCompleted ? readTask.Result : readTask.AsTask().Result;
+                                    int bytesRead = bytesReadTask.IsCompleted ? bytesReadTask.Result : bytesReadTask.AsTask().Result;
 #else
                                     int bytesRead = localStream.Read(packetData, packetBytesRead, packetLength - packetBytesRead);
 #endif
