diff --git a/src/Shared/UnitTests/TestEnvironment.cs b/src/Shared/UnitTests/TestEnvironment.cs
index 4616f8ef28c..ff7d2e6069c 100644
--- a/src/Shared/UnitTests/TestEnvironment.cs
+++ b/src/Shared/UnitTests/TestEnvironment.cs
@@ -454,10 +454,25 @@ private string[] GetMSBuildLogFiles()
             string debugPath = FileUtilities.TempFileDirectory;
             if (debugPath != null)
             {
-                files.AddRange(Directory.GetFiles(debugPath, MSBuildLogFiles));
+                try
+                {
+                    files.AddRange(Directory.GetFiles(debugPath, MSBuildLogFiles));
+                }
+                catch (DirectoryNotFoundException)
+                {
+                    // Temp folder might have been deleted by other TestEnvironment logic
+                }
+            }
+
+            try
+            {
+                files.AddRange(Directory.GetFiles(Path.GetTempPath(), MSBuildLogFiles));
+            }
+            catch (DirectoryNotFoundException)
+            {
+                // Temp folder might have been deleted by other TestEnvironment logic
             }
 
-            files.AddRange(Directory.GetFiles(Path.GetTempPath(), MSBuildLogFiles));
             return files.ToArray();
         }
 
