diff --git a/src/Build/Evaluation/Conditionals/CharacterUtilities.cs b/src/Build/Evaluation/Conditionals/CharacterUtilities.cs
index 8318bdd2d42..ed5502e91ca 100644
--- a/src/Build/Evaluation/Conditionals/CharacterUtilities.cs
+++ b/src/Build/Evaluation/Conditionals/CharacterUtilities.cs
@@ -3,8 +3,6 @@
 
 #nullable disable
 
-using Microsoft.Build.Framework;
-
 namespace Microsoft.Build.Evaluation
 {
     internal static class CharacterUtilities
@@ -26,18 +24,7 @@ internal static bool IsSimpleStringChar(char candidate)
 
         internal static bool IsHexDigit(char candidate)
         {
-            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_14))
-            {
-#if NET
-                return char.IsAsciiHexDigit(candidate);
-#else
-                return (candidate - '0' <= '9' - '0') || ((uint)((candidate | 0x20) - 'a') <= 'f' - 'a');
-#endif
-            }
-            else
-            {
-                return char.IsDigit(candidate) || ((uint)((candidate | 0x20) - 'a') <= 'f' - 'a');
-            }
+            return char.IsDigit(candidate) || ((uint)((candidate | 0x20) - 'a') <= 'f' - 'a');
         }
     }
 }
