diff --git a/src/Tasks/StateFileBase.cs b/src/Tasks/StateFileBase.cs
index b5b34a31b0b..56177138fc3 100644
--- a/src/Tasks/StateFileBase.cs
+++ b/src/Tasks/StateFileBase.cs
@@ -73,7 +73,7 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
             {
                 if (!string.IsNullOrEmpty(stateFile) && FileSystems.Default.FileExists(stateFile))
                 {
-                    using (FileStream s = new FileStream(stateFile, FileMode.Open))
+                    using (FileStream s = File.OpenRead(stateFile))
                     {
                         var formatter = new BinaryFormatter();
                         object deserializedObject = formatter.Deserialize(s);
diff --git a/src/Tasks/SystemState.cs b/src/Tasks/SystemState.cs
index 3f5ea428b0c..3a56a2ce98f 100644
--- a/src/Tasks/SystemState.cs
+++ b/src/Tasks/SystemState.cs
@@ -284,7 +284,7 @@ internal static SystemState DeserializeCacheByTranslator(string stateFile, TaskL
             {
                 if (!string.IsNullOrEmpty(stateFile) && FileSystems.Default.FileExists(stateFile))
                 {
-                    using FileStream s = new FileStream(stateFile, FileMode.Open);
+                    using FileStream s = File.OpenRead(stateFile);
                     var translator = BinaryTranslator.GetReadTranslator(s, buffer:null); // TODO: shared buffering?
 
                     // verify file signature
