diff --git a/src/Framework/FileClassifier.cs b/src/Framework/FileClassifier.cs
index 84c54dc3bc0..290c2b9dd04 100644
--- a/src/Framework/FileClassifier.cs
+++ b/src/Framework/FileClassifier.cs
@@ -86,12 +86,10 @@ public FileClassifier()
             if (!string.IsNullOrEmpty(programFiles32))
             {
                 RegisterImmutableDirectory(Path.Combine(programFiles32, "Reference Assemblies", "Microsoft"));
-                RegisterImmutableDirectory(Path.Combine(programFiles32, "dotnet"));
             }
             if (!string.IsNullOrEmpty(programFiles64))
             {
                 RegisterImmutableDirectory(Path.Combine(programFiles64, "Reference Assemblies", "Microsoft"));
-                RegisterImmutableDirectory(Path.Combine(programFiles64, "dotnet"));
             }
             RegisterImmutableDirectory(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget", "packages"));
             RegisterImmutableDirectory(GetVSInstallationDirectory());
