diff --git a/src/Tasks/AssignCulture.cs b/src/Tasks/AssignCulture.cs
index 9b1cc031ab5..db730f063ae 100644
--- a/src/Tasks/AssignCulture.cs
+++ b/src/Tasks/AssignCulture.cs
@@ -166,6 +166,8 @@ public override bool Execute()
                         {
                             Log.LogWarningWithCodeFromResources("AssignCulture.CultureOverwritten",
                                 existingCulture, AssignedFiles[i].ItemSpec, info.culture);
+
+                            // Here we should have had removed the culture - let's do that behind trait
                         }
 
                         if (!string.IsNullOrEmpty(info.culture))
