diff --git a/src/Directory.Build.props b/src/Directory.Build.props
index d6c6dbbdda8..497b198210a 100644
--- a/src/Directory.Build.props
+++ b/src/Directory.Build.props
@@ -93,10 +93,10 @@
     <PublishWindowsPdb>false</PublishWindowsPdb>
   </PropertyGroup>
 
-  <!-- Ensure ns2.0 ref assemblies are placed under `ref/netstandard2.0` in the NuGet package -->
+  <!-- Ensure ref assemblies are placed under `ref/$(TargetFramework)` in the NuGet package -->
   <Target Name="ShipRefAssembliesToNuGetPackage" BeforeTargets="Pack" Condition="$(IsInnerBuild) == true">
     <ItemGroup>
-      <TfmSpecificPackageFile Include="@(IntermediateRefAssembly);@(FinalDocFile)">
+      <TfmSpecificPackageFile Include="$(TargetRefPath);@(FinalDocFile)">
           <PackagePath>ref/$(TargetFramework)</PackagePath>
       </TfmSpecificPackageFile>
       <!-- ns2.0 builds use `BuiltProjectOutputGroupOutput` for output ref assemblies -->
