diff --git a/src/Framework/Sdk/SdkResultItem.cs b/src/Framework/Sdk/SdkResultItem.cs
index 5c24ad028b8..36b096b8bc5 100644
--- a/src/Framework/Sdk/SdkResultItem.cs
+++ b/src/Framework/Sdk/SdkResultItem.cs
@@ -34,13 +34,18 @@ public override bool Equals(object obj)
         {
             if (obj is SdkResultItem item &&
                    ItemSpec == item.ItemSpec &&
-                   Metadata?.Count == item.Metadata?.Count)
+                   Metadata.Count == item.Metadata?.Count)
             {
-                if (Metadata != null)
+                foreach (var kvp in Metadata)
                 {
-                    foreach (var kvp in Metadata)
+                    var itemValue = "";
+                    if (!item.Metadata.TryGetValue(kvp.Key, out itemValue))
+                    {
+                        return false;
+                    }
+                    else
                     {
-                        if (item.Metadata[kvp.Key] != kvp.Value)
+                        if (kvp.Value != itemValue)
                         {
                             return false;
                         }
