diff --git a/src/MSBuildTaskHost/FileSystem/MSBuildTaskHostFileSystem.cs b/src/MSBuildTaskHost/FileSystem/MSBuildTaskHostFileSystem.cs
index e211857a448..03c73595094 100644
--- a/src/MSBuildTaskHost/FileSystem/MSBuildTaskHostFileSystem.cs
+++ b/src/MSBuildTaskHost/FileSystem/MSBuildTaskHostFileSystem.cs
@@ -20,11 +20,7 @@ internal class MSBuildTaskHostFileSystem : IFileSystem
 
         public bool FileOrDirectoryExists(string path)
         {
-#if NETFRAMEWORK
             return FileExists(path) || DirectoryExists(path);
-#else
-            return Path.Exists(path);
-#endif
         }
 
         public FileAttributes GetAttributes(string path)
@@ -39,11 +35,7 @@ public DateTime GetLastWriteTimeUtc(string path)
 
         public bool DirectoryExists(string path)
         {
-#if NETFRAMEWORK
-            return Microsoft.IO.Directory.Exists(path);
-#else
             return Directory.Exists(path);
-#endif
         }
 
         public IEnumerable<string> EnumerateDirectories(string path, string searchPattern = "*", SearchOption searchOption = SearchOption.TopDirectoryOnly)
@@ -85,11 +77,7 @@ public IEnumerable<string> EnumerateFileSystemEntries(string path, string search
 
         public bool FileExists(string path)
         {
-#if NETFRAMEWORK
-            return Microsoft.IO.File.Exists(path);
-#else
             return File.Exists(path);
-#endif
         }
     }
 }
