diff --git a/src/Tasks.UnitTests/XslTransformation_Tests.cs b/src/Tasks.UnitTests/XslTransformation_Tests.cs
index 41e7185e0a9..6fa575f6670 100644
--- a/src/Tasks.UnitTests/XslTransformation_Tests.cs
+++ b/src/Tasks.UnitTests/XslTransformation_Tests.cs
@@ -892,7 +892,7 @@ public void XsltPreserveWhitespace()
                 using (StreamReader sr = new StreamReader(t.OutputPaths[0].ItemSpec))
                 {
                     string fileContents = sr.ReadToEnd();
-                    Assert.True(fileContents.Equals(
+                    Assert.Equal(
 @"<?xml version=""1.0"" encoding=""utf-8""?>
 <surround>
   <root>
@@ -906,8 +906,8 @@ public void XsltPreserveWhitespace()
       </abc>
     </surround>
   </root>
-</surround>")
-                    );
+</surround>",
+                    fileContents);
                 }
             }
 
