diff --git a/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs b/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs
index 3660ea702b5..853557fc80b 100644
--- a/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs
+++ b/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs
@@ -790,96 +790,6 @@ public void SetParameterizedSwitchTests2()
             Assert.Equal("B", parameters[1]);
         }
 
-        [Fact]
-        public void SettingGetPropertyTurnsDownVerbosity()
-        {
-            using TestEnvironment env = TestEnvironment.Create();
-            string path = env.CreateFile("foo.csproj", string.Empty).Path;
-            CommandLineSwitches switches = new();
-            switches.SetParameterizedSwitch(CommandLineSwitches.ParameterizedSwitch.Verbosity, "diag", "diag", true, true, true);
-            switches.SetParameterizedSwitch(CommandLineSwitches.ParameterizedSwitch.BinaryLogger, "outputPath.binlog", "outputPath.binlog", true, true, true);
-            switches.SetParameterizedSwitch(CommandLineSwitches.ParameterizedSwitch.GetProperty, "MyProp", "MyProp", true, true, true);
-            switches.SetParameterizedSwitch(CommandLineSwitches.ParameterizedSwitch.Project, path, path, true, true, true);
-            switches.SetParameterlessSwitch(CommandLineSwitches.ParameterlessSwitch.NoAutoResponse, "-noautoresponse");
-            string projectFile = string.Empty;
-            string[] targets = Array.Empty<string>();
-            string toolsVersion = string.Empty;
-            Dictionary<string, string> globalProperties = new(0);
-            ILogger[] loggers = Array.Empty<ILogger>();
-            LoggerVerbosity verbosity = LoggerVerbosity.Normal;
-            LoggerVerbosity originalVerbosity = LoggerVerbosity.Normal;
-            List<DistributedLoggerRecord> distributedLoggerRecords = new(0);
-#if FEATURE_XML_SCHEMA_VALIDATION
-            bool needToValidateProject = false;
-            string schemaFile = string.Empty;
-#endif
-            int cpuCount = 0;
-            bool enableNodeReuse = false;
-            TextWriter preprocessWriter = null;
-            TextWriter targetsWriter = null;
-            bool detailedSummary = false;
-            ISet<string> warningsAsErrors = new HashSet<string>(0);
-            ISet<string> warningsNotAsErrors = new HashSet<string>(0);
-            ISet<string> warningsAsMessages = new HashSet<string>(0);
-            bool enableRestore = false;
-            bool interactive = false;
-            ProfilerLogger profilerLogger = null;
-            bool enableProfiler = false;
-            Dictionary<string, string> restoreProperties = new(0);
-            ProjectIsolationMode isolateProjects = ProjectIsolationMode.False;
-            GraphBuildOptions graphBuild = null;
-            string[] inputResultsCaches = Array.Empty<string>();
-            string outputResultsCache = string.Empty;
-            bool lowPriority = false;
-            bool question = false;
-            string[] getProperty = Array.Empty<string>();
-            string[] getItem = Array.Empty<string>();
-            string[] getTargetResult = Array.Empty<string>();
-            bool recursing = false;
-            string commandLine = string.Empty;
-            MSBuildApp.ProcessCommandLineSwitches(
-                new CommandLineSwitches(),
-                switches,
-                ref projectFile,
-                ref targets,
-                ref toolsVersion,
-                ref globalProperties,
-                ref loggers,
-                ref verbosity,
-                ref originalVerbosity,
-                ref distributedLoggerRecords,
-    #if FEATURE_XML_SCHEMA_VALIDATION
-                ref needToValidateProject,
-                ref schemaFile,
-    #endif
-                ref cpuCount,
-                ref enableNodeReuse,
-                ref preprocessWriter,
-                ref targetsWriter,
-                ref detailedSummary,
-                ref warningsAsErrors,
-                ref warningsNotAsErrors,
-                ref warningsAsMessages,
-                ref enableRestore,
-                ref interactive,
-                ref profilerLogger,
-                ref enableProfiler,
-                ref restoreProperties,
-                ref isolateProjects,
-                ref graphBuild,
-                ref inputResultsCaches,
-                ref outputResultsCache,
-                ref lowPriority,
-                ref question,
-                ref getProperty,
-                ref getItem,
-                ref getTargetResult,
-                recursing,
-                commandLine).ShouldBeTrue();
-
-            verbosity.ShouldBe(LoggerVerbosity.Quiet, "Should have been quiet; was " + verbosity);
-        }
-
         [Fact]
         public void SetParameterizedSwitchTests3()
         {
diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index b7daaa6fe89..3195e3dd4bd 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2309,7 +2309,7 @@ private static bool GatherAutoResponseFileSwitchesFromFullPath(string autoRespon
         /// performs deeper error checking on the switches and their parameters.
         /// </summary>
         /// <returns>true, if build can be invoked</returns>
-        internal static bool ProcessCommandLineSwitches(
+        private static bool ProcessCommandLineSwitches(
             CommandLineSwitches switchesFromAutoResponseFile,
             CommandLineSwitches switchesNotFromAutoResponseFile,
             ref string projectFile,
