diff --git a/src/Build/BackEnd/Components/Logging/EventSourceSink.cs b/src/Build/BackEnd/Components/Logging/EventSourceSink.cs
index ef8e14ff2f3..340dfafc495 100644
--- a/src/Build/BackEnd/Components/Logging/EventSourceSink.cs
+++ b/src/Build/BackEnd/Components/Logging/EventSourceSink.cs
@@ -537,7 +537,7 @@ private void RaiseProjectStartedEvent(object sender, ProjectStartedEventArgs bui
                     // if a logger has failed politely, abort immediately
                     // first unregister all loggers, since other loggers may receive remaining events in unexpected orderings
                     // if a fellow logger is throwing in an event handler.
-                    this.UnregisterAllEventHandlers();
+                    UnregisterAllEventHandlers();
                     throw;
                 }
                 catch (Exception exception)
@@ -548,7 +548,6 @@ private void RaiseProjectStartedEvent(object sender, ProjectStartedEventArgs bui
                     }
 
                     InternalLoggerException.Throw(exception, buildEvent, "FatalErrorWhileLogging", false);
-                    UnregisterAllEventHandlers();
                 }
             }
 
@@ -892,7 +891,6 @@ private void RaiseAnyEvent(object sender, BuildEventArgs buildEvent)
                     }
 
                     InternalLoggerException.Throw(exception, buildEvent, "FatalErrorWhileLogging", false);
-                    UnregisterAllEventHandlers();
                 }
             }
         }
