diff --git a/src/Build/BackEnd/Client/MSBuildClient.cs b/src/Build/BackEnd/Client/MSBuildClient.cs
index b2329da56b5..2877a3d3e9d 100644
--- a/src/Build/BackEnd/Client/MSBuildClient.cs
+++ b/src/Build/BackEnd/Client/MSBuildClient.cs
@@ -371,7 +371,7 @@ private bool TrySendPacket(Func<INodePacket> packetResolver)
         /// <returns> Whether MSBuild server was started successfully.</returns>
         private bool TryLaunchServer()
         {
-            string serverLaunchMutexName = $@"Global\server-launch-{_handshake.ComputeHash()}";
+            string serverLaunchMutexName = $@"Global\msbuild-server-launch-{_handshake.ComputeHash()}";
             using var serverLaunchMutex = ServerNamedMutex.OpenOrCreateMutex(serverLaunchMutexName, out bool mutexCreatedNew);
             if (!mutexCreatedNew)
             {
diff --git a/src/Build/BackEnd/Node/OutOfProcServerNode.cs b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
index 6a0b4c94242..5d5cbc20acf 100644
--- a/src/Build/BackEnd/Node/OutOfProcServerNode.cs
+++ b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
@@ -153,10 +153,10 @@ internal static string GetPipeName(ServerNodeHandshake handshake)
             => NamedPipeUtil.GetPlatformSpecificPipeName($"MSBuildServer-{handshake.ComputeHash()}");
 
         internal static string GetRunningServerMutexName(ServerNodeHandshake handshake)
-            => $@"Global\server-running-{handshake.ComputeHash()}";
+            => $@"Global\msbuild-server-running-{handshake.ComputeHash()}";
 
         internal static string GetBusyServerMutexName(ServerNodeHandshake handshake)
-            => $@"Global\server-busy-{handshake.ComputeHash()}";
+            => $@"Global\msbuild-server-busy-{handshake.ComputeHash()}";
 
         #region INodePacketFactory Members
 
