diff --git a/src/Tasks/GetCompatiblePlatform.cs b/src/Tasks/GetCompatiblePlatform.cs
index 3063fd1f8d8..c86c88199d7 100644
--- a/src/Tasks/GetCompatiblePlatform.cs
+++ b/src/Tasks/GetCompatiblePlatform.cs
@@ -134,10 +134,10 @@ public override bool Execute()
 
             foreach (string s in stringTable.Trim().Split(MSBuildConstants.SemicolonChar, StringSplitOptions.RemoveEmptyEntries))
             {
-                string[] keyVal = s.Trim().Split(MSBuildConstants.EqualsChar, StringSplitOptions.RemoveEmptyEntries);
+                string[] keyVal = s.Trim().Split(MSBuildConstants.EqualsChar);
 
                 // Invalid table, don't use it.
-                if (keyVal.Length <= 1 || keyVal.Length > 2)
+                if (keyVal.Length != 2 || string.IsNullOrEmpty(keyVal[0]) || string.IsNullOrEmpty(keyVal[1]))
                 {
                     Log.LogWarningWithCodeFromResources("GetCompatiblePlatform.InvalidLookupTableFormat", stringTable);
                     return null;
