diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index cccf36ed84e..53cdafdb15b 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -4527,8 +4527,8 @@ private static void ShowFeatureAvailability(string[] features)
                     jsonNode[featureName] = Features.CheckFeatureAvailability(featureName).ToString();
                 }
 
-                var s_options = new JsonSerializerOptions() { AllowTrailingCommas = false, WriteIndented = true };
-                Console.WriteLine(jsonNode.ToJsonString(s_options));
+                var options = new JsonSerializerOptions() { AllowTrailingCommas = false, WriteIndented = true };
+                Console.WriteLine(jsonNode.ToJsonString(options));
             }
         }
     }
