diff --git a/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs b/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs
index 633554ab190..ef905d4acb1 100644
--- a/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs
+++ b/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs
@@ -83,7 +83,7 @@ internal virtual IReadOnlyList<SdkResolverManifest> GetResolversManifests(Elemen
             }
             finally
             {
-                MSBuildEventSource.Log.SdkResolverFindResolversManifestsStop(allResolversManifests is null ? 0 : allResolversManifests.Count);
+                MSBuildEventSource.Log.SdkResolverFindResolversManifestsStop(allResolversManifests?.Count ?? 0);
             }
             return allResolversManifests;
         }
