diff --git a/src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs b/src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs
index 9e0909a1e94..9ce1f4caeca 100644
--- a/src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs
+++ b/src/Build/BuildCheck/Acquisition/BuildCheckAcquisitionModule.cs
@@ -32,7 +32,7 @@ internal class BuildCheckAcquisitionModule : IBuildCheckAcquisitionModule
     /// </summary>
     public List<BuildAnalyzerFactory> CreateBuildAnalyzerFactories(AnalyzerAcquisitionData analyzerAcquisitionData, BuildEventContext buildEventContext)
     {
-        var analyzersFactories = new List<BuildAnalyzerFactory?>();
+        var analyzersFactories = new List<BuildAnalyzerFactory>();
 
         try
         {
