diff --git a/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFile.cs b/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFile.cs
index 5f3f75ab761..6c6a9daf26c 100644
--- a/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFile.cs
+++ b/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFile.cs
@@ -8,7 +8,7 @@
 
 namespace Microsoft.Build.Logging
 {
-    public class ArchiveFile
+    public sealed class ArchiveFile
     {
         public ArchiveFile(string fullPath, Stream contentStream)
         {
diff --git a/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFileEventArgs.cs b/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFileEventArgs.cs
index 7045f261d86..45bc252a066 100644
--- a/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFileEventArgs.cs
+++ b/src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFileEventArgs.cs
@@ -6,7 +6,7 @@
 
 namespace Microsoft.Build.Logging;
 
-public class ArchiveFileEventArgs : EventArgs
+public sealed class ArchiveFileEventArgs : EventArgs
 {
     private ArchiveFile  _archiveFile;
     private bool _resultSet;
