diff --git a/src/Shared/TaskParameter.cs b/src/Shared/TaskParameter.cs
index 0a4ac9d4927..4595adaba62 100644
--- a/src/Shared/TaskParameter.cs
+++ b/src/Shared/TaskParameter.cs
@@ -685,8 +685,9 @@ private void TranslateValueType(ITranslator translator)
 
             if (translator.Mode == TranslationDirection.ReadFromStream)
             {
-                // We don't know how to convert the string back to the original value type.
-                // This is fine because the engine would eventually convert it to string anyway.
+                // We don't know how to convert the string back to the original value type. This is fine because output
+                // task parameters are anyway converted to strings by the engine (see TaskExecutionHost.GetValueOutputs)
+                // and input task parameters of custom value types are not supported.
                 _wrappedParameter = valueString;
             }
         }
