diff --git a/src/Tasks/GenerateResource.cs b/src/Tasks/GenerateResource.cs
index b888da26a59..d5c61ecd8f6 100644
--- a/src/Tasks/GenerateResource.cs
+++ b/src/Tasks/GenerateResource.cs
@@ -1705,8 +1705,7 @@ private void UpdateNewestUncorrelatedInputWriteTime()
 
             // Check the timestamp of each of the passed-in references to find the newest;
             // and then the additional inputs
-            ITaskItem[] inputs = this.References ?? [.. (this.AdditionalInputs ?? [])];
-
+            var inputs = (this.References ?? []).Concat(this.AdditionalInputs ?? []);
             foreach (ITaskItem input in inputs)
             {
                 DateTime time = NativeMethodsShared.GetLastWriteFileUtcTime(input.ItemSpec);
