diff --git a/src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs b/src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs
index 36f59dbce2f..243d54d799c 100644
--- a/src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs
+++ b/src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs
@@ -891,7 +891,15 @@ public void ReadingCorruptedStreamThrows()
                 var binaryReader = new BinaryReader(memoryStream);
                 using var buildEventArgsReader = new BuildEventArgsReader(binaryReader, BinaryLogger.FileFormatVersion);
 
-                Assert.Throws<EndOfStreamException>(() => buildEventArgsReader.Read());
+                try
+                {
+                    buildEventArgsReader.Read();
+                }
+                catch (Exception e)
+                {
+                    // if the EndOfStreamException is received during EventArgs parsing - the parsing code will translate it to InvalidDataException
+                    Assert.True(e is InvalidDataException or EndOfStreamException, "Abruptly ended stream should lead to InvalidDataException or EndOfStreamException");
+                }
             }
         }
 
