diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index ba417e063f2..8dc4c5dd9d0 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -892,7 +892,7 @@ dotnet_diagnostic.SA1212.severity = none
 dotnet_diagnostic.SA1214.severity = suggestion
 
 # Using static directives should be placed at the correct location
-dotnet_diagnostic.SA1216.severity = suggestion
+dotnet_diagnostic.SA1216.severity = warning
 
 # The using static directives within a C# code file are not sorted alphabetically by full type name.
 dotnet_diagnostic.SA1217.severity = suggestion
diff --git a/src/Build.UnitTests/Graph/ProjectGraph_Tests.cs b/src/Build.UnitTests/Graph/ProjectGraph_Tests.cs
index 4c220e35177..934c2ec3da9 100644
--- a/src/Build.UnitTests/Graph/ProjectGraph_Tests.cs
+++ b/src/Build.UnitTests/Graph/ProjectGraph_Tests.cs
@@ -11,12 +11,12 @@
 using Microsoft.Build.Exceptions;
 using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
-using static Microsoft.Build.Graph.UnitTests.GraphTestingUtilities;
 using Microsoft.Build.Shared;
 using Microsoft.Build.UnitTests;
 using Shouldly;
 using Xunit;
 using Xunit.Abstractions;
+using static Microsoft.Build.Graph.UnitTests.GraphTestingUtilities;
 
 #nullable disable
 
