diff --git a/src/Framework/Microsoft.Build.Framework.csproj b/src/Framework/Microsoft.Build.Framework.csproj
index cb83eb4848c..b31edfef74b 100644
--- a/src/Framework/Microsoft.Build.Framework.csproj
+++ b/src/Framework/Microsoft.Build.Framework.csproj
@@ -57,9 +57,6 @@
     <Compile Include="..\Shared\IMSBuildElementLocation.cs">
       <Link>Shared\IMSBuildElementLocation.cs</Link>
     </Compile>
-    <Compile Include="..\Shared\ConversionUtilities.cs">
-      <Link>Shared\ConversionUtilities.cs</Link>
-    </Compile>
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETFramework'">
diff --git a/src/Framework/Telemetry/OpenTelemetryManager.cs b/src/Framework/Telemetry/OpenTelemetryManager.cs
index ad463a667ab..5171ee40f79 100644
--- a/src/Framework/Telemetry/OpenTelemetryManager.cs
+++ b/src/Framework/Telemetry/OpenTelemetryManager.cs
@@ -13,7 +13,7 @@
 using System;
 using System.Diagnostics;
 using System.Threading;
-using Microsoft.Build.Shared;
+using System.Globalization;
 
 namespace Microsoft.Build.Framework.Telemetry
 {
@@ -228,7 +228,7 @@ private bool IsSampled()
         {
             string? sampleRateString =
                 Environment.GetEnvironmentVariable(name);
-            return ConversionUtilities.TryConvertDecimalOrHexToDouble(sampleRateString, out double result) ? result : null;
+            return double.TryParse(sampleRateString, NumberStyles.AllowDecimalPoint | NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture.NumberFormat, out double result) ? result : null;
         }
 
         /// <summary>
