diff --git a/src/Build/Instance/TaskRegistry.cs b/src/Build/Instance/TaskRegistry.cs
index 128b4b4774b..91d6d55c8d8 100644
--- a/src/Build/Instance/TaskRegistry.cs
+++ b/src/Build/Instance/TaskRegistry.cs
@@ -619,12 +619,19 @@ private Dictionary<RegisteredTaskIdentity, List<RegisteredTaskRecord>> GetReleva
 
             List<RegisteredTaskRecord> taskAssemblies;
 
-            if (exactMatchRequired && _taskRegistrations.TryGetValue(taskIdentity, out taskAssemblies))
+            // if we find an exact match
+            if (_taskRegistrations.TryGetValue(taskIdentity, out taskAssemblies))
             {
+                // we're done
                 relevantTaskRegistrations[taskIdentity] = taskAssemblies;
                 return relevantTaskRegistrations;
             }
 
+            if (exactMatchRequired)
+            {
+                return relevantTaskRegistrations;
+            }
+
             // look through all task declarations for partial matches
             foreach (KeyValuePair<RegisteredTaskIdentity, List<RegisteredTaskRecord>> taskRegistration in _taskRegistrations)
             {
