diff --git a/src/Package/Localization/Microsoft.Build.Localization.nuspec b/src/Package/Localization/Microsoft.Build.Localization.nuspec
index 6d6e0e605c4..3a345c0a1ee 100644
--- a/src/Package/Localization/Microsoft.Build.Localization.nuspec
+++ b/src/Package/Localization/Microsoft.Build.Localization.nuspec
@@ -31,7 +31,7 @@
     <file src="pt-BR\Microsoft.Build.Framework.resources.dll" target="lib\netstandard1.3\pt-BR\Microsoft.Build.Framework.resources.dll" />
     <file src="ru\Microsoft.Build.Framework.resources.dll" target="lib\netstandard1.3\ru\Microsoft.Build.Framework.resources.dll" />
     <file src="tr\Microsoft.Build.Framework.resources.dll" target="lib\netstandard1.3\tr\Microsoft.Build.Framework.resources.dll" />
-    <file src="zh-Hans\Microsoft.Build.Framework..resources.dll" target="lib\netstandard1.3\zh-Hans\Microsoft.Build.Framework.resources.dll" />
+    <file src="zh-Hans\Microsoft.Build.Framework.resources.dll" target="lib\netstandard1.3\zh-Hans\Microsoft.Build.Framework.resources.dll" />
     <file src="zh-Hant\Microsoft.Build.Framework.resources.dll" target="lib\netstandard1.3\zh-Hant\Microsoft.Build.Framework.resources.dll" />
 
   </files>
