diff --git a/src/Utilities/TaskItem.cs b/src/Utilities/TaskItem.cs
index 919b460d538..df52ab93423 100644
--- a/src/Utilities/TaskItem.cs
+++ b/src/Utilities/TaskItem.cs
@@ -94,7 +94,7 @@ public TaskItem()
         /// </param>
         public TaskItem(
             string itemSpec,
-            bool treatAsFilePath = true)
+            bool treatAsFilePath)
         {
             ErrorUtilities.VerifyThrowArgumentNull(itemSpec);
 
