diff --git a/src/Shared/TempFileUtilities.cs b/src/Shared/TempFileUtilities.cs
index fb80ce2fe49..1f42417662a 100644
--- a/src/Shared/TempFileUtilities.cs
+++ b/src/Shared/TempFileUtilities.cs
@@ -152,8 +152,6 @@ internal static string GetTemporaryFile(string directory, string fileName, strin
 
                 string file = Path.Combine(directory, $"{fileName}{extension}");
 
-                ErrorUtilities.VerifyThrow(!FileSystems.Default.FileExists(file), "Guid should be unique");
-
                 if (createFile)
                 {
                     File.WriteAllText(file, string.Empty);
