diff --git a/src/Shared/NativeMethodsShared.cs b/src/Shared/NativeMethodsShared.cs
index b8b5fb8de0e..9eb9ce25aeb 100644
--- a/src/Shared/NativeMethodsShared.cs
+++ b/src/Shared/NativeMethodsShared.cs
@@ -525,7 +525,8 @@ public unsafe static int GetPhysicalCoreCount()
                         // Walk each SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX in the buffer, where the Size of each dictates how
                         // much space it's consuming.  For each group relation, count the number of active processors in each of its group infos.
                         int processorCount = 0;
-                        byte* ptr = bufferPtr, endPtr = bufferPtr + len;
+                        byte* ptr = bufferPtr;
+                        byte* endPtr = bufferPtr + len;
                         while (ptr < endPtr)
                         {
                             var current = (SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*)ptr;
