diff --git a/eng/Versions.props b/eng/Versions.props
index 25a4c5803c5..db8bad27935 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -34,6 +34,7 @@
   <!-- Toolset Dependencies -->
   <PropertyGroup>
     <DotNetCliVersion>3.1.100</DotNetCliVersion>
+    <DropAppVersion>18.165.29912-buildid11693003</DropAppVersion>
   </PropertyGroup>
   <Target Name="OverrideArcadeFileVersion" AfterTargets="_InitializeAssemblyVersion">
     <!-- See https://github.com/dotnet/arcade/issues/3386
