diff --git a/src/Build/BackEnd/Node/NodeConfiguration.cs b/src/Build/BackEnd/Node/NodeConfiguration.cs
index dc6274d5ff5..d6c27aa9afa 100644
--- a/src/Build/BackEnd/Node/NodeConfiguration.cs
+++ b/src/Build/BackEnd/Node/NodeConfiguration.cs
@@ -164,7 +164,7 @@ public void Translate(ITranslator translator)
 #if FEATURE_APPDOMAIN
             if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10) || !Traits.Instance.EscapeHatches.IsBinaryFormatterSerializationAllowed)
             {
-                byte[] _appDomainConfigBytes = null;
+                byte[] appDomainConfigBytes = null;
 
                 // Set the configuration bytes just before serialization in case the SetConfigurationBytes was invoked during lifetime of this instance.
                 if (translator.Mode == TranslationDirection.WriteToStream)
