diff --git a/src/Build/Evaluation/ProjectRootElementCache.cs b/src/Build/Evaluation/ProjectRootElementCache.cs
index 87164c522b3..66cab298d8f 100644
--- a/src/Build/Evaluation/ProjectRootElementCache.cs
+++ b/src/Build/Evaluation/ProjectRootElementCache.cs
@@ -446,16 +446,18 @@ internal override void DiscardImplicitReferences()
                         {
                             _weakCache[projectPath] = rootElement;
                         }
+                    }
+                }
 
-                        if (rootElement.IsExplicitlyLoaded && oldStrongCache.Contains(rootElement))
-                        {
-                            _strongCache.AddFirst(rootElement);
-                        }
-                        else
-                        {
-                            _strongCache.Remove(rootElement);
-                            RaiseProjectRootElementRemovedFromStrongCache(rootElement);
-                        }
+                for (LinkedListNode<ProjectRootElement> node = oldStrongCache.First; node is not null; node = node.Next)
+                {
+                    if (node.Value.IsExplicitlyLoaded)
+                    {
+                        _strongCache.AddFirst(node);
+                    }
+                    else
+                    {
+                        RaiseProjectRootElementRemovedFromStrongCache(node.Value);
                     }
                 }
             }
