diff --git a/src/Build/Logging/BinaryLogger/BinaryLogger.cs b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
index 99887549606..d4c37461938 100644
--- a/src/Build/Logging/BinaryLogger/BinaryLogger.cs
+++ b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
@@ -72,7 +72,7 @@ public sealed class BinaryLogger : ILogger
         // version 20:
         //   - TaskStartedEventArgs: Added TaskAssemblyLocation property
         // version 21:
-        //   - TaskParameterEventArgs: Added ParameterName property
+        //   - TaskParameterEventArgs: Added ParameterName and PropertyName properties
 
         // This should be never changed.
         // The minimum version of the binary log reader that can read log of above version.
