diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index aa6fd17b06e..5e3b1ecd653 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -120,7 +120,7 @@ public Copy()
 
         /// <summary>
         /// Gets or sets a value that indicates whether to create Symbolic Links for the copied files
-        // rather than copy the files, if it's possible to do so.
+        /// rather than copy the files, if it's possible to do so.
         /// </summary>
         public bool UseSymboliclinksIfPossible { get; set; } = s_forceSymlinks;
 
