diff --git a/src/Framework/BuildEventArgs.cs b/src/Framework/BuildEventArgs.cs
index 4484326cc14..307b0bed35b 100644
--- a/src/Framework/BuildEventArgs.cs
+++ b/src/Framework/BuildEventArgs.cs
@@ -181,7 +181,7 @@ public BuildEventContext BuildEventContext
         /// </summary>
         /// <param name="writer">Binary writer which is attached to the stream the event will be serialized into</param>
         /// <param name="messageToWrite">The message to write to the stream.</param>
-        private protected void WriteToStream(BinaryWriter writer, string messageToWrite)
+        private protected void WriteToStreamWithExplicitMessage(BinaryWriter writer, string messageToWrite)
         {
             writer.WriteOptionalString(messageToWrite);
             writer.WriteOptionalString(helpKeyword);
@@ -197,7 +197,7 @@ private protected void WriteToStream(BinaryWriter writer, string messageToWrite)
         /// <param name="writer">Binary writer which is attached to the stream the event will be serialized into</param>
         internal virtual void WriteToStream(BinaryWriter writer)
         {
-            WriteToStream(writer, message);
+            WriteToStreamWithExplicitMessage(writer, message);
         }
 
         /// <summary>
diff --git a/src/Framework/LazyFormattedBuildEventArgs.cs b/src/Framework/LazyFormattedBuildEventArgs.cs
index 7d3d2fd8cf0..23529cc81db 100644
--- a/src/Framework/LazyFormattedBuildEventArgs.cs
+++ b/src/Framework/LazyFormattedBuildEventArgs.cs
@@ -130,7 +130,7 @@ internal override void WriteToStream(BinaryWriter writer)
             object argsOrMessage = argumentsOrFormattedMessage;
             if (argsOrMessage is object[] arguments && arguments.Length > 0)
             {
-                base.WriteToStream(writer, RawMessage);
+                base.WriteToStream(writer);
                 writer.Write(arguments.Length);
 
                 foreach (object argument in arguments)
@@ -142,7 +142,7 @@ internal override void WriteToStream(BinaryWriter writer)
             }
             else
             {
-                base.WriteToStream(writer, (argsOrMessage is string formattedMessage) ? formattedMessage : RawMessage);
+                base.WriteToStreamWithExplicitMessage(writer, (argsOrMessage is string formattedMessage) ? formattedMessage : RawMessage);
                 writer.Write(-1);
             }
 
