diff --git a/src/Framework/FileClassifier.cs b/src/Framework/FileClassifier.cs
index 5d7512c0687..b9c72fdcd1f 100644
--- a/src/Framework/FileClassifier.cs
+++ b/src/Framework/FileClassifier.cs
@@ -24,7 +24,7 @@ namespace Microsoft.Build.Framework
     ///         </list>
     ///     </para>
     /// </remarks>
-    internal class FileClassifier     
+    internal class FileClassifier
     {
         private const StringComparison PathComparison = StringComparison.OrdinalIgnoreCase;
 
@@ -33,6 +33,11 @@ internal class FileClassifier
         /// </summary>
         private static readonly char[] s_semicolonDelimiter = {';'};
 
+        /// <summary>
+        ///     Singe, static <see cref="Lazy{T}"/> instance of shared file FileClassifier for <see cref="Shared"/> member.
+        /// </summary>
+        private static readonly Lazy<FileClassifier> s_sharedInstance = new(() => new FileClassifier());
+
         private readonly ConcurrentDictionary<string, string> _knownImmutableDirectory = new(StringComparer.OrdinalIgnoreCase);
 
         /// <summary>
@@ -79,12 +84,13 @@ static string GetVSInstallationDirectory()
         }
 
         /// <summary>
-        ///     Shared singleton instance
+        ///     Shared singleton instance.
         /// </summary>
-        public static FileClassifier Shared { get; } = new();
+        public static FileClassifier Shared => s_sharedInstance.Value;
 
         /// <summary>
-        ///     Try add paths found in the <c>NuGetPackageFolders</c> property value for a project into set of known immutable paths.
+        ///     Try add paths found in the <c>NuGetPackageFolders</c> property value for a project into set of known immutable
+        ///     paths.
         ///     Project files under any of these folders are considered non-modifiable.
         /// </summary>
         /// <remarks>
