diff --git a/src/Build/Graph/GraphBuilder.cs b/src/Build/Graph/GraphBuilder.cs
index d28025a8f36..cc54f99ccab 100644
--- a/src/Build/Graph/GraphBuilder.cs
+++ b/src/Build/Graph/GraphBuilder.cs
@@ -318,8 +318,8 @@ meanwhile another thread passes the same checks with the same data and inserts i
 
         internal static string FormatCircularDependencyError(List<string> projectsInCycle)
         {
-            const int MAX_PATH = 260;
-            var errorMessage = new StringBuilder(projectsInCycle.Count * MAX_PATH);
+            var errorMessage = new StringBuilder(projectsInCycle.Select(p => p.Length).Sum());
+
             errorMessage.AppendLine();
             for (var i = projectsInCycle.Count - 1; i >= 0; i--)
             {
