diff --git a/src/MSBuild/TerminalLogger/TerminalLogger.cs b/src/MSBuild/TerminalLogger/TerminalLogger.cs
index f73ab039139..cff8950efb9 100644
--- a/src/MSBuild/TerminalLogger/TerminalLogger.cs
+++ b/src/MSBuild/TerminalLogger/TerminalLogger.cs
@@ -1011,8 +1011,9 @@ private string FormatEventMessage(
                 }
                 else
                 {
+                    builder.AppendLine();
                     builder.Append(indent);
-                    builder.AppendLine(line);
+                    builder.Append(line);
                 }
             }
         }
@@ -1030,8 +1031,9 @@ private static void WrapText(StringBuilder sb, string text, int maxLength, strin
         while (start < text.Length)
         {
             int length = Math.Min(maxLength - indent.Length, text.Length - start);
+            sb.AppendLine();
             sb.Append(indent);
-            sb.AppendLine(text.Substring(start, length));
+            sb.Append(text.Substring(start, length));
 
             start += length;
         }
