diff --git a/src/Framework/BuildEventArgs.cs b/src/Framework/BuildEventArgs.cs
index 014a1ce3111..e98e2d5bc16 100644
--- a/src/Framework/BuildEventArgs.cs
+++ b/src/Framework/BuildEventArgs.cs
@@ -201,16 +201,6 @@ internal virtual void WriteToStream(BinaryWriter writer)
         }
 
 
-        /// <summary>
-        /// Convenience access point for CreateFromStream method to avoid making everything public.
-        /// </summary>
-        /// <param name="reader"></param>
-        /// <param name="version"></param>
-        public void PublicCreateFromStream(BinaryReader reader, int version)
-        {
-            CreateFromStream(reader, version);
-        }
-
         /// <summary>
         /// Deserializes from a stream through a binary reader
         /// </summary>
diff --git a/src/Shared/LogMessagePacketBase.cs b/src/Shared/LogMessagePacketBase.cs
index 4bada688cc5..15e838eef7b 100644
--- a/src/Shared/LogMessagePacketBase.cs
+++ b/src/Shared/LogMessagePacketBase.cs
@@ -496,7 +496,7 @@ internal void ReadFromStream(ITranslator translator)
                 readerMethod(translator.Reader, packetVersion);
 
 #else
-                _buildEvent.PublicCreateFromStream(translator.Reader, packetVersion);
+                _buildEvent.CreateFromStream(translator.Reader, packetVersion);
 #endif
 
                 
