diff --git a/src/Build/Logging/FancyLogger/FancyLogger.cs b/src/Build/Logging/FancyLogger/FancyLogger.cs
index e4fdde2078e..acafeaa3df9 100644
--- a/src/Build/Logging/FancyLogger/FancyLogger.cs
+++ b/src/Build/Logging/FancyLogger/FancyLogger.cs
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Threading.Tasks;
 using Microsoft.Build.Framework;
 
 namespace Microsoft.Build.Logging.FancyLogger
@@ -48,12 +49,18 @@ public void Initialize(IEventSource eventSource)
             FancyLoggerBuffer.WriteNewLine("");
             // for (int i = 0; i < 120; i++) FancyLoggerBuffer.WriteNewLine( new string(Convert.ToString(i, 16)[0], 200) );
             FancyLoggerBuffer.Render();
+
+
+            Task.Run(() =>
+            {
+            });
         }
 
         // Build
         void eventSource_BuildStarted(object sender, BuildStartedEventArgs e)
         {
         }
+
         void eventSource_BuildFinished(object sender, BuildFinishedEventArgs e)
         {
             Succeeded = e.Succeeded;
@@ -89,7 +96,7 @@ void eventSource_TargetStarted(object sender, TargetStartedEventArgs e)
             if (!projects.TryGetValue(id, out FancyLoggerProjectNode? node)) return;
             // Update
             node.AddTarget(e);
-            node.Log();
+            // node.Log();
         }
         void eventSource_TargetFinished(object sender, TargetFinishedEventArgs e)
         {
@@ -98,7 +105,7 @@ void eventSource_TargetFinished(object sender, TargetFinishedEventArgs e)
             if (!projects.TryGetValue(id, out FancyLoggerProjectNode? node)) return;
             // Update
             node.FinishedTargets++;
-            node.Log();
+            // node.Log();
         }
 
         // Task
@@ -110,7 +117,7 @@ void eventSource_TaskStarted(object sender, TaskStartedEventArgs e)
             if (!projects.TryGetValue(id, out FancyLoggerProjectNode? node)) return;
             // Update
             node.AddTask(e);
-            node.Log();
+            // node.Log();
             existingTasks++;
         }
 
@@ -126,7 +133,7 @@ void eventSource_MessageRaised(object sender, BuildMessageEventArgs e)
             if (!projects.TryGetValue(id, out FancyLoggerProjectNode? node)) return;
             // Update
             node.AddMessage(e);
-            node.Log();
+            // node.Log();
         }
         void eventSource_WarningRaised(object sender, BuildWarningEventArgs e)
         {
@@ -135,7 +142,7 @@ void eventSource_WarningRaised(object sender, BuildWarningEventArgs e)
             if (!projects.TryGetValue(id, out FancyLoggerProjectNode? node)) return;
             // Update
             node.AddWarning(e);
-            node.Log();
+            // node.Log();
         }
         void eventSource_ErrorRaised(object sender, BuildErrorEventArgs e)
         {
@@ -144,7 +151,7 @@ void eventSource_ErrorRaised(object sender, BuildErrorEventArgs e)
             if (!projects.TryGetValue(id, out FancyLoggerProjectNode? node)) return;
             // Update
             node.AddError(e);
-            node.Log();
+            // node.Log();
         }
 
 
diff --git a/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs b/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
index 0e6a88270ae..6fc1b030cdf 100644
--- a/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
+++ b/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
@@ -41,7 +41,7 @@ public class FancyLoggerBuffer
     {
         private static List<FancyLoggerBufferLine> Lines = new();
         private static int TopLineIndex = 0;
-        private static bool AutoScrollEnabled = true;
+        // private static bool AutoScrollEnabled = true;
         public static void Initialize()
         {
             // Use alternate buffer
@@ -59,7 +59,7 @@ public static void Initialize()
 
             Task.Run(() =>
             {
-                while (true)
+                /*while (true)
                 {
                     switch (Console.ReadKey().Key)
                     {
@@ -76,7 +76,7 @@ public static void Initialize()
                             AutoScrollEnabled = !AutoScrollEnabled;
                             break;
                     }
-                }
+                }*/
             });
         }
 
diff --git a/src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs b/src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
index 324785cbbf8..7d771aaf336 100644
--- a/src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
+++ b/src/Build/Logging/FancyLogger/FancyLoggerProjectNode.cs
@@ -5,6 +5,7 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Threading.Tasks;
 using Microsoft.Build.Framework;
 
 namespace Microsoft.Build.Logging.FancyLogger
@@ -87,7 +88,10 @@ public void Log()
             }
 
             // Update buffer
-            FancyLoggerBuffer.Render();
+            Task.Run(() =>
+            {
+                FancyLoggerBuffer.Render();
+            });
         }
 
         public void AddTarget(TargetStartedEventArgs args)
