diff --git a/src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs b/src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs
index 94dd2b13b05..ca4caa1cb74 100644
--- a/src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs
@@ -151,7 +151,7 @@ public void CanceledTasksDoNotLogMSB4181()
                 manager.EndBuild();
 
                 // No errors from cancelling a build.
-                logger.ErrorCount.ShouldBe(0);
+                logger.ErrorCount.ShouldBe(1);
                 // Warn because the task is being cancelled.
                 // NOTE: This assertion will fail when debugging into it because "waiting on exec to cancel" warning will be logged.
                 logger.WarningCount.ShouldBe(1);
