diff --git a/src/Build/BackEnd/Node/OutOfProcServerNode.cs b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
index 5ec25163a82..95637d7b434 100644
--- a/src/Build/BackEnd/Node/OutOfProcServerNode.cs
+++ b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
@@ -292,9 +292,6 @@ private void HandlePacket(INodePacket packet)
                 case NodePacketType.ServerNodeBuildCommand:
                     HandleServerNodeBuildCommand((ServerNodeBuildCommand)packet);
                     break;
-                case NodePacketType.NodeBuildComplete:
-                    HandleNodeBuildComplete((NodeBuildComplete)packet);
-                    break;
             }
         }
 
@@ -394,15 +391,6 @@ public override void Flush()
             }
         }
 
-        /// <summary>
-        /// Handles the NodeBuildComplete packet.
-        /// </summary>
-        private void HandleNodeBuildComplete(NodeBuildComplete buildComplete)
-        {
-            _shutdownReason = buildComplete.PrepareForReuse ? NodeEngineShutdownReason.BuildCompleteReuse : NodeEngineShutdownReason.BuildComplete;
-            _shutdownEvent.Set();
-        }
-
         internal sealed class ServerNamedMutex : IDisposable
         {
             public readonly Mutex _serverMutex;
diff --git a/src/Shared/NodeEndpointOutOfProcBase.cs b/src/Shared/NodeEndpointOutOfProcBase.cs
index 3589785563f..f402ecac71a 100644
--- a/src/Shared/NodeEndpointOutOfProcBase.cs
+++ b/src/Shared/NodeEndpointOutOfProcBase.cs
@@ -14,6 +14,7 @@
 using Microsoft.Build.Shared;
 #if FEATURE_SECURITY_PERMISSIONS || FEATURE_PIPE_SECURITY
 using System.Security.AccessControl;
+using System.Linq;
 #endif
 #if FEATURE_PIPE_SECURITY && FEATURE_NAMED_PIPE_SECURITY_CONSTRUCTOR
 using System.Security.Principal;
