diff --git a/src/Build/BackEnd/Components/Logging/LoggingService.cs b/src/Build/BackEnd/Components/Logging/LoggingService.cs
index 36826b8a70e..213b3009be4 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -942,18 +942,13 @@ private void WarnOnDeprecatedCustomArgsSerialization(LogMessagePacket loggingPac
                 warning.BuildEventContext = buildEventContext;
                 if (warning.ProjectFile == null && buildEventContext.ProjectContextId != BuildEventContext.InvalidProjectContextId)
                 {
-                    if (buildEvent is BuildMessageEventArgs buildMessageEvent)
+                    warning.ProjectFile = buildEvent switch
                     {
-                        warning.ProjectFile = buildMessageEvent.ProjectFile;
-                    }
-                    else if (buildEvent is BuildErrorEventArgs buildErrorEvent)
-                    {
-                        warning.ProjectFile = buildErrorEvent.ProjectFile;
-                    }
-                    else if (buildEvent is BuildWarningEventArgs buildWarningEvent)
-                    {
-                        warning.ProjectFile = buildWarningEvent.ProjectFile;
-                    }
+                        BuildMessageEventArgs buildMessageEvent => buildMessageEvent.ProjectFile,
+                        BuildErrorEventArgs buildErrorEvent => buildErrorEvent.ProjectFile,
+                        BuildWarningEventArgs buildWarningEvent => buildWarningEvent.ProjectFile,
+                        _ => null,
+                    };
                 }
 
                 ProcessLoggingEvent(warning);
