diff --git a/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs b/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs
index 29dce609aed..47514c4ff4d 100644
--- a/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs
+++ b/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs
@@ -8637,9 +8637,6 @@ public void LogsParentAssemblyForEveryConsideredAndRejectedSearchPath()
 
             rar.SearchPaths = new string[]
             {
-                "{CandidateAssemblyFiles}",
-                "{HintPathFromItem}",
-                "{TargetFrameworkDirectory}",
                 "{RawFileName}",
             };
 
