diff --git a/src/Framework/Traits.cs b/src/Framework/Traits.cs
index a19ff522759..32557ee22d7 100644
--- a/src/Framework/Traits.cs
+++ b/src/Framework/Traits.cs
@@ -163,9 +163,15 @@ private static int ParseIntFromEnvironmentVariableOrDefault(string environmentVa
                 : defaultValue;
         }
 
+        /// <summary>
+        /// Parse a double from an environment variable with invariant culture.
+        /// </summary>
         private static double? ParseDoubleFromEnvironmentVariable(string environmentVariable)
         {
-            return double.TryParse(Environment.GetEnvironmentVariable(environmentVariable), out double result)
+            return double.TryParse(Environment.GetEnvironmentVariable(environmentVariable),
+                                  NumberStyles.Float,
+                                  CultureInfo.InvariantCulture,
+                                  out double result)
                 ? result
                 : null;
         }
