diff --git a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
index cd0183d5f1c..b9e32298b17 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
@@ -703,13 +703,6 @@ public void ReleaseCores(int coresToRelease)
         {
             var rms = _host.GetComponent(BuildComponentType.TaskResourceManager) as ResourceManagerService;
 
-            if (coresToRelease > runningTotal)
-            {
-                // TODO: log
-            }
-
-            coresToRelease = Math.Min(runningTotal, coresToRelease);
-
             if (coresToRelease >= 1)
             {
                 rms.ReleaseCores(coresToRelease, _taskLoggingContext);
diff --git a/src/Build/BackEnd/Components/ResourceManager/ResourceManagerService.cs b/src/Build/BackEnd/Components/ResourceManager/ResourceManagerService.cs
index 75079f52adc..d58cb6e9ce7 100644
--- a/src/Build/BackEnd/Components/ResourceManager/ResourceManagerService.cs
+++ b/src/Build/BackEnd/Components/ResourceManager/ResourceManagerService.cs
@@ -118,6 +118,8 @@ public void ReleaseCores(int coresToRelease, TaskLoggingContext _taskLoggingCont
             if (coresToRelease > TotalNumberHeld)
             {
                 _loggingService?.LogWarning(_taskLoggingContext.BuildEventContext, null, null, "ResourceManagerExcessRelease", coresToRelease);
+
+                coresToRelease = TotalNumberHeld;
             }
 
             s.Release(coresToRelease);
