diff --git a/Directory.Build.props b/Directory.Build.props
index b3524ccc277..711b765ffae 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -87,4 +87,12 @@
     <EnableNETAnalyzers>true</EnableNETAnalyzers>
   </PropertyGroup>
 
+  <!-- TODO: Freeze the version for the current upgrade of the sdk. Delete the lines below once the Retargeting merged. -->
+  <PropertyGroup>
+      <XUnitVersion>2.4.2</XUnitVersion>
+      <XUnitAnalyzersVersion>1.1.0</XUnitAnalyzersVersion>
+      <XUnitRunnerConsoleVersion>$(XUnitVersion)</XUnitRunnerConsoleVersion>
+      <XUnitRunnerVisualStudioVersion>2.4.3</XUnitRunnerVisualStudioVersion>
+  </PropertyGroup>
+
 </Project>
diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index 32fcaf2814c..8aaa1b94670 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -1108,11 +1108,4 @@ dotnet_diagnostic.SA1649.severity = suggestion
 dotnet_diagnostic.SA1652.severity = none
 
 # Do not use equality check to check for collection size.
-dotnet_diagnostic.xUnit2013.severity = none
-
-#TODO: fix xUnit new warning 
-dotnet_diagnostic.xUnit1031.severity = suggestion
-dotnet_diagnostic.xUnit1048.severity = suggestion
-dotnet_diagnostic.xUnit2020.severity = suggestion
-dotnet_diagnostic.xUnit1012.severity = suggestion
-dotnet_diagnostic.xUnit2029.severity = suggestion
\ No newline at end of file
+dotnet_diagnostic.xUnit2013.severity = none
\ No newline at end of file
diff --git a/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs b/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
index f9de344a2a0..a2f5b2b7776 100644
--- a/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
@@ -1516,7 +1516,7 @@ public void CancelledBuildWithUnexecutedSubmission()
         /// <summary>
         /// A canceled build
         /// </summary>
-        [Fact]
+        [Fact(Timeout = 20_000)]
         public void CancelledBuild()
         {
             Console.WriteLine("Starting CancelledBuild test that is known to hang.");
