diff --git a/src/Build/Evaluation/LazyItemEvaluator.cs b/src/Build/Evaluation/LazyItemEvaluator.cs
index 8e35e537bc3..3d8db55d65c 100644
--- a/src/Build/Evaluation/LazyItemEvaluator.cs
+++ b/src/Build/Evaluation/LazyItemEvaluator.cs
@@ -434,7 +434,10 @@ public void RemoveMatchingItems(ItemSpec<P, I> itemSpec)
                                 if (fragment.IsMatchNormalized(kvp.Key))
                                 {
                                     items ??= new HashSet<I>();
-                                    items.UnionWith(kvp.Value);
+                                    foreach (I item in kvp.Value)
+                                    {
+                                        items.Add(item);
+                                    }
                                     keysToRemove ??= new List<string>();
                                     keysToRemove.Add(kvp.Key);
                                 }
