diff --git a/src/Build/Logging/FancyLogger/FancyLoggerNode.cs b/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
index 92f672d5bda..98f2ff4be7e 100644
--- a/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
+++ b/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
@@ -161,7 +161,7 @@ public FancyLoggerMessageNode(LazyFormattedBuildEventArgs args)
         public void Log()
         {
             if (Line == null) return;
-            FancyLoggerBuffer.UpdateLine(Line.Id, $"   {ANSIBuilder.Formatting.Italic(Message)}");
+            FancyLoggerBuffer.UpdateLine(Line.Id, $"   └── {ANSIBuilder.Formatting.Italic(Message)}");
         }
     }
 }
