diff --git a/src/Tasks/GenerateResource.cs b/src/Tasks/GenerateResource.cs
index a3c69363cbd..f61b4b1b1c4 100644
--- a/src/Tasks/GenerateResource.cs
+++ b/src/Tasks/GenerateResource.cs
@@ -3151,7 +3151,7 @@ internal void ReadAssemblyResources(String name, String outFileOrDir)
                 // We can't easily filter those.  We can simply skip them.
                 return;
             }
-            catch (ArgumentException)
+            catch (ArgumentException e) when (e.InnerException is BadImageFormatException)
             {
                 // BadImageFormatExceptions can be wrapped in ArgumentExceptions, so catch those, too.
                 return;
