diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index 46b9c54b701..a79eb929686 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -1110,14 +1110,10 @@ public GraphBuildResult Build(BuildParameters parameters, GraphBuildRequestData
         /// </summary>
         public void ShutdownAllNodes()
         {
-            if (_nodeManager == null)
-            {
-                _nodeManager = (INodeManager)((IBuildComponentHost)this).GetComponent(BuildComponentType.NodeManager);
-            }
+            MSBuildClient.ShutdownServer(CancellationToken.None);
 
+            _nodeManager ??= (INodeManager)((IBuildComponentHost)this).GetComponent(BuildComponentType.NodeManager);
             _nodeManager.ShutdownAllNodes();
-
-            MSBuildClient.ShutdownServer(CancellationToken.None);
         }
 
         /// <summary>
diff --git a/src/MSBuild.UnitTests/MSBuildServer_Tests.cs b/src/MSBuild.UnitTests/MSBuildServer_Tests.cs
index cbd8113e7d2..06a22afd6ff 100644
--- a/src/MSBuild.UnitTests/MSBuildServer_Tests.cs
+++ b/src/MSBuild.UnitTests/MSBuildServer_Tests.cs
@@ -21,6 +21,7 @@
 using Shouldly;
 using Xunit;
 using Xunit.Abstractions;
+using Path = System.IO.Path;
 
 namespace Microsoft.Build.Engine.UnitTests
 {
@@ -224,7 +225,7 @@ public void CanShutdownServerProcess(bool byBuildManager)
             // TODO: delete
             var debugFolder = _env.CreateFolder();
             _env.SetEnvironmentVariable("MSBUILDDEBUGENGINE", "1");
-            _env.SetEnvironmentVariable("MSBUILDDEBUGPATH", debugFolder.Path);
+            _env.SetEnvironmentVariable("MSBUILDDEBUGPATH", Path.GetTempPath());
 
             // Just for sure close server, so previous server instances does not effect this run.
             BuildManager.DefaultBuildManager.ShutdownAllNodes();
