diff --git a/src/Build/BuildCheck/Checks/ExecCliBuildCheck.cs b/src/Build/BuildCheck/Checks/ExecCliBuildCheck.cs
index 5d148290ac4..39288fdedbb 100644
--- a/src/Build/BuildCheck/Checks/ExecCliBuildCheck.cs
+++ b/src/Build/BuildCheck/Checks/ExecCliBuildCheck.cs
@@ -133,7 +133,7 @@ private static bool TryGetMatchingKnownBuildCommand(ReadOnlySpan<char> command,
             }
         }
 
-        knownBuildCommand = null;
+        knownBuildCommand = string.Empty;
         return false;
     }
 
