diff --git a/src/Build/Logging/BinaryLogger/BinaryLogger.cs b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
index fd3c8d5e3ca..9da915cb2d6 100644
--- a/src/Build/Logging/BinaryLogger/BinaryLogger.cs
+++ b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
@@ -66,6 +66,7 @@ public sealed class BinaryLogger : ILogger
         // version 18:
         //   - Making ProjectStartedEventArgs, ProjectEvaluationFinishedEventArgs, AssemblyLoadBuildEventArgs equal
         //     between de/serialization roundtrips.
+        //   - Adding serialized events lengths - to support forward compatible reading
         internal const int FileFormatVersion = 18;
 
         private Stream stream;
