diff --git a/src/Tasks/Exec.cs b/src/Tasks/Exec.cs
index 95ef57a6837..64dcc470f86 100644
--- a/src/Tasks/Exec.cs
+++ b/src/Tasks/Exec.cs
@@ -410,7 +410,10 @@ protected override void LogEventsFromTextOutput(string singleLine, MessageImport
 
             if (ConsoleToMSBuild)
             {
-                string trimmedTextLine = singleLine.TrimEnd();
+                string trimmedTextLine = ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10) ?
+                    singleLine.TrimEnd() :
+                    singleLine.Trim();
+
                 if (trimmedTextLine.Length > 0)
                 {
                     // The lines read may be unescaped, so we need to escape them
