diff --git a/src/Build/Construction/Solution/SolutionProjectGenerator.cs b/src/Build/Construction/Solution/SolutionProjectGenerator.cs
index 28ce38d434b..308f5716534 100644
--- a/src/Build/Construction/Solution/SolutionProjectGenerator.cs
+++ b/src/Build/Construction/Solution/SolutionProjectGenerator.cs
@@ -1971,10 +1971,7 @@ private static void AddTraversalReferencesTarget(ProjectInstance traversalProjec
 
             string correctedTargetName = targetName ?? "Build";
 
-            if (traversalProject.Targets.Select(target => target.Key).Contains(correctedTargetName))
-            {
-                traversalProject.RemoveTarget(correctedTargetName);
-            }
+            traversalProject.RemoveTarget(correctedTargetName);
             ProjectTargetInstance target = traversalProject.AddTarget(correctedTargetName, string.Empty, string.Empty, outputItemAsItem, null, string.Empty, string.Empty, string.Empty, string.Empty, false /* legacy target returns behaviour */);
             AddReferencesBuildTask(target, targetName, outputItem);
         }
