diff --git a/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs b/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs
index d33176bb73f..957300c06ac 100644
--- a/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs
+++ b/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs
@@ -645,7 +645,7 @@ public async ValueTask DisposeAsync()
                 });
             }
 
-            await Task.WhenAll(cleanupTasks);
+            await Task.WhenAll(cleanupTasks).ConfigureAwait(false);
 
             if (pluginLogger.HasLoggedErrors)
             {
