diff --git a/ref/Microsoft.Build/net/Microsoft.Build.cs b/ref/Microsoft.Build/net/Microsoft.Build.cs
index 5152392b1a3..83b81af5ad6 100644
--- a/ref/Microsoft.Build/net/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/net/Microsoft.Build.cs
@@ -1531,6 +1531,7 @@ internal MSBuildGlob() { }
         public string FilenamePart { get { throw null; } }
         public string FixedDirectoryPart { get { throw null; } }
         public bool IsLegal { get { throw null; } }
+        public System.Text.RegularExpressions.Regex Regex { get { throw null; } }
         public string WildcardDirectoryPart { get { throw null; } }
         public bool IsMatch(string stringToMatch) { throw null; }
         public Microsoft.Build.Globbing.MSBuildGlob.MatchInfoResult MatchInfo(string stringToMatch) { throw null; }
diff --git a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
index cffbea13779..c977929fe38 100644
--- a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
@@ -1525,6 +1525,7 @@ internal MSBuildGlob() { }
         public string FilenamePart { get { throw null; } }
         public string FixedDirectoryPart { get { throw null; } }
         public bool IsLegal { get { throw null; } }
+        public System.Text.RegularExpressions.Regex Regex { get { throw null; } }
         public string WildcardDirectoryPart { get { throw null; } }
         public bool IsMatch(string stringToMatch) { throw null; }
         public Microsoft.Build.Globbing.MSBuildGlob.MatchInfoResult MatchInfo(string stringToMatch) { throw null; }
diff --git a/src/Build/Globbing/MSBuildGlob.cs b/src/Build/Globbing/MSBuildGlob.cs
index 9460958a40f..e671c976a27 100644
--- a/src/Build/Globbing/MSBuildGlob.cs
+++ b/src/Build/Globbing/MSBuildGlob.cs
@@ -70,6 +70,11 @@ public GlobState(string globRoot, string fileSpec, bool isLegal, string fixedDir
         /// </summary>
         public string FilenamePart => _state.Value.FilenamePart;
 
+        /// <summary>
+        ///     The regex
+        /// </summary>
+        public Regex Regex => _state.Value.Regex;
+
         /// <summary>
         ///     Whether the glob was parsed sucsesfully from a string.
         ///     Illegal glob strings contain:
