diff --git a/src/MSBuild.UnitTests/LiveLogger_Tests.cs b/src/MSBuild.UnitTests/LiveLogger_Tests.cs
index 22d3280622a..c60b54f8a79 100644
--- a/src/MSBuild.UnitTests/LiveLogger_Tests.cs
+++ b/src/MSBuild.UnitTests/LiveLogger_Tests.cs
@@ -36,18 +36,31 @@ public LiveLogger_Tests()
 
 #pragma warning disable CS0067
         public event BuildMessageEventHandler? MessageRaised;
+
         public event BuildErrorEventHandler? ErrorRaised;
+
         public event BuildWarningEventHandler? WarningRaised;
+
         public event BuildStartedEventHandler? BuildStarted;
+
         public event BuildFinishedEventHandler? BuildFinished;
+
         public event ProjectStartedEventHandler? ProjectStarted;
+
         public event ProjectFinishedEventHandler? ProjectFinished;
+
         public event TargetStartedEventHandler? TargetStarted;
+
         public event TargetFinishedEventHandler? TargetFinished;
+
         public event TaskStartedEventHandler? TaskStarted;
+
         public event TaskFinishedEventHandler? TaskFinished;
+
         public event CustomBuildEventHandler? CustomEventRaised;
+
         public event BuildStatusEventHandler? StatusEventRaised;
+
         public event AnyEventHandler? AnyEventRaised;
 #pragma warning restore CS0067
 
