diff --git a/src/Shared/NativeMethodsShared.cs b/src/Shared/NativeMethodsShared.cs
index 9eb9ce25aeb..959ea8aad8c 100644
--- a/src/Shared/NativeMethodsShared.cs
+++ b/src/Shared/NativeMethodsShared.cs
@@ -530,13 +530,20 @@ public unsafe static int GetPhysicalCoreCount()
                         while (ptr < endPtr)
                         {
                             var current = (SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*)ptr;
-                            if (current->Relationship == LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup)
+                            if (current != null && current->Relationship == LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup)
                             {
                                 PROCESSOR_GROUP_INFO* groupInfo = &current->Group.GroupInfo;
-                                int groupCount = current->Group.ActiveGroupCount;
-                                for (int i = 0; i < groupCount; i++)
+                                if (groupInfo != null)
                                 {
-                                    processorCount += (groupInfo + i)->ActiveProcessorCount;
+                                    int groupCount = current->Group.ActiveGroupCount;
+                                    for (int i = 0; i < groupCount; i++)
+                                    {
+                                        var nextGroupInfo = (groupInfo + i);
+                                        if (nextGroupInfo != null)
+                                        {
+                                            processorCount += nextGroupInfo->ActiveProcessorCount;
+                                        }
+                                    }
                                 }
                             }
                             ptr += current->Size;
