diff --git a/src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs b/src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs
index a903ee5b609..6041fbc45ac 100644
--- a/src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs
+++ b/src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs
@@ -451,6 +451,8 @@ public void WarningsChangeWaveTest(string property, string propertyData, bool tr
                     // Since the "no prefix" variations can't do anything with the change wave disabled, this should always fail.
                     MockLogger logger = proj.BuildProjectExpectFailure();
                     logger.ErrorCount.ShouldBe(1);
+                    logger.AssertLogContains($"error {warningCode}");
+
                     logger.AssertLogContains(warningCode);
                 }
                 else
@@ -458,11 +460,11 @@ public void WarningsChangeWaveTest(string property, string propertyData, bool tr
                     MockLogger logger = proj.BuildProjectExpectSuccess();
 
                     logger.WarningCount.ShouldBe(1);
+                    logger.AssertLogContains($"warning {warningCode}");
                     logger.ErrorCount.ShouldBe(0);
 
                     logger.AssertLogContains(warningCode);
                 }
-                ChangeWaves.ResetStateForTests();
             }
         }
 
