diff --git a/src/Framework/Sdk/SdkResolverContext.cs b/src/Framework/Sdk/SdkResolverContext.cs
index 440587b0c47..d932cd2ba9b 100644
--- a/src/Framework/Sdk/SdkResolverContext.cs
+++ b/src/Framework/Sdk/SdkResolverContext.cs
@@ -42,7 +42,7 @@ public abstract class SdkResolverContext
         ///    <format type="text/markdown"><![CDATA[
         /// ## Remarks
         ///     
-        /// File version is based on commit height from our public git repository. This is informational and not equal to the assembly version.
+        /// File version is informational and not equal to the assembly version.
         /// ]]></format>
         /// </remarks>
         public virtual Version MSBuildVersion { get; protected set; }
diff --git a/src/Tasks/CodeTaskFactory.cs b/src/Tasks/CodeTaskFactory.cs
index 90eb5e09eb6..983e9547fe9 100644
--- a/src/Tasks/CodeTaskFactory.cs
+++ b/src/Tasks/CodeTaskFactory.cs
@@ -170,7 +170,7 @@ public TaskPropertyInfo[] GetTaskParameters()
         }
 
         /// <summary>
-        /// Initialzes the task factory.
+        /// Initializes the task factory.
         /// </summary>
         public bool Initialize(string taskName, IDictionary<string, TaskPropertyInfo> taskParameters, string taskElementContents, IBuildEngine taskFactoryLoggingHost)
         {
diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index 5e3b1ecd653..3e67d20a147 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -113,13 +113,13 @@ public Copy()
         public int RetryDelayMilliseconds { get; set; }
 
         /// <summary>
-        /// Gets or sets a value that indicates whether to use Hard Links for the copied files
+        /// Gets or sets a value that indicates whether to use hard links for the copied files
         /// rather than copy the files, if it's possible to do so.
         /// </summary>
         public bool UseHardlinksIfPossible { get; set; }
 
         /// <summary>
-        /// Gets or sets a value that indicates whether to create Symbolic Links for the copied files
+        /// Gets or sets a value that indicates whether to create symbolic links for the copied files
         /// rather than copy the files, if it's possible to do so.
         /// </summary>
         public bool UseSymboliclinksIfPossible { get; set; } = s_forceSymlinks;
