diff --git a/src/Shared/ExceptionHandling.cs b/src/Shared/ExceptionHandling.cs
index d1eaf67c730..4426afebf01 100644
--- a/src/Shared/ExceptionHandling.cs
+++ b/src/Shared/ExceptionHandling.cs
@@ -1,9 +1,9 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-#if BUILDINGAPPXTASKS
 #nullable disable
 
+#if BUILDINGAPPXTASKS
 namespace Microsoft.Build.AppxPackage.Shared
 #else
 using System;
@@ -26,8 +26,6 @@ namespace Microsoft.Build.AppxPackage.Shared
 #endif
 using Microsoft.Build.Framework;
 
-#nullable disable
-
 namespace Microsoft.Build.Shared
 #endif
 {
