diff --git a/src/Shared/FileUtilities.cs b/src/Shared/FileUtilities.cs
index e237129e652..32272f2d001 100644
--- a/src/Shared/FileUtilities.cs
+++ b/src/Shared/FileUtilities.cs
@@ -1075,7 +1075,7 @@ internal static string MakeRelative(string basePath, string path)
             }
             for (int i = index; i < splitPath.Length; i++)
             {
-                sb.Append(splitPath[index]).Append(Path.DirectorySeparatorChar);
+                sb.Append(splitPath[i]).Append(Path.DirectorySeparatorChar);
             }
             sb.Length--;
             return StringBuilderCache.GetStringAndRelease(sb);
