diff --git a/src/Build.UnitTests/BackEnd/OpenTelemetryManager_Tests.cs b/src/Build.UnitTests/BackEnd/OpenTelemetryManager_Tests.cs
index 3132af1fb30..bb3545a838c 100644
--- a/src/Build.UnitTests/BackEnd/OpenTelemetryManager_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/OpenTelemetryManager_Tests.cs
@@ -15,16 +15,16 @@ namespace Microsoft.Build.Framework.Telemetry.Tests
     public class OpenTelemetryManagerTests : IDisposable
     {
         // Store original environment variables so we can restore after each test
-        private readonly string _originalDotnetOptOut;
-        private readonly string _originalMsBuildTelemetryOptOut;
-        private readonly string _originalSampleRateOverride;
+        private readonly string? _originalDotnetOptOut;
+        private readonly string? _originalMsBuildTelemetryOptOut;
+        private readonly string? _originalSampleRateOverride;
 
         public OpenTelemetryManagerTests()
         {
             // Capture existing env vars
-            _originalDotnetOptOut = Environment.GetEnvironmentVariable(TelemetryConstants.DotnetOptOut) ?? "";
-            _originalMsBuildTelemetryOptOut = Environment.GetEnvironmentVariable(TelemetryConstants.TelemetryFxOptoutEnvVarName) ?? "";
-            _originalSampleRateOverride = Environment.GetEnvironmentVariable(TelemetryConstants.TelemetrySampleRateOverrideEnvVarName) ?? "";
+            _originalDotnetOptOut = Environment.GetEnvironmentVariable(TelemetryConstants.DotnetOptOut);
+            _originalMsBuildTelemetryOptOut = Environment.GetEnvironmentVariable(TelemetryConstants.TelemetryFxOptoutEnvVarName);
+            _originalSampleRateOverride = Environment.GetEnvironmentVariable(TelemetryConstants.TelemetrySampleRateOverrideEnvVarName);
 
             // Ensure a clean manager state before each test
             ResetManagerState();
@@ -78,6 +78,7 @@ public void Initialize_ShouldSetStateToUnsampled_WhenNoOverrideOnNetCore()
             OpenTelemetryManager.Instance.DefaultActivitySource.ShouldBeNull();
         }
 #endif
+        
         [WindowsFullFrameworkOnlyTheory]
         [InlineData(true)]
         [InlineData(false)]
diff --git a/src/Framework/Telemetry/ActivityExtensions.cs b/src/Framework/Telemetry/ActivityExtensions.cs
index 3ac32fa6e18..8147b1cfe85 100644
--- a/src/Framework/Telemetry/ActivityExtensions.cs
+++ b/src/Framework/Telemetry/ActivityExtensions.cs
@@ -64,7 +64,6 @@ private static object GetHashed(object value)
 #endif
         }
 
-
         // https://github.com/dotnet/sdk/blob/8bd19a2390a6bba4aa80d1ac3b6c5385527cc311/src/Cli/Microsoft.DotNet.Cli.Utils/Sha256Hasher.cs + workaround for netstandard2.0
 #if NET || NETSTANDARD2_0_OR_GREATER
         private static class Sha256Hasher
