diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index 90398362087..a9493f89bc8 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -821,6 +821,8 @@ void Callback(object? state)
 
         /// <summary>
         /// Point in time snapshot of all worker processes leveraged by this BuildManager.
+        /// This is meant to be used by VS. External users should not this is only best-effort, point-in-time functionality
+        ///  without guarantee of 100% correctness and safety.
         /// </summary>
         /// <returns>Enumeration of <see cref="Process"/> objects that were valid during the time of call to this function.</returns>
         public IEnumerable<Process> GetWorkerProcesses()
