diff --git a/global.json b/global.json
index 2a05e07afeb..c3c79bb7bb8 100644
--- a/global.json
+++ b/global.json
@@ -1,6 +1,8 @@
 {
   "sdk": {
-    "allowPrerelease": true
+    "allowPrerelease": false,
+    "rollForward": "disable",
+    "version": "9.0.100-preview.6.24328.19"
   },
   "tools": {
     "dotnet": "9.0.100-preview.6.24328.19",
diff --git a/src/BuildCheck.UnitTests/TestAssets/InvalidCustomCheck/InvalidCheck.cs b/src/BuildCheck.UnitTests/TestAssets/InvalidCustomCheck/InvalidCheck.cs
index 0b727549e95..e2bcfb57a88 100644
--- a/src/BuildCheck.UnitTests/TestAssets/InvalidCustomCheck/InvalidCheck.cs
+++ b/src/BuildCheck.UnitTests/TestAssets/InvalidCustomCheck/InvalidCheck.cs
@@ -1,4 +1,7 @@
-﻿using System.Collections.Generic;
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
 
 namespace InvalidCustomCheck
 {
