diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index be36680b2e2..fcd2960a715 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2366,6 +2366,7 @@ string commandLine
                         out enableProfiler
                         );
 
+                    // We're finished with defining individual loggers' verbosity at this point, so we don't need to worry about messing them up.
                     if (Traits.Instance.DebugEngine)
                     {
                         verbosity = LoggerVerbosity.Diagnostic;
