diff --git a/eng/Versions.props b/eng/Versions.props
index 2109d36bf19..2bca78395c4 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -18,9 +18,6 @@
     <UsingToolMicrosoftNetCompilers>true</UsingToolMicrosoftNetCompilers>
     <UsingToolSymbolUploader>true</UsingToolSymbolUploader>
     <UsingToolVSSDK>true</UsingToolVSSDK>
-    <!-- Override Arcade's default VSSDK version with one that supports client enablement.
-         Can be removed after Arcade moves up. -->
-    <MicrosoftVSSDKBuildToolsVersion>16.7.13</MicrosoftVSSDKBuildToolsVersion>
   </PropertyGroup>
   <!-- Production Dependencies -->
   <PropertyGroup>
