diff --git a/eng/Versions.props b/eng/Versions.props
index 91ab29ccea6..dcfb8e665c7 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -2,7 +2,7 @@
 <!-- Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the MIT license. See License.txt in the project root for full license information. -->
 <Project>
   <PropertyGroup>
-    <VersionPrefix>17.2.0</VersionPrefix><DotNetFinalVersionKind>release</DotNetFinalVersionKind>
+    <VersionPrefix>17.2.1</VersionPrefix><DotNetFinalVersionKind>release</DotNetFinalVersionKind>
     <AssemblyVersion>15.1.0.0</AssemblyVersion>
     <PreReleaseVersionLabel>preview</PreReleaseVersionLabel>
     <DotNetUseShippingVersions>true</DotNetUseShippingVersions>
diff --git a/src/Build.OM.UnitTests/Construction/ProjectRootElement_Tests.cs b/src/Build.OM.UnitTests/Construction/ProjectRootElement_Tests.cs
index ba0defea01b..618881403c1 100644
--- a/src/Build.OM.UnitTests/Construction/ProjectRootElement_Tests.cs
+++ b/src/Build.OM.UnitTests/Construction/ProjectRootElement_Tests.cs
@@ -1932,6 +1932,11 @@ private static void ValidateDeepCloneAndCopyFrom(ProjectRootElement pre)
             pre3.AddPropertyGroup(); // this should get wiped out in the DeepCopyFrom
             pre3.DeepCopyFrom(pre);
             Assert.Equal(pre.RawXml, pre3.RawXml);
+
+            // DeepCopying from a DeepCopy should work fine (regression test)
+            var pre4 = ProjectRootElement.Create();
+            pre4.DeepCopyFrom(pre2);
+            Assert.Equal(pre.RawXml, pre4.RawXml);
         }
 
         /// <summary>
diff --git a/src/Build/Construction/ProjectElementContainer.cs b/src/Build/Construction/ProjectElementContainer.cs
index 5a2d0da07c9..7348a4cc7ce 100644
--- a/src/Build/Construction/ProjectElementContainer.cs
+++ b/src/Build/Construction/ProjectElementContainer.cs
@@ -433,17 +433,18 @@ private void AppendAttributesAndChildren(XmlNode appendTo, XmlNode appendFrom)
                 }
             }
 
-            // If this element has pure text content, copy that over.
-            if (appendFrom.ChildNodes.Count == 1 && appendFrom.FirstChild.NodeType == XmlNodeType.Text)
-            {
-                appendTo.AppendChild(appendTo.OwnerDocument.CreateTextNode(appendFrom.FirstChild.Value));
-            }
-
             foreach (XmlNode child in appendFrom.ChildNodes)
             {
-                XmlNode childClone = appendTo.OwnerDocument.CreateNode(child.NodeType, child.Prefix, child.Name, child.NamespaceURI);
-                appendTo.AppendChild(childClone);
-                AppendAttributesAndChildren(childClone, child);
+                if (child.NodeType == XmlNodeType.Text)
+                {
+                    appendTo.AppendChild(appendTo.OwnerDocument.CreateTextNode(child.Value));
+                }
+                else
+                {
+                    XmlNode childClone = appendTo.OwnerDocument.CreateNode(child.NodeType, child.Prefix, child.Name, child.NamespaceURI);
+                    appendTo.AppendChild(childClone);
+                    AppendAttributesAndChildren(childClone, child);
+                }
             }
         }
 
