diff --git a/src/Framework/GeneratedFileUsedEventArgs.cs b/src/Framework/GeneratedFileUsedEventArgs.cs
index 3c7c1f007e4..f64ba9521ff 100644
--- a/src/Framework/GeneratedFileUsedEventArgs.cs
+++ b/src/Framework/GeneratedFileUsedEventArgs.cs
@@ -20,7 +20,9 @@ public GeneratedFileUsedEventArgs()
         /// </summary>
         /// 
         public GeneratedFileUsedEventArgs(string filePath, string content)
-            : base(nameof(GeneratedFileUsedEventArgs) + ": " + filePath, null, null, MessageImportance.Low)
+        // We are not sending the event to binlog (just the file), so we do not want it
+        // to have any stringified representation for other logs either.
+            : base(string.Empty, null, null, MessageImportance.Low)
         {
             FilePath = filePath;
             Content = content;
