diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index 2547600886c..550c4a395fd 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -300,14 +300,11 @@ FileState destinationFileState  // The destination file
                 Log.LogMessage(MessageImportance.Normal, FileComment, sourceFilePath, destinationFilePath);
 
                 File.Copy(sourceFileState.Name, destinationFileState.Name, true);
-                CopiedAtLeastOneFile = true;
-            }
-            else
-            {
-                // Properly linked files will be considered successfully copied files.
-                CopiedAtLeastOneFile = true;
             }
 
+            // Files were successfully copied or linked. Those are equivalent here.
+           CopiedAtLeastOneFile = true;
+
             destinationFileState.Reset();
 
             // If the destinationFile file exists, then make sure it's read-write.
