diff --git a/src/Tasks/GetAssembliesMetadata.cs b/src/Tasks/GetAssembliesMetadata.cs
index 127a47f1605..d5f67c62f36 100644
--- a/src/Tasks/GetAssembliesMetadata.cs
+++ b/src/Tasks/GetAssembliesMetadata.cs
@@ -47,7 +47,8 @@ public override bool Execute()
             var assembliesMetadata = new List<ITaskItem>();
             foreach (string assemblyPath in AssemblyPaths)
             {
-                if (Directory.Exists(assemblyPath))
+                // During DTB the referenced project may not has been built yet, so we need to check if the assembly already exists.
+                if (File.Exists(assemblyPath))
                 {
                     AssemblyInformation assemblyInformation = new(assemblyPath);
                     AssemblyAttributes attributes = assemblyInformation.GetAssemblyMetadata();
diff --git a/src/Tasks/Microsoft.Common.CurrentVersion.targets b/src/Tasks/Microsoft.Common.CurrentVersion.targets
index 115a91b36dd..9aaf2d51bfa 100644
--- a/src/Tasks/Microsoft.Common.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.Common.CurrentVersion.targets
@@ -3049,7 +3049,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     Resolve Assembly attributes for assemblies
     The ResolveComReferences dependency is needed for defining COM-referenced wrappers only,
     but this target handles all the assembly references from ResolveAssemblyReferences
-    The target is needed by the Visual Studio legacy project system
+    This target is needed by the Visual Studio legacy project system during design time build only
 
         [IN]
         @(ReferencePath) - Paths to COM-referenced wrappers and other types of assembly references.
@@ -3061,7 +3061,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
   <Target
       Name="_GetAssembliesMetadata"
-      Condition="'@(ReferencePath)'!=''"
+      Condition="'@(ReferencePath)'!=''and '$(DesignTimeBuild)' == 'true' "
       Returns="@(_AssembliesMetadata)"
       DependsOnTargets="ResolveComReferences"
         >
diff --git a/src/Tasks/Microsoft.Common.tasks b/src/Tasks/Microsoft.Common.tasks
index 68fdc448699..cf67254502d 100644
--- a/src/Tasks/Microsoft.Common.tasks
+++ b/src/Tasks/Microsoft.Common.tasks
@@ -48,7 +48,7 @@
 
   <UsingTask TaskName="Microsoft.Build.Tasks.GenerateTrustInfo"                     AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
   <UsingTask TaskName="Microsoft.Build.Tasks.GetAssemblyIdentity"                   AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
-  <UsingTask TaskName="Microsoft.Build.Tasks.GetAssembliesMetadata"                 AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'"/>
+  <UsingTask TaskName="Microsoft.Build.Tasks.GetAssembliesMetadata"                 AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" Condition="'$(MSBuildRuntimeType)' == 'Full'"/>
   <UsingTask TaskName="Microsoft.Build.Tasks.GetCompatiblePlatform"                 AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
   <UsingTask TaskName="Microsoft.Build.Tasks.GetFileHash"                           AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
   <UsingTask TaskName="Microsoft.Build.Tasks.GetFrameworkPath"                      AssemblyName="Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
