diff --git a/src/StringTools/InternableString.cs b/src/StringTools/InternableString.cs
index ca8fa75ef48..cba8b9088fe 100644
--- a/src/StringTools/InternableString.cs
+++ b/src/StringTools/InternableString.cs
@@ -265,6 +265,14 @@ public unsafe string ExpensiveConvertToString()
                         }
                     }
                 }
+
+                // The invariant that Length is the sum of span lengths is critical in this unsafe method.
+                // Violating it may lead to memory corruption and, since this code tends to run under a lock,
+                // to hangs caused by the lock getting orphaned.
+                if (destPtr != resultPtr + Length)
+                {
+                    throw new InvalidOperationException("Length property does not match sum of span lengths");
+                }
             }
             return result;
         }
