diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index ef032ea7791..962db7514ec 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -1628,12 +1628,16 @@ private void HandleExecuteSubmissionException(GraphBuildSubmission submission, E
                 ? ae.InnerExceptions.First()
                 : ex;
 
-            if (submission.IsStarted)
+            lock (_syncLock)
             {
-                submission.CompleteResults(new GraphBuildResult(submission.SubmissionId, ex));
+                if (submission.IsStarted)
+                {
+                    submission.CompleteResults(new GraphBuildResult(submission.SubmissionId, ex));
+                }
+
+                _overallBuildSuccess = false;
             }
 
-            _overallBuildSuccess = false;
             CheckSubmissionCompletenessAndRemove(submission);
         }
 
@@ -1684,20 +1688,24 @@ private void IssueBuildSubmissionToScheduler(BuildSubmission submission, bool al
                     throw;
                 }
 
-                if (resetMainThreadOnFailure)
+                lock (_syncLock)
                 {
-                    _legacyThreadingData.MainThreadSubmissionId = -1;
-                }
 
-                if (projectException == null)
-                {
-                    BuildEventContext buildEventContext = new BuildEventContext(submission.SubmissionId, 1, BuildEventContext.InvalidProjectInstanceId, BuildEventContext.InvalidProjectContextId, BuildEventContext.InvalidTargetId, BuildEventContext.InvalidTaskId);
-                    ((IBuildComponentHost)this).LoggingService.LogFatalBuildError(buildEventContext, ex, new BuildEventFileInfo(submission.BuildRequestData.ProjectFullPath));
-                }
+                    if (resetMainThreadOnFailure)
+                    {
+                        _legacyThreadingData.MainThreadSubmissionId = -1;
+                    }
 
-                submission.CompleteLogging(true);
-                ReportResultsToSubmission(new BuildResult(submission.BuildRequest, ex));
-                _overallBuildSuccess = false;
+                    if (projectException == null)
+                    {
+                        BuildEventContext buildEventContext = new BuildEventContext(submission.SubmissionId, 1, BuildEventContext.InvalidProjectInstanceId, BuildEventContext.InvalidProjectContextId, BuildEventContext.InvalidTargetId, BuildEventContext.InvalidTaskId);
+                        ((IBuildComponentHost)this).LoggingService.LogFatalBuildError(buildEventContext, ex, new BuildEventFileInfo(submission.BuildRequestData.ProjectFullPath));
+                    }
+
+                    submission.CompleteLogging(true);
+                    ReportResultsToSubmission(new BuildResult(submission.BuildRequest, ex));
+                    _overallBuildSuccess = false;
+                }
             }
         }
 
@@ -1824,7 +1832,10 @@ private void ExecuteGraphBuildScheduler(GraphBuildSubmission submission)
 
                 ReportResultsToSubmission(result);
 
-                _overallBuildSuccess = false;
+                lock (_syncLock)
+                {
+                    _overallBuildSuccess = false;
+                }
             }
         }
 
