diff --git a/.editorconfig b/.editorconfig
index 4a47432ae4f..69a8cdc0183 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -210,7 +210,7 @@ dotnet_analyzer_diagnostic.category-Style.severity = warning
 dotnet_diagnostic.IDE0004.severity = suggestion
 
 # IDE0005: Remove unnecessary usings/imports
-dotnet_diagnostic.IDE0005.severity = warning
+dotnet_diagnostic.IDE0005.severity = none
 
 # Use explicit type instead of 'var'
 dotnet_diagnostic.IDE0008.severity = suggestion
