diff --git a/src/Tasks/BuildUri.cs b/src/Tasks/BuildUri.cs
index ff2dbfac172..7688dce655e 100644
--- a/src/Tasks/BuildUri.cs
+++ b/src/Tasks/BuildUri.cs
@@ -83,7 +83,8 @@ private ITaskItem CreateUriTaskItem(ITaskItem item)
                 // The Scheme property setter throws an ArgumentException for an invalid scheme.
                 builder.Scheme = UriScheme;
                 // If a scheme has been provided and a port has not, use the default port for the scheme.
-                if (UriPort == UseDefaultPortForScheme)
+                // (This is for the case where the UriBuilder was constructed with an ItemSpec. The port will have been set for the scheme used in the ItemSpec.)
+                if (UriPort == UseDefaultPortForScheme && builder.Port != UseDefaultPortForScheme)
                 {
                     builder.Port = UseDefaultPortForScheme;
                 }
