diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 2bf0ac9fdc0..94a9409660d 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2686,8 +2686,13 @@ bool TrueOrFalse()
                 if (bool.TryParse(terminalLoggerArg, out bool result))
                 {
                     useTerminalLogger = result;
-                    // This needs to be called so Ansi Color Codes are enabled for the terminal logger.
-                    (_, _, s_originalConsoleMode) =  NativeMethodsShared.QueryIsScreenAndTryEnableAnsiColorCodes();
+
+                    // Try Enable Ansi Color Codes when terminal logger is enabled/enforced.
+                    if (result)
+                    {
+                        // This needs to be called so Ansi Color Codes are enabled for the terminal logger.
+                        (_, _, s_originalConsoleMode) = NativeMethodsShared.QueryIsScreenAndTryEnableAnsiColorCodes();
+                    }
 
                     return true;
                 }
