diff --git a/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs b/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs
index 48742cc58bd..75ea877da61 100644
--- a/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs
+++ b/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs
@@ -3229,6 +3229,6 @@ public override bool Execute()
                 p => ReferenceTable.ReadMachineTypeFromPEHeader(p));
         }
 
-#endregion
+        #endregion
     }
 }
