diff --git a/src/StringTools/StringTools.cs b/src/StringTools/StringTools.cs
index 675e36ec599..fbe794342b9 100644
--- a/src/StringTools/StringTools.cs
+++ b/src/StringTools/StringTools.cs
@@ -93,10 +93,6 @@ public static string CreateDiagnosticReport()
             return WeakStringCacheInterner.Instance.FormatStatistics();
         }
 
-#if NETSTANDARD2_0
-        public string AsSpan(string s) => s;
-#endif // NETSTANDARD2_0
-
         #endregion
 
         /// <summary>
