diff --git a/src/MSBuild/CommandLineSwitches.cs b/src/MSBuild/CommandLineSwitches.cs
index 70efdfe7b1a..e99faefea4b 100644
--- a/src/MSBuild/CommandLineSwitches.cs
+++ b/src/MSBuild/CommandLineSwitches.cs
@@ -398,7 +398,7 @@ private struct DetectedParameterizedSwitch
         private DetectedParameterizedSwitch[] _parameterizedSwitches;
         // NOTE: the above arrays are instance members because this class is not required to be a singleton
 
-        internal static List<(string path, string contents)> SwitchesFromResponseFiles;
+        internal static List<(string path, string contents)> SwitchesFromResponseFiles = new();
 
         /// <summary>
         /// Default constructor.
@@ -411,8 +411,6 @@ internal CommandLineSwitches()
             Debug.Assert(s_parameterizedSwitchesMap.Length == (int)ParameterizedSwitch.NumberOfParameterizedSwitches,
                 "The map of parameterized switches must have an entry for each switch in the ParameterizedSwitch enumeration.");
 
-            SwitchesFromResponseFiles ??= new();
-
             for (int i = 0; i < s_parameterlessSwitchesMap.Length; i++)
             {
                 Debug.Assert(i == (int)(s_parameterlessSwitchesMap[i].parameterlessSwitch),
