diff --git a/src/Shared/TaskParameter.cs b/src/Shared/TaskParameter.cs
index 8cfceba1254..4c7f2cc7b1c 100644
--- a/src/Shared/TaskParameter.cs
+++ b/src/Shared/TaskParameter.cs
@@ -746,6 +746,25 @@ public void CopyMetadataTo(ITaskItem destinationItem)
                 // between items, and need to know the source item where the metadata came from
                 string originalItemSpec = destinationItem.GetMetadata("OriginalItemSpec");
 
+#if !TASKHOST
+                if (_customEscapedMetadata != null && destinationItem is IMetadataContainer destinationItemAsMetadataContainer)
+                {
+                    // The destination implements IMetadataContainer so we can use the ImportMetadata bulk-set operation.
+                    IEnumerable<KeyValuePair<string, string>> metadataToImport = _customEscapedMetadata
+                        .Where(metadatum => string.IsNullOrEmpty(destinationItem.GetMetadata(metadatum.Key)));
+
+#if FEATURE_APPDOMAIN
+                    if (!AppDomain.CurrentDomain.IsDefaultAppDomain())
+                    {
+                        // Linq is not serializable so materialize the collection before making the call.
+                        metadataToImport = metadataToImport.ToList();
+                    }
+#endif
+
+                    destinationItemAsMetadataContainer.ImportMetadata(metadataToImport);
+                }
+                else
+#endif
                 if (_customEscapedMetadata != null)
                 {
                     foreach (KeyValuePair<string, string> entry in _customEscapedMetadata)
