diff --git a/src/Shared/PlatformNegotiation.cs b/src/Shared/PlatformNegotiation.cs
index 71741186e12..804f648d9a5 100644
--- a/src/Shared/PlatformNegotiation.cs
+++ b/src/Shared/PlatformNegotiation.cs
@@ -21,7 +21,7 @@ internal static string GetNearestPlatform(string referencedProjectPlatform, stri
             if (string.IsNullOrEmpty(projectReferencePlatformsMetadata) && string.IsNullOrEmpty(referencedProjectPlatform))
                 {
                     log?.LogWarningWithCodeFromResources("GetCompatiblePlatform.NoPlatformsListed", projectPath);
-                    return "";
+                    return string.Empty;
                 }
 
                 // Pull platformLookupTable metadata from the referenced project. This allows custom
