diff --git a/src/Shared/UnitTests/TestEnvironment.cs b/src/Shared/UnitTests/TestEnvironment.cs
index 88cd3d9518a..92eff281743 100644
--- a/src/Shared/UnitTests/TestEnvironment.cs
+++ b/src/Shared/UnitTests/TestEnvironment.cs
@@ -448,15 +448,25 @@ public BuildFailureLogInvariant()
 
         private string[] GetMSBuildLogFiles()
         {
-            List<string> files = new();
-            string debugPath = FileUtilities.TempFileDirectory;
-            if (debugPath != null)
+            try
             {
-                files.AddRange(Directory.GetFiles(debugPath, MSBuildLogFiles));
-            }
-            files.AddRange(Directory.GetFiles(Path.GetTempPath(), MSBuildLogFiles));
+                List<string> files = new();
+                string debugPath = FileUtilities.TempFileDirectory;
+                if (debugPath != null)
+                {
+                    files.AddRange(Directory.GetFiles(debugPath, MSBuildLogFiles));
+                }
 
-            return files.ToArray();
+                files.AddRange(Directory.GetFiles(Path.GetTempPath(), MSBuildLogFiles));
+                return files.ToArray();
+            }
+            catch (Exception ex)
+            {
+                throw new AggregateException(
+                    ex,
+                    new InvalidOperationException($"Weird OSX error: debugPath: {FileUtilities.TempFileDirectory}, Path.GetTempPath(): {Path.GetTempPath()}, [TMP]: {Environment.GetEnvironmentVariable("TMP")}, [TMPDIR]: {Environment.GetEnvironmentVariable("TMPDIR")}")
+                );
+            }
         }
 
         public override void AssertInvariant(ITestOutputHelper output)
