diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 15ccab46b32..3e7a6f19c74 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -39,9 +39,6 @@
 using Microsoft.Build.Experimental;
 using Microsoft.Build.Framework.Telemetry;
 using Microsoft.Build.Internal;
-using Microsoft.Build.Logging.LiveLogger;
-using System.Runtime.InteropServices;
-using Microsoft.Win32;
 
 #nullable disable
 
@@ -1275,7 +1272,7 @@ internal static bool BuildProject(
                     }
                     else
                     {
-                        Evaluation.Project project = projectCollection.LoadProject(projectFile, globalProperties, toolsVersion);
+                        Project project = projectCollection.LoadProject(projectFile, globalProperties, toolsVersion);
 
                         project.SaveLogicalProject(preprocessWriter);
 
@@ -1521,7 +1518,7 @@ private static bool PrintTargets(string projectFile, string toolsVersion, Dictio
         {
             try
             {
-                Evaluation.Project project = projectCollection.LoadProject(projectFile, globalProperties, toolsVersion);
+                Project project = projectCollection.LoadProject(projectFile, globalProperties, toolsVersion);
 
                 foreach (string target in project.Targets.Keys)
                 {
