diff --git a/src/Shared/TaskLoggingHelper.cs b/src/Shared/TaskLoggingHelper.cs
index 8b5748932ee..3d12cb49e25 100644
--- a/src/Shared/TaskLoggingHelper.cs
+++ b/src/Shared/TaskLoggingHelper.cs
@@ -246,7 +246,8 @@ public virtual string GetResourceMessage(string resourceName)
         /// <returns>True if messages of the given importance should be logged, false if it's guaranteed that such messages would be ignored.</returns>
         public bool ShouldLogMessage(MessageImportance importance)
         {
-            return importance <= ((IBuildEngine10)BuildEngine).EngineInterface.MinimumRequiredMessageImportance;
+            return BuildEngine is not IBuildEngine10 buildEngine10
+                || importance <= buildEngine10.EngineInterface.MinimumRequiredMessageImportance;
         }
 
         /// <summary>
