diff --git a/src/Build/Instance/ProjectItemInstance.cs b/src/Build/Instance/ProjectItemInstance.cs
index 4c2b4beb6ac..a13583f1040 100644
--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -1568,7 +1568,14 @@ public bool Equals(TaskItem other)
                     }
                 }
 
-                foreach (ProjectMetadataInstance metadatum in other.MetadataCollection)
+                CopyOnWritePropertyDictionary<ProjectMetadataInstance> otherMetadata = other.MetadataCollection;
+
+                if (otherMetadata.Count != thisNames.Count)
+                {
+                    return false;
+                }
+
+                foreach (ProjectMetadataInstance metadatum in otherMetadata)
                 {
                     string name = metadatum.Name;
 
