diff --git a/.editorconfig b/.editorconfig
index d99d59c5c61..76f52d190e5 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -431,4 +431,6 @@ dotnet_diagnostic.IDE0330.severity = suggestion
 dotnet_diagnostic.xUnit2020.severity = none
 dotnet_diagnostic.xUnit1031.severity = none
 dotnet_diagnostic.xUnit1012.severity = none
-dotnet_diagnostic.xUnit2029.severity = none
\ No newline at end of file
+dotnet_diagnostic.xUnit2029.severity = none
+# Do not use equality check to check for collection size.
+dotnet_diagnostic.xUnit2013.severity = none
\ No newline at end of file
diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index 57c789f0371..56bba38f245 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -1102,7 +1102,4 @@ dotnet_diagnostic.SA1643.severity = suggestion
 # File name should match first type name
 dotnet_diagnostic.SA1649.severity = suggestion
 
-dotnet_diagnostic.SA1652.severity = none
-
-# Do not use equality check to check for collection size.
-dotnet_diagnostic.xUnit2013.severity = none
\ No newline at end of file
+dotnet_diagnostic.SA1652.severity = none
\ No newline at end of file
