diff --git a/eng/dependabot/Packages.props b/eng/dependabot/Packages.props
index ea67de5f5d4..e8fcf19704b 100644
--- a/eng/dependabot/Packages.props
+++ b/eng/dependabot/Packages.props
@@ -17,7 +17,7 @@
     <PackageReference Update="Shouldly" Version="3.0.0" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(DotNetBuildFromSource)' != 'true'">
+  <ItemGroup Condition="'$(DotNetBuildFromSource)' != 'true' AND !$([System.String]::Copy($(MSBuildProjectDirectoryNoRoot)).StartsWith('src\Deprecated\')) ">
     <GlobalPackageReference Include="Microsoft.CodeAnalysis.BannedApiAnalyzers" Version="3.3.3" />
 	<GlobalPackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="6.0.0" PrivateAssets="all" />
 	<GlobalPackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.354" PrivateAssets="all"/>
