diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 62790956790..505c4a46899 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1550,7 +1550,8 @@ private static void VerifyThrowSupportedOS()
             {
                 // If we're running on any of the unsupported OS's, fail immediately.  This way,
                 // we don't run into some obscure error down the line, totally confusing the user.
-                InitializationException.VerifyThrow(false, "UnsupportedOS");
+                InitializationException.Throw($"Platform {Environment.OSVersion.Platform}, version {Environment.OSVersion.Version}", "");
+                //InitializationException.VerifyThrow(false, "UnsupportedOS");
             }
 #endif
         }
