diff --git a/src/UnitTests.Shared/TestEnvironment.cs b/src/UnitTests.Shared/TestEnvironment.cs
index 79fbe9a02b6..6a8f72783bd 100644
--- a/src/UnitTests.Shared/TestEnvironment.cs
+++ b/src/UnitTests.Shared/TestEnvironment.cs
@@ -87,6 +87,8 @@ private void Cleanup()
             {
                 _disposed = true;
 
+                // It should be in reverse order revert to get back to original state.
+                _variants.Reverse();
                 // Reset test variants
                 foreach (var variant in _variants)
                 {
