diff --git a/src/Build/Evaluation/LazyItemEvaluator.cs b/src/Build/Evaluation/LazyItemEvaluator.cs
index 1c91e6c1eca..7a3674ce605 100644
--- a/src/Build/Evaluation/LazyItemEvaluator.cs
+++ b/src/Build/Evaluation/LazyItemEvaluator.cs
@@ -328,7 +328,8 @@ internal Builder(int version, ImmutableList<ItemData>.Builder listBuilder, Immut
 
                 #region IEnumerable implementation
 
-                public IEnumerator<ItemData> GetEnumerator() => _listBuilder.GetEnumerator();
+                ImmutableList<ItemData>.Enumerator GetEnumerator() => _listBuilder.GetEnumerator();
+                IEnumerator<ItemData> IEnumerable<T>.GetEnumerator() => _listBuilder.GetEnumerator();
 
                 System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => _listBuilder.GetEnumerator();
 
