diff --git a/src/Build/ElementLocation/ElementLocation.cs b/src/Build/ElementLocation/ElementLocation.cs
index 7dbf228bbd9..fca729ed64a 100644
--- a/src/Build/ElementLocation/ElementLocation.cs
+++ b/src/Build/ElementLocation/ElementLocation.cs
@@ -26,7 +26,7 @@ public abstract class ElementLocation : IElementLocation, ITranslatable, IImmuta
         private static readonly ElementLocation s_emptyElementLocation = new SmallElementLocation("", 0, 0);
 
         /// <summary>
-        /// The file from which this particular element originated.  It may
+        /// Gets the file from which this particular element originated.  It may
         /// differ from the ProjectFile if, for instance, it was part of
         /// an import or originated in a targets file.
         /// If not known, returns empty string.
@@ -38,7 +38,7 @@ public abstract string File
         }
 
         /// <summary>
-        /// The line number where this element exists in its file.
+        /// Gets the line number where this element exists in its file.
         /// The first line is numbered 1.
         /// Zero indicates "unknown location".
         /// </summary>
@@ -49,7 +49,7 @@ public abstract int Line
         }
 
         /// <summary>
-        /// The column number where this element exists in its file.
+        /// Gets the column number where this element exists in its file.
         /// The first column is numbered 1.
         /// Zero indicates "unknown location".
         /// </summary>
@@ -60,7 +60,7 @@ public abstract int Column
         }
 
         /// <summary>
-        /// The location in a form suitable for replacement
+        /// Gets the location in a form suitable for replacement
         /// into a message.
         /// Example: "c:\foo\bar.csproj (12,34)"
         /// Calling this creates and formats a new string.
