diff --git a/src/Build/BuildCop/Infrastructure/BuildCopManagerProvider.cs b/src/Build/BuildCop/Infrastructure/BuildCopManagerProvider.cs
index c826f3854bc..ada79d28df3 100644
--- a/src/Build/BuildCop/Infrastructure/BuildCopManagerProvider.cs
+++ b/src/Build/BuildCop/Infrastructure/BuildCopManagerProvider.cs
@@ -47,7 +47,9 @@ public void InitializeComponent(IBuildComponentHost host)
 
         if (s_isInitialized)
         {
-            throw new InvalidOperationException("BuildCopManagerProvider is already initialized");
+            // TODO: change to interlocked
+            return;
+            // throw new InvalidOperationException("BuildCopManagerProvider is already initialized");
         }
         s_isInitialized = true;
 
