diff --git a/src/BuildCheck.UnitTests/EndToEndTests.cs b/src/BuildCheck.UnitTests/EndToEndTests.cs
index 0a2ba6d8364..be460ba4062 100644
--- a/src/BuildCheck.UnitTests/EndToEndTests.cs
+++ b/src/BuildCheck.UnitTests/EndToEndTests.cs
@@ -37,10 +37,10 @@ public EndToEndTests(ITestOutputHelper output)
 
     public void Dispose() => _env.Dispose();
 
-    [WindowsFullFrameworkOnlyTheory]
+    [Theory]
     [InlineData(true)]
     [InlineData(false)]
-    public void PropertiesUsageAnalyzerTest_FullFramework(bool buildInOutOfProcessNode)
+    public void PropertiesUsageAnalyzerTest(bool buildInOutOfProcessNode)
     {
         PrepareSampleProjectsAndConfig(
             buildInOutOfProcessNode,
@@ -62,30 +62,6 @@ public void PropertiesUsageAnalyzerTest_FullFramework(bool buildInOutOfProcessNo
         Regex.Matches(output, "BC0203 .* Property").Count.ShouldBe(2);
     }
 
-    [DotNetOnlyTheory]
-    [InlineData(true)]
-    [InlineData(false)]
-    public void PropertiesUsageAnalyzerTest_NetCore(bool buildInOutOfProcessNode)
-    {
-        PrepareSampleProjectsAndConfig(
-            buildInOutOfProcessNode,
-            out TransientTestFile projectFile,
-            "PropsCheckTest.csproj");
-
-        string output = RunnerUtilities.ExecBootstrapedMSBuild($"{projectFile.Path} -check", out bool success);
-        _env.Output.WriteLine(output);
-        _env.Output.WriteLine("=========================");
-        success.ShouldBeTrue(output);
-
-        output.ShouldMatch(@"BC0201: .* Property: \[MyProp11\]");
-        output.ShouldMatch(@"BC0202: .* Property: \[MyPropT2\]");
-        output.ShouldMatch(@"BC0203: .* Property: \[MyProp13\]");
-
-        Regex.Matches(output, "BC0201: .* Property").Count.ShouldBe(1);
-        Regex.Matches(output, "BC0202: .* Property").Count.ShouldBe(1);
-        Regex.Matches(output, "BC0203 .* Property").Count.ShouldBe(1);
-    }
-
     [Theory]
     [InlineData(true, true)]
     [InlineData(false, true)]
diff --git a/src/Shared/UnitTests/FileMatcher_Tests.cs b/src/Shared/UnitTests/FileMatcher_Tests.cs
index 9576c8b5c78..abf4e350cc4 100644
--- a/src/Shared/UnitTests/FileMatcher_Tests.cs
+++ b/src/Shared/UnitTests/FileMatcher_Tests.cs
@@ -1098,7 +1098,7 @@ public void ParentWithoutSlash()
                 true);
         }
 
-        [Fact]
+        [WindowsOnlyFact]
         public void Unc()
         {
             // Check UNC functionality
