diff --git a/src/MSBuild/TerminalLogger/TerminalLogger.cs b/src/MSBuild/TerminalLogger/TerminalLogger.cs
index 90e13354e94..c9fe0c5a24e 100644
--- a/src/MSBuild/TerminalLogger/TerminalLogger.cs
+++ b/src/MSBuild/TerminalLogger/TerminalLogger.cs
@@ -474,11 +474,11 @@ private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
                             else
                             {
                                 Terminal.Write(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("TestProjectFinished_WithTF",
-                                    Indentation, Indentation,
-                                    projectFile, projectFile,
+                                    Indentation,
+                                    projectFile,
                                     AnsiCodes.Colorize(project.TargetFramework, TargetFrameworkColor), AnsiCodes.Colorize(project.TargetFramework, TargetFrameworkColor),
-                                    buildResult, buildResult,
-                                    duration)); duration));
+                                    buildResult,
+                                    duration));
                             }
                         }
                     }
