diff --git a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
index 3b25bfdb8dd..ef21df23454 100644
--- a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
@@ -30,12 +30,6 @@
 using Microsoft.Build.Framework;
 using Microsoft.Build.BackEnd.Logging;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.BackEnd
 {
     /// <summary>
@@ -259,8 +253,8 @@ protected NodeContext GetNode(string msbuildLocation, string commandLineArgs, in
                 // Failure to detect this has been known to stall builds when Windows pops up a related dialog.
                 // It's also a waste of time when we attempt several times to launch multiple MSBuildTaskHost.exe (CLR2 TaskHost)
                 // nodes because we should never be able to connect in this case.
-                var taskHostNameForClr2TaskHost = Path.GetFileNameWithoutExtension(NodeProviderOutOfProcTaskHost.TaskHostNameForClr2TaskHost.AsSpan());
-                if (MemoryExtensions.Equals(Path.GetFileNameWithoutExtension(msbuildLocation.AsSpan()), taskHostNameForClr2TaskHost, StringComparison.OrdinalIgnoreCase))
+                string taskHostNameForClr2TaskHost = Path.GetFileNameWithoutExtension(NodeProviderOutOfProcTaskHost.TaskHostNameForClr2TaskHost);
+                if (Path.GetFileNameWithoutExtension(msbuildLocation).Equals(taskHostNameForClr2TaskHost, StringComparison.OrdinalIgnoreCase))
                 {
                     if (FrameworkLocationHelper.GetPathToDotNetFrameworkV35(DotNetFrameworkArchitecture.Current) == null)
                     {
diff --git a/src/Build/BackEnd/Shared/BuildRequestConfiguration.cs b/src/Build/BackEnd/Shared/BuildRequestConfiguration.cs
index 6cf192117b6..2cc6ab1f7b4 100644
--- a/src/Build/BackEnd/Shared/BuildRequestConfiguration.cs
+++ b/src/Build/BackEnd/Shared/BuildRequestConfiguration.cs
@@ -16,12 +16,6 @@
 using Microsoft.Build.Globbing;
 using Microsoft.Build.Shared.FileSystem;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.BackEnd
 {
     /// <summary>
@@ -296,7 +290,7 @@ public bool IsTraversal
             {
                 if (!_isTraversalProject.HasValue)
                 {
-                    if (MemoryExtensions.Equals(Path.GetFileName(ProjectFullPath.AsSpan()), "dirs.proj".AsSpan(), StringComparison.OrdinalIgnoreCase))
+                    if (String.Equals(Path.GetFileName(ProjectFullPath), "dirs.proj", StringComparison.OrdinalIgnoreCase))
                     {
                         // dirs.proj are assumed to be traversals
                         _isTraversalProject = true;
diff --git a/src/Build/ElementLocation/XmlDocumentWithLocation.cs b/src/Build/ElementLocation/XmlDocumentWithLocation.cs
index 35fff7227c3..15c91a6f68e 100644
--- a/src/Build/ElementLocation/XmlDocumentWithLocation.cs
+++ b/src/Build/ElementLocation/XmlDocumentWithLocation.cs
@@ -7,12 +7,6 @@
 using Microsoft.Build.Internal;
 using Microsoft.Build.Shared;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.Construction
 {
     /// <summary>
@@ -356,23 +350,23 @@ private void DetermineWhetherToLoadReadOnly(string fullPath)
                 else if (s_readOnlyFlags == ReadOnlyLoadFlags.LoadReadOnlyIfAppropriate && fullPath is object)
                 {
                     // Only files from Microsoft
-                    if (Path.GetFileName(fullPath.AsSpan()).StartsWith("Microsoft.".AsSpan(), StringComparison.OrdinalIgnoreCase))
+                    if (Path.GetFileName(fullPath).StartsWith("Microsoft.", StringComparison.OrdinalIgnoreCase))
                     {
                         // If we are loading devdiv targets, we're in razzle
-                        if (Path.GetFileName(fullPath.AsSpan()).StartsWith("Microsoft.DevDiv".AsSpan(), StringComparison.OrdinalIgnoreCase))
+                        if (Path.GetFileName(fullPath).StartsWith("Microsoft.DevDiv", StringComparison.OrdinalIgnoreCase))
                         {
                             _loadAsReadOnly = true;
                         }
                         else // Else, only load if they're in program files or windows directories
                         {
                             ErrorUtilities.VerifyThrow(Path.IsPathRooted(fullPath), "should be full path");
-                            var directory = Path.GetDirectoryName(fullPath.AsSpan());
+                            string directory = Path.GetDirectoryName(fullPath);
 
-                            var windowsFolder = Environment.GetFolderPath(Environment.SpecialFolder.Windows).AsSpan();
+                            string windowsFolder = Environment.GetFolderPath(Environment.SpecialFolder.Windows);
 
-                            if ((windowsFolder.Length > 0 && directory.StartsWith(windowsFolder, StringComparison.OrdinalIgnoreCase)) ||
-                                (!string.IsNullOrEmpty(FrameworkLocationHelper.programFiles32) && directory.StartsWith(FrameworkLocationHelper.programFiles32.AsSpan(), StringComparison.OrdinalIgnoreCase)) ||
-                                (!string.IsNullOrEmpty(FrameworkLocationHelper.programFiles64) && directory.StartsWith(FrameworkLocationHelper.programFiles64.AsSpan(), StringComparison.OrdinalIgnoreCase)))
+                            if ((!String.IsNullOrEmpty(windowsFolder) && directory.StartsWith(windowsFolder, StringComparison.OrdinalIgnoreCase)) ||
+                                (!String.IsNullOrEmpty(FrameworkLocationHelper.programFiles32) && directory.StartsWith(FrameworkLocationHelper.programFiles32, StringComparison.OrdinalIgnoreCase)) ||
+                                (!String.IsNullOrEmpty(FrameworkLocationHelper.programFiles64) && directory.StartsWith(FrameworkLocationHelper.programFiles64, StringComparison.OrdinalIgnoreCase)))
                             {
                                 _loadAsReadOnly = true;
                             }
diff --git a/src/Build/Graph/ProjectGraph.cs b/src/Build/Graph/ProjectGraph.cs
index 340c2dcf3f8..40cf6aee0ed 100644
--- a/src/Build/Graph/ProjectGraph.cs
+++ b/src/Build/Graph/ProjectGraph.cs
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.IO;
 using System.Linq;
 using System.Text;
 using System.Threading;
@@ -17,12 +18,6 @@
 using Microsoft.Build.Shared.Debugging;
 using Microsoft.Build.Utilities;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.Graph
 {
     /// <summary>
@@ -508,11 +503,8 @@ internal string ToDot(
             {
                 var nodeId = GetNodeId(node);
 
-#if FEATURE_MSIOREDIST
                 var nodeName = Path.GetFileNameWithoutExtension(node.ProjectInstance.FullPath);
-#else
-                var nodeName = Path.GetFileNameWithoutExtension(node.ProjectInstance.FullPath.AsSpan());
-#endif
+
                 var globalPropertiesString = string.Join(
                     "<br/>",
                     node.ProjectInstance.GlobalProperties.OrderBy(kvp => kvp.Key)
@@ -520,9 +512,7 @@ internal string ToDot(
 
                 var targetListString = GetTargetListString(node);
 
-                sb.Append($"\t{nodeId} [label=<");
-                sb.Append(nodeName);
-                sb.AppendLine($"<br/>({targetListString})<br/>{globalPropertiesString}>]");
+                sb.AppendLine($"\t{nodeId} [label=<{nodeName}<br/>({targetListString})<br/>{globalPropertiesString}>]");
 
                 foreach (var reference in node.ProjectReferences)
                 {
diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 918660bd0ed..c4408477cfa 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -40,12 +40,6 @@
 using ForwardingLoggerRecord = Microsoft.Build.Logging.ForwardingLoggerRecord;
 using BinaryLogger = Microsoft.Build.Logging.BinaryLogger;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.CommandLine
 {
     /// <summary>
@@ -2840,12 +2834,10 @@ DirectoryGetFiles getFiles
                 if (actualProjectFiles.Count == 1 && actualSolutionFiles.Count == 1)
                 {
                     // Grab the name of both project and solution without extensions
-                    var solutionName = Path.GetFileNameWithoutExtension(actualSolutionFiles[0].AsSpan());
-                    var projectName = Path.GetFileNameWithoutExtension(actualProjectFiles[0].AsSpan());
+                    string solutionName = Path.GetFileNameWithoutExtension(actualSolutionFiles[0]);
+                    string projectName = Path.GetFileNameWithoutExtension(actualProjectFiles[0]);
                     // Compare the names and error if they are not identical
-                    InitializationException.VerifyThrow(
-                        MemoryExtensions.Equals(solutionName, projectName, StringComparison.OrdinalIgnoreCase),
-                        projectDirectory == null ? "AmbiguousProjectError" : "AmbiguousProjectDirectoryError", null, projectDirectory);
+                    InitializationException.VerifyThrow(string.Equals(solutionName, projectName, StringComparison.OrdinalIgnoreCase), projectDirectory == null ? "AmbiguousProjectError" : "AmbiguousProjectDirectoryError", null, projectDirectory);
                     projectFile = actualSolutionFiles[0];
                 }
                 // If there is more than one solution file in the current directory we have no idea which one to use
diff --git a/src/Shared/FileMatcher.cs b/src/Shared/FileMatcher.cs
index 4da5a82e462..22a9572fc82 100644
--- a/src/Shared/FileMatcher.cs
+++ b/src/Shared/FileMatcher.cs
@@ -13,12 +13,6 @@
 using Microsoft.Build.Shared.FileSystem;
 using Microsoft.Build.Utilities;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.Shared
 {
     /// <summary>
@@ -1678,7 +1672,9 @@ internal Result()
         internal static bool IsFileNameMatch(string path, string pattern)
         {
             // Use a span-based Path.GetFileName if it is available.
-#if NETSTANDARD2_0
+#if FEATURE_MSIOREDIST
+            return IsMatch(Microsoft.IO.Path.GetFileName(path.AsSpan()), pattern);
+#elif NETSTANDARD2_0
             return IsMatch(Path.GetFileName(path), pattern);
 #else
             return IsMatch(Path.GetFileName(path.AsSpan()), pattern);
diff --git a/src/Shared/FrameworkLocationHelper.cs b/src/Shared/FrameworkLocationHelper.cs
index df9591df8f1..afdbc7fefec 100644
--- a/src/Shared/FrameworkLocationHelper.cs
+++ b/src/Shared/FrameworkLocationHelper.cs
@@ -11,12 +11,6 @@
 
 using Microsoft.Build.Shared.FileSystem;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.Shared
 {
     /// <summary>
@@ -787,13 +781,8 @@ DotNetFrameworkArchitecture architecture
 
             // If the current runtime starts with correct prefix, then this is the runtime we want to use.
             // However, only if we're requesting current architecture -- otherwise, the base path may be different, so we'll need to look it up.
-#if NETSTANDARD2_0
-            var leaf = Path.GetFileName(currentRuntimePath);
+            string leaf = Path.GetFileName(currentRuntimePath);
             if (leaf.StartsWith(prefix, StringComparison.Ordinal) && architecture == DotNetFrameworkArchitecture.Current)
-#else
-            var leaf = Path.GetFileName(currentRuntimePath.AsSpan());
-            if (leaf.StartsWith(prefix.AsSpan(), StringComparison.Ordinal) && architecture == DotNetFrameworkArchitecture.Current)
-#endif
             {
                 return currentRuntimePath;
             }
diff --git a/src/Utilities/ToolTask.cs b/src/Utilities/ToolTask.cs
index 61ba8c8e8d5..e7766038d11 100644
--- a/src/Utilities/ToolTask.cs
+++ b/src/Utilities/ToolTask.cs
@@ -16,12 +16,6 @@
 using Microsoft.Build.Shared;
 using Microsoft.Build.Shared.FileSystem;
 
-#if FEATURE_MSIOREDIST
-using Path = Microsoft.IO.Path;
-#else
-using Path = System.IO.Path;
-#endif
-
 namespace Microsoft.Build.Utilities
 {
     /// <summary>
@@ -488,11 +482,7 @@ private string ComputePathToTool()
             // look for it in the path
             if (pathToTool != null)
             {
-#if NETSTANDARD2_0
                 bool isOnlyFileName = Path.GetFileName(pathToTool).Length == pathToTool.Length;
-#else
-                bool isOnlyFileName = Path.GetFileName(pathToTool.AsSpan()).Length == pathToTool.AsSpan().Length;
-#endif
                 if (!isOnlyFileName)
                 {
                     bool isExistingFile = FileSystems.Default.FileExists(pathToTool);
