diff --git a/src/Tasks/Delete.cs b/src/Tasks/Delete.cs
index 8cf304d4729..bd782dd8bae 100644
--- a/src/Tasks/Delete.cs
+++ b/src/Tasks/Delete.cs
@@ -17,7 +17,7 @@ namespace Microsoft.Build.Tasks
     /// <summary>
     /// Delete files from disk.
     /// </summary>
-    public class Delete : TaskExtension, ICancelableTask, IIncrementalTask
+    public class Delete : TaskExtension, ICancelableTask
     {
         #region Properties
 
@@ -57,12 +57,6 @@ public ITaskItem[] Files
 
         #endregion
 
-        /// <summary>
-        /// Set question parameter to verify if this is incremental.
-        /// </summary>
-        /// <remarks></remarks>
-        public bool FailIfNotIncremental { get; set; }
-
         /// <summary>
         /// Verify that the inputs are correct.
         /// </summary>
@@ -121,11 +115,6 @@ public override bool Execute()
                     {
                         if (FileSystems.Default.FileExists(file.ItemSpec))
                         {
-                            if (FailIfNotIncremental)
-                            {
-                                Log.LogWarningFromResources("Delete.DeletingFile", file.ItemSpec);
-                            }
-
                             // Do not log a fake command line as well, as it's superfluous, and also potentially expensive
                             Log.LogMessageFromResources(MessageImportance.Normal, "Delete.DeletingFile", file.ItemSpec);
 
diff --git a/src/Tasks/Touch.cs b/src/Tasks/Touch.cs
index 8f887fe5000..a73bda8470b 100644
--- a/src/Tasks/Touch.cs
+++ b/src/Tasks/Touch.cs
@@ -225,8 +225,10 @@ private bool TouchFile(
             {
                 Log.LogWarningFromResources("Touch.Touching", file);
             }
-
-            Log.LogMessageFromResources(messageImportance, "Touch.Touching", file);
+            else
+            {
+                Log.LogMessageFromResources(messageImportance, "Touch.Touching", file);
+            }
 
             // If the file is read only then we must either issue an error, or, if the user so 
             // specified, make the file temporarily not read only.
