diff --git a/src/Shared/CommunicationsUtilities.cs b/src/Shared/CommunicationsUtilities.cs
index 3f30add4415..12061206c4d 100644
--- a/src/Shared/CommunicationsUtilities.cs
+++ b/src/Shared/CommunicationsUtilities.cs
@@ -541,20 +541,21 @@ internal static int ReadIntForHandshake(this PipeStream stream, byte? byteToAcce
 #endif
             {
                 int bytesRead = stream.Read(bytes, 0, bytes.Length);
-                if (bytesRead != bytes.Length)
-                {
-                    // We've unexpectly reached end of stream.
-                    // We are now in a bad state, disconnect on our end
-                    throw new IOException(String.Format(CultureInfo.InvariantCulture, "Unexpected end of stream while reading for handshake"));
-                }
 
-                // Legacy approach with an early-abort for connection attempts from ancient MSBuild.exes
-                if (byteToAccept != null && byteToAccept != bytes[0])
+                // Abort for connection attempts from ancient MSBuild.exes
+                if (byteToAccept != null && bytesRead > 0 && byteToAccept != bytes[0])
                 {
                     stream.WriteIntForHandshake(0x0F0F0F0F);
                     stream.WriteIntForHandshake(0x0F0F0F0F);
                     throw new InvalidOperationException(String.Format(CultureInfo.InvariantCulture, "Client: rejected old host. Received byte {0} instead of {1}.", bytes[0], byteToAccept));
                 }
+
+                if (bytesRead != bytes.Length)
+                {
+                    // We've unexpectly reached end of stream.
+                    // We are now in a bad state, disconnect on our end
+                    throw new IOException(String.Format(CultureInfo.InvariantCulture, "Unexpected end of stream while reading for handshake"));
+                }
             }
 
             int result;
