diff --git a/src/Build.UnitTests/Evaluation/Expander_Tests.cs b/src/Build.UnitTests/Evaluation/Expander_Tests.cs
index 7386075a724..d5fefad3e1e 100644
--- a/src/Build.UnitTests/Evaluation/Expander_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Expander_Tests.cs
@@ -4336,8 +4336,6 @@ public void PropertyFunctionGuidNewGuid()
         // TODO: update features list
         [Theory]
         [InlineData("NonExistingFeature", "NotFound")]
-        [InlineData("Feature1", "Available")]
-        [InlineData("Feature2", "NotAvailable")]
         public void PropertyFunctionCheckFeatureAvailability(string featureName, string availability)
         {
             var expander = new Expander<ProjectPropertyInstance, ProjectItemInstance>(new PropertyDictionary<ProjectPropertyInstance>(), FileSystems.Default);
