diff --git a/src/MSBuild.UnitTests/XMake_Tests.cs b/src/MSBuild.UnitTests/XMake_Tests.cs
index 56b3811f249..6d2eddab546 100644
--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -1819,7 +1819,7 @@ public void TestProcessProjectSwitchSlnProjDifferentNames()
         [Fact]
         public void TestProcessProjectSwitchSlnxProjDifferentNames()
         {
-            string[] projects = { "test.proj", "Different.slnx" };
+            string[] projects = ["test.proj", "Different.slnx"];
             string[] extensionsToIgnore = null;
 
             Should.Throw<InitializationException>(() =>
