diff --git a/src/Utilities/ToolLocationHelper.cs b/src/Utilities/ToolLocationHelper.cs
index 8920c8868a0..13a9054c93b 100644
--- a/src/Utilities/ToolLocationHelper.cs
+++ b/src/Utilities/ToolLocationHelper.cs
@@ -1711,6 +1711,7 @@ public static string GetPathToDotNetFramework(TargetDotNetFrameworkVersion versi
         /// that specifies a VisualStudioVersion.
         /// </summary>
         /// <param name="version">Version of the targeted .NET Framework</param>
+        /// <param name="output">asdf</param>
         /// <returns>Path string.</returns>
         public static string GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion version, Action<string> output = null) => GetPathToDotNetFrameworkSdk(version, VisualStudioVersion.VersionLatest, output);
 
@@ -1719,6 +1720,7 @@ public static string GetPathToDotNetFramework(TargetDotNetFrameworkVersion versi
         /// </summary>
         /// <param name="version">The <see cref="TargetDotNetFrameworkVersion"/> of the .NET Framework.</param>
         /// <param name="visualStudioVersion">The <see cref="VisualStudioVersion"/> of Visual Studio.</param>
+        /// <param name="output">asdf</param>
         /// <returns></returns>
         public static string GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion version, VisualStudioVersion visualStudioVersion, Action<string> output = null)
         {
