diff --git a/.editorconfig b/.editorconfig
index 2927cad5369..c61593e8297 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -428,6 +428,9 @@ dotnet_diagnostic.IDE0073.severity = error
 # Use 'System.Threading.Lock'
 dotnet_diagnostic.IDE0330.severity = suggestion
 
+# Use implicitly typed lambda
+dotnet_diagnostic.IDE0350.severity = none
+
 # Value types are incompatible with null values. https://xunit.net/xunit.analyzers/rules/xUnit1012
 dotnet_diagnostic.xUnit1012.severity = warning
 
