diff --git a/src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs b/src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs
index 40d6fcb1a66..8ef00a712e8 100644
--- a/src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs
+++ b/src/Build/BuildCheck/Infrastructure/BuildCheckConnectorLogger.cs
@@ -107,7 +107,7 @@ private void LogAnalyzerStats(LoggingContext loggingContext)
         {
             if (stat.Key.StartsWith(BuildCheckConstants.infraStatPrefix))
             {
-                string newKey = stat.Key.Replace(BuildCheckConstants.infraStatPrefix, string.Empty);
+                string newKey = stat.Key.Substring(BuildCheckConstants.infraStatPrefix.Length);
                 infraStats[newKey] = stat.Value;
             }
             else
