diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 5444c6e40a9..202017f2c76 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2586,6 +2586,14 @@ static bool DoesEnvironmentSupportTerminalLogger()
                         new BuildManager.DeferredBuildMessage("TerminalLogger was not used because the output is not supported.", MessageImportance.Low));
                     return false;
                 }
+
+                if (Traits.Instance.EscapeHatches.EnsureStdOutForChildNodesIsPrimaryStdout)
+                {
+                    s_globalMessagesToLogInBuildLoggers.Add(
+                        new BuildManager.DeferredBuildMessage("TerminalLogger was not used because build is run in context of process (e.g. 'dotnet test') that requests direct access to stdout stream.", MessageImportance.Low));
+                    return false;
+                }
+
                 return true;
             }
         }
