diff --git a/src/Shared/Debugging/DebugUtils.cs b/src/Shared/Debugging/DebugUtils.cs
index 3ae6cf16891..1aa30656305 100644
--- a/src/Shared/Debugging/DebugUtils.cs
+++ b/src/Shared/Debugging/DebugUtils.cs
@@ -28,6 +28,16 @@ static DebugUtils()
 
             if (Traits.Instance.DebugEngine)
             {
+                string debugEngineValue = FileUtilities.TrimAndStripAnyQuotes(Environment.GetEnvironmentVariable("MSBuildDebugEngine"));
+
+                if (string.IsNullOrWhiteSpace(debugDirectory)
+                    && !string.IsNullOrWhiteSpace(debugEngineValue)
+                    && Path.IsPathRooted(debugEngineValue))
+                {
+                    // DebugEngine value is a path, so use it as the directory, unless DEBUGPATH was set.
+                    debugDirectory = debugEngineValue;
+                }
+
                 if (!string.IsNullOrWhiteSpace(debugDirectory) && FileUtilities.CanWriteToDirectory(debugDirectory))
                 {
                     // Debug directory is writable; no need for fallbacks
