diff --git a/src/Tasks.UnitTests/CodeTaskFactoryTests.cs b/src/Tasks.UnitTests/CodeTaskFactoryTests.cs
index 4bf1728b5a1..b0e7a030f9d 100644
--- a/src/Tasks.UnitTests/CodeTaskFactoryTests.cs
+++ b/src/Tasks.UnitTests/CodeTaskFactoryTests.cs
@@ -13,7 +13,7 @@
 
 namespace Microsoft.Build.UnitTests
 {
-
+#if FEATURE_CODETASKFACTORY
 
     using System.CodeDom.Compiler;
     using System.IO.Compression;
@@ -1249,4 +1249,34 @@ public void RedundantMSBuildReferences()
             mockLogger.AssertLogContains("Hello, World!");
         }
     }
+#else
+    public sealed class CodeTaskFactoryTests
+    {
+        [Fact]
+        public void CodeTaskFactoryNotSupported()
+        {
+            string projectFileContents = @"
+                    <Project ToolsVersion='msbuilddefaulttoolsversion'>
+                        <UsingTask TaskName=`CustomTaskFromCodeFactory_BuildTaskSimpleCodeFactory` TaskFactory=`CodeTaskFactory` AssemblyFile=`$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll` >
+                         <ParameterGroup>
+                             <Text/>
+                          </ParameterGroup>
+                            <Task>
+                                <Code>
+                                     Log.LogMessage(MessageImportance.High, Text);
+                                </Code>
+                            </Task>
+                        </UsingTask>
+                        <Target Name=`Build`>
+                            <CustomTaskFromCodeFactory_BuildTaskSimpleCodeFactory Text=`Hello, World!` />
+                        </Target>
+                    </Project>";
+            MockLogger mockLogger = Helpers.BuildProjectWithNewOMExpectFailure(projectFileContents, allowTaskCrash: false);
+            BuildErrorEventArgs error = mockLogger.Errors.FirstOrDefault();
+            Assert.NotNull(error);
+            Assert.Equal("MSB4801", error.Code);
+            Assert.Contains("CodeTaskFactory", error.Message);
+        }
+    }
+#endif
 }
