diff --git a/src/Framework/ChangeWaves.cs b/src/Framework/ChangeWaves.cs
index 3ebe0b75ef8..9b5a15e2aaa 100644
--- a/src/Framework/ChangeWaves.cs
+++ b/src/Framework/ChangeWaves.cs
@@ -26,8 +26,8 @@ internal class ChangeWaves
     {
         internal static readonly Version Wave16_10 = new Version(16, 10);
         internal static readonly Version Wave17_0 = new Version(17, 0);
-        internal static readonly Version[] AllWaves = { Wave16_10, Wave17_0 };
-
+        internal static readonly Version Wave17_2 = new Version(17, 2);
+        internal static readonly Version[] AllWaves = { Wave16_10, Wave17_0, Wave17_2 };
         /// <summary>
         /// Special value indicating that all features behind all Change Waves should be enabled.
         /// </summary>
