diff --git a/src/Build/Graph/GraphBuilder.cs b/src/Build/Graph/GraphBuilder.cs
index 97e6aef1e10..7354a622c9a 100644
--- a/src/Build/Graph/GraphBuilder.cs
+++ b/src/Build/Graph/GraphBuilder.cs
@@ -101,6 +101,7 @@ private static List<ConfigurationMetadata> AddGraphBuildPropertyToEntryPoints(IE
                     var configurationMetadata = new ConfigurationMetadata(FileUtilities.NormalizePath(entryPoint.ProjectFile), globalPropertyDictionary);
                     entryPointConfigurationMetadata.Add(configurationMetadata);
                 }
+
                 return entryPointConfigurationMetadata;
             }
 
