diff --git a/src/Tasks/AssemblyDependency/ReferenceTable.cs b/src/Tasks/AssemblyDependency/ReferenceTable.cs
index dab85a3173a..8a48820d777 100644
--- a/src/Tasks/AssemblyDependency/ReferenceTable.cs
+++ b/src/Tasks/AssemblyDependency/ReferenceTable.cs
@@ -1282,8 +1282,13 @@ private void ResolveReference(
             var parentReferenceDirectoriesMap = new Dictionary<string, List<string>>();
             foreach (Reference parentReference in reference.GetDependees())
             {
-                parentReferenceDirectoriesMap[parentReference.FullPath] = new List<string>();
-                CalculateParentAssemblyDirectories(parentReferenceDirectoriesMap[parentReference.FullPath], parentReference);
+                if (!parentReferenceDirectoriesMap.TryGetValue(parentReference.FullPath, out List<string> value))
+                {
+                    value = new List<string>();
+                    parentReferenceDirectoriesMap[parentReference.FullPath] = value;
+                }
+
+                CalculateParentAssemblyDirectories(value, parentReference);
             }
 
             // Build the set of resolvers.
