diff --git a/src/Build/BuildCheck/Infrastructure/ConfigurationProvider.cs b/src/Build/BuildCheck/Infrastructure/ConfigurationProvider.cs
index 3ed01264544..b71b9f73d6a 100644
--- a/src/Build/BuildCheck/Infrastructure/ConfigurationProvider.cs
+++ b/src/Build/BuildCheck/Infrastructure/ConfigurationProvider.cs
@@ -59,7 +59,7 @@ public CustomConfigurationData GetCustomConfiguration(string projectFullPath, st
     {
         var configuration = GetConfiguration(projectFullPath, ruleId);
 
-        if (configuration is null || !configuration.Any())
+        if (configuration is null)
         {
             return CustomConfigurationData.Null;
         }
@@ -70,6 +70,11 @@ public CustomConfigurationData GetCustomConfiguration(string projectFullPath, st
             configuration.Remove(infraConfigurationKey);
         }
 
+        if (!configuration.Any())
+        {
+            return CustomConfigurationData.Null;
+        }
+
         var data = new CustomConfigurationData(ruleId, configuration);
 
         if (!_customConfigurationData.ContainsKey(ruleId))
@@ -210,7 +215,7 @@ private Dictionary<string, string> FetchEditorConfigRules(string projectFullPath
 
         // clear the dictionary from the key-value pairs not BuildCheck related and
         // store the data so there is no need to parse the .editorconfigs all over again
-        _editorConfigData[projectFullPath] = FilterDictionaryByKeys($"{BuildCheck_ConfigurationKey}.",  config);
+        _editorConfigData[projectFullPath] = FilterDictionaryByKeys($"{BuildCheck_ConfigurationKey}.", config);
 
         return _editorConfigData[projectFullPath];
     }
