diff --git a/src/MSBuild.UnitTests/TerminalLogger_Tests.cs b/src/MSBuild.UnitTests/TerminalLogger_Tests.cs
index 5120f4ce1ae..d22dc0a5586 100644
--- a/src/MSBuild.UnitTests/TerminalLogger_Tests.cs
+++ b/src/MSBuild.UnitTests/TerminalLogger_Tests.cs
@@ -335,7 +335,7 @@ public Task PrintImmediateWarningMessage_Succeeded()
         [Fact]
         public Task PrintCopyTaskRetryWarningAsImmediateMessage_Failed()
         {
-            InvokeLoggerCallbacksForSimpleProject(succeeded: false, () =>
+            InvokeLoggerCallbacksForSimpleProject(succeeded: false, _projectFile, () =>
             {
                 WarningRaised?.Invoke(_eventSender, MakeCopyRetryWarning(1));
                 WarningRaised?.Invoke(_eventSender, MakeCopyRetryWarning(2));
