diff --git a/src/Framework/Features.cs b/src/Framework/Features.cs
index c28036dc113..a7b1ec4b16c 100644
--- a/src/Framework/Features.cs
+++ b/src/Framework/Features.cs
@@ -40,7 +40,7 @@ public enum FeatureStatus
     /// </summary>
     public static class Features
     {
-        private static Dictionary<string, FeatureStatus> _featureStatusMap = new Dictionary<string, FeatureStatus>
+        private static readonly Dictionary<string, FeatureStatus> _featureStatusMap = new Dictionary<string, FeatureStatus>
         {
             // TODO: Fill in the dictionary with the features and their status
             { "Feature1", FeatureStatus.Available },
