diff --git a/src/MSBuild/LiveLogger/LiveLogger.cs b/src/MSBuild/LiveLogger/LiveLogger.cs
index e5fad94d167..adb4dc39459 100644
--- a/src/MSBuild/LiveLogger/LiveLogger.cs
+++ b/src/MSBuild/LiveLogger/LiveLogger.cs
@@ -271,6 +271,12 @@ private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
             return;
         }
 
+        // Mark node idle until something uses it again
+        lock (_lock)
+        {
+            _nodes[NodeIndexForContext(buildEventContext)] = null;
+        }
+
         ProjectContext c = new(buildEventContext);
 
         // First check if we're done restoring.
