diff --git a/src/Build/Construction/Solution/SolutionFile.cs b/src/Build/Construction/Solution/SolutionFile.cs
index aca9c13f723..8b47c205990 100644
--- a/src/Build/Construction/Solution/SolutionFile.cs
+++ b/src/Build/Construction/Solution/SolutionFile.cs
@@ -803,7 +803,7 @@ private void ParseProject(string firstLine)
             while ((line = ReadLine()) != null)
             {
                 // If we see an "EndProject", well ... that's the end of this project!
-                if (line == "EndProject")
+                if (string.Equals(line, "EndProject", StringComparison.Ordinal))
                 {
                     break;
                 }
