diff --git a/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs b/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs
index d08aa8aab6e..4c7f50a76fb 100644
--- a/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs
@@ -725,7 +725,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
 
             ProjectUsingTaskElement element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
             element.Runtime = "CLR4";
-            element.Architecture = "*";
+            element.Architecture = "x86";
             elementList.Add(element);
 
             TaskRegistry registry = CreateTaskRegistryAndRegisterTasks(elementList);
@@ -740,7 +740,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // parameters that are a successful exact match - should retrieve from cache
@@ -749,11 +749,11 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
                     registry,
                     exactMatchRequired: false,
                     runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.any,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // parameters that would be a successful fuzzy match - should still be retrieved from the cache
@@ -762,11 +762,11 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
                     registry,
                     exactMatchRequired: false,
                     runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.any,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // parameters that do not match -- but would match the previous fuzzy match request. Should NOT retrieve anything
@@ -774,7 +774,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: false,
                     shouldBeRetrievedFromCache: false
@@ -786,11 +786,11 @@ public void RetrieveFromCacheFuzzyMatchingParameters()
                     registry,
                     exactMatchRequired: false,
                     runtime: XMakeAttributes.MSBuildRuntimeValues.any,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.any,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
         }
 
@@ -808,7 +808,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_RecoverFromFailure()
 
             ProjectUsingTaskElement element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
             element.Runtime = "CLR4";
-            element.Architecture = null;
+            element.Architecture = "x86";
             elementList.Add(element);
 
             TaskRegistry registry = CreateTaskRegistryAndRegisterTasks(elementList);
@@ -823,7 +823,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_RecoverFromFailure()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // parameters that do not match at all - shouldn't retrieve anything
@@ -843,11 +843,11 @@ public void RetrieveFromCacheFuzzyMatchingParameters_RecoverFromFailure()
                     registry,
                     exactMatchRequired: false,
                     runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.any,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
         }
 
@@ -866,12 +866,12 @@ public void RetrieveFromCacheFuzzyMatchingParameters_MultipleUsingTasks()
 
             ProjectUsingTaskElement element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
             element.Runtime = "CLR4";
-            element.Architecture = "*"; // was x86
+            element.Architecture = "x86";
             elementList.Add(element);
 
             element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
-            element.Runtime = "CLR2"; // was *
-            element.Architecture = "*"; // was x64
+            element.Runtime = "*";
+            element.Architecture = "x64";
             elementList.Add(element);
 
             TaskRegistry registry = CreateTaskRegistryAndRegisterTasks(elementList);
@@ -886,10 +886,10 @@ public void RetrieveFromCacheFuzzyMatchingParameters_MultipleUsingTasks()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
-            // parameters that are a successful exact match for CLR4|* -- should come from cache
+            // parameters that are a successful exact match for CLR4|x86 -- should come from cache
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
@@ -899,49 +899,49 @@ public void RetrieveFromCacheFuzzyMatchingParameters_MultipleUsingTasks()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
-            // parameters that would be a successful fuzzy match for either, so should get the one in the cache (CLR4|*)
+            // parameters that would be a successful fuzzy match for either, so should get the one in the cache (CLR4|x86)
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.any,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
-            // parameters that match CLR2|* - should retrieve that
+            // parameters that match *|x64 - should retrieve that
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
-                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr2,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x64
                 );
 
-            // match CLR4|* again - comes from the cache
+            // match CLR4|x86 again - comes from the cache
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
-            // match clr2|* again, should be a fuzzy cache hit
+            // match *|x64 again
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
@@ -950,21 +950,32 @@ public void RetrieveFromCacheFuzzyMatchingParameters_MultipleUsingTasks()
                     architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
-                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr2,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x64
                 );
 
-            // match clr4|* again -- should still be a cache hit
+            // CLR2|x86 should not match either task record
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.any,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
+                    shouldBeRetrieved: false,
+                    shouldBeRetrievedFromCache: false
+                );
+
+            // match *|x64 again -- should still be a cache hit
+            RetrieveAndValidateRegisteredTaskRecord
+                (
+                    registry,
+                    exactMatchRequired: false,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
-                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x64
                 );
         }
 
@@ -985,12 +996,12 @@ public void RetrieveFromCacheFuzzyMatchingParameters_MultipleUsingTasks_PreferCa
 
             ProjectUsingTaskElement element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
             element.Runtime = "CLR4";
-            element.Architecture = "*";
+            element.Architecture = "x86";
             elementList.Add(element);
 
             element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
-            element.Runtime = "CLR2";
-            element.Architecture = "*";
+            element.Runtime = "*";
+            element.Architecture = "x64";
             elementList.Add(element);
 
             TaskRegistry registry = CreateTaskRegistryAndRegisterTasks(elementList);
@@ -1000,26 +1011,26 @@ public void RetrieveFromCacheFuzzyMatchingParameters_MultipleUsingTasks_PreferCa
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
-                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr2,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x64
                 );
 
-            // *|* -- could be filled by either, would normally be filled by CLR4|* (since it was registered first),
-            // but since CLR2|* is in the cache already, we return that one.
+            // CLR4|* -- could be filled by either, would normally be filled by CLR4|x86 (since it was registered first),
+            // but since *|x64 is in the cache already, we return that one.
             RetrieveAndValidateRegisteredTaskRecord
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.any,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
-                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr2,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.any,
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x64
                 );
         }
 
@@ -1037,7 +1048,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_ExactMatches()
 
             ProjectUsingTaskElement element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
             element.Runtime = "CLR4";
-            element.Architecture = "*";
+            element.Architecture = "x86";
             elementList.Add(element);
 
             TaskRegistry registry = CreateTaskRegistryAndRegisterTasks(elementList);
@@ -1048,11 +1059,11 @@ public void RetrieveFromCacheFuzzyMatchingParameters_ExactMatches()
                     registry,
                     exactMatchRequired: false,
                     runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    architecture: XMakeAttributes.MSBuildArchitectureValues.x86,
+                    architecture: XMakeAttributes.MSBuildArchitectureValues.any,
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // CLR4|x64 should not match
@@ -1060,7 +1071,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_ExactMatches()
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: false,
                     shouldBeRetrievedFromCache: false
@@ -1076,7 +1087,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_ExactMatches()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // try CLR4|x64 again -- should also come from the catch (but needless to say, still not be a match)
@@ -1084,7 +1095,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_ExactMatches()
                 (
                     registry,
                     exactMatchRequired: false,
-                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr2,
+                    runtime: XMakeAttributes.MSBuildRuntimeValues.clr4,
                     architecture: XMakeAttributes.MSBuildArchitectureValues.x64,
                     shouldBeRetrieved: false,
                     shouldBeRetrievedFromCache: true
@@ -1107,7 +1118,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_AdditionalParameters()
 
             ProjectUsingTaskElement element = project.AddUsingTask(TestTaskName, _testTaskLocation, null);
             element.Runtime = "CLR4";
-            element.Architecture = "*";
+            element.Architecture = "x86";
             elementList.Add(element);
 
             TaskRegistry registry = CreateTaskRegistryAndRegisterTasks(elementList);
@@ -1126,7 +1137,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_AdditionalParameters()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: false,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             // And if we try again, we should get it from the cache this time.
@@ -1138,7 +1149,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_AdditionalParameters()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
 
             taskParameters = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
@@ -1157,7 +1168,7 @@ public void RetrieveFromCacheFuzzyMatchingParameters_AdditionalParameters()
                     shouldBeRetrieved: true,
                     shouldBeRetrievedFromCache: true,
                     expectedRuntime: XMakeAttributes.MSBuildRuntimeValues.clr4,
-                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.any
+                    expectedArchitecture: XMakeAttributes.MSBuildArchitectureValues.x86
                 );
         }
 
