diff --git a/src/Tasks/BootstrapperUtil/ResourceUpdater.cs b/src/Tasks/BootstrapperUtil/ResourceUpdater.cs
index 6c38bab0b9c..b8bfa0173f1 100644
--- a/src/Tasks/BootstrapperUtil/ResourceUpdater.cs
+++ b/src/Tasks/BootstrapperUtil/ResourceUpdater.cs
@@ -99,11 +99,14 @@ public bool UpdateResources(string filename, BuildResults results)
                             {
                                 fileLength = (int)fs.Length;
                                 fileContent = new byte[fileLength];
-
+#if NET7_0_OR_GREATER
+                                fs.ReadExactly(fileContent, 0, fileLength);
+#else
 #pragma warning disable CA2022 // Avoid inexact read with 'Stream.Read'
-                               // TODO: Read the count of read bytes and check if it matches the expected length, if not raise an exception
-                               fs.Read(fileContent, 0, fileLength);
+                                // TODO: Read the count of read bytes and check if it matches the expected length, if not raise an exception
+                                fs.Read(fileContent, 0, fileLength);
 #pragma warning restore CA2022 // Avoid inexact read with 'Stream.Read'
+#endif
                             }
 
                             // Update the resources to include this file's data
diff --git a/src/Tasks/ResolveKeySource.cs b/src/Tasks/ResolveKeySource.cs
index f72c71a6c5e..ff6114d8eed 100644
--- a/src/Tasks/ResolveKeySource.cs
+++ b/src/Tasks/ResolveKeySource.cs
@@ -127,10 +127,15 @@ private bool ResolveAssemblyKey()
                             fs = File.OpenRead(KeyFile);
                             int fileLength = (int)fs.Length;
                             var keyBytes = new byte[fileLength];
+#if NET7_0_OR_GREATER
+                            // fail fast in case the file is not read till the end
+                            fs.ReadExactly(keyBytes, 0, fileLength);
+#else
 #pragma warning disable CA2022 // Avoid inexact read with 'Stream.Read'
                             // TODO: Read the count of read bytes and check if it matches the expected length, if not raise an exception
                             fs.Read(keyBytes, 0, fileLength);
 #pragma warning restore CA2022 // Avoid inexact read with 'Stream.Read'
+#endif
 
                             UInt64 hash = HashFromBlob(keyBytes);
                             hash ^= HashFromBlob(userNameBytes); // modify it with the username hash, so each user would get different hash for the same key
