diff --git a/Directory.Build.props b/Directory.Build.props
index d9e3ec92d8d..889acdc0e67 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -57,7 +57,7 @@
         RS0016 & RS0017: Roslyn analyzers seem to be bugged, claiming that API's that exist don't and vise-versa: https://github.com/dotnet/msbuild/issues/7903
     -->
 
-    <NoWarn>$(NoWarn);NU1507;NU1603;NU1605;NU5105;NU5100;NU1701;1702;SYSLIB0011;SYSLIB0037;SYSLIB0044;RS0016;RS0017;</NoWarn>
+    <NoWarn>$(NoWarn);NU1507;NU1603;NU5105;1702;SYSLIB0011;SYSLIB0037;SYSLIB0044;RS0016;RS0017;</NoWarn>
     <!-- Loading certificate data through the constructor or Import is obsolete. Use X509CertificateLoader instead to load certific
 ates https://learn.microsoft.com/en-gb/dotnet/fundamentals/syslib-diagnostics/syslib0057 -->
     <NoWarn>$(NoWarn);SYSLIB0057;</NoWarn>
