diff --git a/src/Framework.UnitTests/OperatingSystem_Tests.cs b/src/Framework.UnitTests/OperatingSystem_Tests.cs
index 80c39e3e6ca..4fbaf7af615 100644
--- a/src/Framework.UnitTests/OperatingSystem_Tests.cs
+++ b/src/Framework.UnitTests/OperatingSystem_Tests.cs
@@ -51,7 +51,6 @@ public void IsOtherThanWindows()
             Microsoft.Build.Framework.OperatingSystem.IsLinux().ShouldBeFalse();
             Microsoft.Build.Framework.OperatingSystem.IsMacOS().ShouldBeFalse();
             Microsoft.Build.Framework.OperatingSystem.IsMacOSVersionAtLeast(0).ShouldBeFalse();
-            Microsoft.Build.Framework.OperatingSystem.IsOSXLike().ShouldBeFalse();
         }
 #endif
     }
diff --git a/src/Framework/OperatingSystem.cs b/src/Framework/OperatingSystem.cs
index e9715c6d442..883ec55b924 100644
--- a/src/Framework/OperatingSystem.cs
+++ b/src/Framework/OperatingSystem.cs
@@ -30,8 +30,6 @@ public static bool IsOSPlatformVersionAtLeast(string platform, int major, int mi
 
         public static bool IsMacOS() => false;
 
-        public static bool IsOSXLike() => false;
-
         public static bool IsMacOSVersionAtLeast(int major, int minor = 0, int build = 0) => false;
 
         public static bool IsWindows() => true;
