diff --git a/eng/Versions.props b/eng/Versions.props
index fb953891790..3e900cbedba 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -2,7 +2,7 @@
 <!-- Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the MIT license. See License.txt in the project root for full license information. -->
 <Project>
   <PropertyGroup>
-    <VersionPrefix>17.2.9</VersionPrefix>
+    <VersionPrefix>17.2.10</VersionPrefix>
     <DotNetFinalVersionKind>release</DotNetFinalVersionKind>
     <AssemblyVersion>15.1.0.0</AssemblyVersion>
     <PreReleaseVersionLabel>preview</PreReleaseVersionLabel>
diff --git a/global.json b/global.json
index 6ea0d4f6d96..28ab997cd2a 100644
--- a/global.json
+++ b/global.json
@@ -5,9 +5,9 @@
   "tools": {
     "dotnet": "6.0.311",
     "vs": {
-      "version": "17.0"
+      "version": "17.2.1"
     },
-    "xcopy-msbuild": "17.0"
+    "xcopy-msbuild": "17.2.1"
   },
   "msbuild-sdks": {
     "Microsoft.Build.CentralPackageVersions": "2.0.1",
