diff --git a/src/Build/Graph/GraphBuilder.cs b/src/Build/Graph/GraphBuilder.cs
index 8ac5c1edf55..8ff12fa2fe4 100644
--- a/src/Build/Graph/GraphBuilder.cs
+++ b/src/Build/Graph/GraphBuilder.cs
@@ -178,21 +178,17 @@ void GetTransitiveProjectReferencesExcludingSelfHelper(ParsedProject parsedProje
                     HashSet<ProjectGraphNode> toCache = new();
                     foreach (ProjectInterpretation.ReferenceInfo referenceInfo in parsedProject.ReferenceInfos)
                     {
-                        if (transitiveReferences.Contains(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode))
+                        if (transitiveReferences.Add(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode))
                         {
-                            if (transitiveReferenceCache.TryGetValue(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode, out cachedTransitiveReferences))
-                            {
-                                toCache = toCache.Concat(cachedTransitiveReferences).ToHashSet();
-                            }
-                            else
-                            {
-                                toCache.Add(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode);
-                            }
+                            GetTransitiveProjectReferencesExcludingSelfHelper(allParsedProjects[referenceInfo.ReferenceConfiguration], transitiveReferences, toCache);
+                        }
+                        else if (transitiveReferenceCache.TryGetValue(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode, out cachedTransitiveReferences))
+                        {
+                            toCache = toCache.Concat(cachedTransitiveReferences).ToHashSet();
                         }
                         else
                         {
-                            transitiveReferences.Add(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode);
-                            GetTransitiveProjectReferencesExcludingSelfHelper(allParsedProjects[referenceInfo.ReferenceConfiguration], transitiveReferences, toCache);
+                            toCache.Add(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode);
                         }
                     }
 
