diff --git a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
index 46c2ae5ce6d..ce93330897a 100644
--- a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
@@ -771,7 +771,7 @@ public void SendData(INodePacket packet)
 #if FEATURE_APM
                             _serverToClientStream.BeginWrite(writeStreamBuffer, i, lengthToWrite, PacketWriteComplete, null);
 #else
-                            _serverToClientStream.WriteAsync(writeStreamBuffer, i, lengthToWrite);
+                            _serverToClientStream.Write(writeStreamBuffer, i, lengthToWrite);
 #endif
                             return;
                         }
