diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index 9ee72433bef..4b9a3dc8716 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -769,6 +769,12 @@ private bool DoCopyIfNecessary(FileState sourceFileState, FileState destinationF
                 }
                 else
                 {
+                    if (!sourceFileState.FileExists)
+                    {
+                        Log.LogErrorWithCodeFromResources("Copy.SourceFileNotFound", sourceFileState.Name);
+                        success = false;
+                    }
+
                     MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name, true);
                 }
             }
