diff --git a/src/Shared/UnitTests/ObjectModelHelpers.cs b/src/Shared/UnitTests/ObjectModelHelpers.cs
index 54ccee25eef..d3091495f80 100644
--- a/src/Shared/UnitTests/ObjectModelHelpers.cs
+++ b/src/Shared/UnitTests/ObjectModelHelpers.cs
@@ -687,7 +687,10 @@ internal static Project CreateInMemoryProject
             ILogger logger /* May be null */,
             string toolsVersion /* may be null */
             )
-            => CreateInMemoryProjectWithLoggers(projectCollection, xml, toolsVersion, logger);
+        {
+            var loggers = logger == null ? null : new[] { logger };
+            return CreateInMemoryProjectWithLoggers(projectCollection, xml, toolsVersion, loggers);
+        }
 
         /// <summary>
         /// Create an in-memory project and attach it to the passed-in engine.
@@ -747,7 +750,8 @@ params ILogger[] loggers
             )
         {
             Project project = CreateInMemoryProjectWithLoggers(projectContents, loggers);
-            project.Build(loggers).ShouldBeTrue();
+            project.Build().ShouldBeTrue();
+            project.ProjectCollection.Dispose();
         }
 
         /// <summary>
