diff --git a/src/MSBuild/PerformanceLogEventListener.cs b/src/MSBuild/PerformanceLogEventListener.cs
index 3659dd535de..3e08da0d512 100644
--- a/src/MSBuild/PerformanceLogEventListener.cs
+++ b/src/MSBuild/PerformanceLogEventListener.cs
@@ -55,16 +55,10 @@ internal static PerformanceLogEventListener Create()
                 // Check to see if we should enable the event listener.
                 string logDirectory = Environment.GetEnvironmentVariable(PerfLogDirEnvVar);
 
-                if (!string.IsNullOrEmpty(logDirectory))
+                if (!string.IsNullOrEmpty(logDirectory) && Directory.CreateDirectory(logDirectory).Exists)
                 {
-                    if (Directory.CreateDirectory(logDirectory).Exists)
-                    {
-                        eventListener = new PerformanceLogEventListener();
-                        eventListener.Initialize(logDirectory);
-                    }
-
-                    // Delete env var so new starting processes will not duplicate Event listening
-                    Environment.SetEnvironmentVariable(PerfLogDirEnvVar, null);
+                    eventListener = new PerformanceLogEventListener();
+                    eventListener.Initialize(logDirectory);
                 }
             }
             catch
