diff --git a/src/Tasks/ResGenDependencies.cs b/src/Tasks/ResGenDependencies.cs
index e86e81f4e50..f7bb0c906cb 100644
--- a/src/Tasks/ResGenDependencies.cs
+++ b/src/Tasks/ResGenDependencies.cs
@@ -186,7 +186,6 @@ internal static ResGenDependencies DeserializeCache(string stateFile, bool useSo
         /// 
         /// This is an on-disk serialization format, don't change field names or types or use readonly.
         /// </remarks>
-        [Serializable]
         internal sealed class ResXFile : DependencyFile
         {
             // Files contained within this resx file.
@@ -283,7 +282,6 @@ private static string[] GetLinkedFiles(string filename, string baseLinkedFileDir
         /// 
         /// This is an on-disk serialization format, don't change field names or types or use readonly.
         /// </remarks>
-        [Serializable]
         internal sealed class PortableLibraryFile : DependencyFile
         {
             internal string[] outputFiles;
