diff --git a/src/Build/BackEnd/Client/MSBuildClient.cs b/src/Build/BackEnd/Client/MSBuildClient.cs
index a7045fe956d..169ea8fbeab 100644
--- a/src/Build/BackEnd/Client/MSBuildClient.cs
+++ b/src/Build/BackEnd/Client/MSBuildClient.cs
@@ -169,7 +169,7 @@ public MSBuildClientExitResult Execute(CancellationToken cancellationToken)
                 string.Join(" ", _commandLine);
 #endif
 
-            CommunicationsUtilities.Trace("Trying execute build at server with command line '{0}'", descriptiveCommandLine);
+            CommunicationsUtilities.Trace("Executing build using MSBuild Server with command line '{0}'", descriptiveCommandLine);
 
             try
             {
