diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index 89df1804592..4b0119bf576 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -404,7 +404,7 @@ public DeferredBuildMessage(string text, MessageImportance importance)
             {
                 Importance = importance;
                 Text = text;
-                FilePath = "";
+                FilePath = null;
             }
             public DeferredBuildMessage(string text, MessageImportance importance, string filePath)
             {
@@ -2882,7 +2882,7 @@ private static void LogDeferredMessages(ILoggingService loggingService, IEnumera
             {
                 loggingService.LogCommentFromText(BuildEventContext.Invalid, message.Importance, message.Text);
                 // If message includes a file path, include that file
-                if (message.FilePath != "")
+                if (message.FilePath is not null)
                 {
                     loggingService.LogIncludeFile(BuildEventContext.Invalid, message.FilePath);
                 }
