diff --git a/src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs b/src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs
index d7f72956f36..d5b73f52998 100644
--- a/src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs
+++ b/src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs
@@ -9,6 +9,9 @@
 
 namespace Microsoft.Build.UnitTests
 {
+    /// <summary>
+    /// A custom <see cref="FactAttribute"/> that skips the test if the OS doesn't support creating symlinks.
+    /// </summary>
     public sealed class RequiresSymbolicLinksFactAttribute : FactAttribute
     {
         public RequiresSymbolicLinksFactAttribute()
@@ -18,6 +21,8 @@ public RequiresSymbolicLinksFactAttribute()
                 return;
             }
 
+            // In Windows, a process can create symlinks only if it has sufficient permissions.
+            // We simply try to create one and if it fails we skip the test.
             string sourceFile = FileUtilities.GetTemporaryFile();
             string destinationFile = FileUtilities.GetTemporaryFile();
             try
