diff --git a/src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs b/src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs
index 4c9d52b9dc1..d1123054a3a 100644
--- a/src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs
+++ b/src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs
@@ -83,6 +83,7 @@ private void AlignAndIndentLineOfMessage(StringBuilder sb, bool prefixAlreadyWri
 
             bool bufferIsLargerThanPrefix = bufferWidthMinusNewLine > prefixWidth;
             if (bufferIsLargerThanPrefix && _alignMessages)
+            if (_alignMessages && bufferIsLargerThanPrefix && count > 0)
             {
                 // If the buffer is larger then the prefix information (timestamp and key) then reformat the messages.
 
