diff --git a/src/Build.UnitTests/ChangeWaves_Tests.cs b/src/Build.UnitTests/ChangeWaves_Tests.cs
index 44c0e7a45f9..d561fcaf9ef 100644
--- a/src/Build.UnitTests/ChangeWaves_Tests.cs
+++ b/src/Build.UnitTests/ChangeWaves_Tests.cs
@@ -69,6 +69,8 @@ private void buildSimpleProjectAndValidateChangeWave(TestEnvironment testEnviron
                 log.WarningCount.ShouldBe(warningCodesLogShouldContain.Length);
                 log.AssertLogContains(warningCodesLogShouldContain);
             }
+
+            ChangeWaves.ResetStateForTests();
         }
 
         [Fact]
diff --git a/src/Build.UnitTests/Evaluation/Expander_Tests.cs b/src/Build.UnitTests/Evaluation/Expander_Tests.cs
index b53348ae690..76dd7380d59 100644
--- a/src/Build.UnitTests/Evaluation/Expander_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Expander_Tests.cs
@@ -1613,6 +1613,7 @@ public void ExpandAllIntoStringNotTruncated()
                 // but that goes back to MSBuild 4.something so I'm codifying it in this test. If you're here because you cleaned it up
                 // and want to fix the test my current opinion is that's fine.
                 actual.ShouldBe(expected);
+                ChangeWaves.ResetStateForTests();
             }
         }
 
diff --git a/src/Build.UnitTests/Scanner_Tests.cs b/src/Build.UnitTests/Scanner_Tests.cs
index 3ac9383f3e7..c965755637a 100644
--- a/src/Build.UnitTests/Scanner_Tests.cs
+++ b/src/Build.UnitTests/Scanner_Tests.cs
@@ -132,6 +132,7 @@ public void SpacePropertyOptOutWave16_10()
             lexer = new Scanner("$( x)", ParserOptions.AllowProperties);
             AdvanceToScannerError(lexer);
             Assert.Null(lexer.UnexpectedlyFound);
+            ChangeWaves.ResetStateForTests();
         }
 
         /// <summary>
