diff --git a/src/BuildCheck.UnitTests/ExecCliBuildCheck_Tests.cs b/src/BuildCheck.UnitTests/ExecCliBuildCheck_Tests.cs
index c266e318982..b74ccfbd75c 100644
--- a/src/BuildCheck.UnitTests/ExecCliBuildCheck_Tests.cs
+++ b/src/BuildCheck.UnitTests/ExecCliBuildCheck_Tests.cs
@@ -60,6 +60,7 @@ public void ExecTask_WithCommandExecutingBuild_ShouldShowWarning(string command)
         [InlineData("dotnet help")]
         [InlineData("where dotnet")]
         [InlineData("where msbuild")]
+        [InlineData("msbuild --help")]
         [InlineData("where nuget")]
         [InlineData("dotnet bin/net472/project.dll")]
         [InlineData("")]
