diff --git a/src/Package/MSBuild.VSSetup/files.swr b/src/Package/MSBuild.VSSetup/files.swr
index 4fca187cc6a..3b4fbd1eee9 100644
--- a/src/Package/MSBuild.VSSetup/files.swr
+++ b/src/Package/MSBuild.VSSetup/files.swr
@@ -194,7 +194,7 @@ folder InstallDir:\MSBuild\Current\Bin\amd64
   file source=$(X86BinPath)System.Reflection.MetadataLoadContext.dll
   file source=$(X86BinPath)System.Text.Json.dll
   file source=$(X86BinPath)Microsoft.Bcl.AsyncInterfaces.dll
-  file source=$(X86BinPath)Microsoft.IO.Redist.dll
+  file source=$(X86BinPath)Microsoft.IO.Redist.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Text.Encodings.Web.dll
   file source=$(X86BinPath)System.Threading.Tasks.Extensions.dll
   file source=$(X86BinPath)System.ValueTuple.dll
