diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index fdb7401c29c..bef125b8bcd 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -4269,7 +4269,7 @@ private static bool CreateAndConfigureLogger(
             }
             catch (Exception e) when (loggerDescription.IsOptional)
             {
-                Console.WriteLine(loggerDescription.Name ": " + ResourceUtilities.FormatResourceStringStripCodeAndKeyword("OptionalLoggerCreationMessage", e.Message));
+                Console.WriteLine(loggerDescription.Name + ": " + ResourceUtilities.FormatResourceStringStripCodeAndKeyword("OptionalLoggerCreationMessage", e.Message));
                 return false;
             }
 
