diff --git a/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs b/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs
index 7cef9303fa2..6a9f2e9db83 100644
--- a/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs
+++ b/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs
@@ -68,7 +68,7 @@ public override bool Execute()
                                                 {
                                                     if (!(name.Equals("System.ValueTuple", System.StringComparer.OrdinalIgnoreCase) && version.Equals("4.0.0.0") && assemblyVersion.Equals("4.0.3.0")))
                                                     {
-                                                        Log.LogError($"Binding redirect for '{name} redirects to a different version ({version}) than MSBuild ships ({assemblyVersion}).");
+                                                        Log.LogError($"Binding redirect for '{name}' redirects to a different version ({version}) than MSBuild ships ({assemblyVersion}).");
                                                     }
                                                 }
                                             }
