diff --git a/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs b/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs
index 68845ff5b2f..8e2077e54a6 100644
--- a/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs
+++ b/src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs
@@ -752,25 +752,6 @@ public void TargetsSwitchIdentificationTests(string @switch)
             emptyParametersAllowed.ShouldBeFalse();
         }
 
-        [Fact]
-        public void TargetsSwitchParameter()
-        {
-            CommandLineSwitches switches = new CommandLineSwitches();
-            MSBuildApp.GatherCommandLineSwitches(new ArrayList() { "/targets:targets.txt" }, switches);
-
-            switches.HaveErrors().ShouldBeFalse();
-            switches[CommandLineSwitches.ParameterizedSwitch.Targets].ShouldBe(new[] { "targets.txt" });
-        }
-
-        [Fact]
-        public void TargetsSwitchDoesNotSupportMultipleOccurrences()
-        {
-            CommandLineSwitches switches = new CommandLineSwitches();
-            MSBuildApp.GatherCommandLineSwitches(new ArrayList() { "/targets /targets" }, switches);
-
-            switches.HaveErrors().ShouldBeTrue();
-        }
-
         [Fact]
         public void IsolateProjectsSwitchIdentificationTests()
         {
