diff --git a/src/Deprecated/Engine/LocalProvider/SharedMemory.cs b/src/Deprecated/Engine/LocalProvider/SharedMemory.cs
index 10752cb9136..30bb926fc11 100644
--- a/src/Deprecated/Engine/LocalProvider/SharedMemory.cs
+++ b/src/Deprecated/Engine/LocalProvider/SharedMemory.cs
@@ -89,7 +89,7 @@ private SharedMemory()
         ///  This type however does not set the type of the memory mapped section,
         ///  the memory mapped section itself is created
         ///  with READWRITE access.
-        /// </param>
+        ///</param>
         /// <param name="allowExistingMapping">
         ///  The shared memory is given a parameter to determine whether or not to
         ///  reuse an existing mapped memory secion. When the node is first created
@@ -98,7 +98,7 @@ private SharedMemory()
         ///  is created, at this point the there should be no shared memory with the
         ///  same name. However when we create the reader and writer threads
         ///  (which happens on node reuse) we want to reuse the memory.
-        /// </param>
+        ///</param>
         internal SharedMemory(string name, SharedMemoryType type, bool allowExistingMapping)
         {
             this.type = type;
