diff --git a/src/Tasks/ResolveSDKReference.cs b/src/Tasks/ResolveSDKReference.cs
index 6f2fe986a12..9936453c588 100644
--- a/src/Tasks/ResolveSDKReference.cs
+++ b/src/Tasks/ResolveSDKReference.cs
@@ -30,7 +30,7 @@ public class ResolveSDKReference : TaskExtension
         /// </summary>
         private static readonly Dictionary<string, string> PlatformAliases = new(StringComparer.OrdinalIgnoreCase)
         {
-            { "UAP", "windows" }
+            { "UAP", "Windows" }
         };
 
         /// <summary>
