diff --git a/src/Framework/CriticalTaskException.cs b/src/Framework/CriticalTaskException.cs
index e372866c74a..6bd36d576f6 100644
--- a/src/Framework/CriticalTaskException.cs
+++ b/src/Framework/CriticalTaskException.cs
@@ -11,7 +11,9 @@
 namespace Microsoft.Build.Framework
 {
     /// <summary>
-    /// A wrapper exception for exceptions thrown by tasks (in TaskBuilder) that are critical to the build process.
+    /// A wrapper exception for exceptions thrown by MsBuild Tasks (in TaskBuilder) that are critical to the task run and overall to the build process.
+    /// However such exception desn't indicate problem within the MsBuild engine, but rather in the Task itself - for this reason we wrap the exception,
+    ///  so that we can properly log it up the stack (and not assume it is a bug within the build engine)
     /// </summary>
     internal sealed class CriticalTaskException : BuildExceptionBase
     {
