diff --git a/src/Build/Definition/Project.cs b/src/Build/Definition/Project.cs
index 254d11158b4..482910cb4a7 100644
--- a/src/Build/Definition/Project.cs
+++ b/src/Build/Definition/Project.cs
@@ -56,7 +56,7 @@ public class Project : ILinkableObject
         /// <summary>
         /// * and ? are invalid file name characters, but they occur in globs as wild cards.
         /// </summary>
-        private static readonly char[] s_invalidGlobChars = FileUtilities.InvalidFileNameChars.Where(c => c != '*' && c != '?' && c!= '/' && c != '\\' && c != ':').ToArray();
+        private static readonly char[] s_invalidGlobChars = FileUtilities.InvalidFileNameChars.Where(c => c != '*' && c != '?' && c != '/' && c != '\\' && c != ':').ToArray();
 
         /// <summary>
         /// Context to log messages and events in.
@@ -2782,17 +2782,18 @@ private List<ProvenanceResult> GetItemProvenance(string itemToMatch, IEnumerable
                     return new List<ProvenanceResult>();
                 }
 
-                return
-                    projectItemElements
+                return projectItemElements
                     .AsParallel()
-                    .AsOrdered()
-                    .Select(i => ComputeProvenanceResult(itemToMatch, i))
+                    .Select((item, index) => ComputeProvenanceResult(itemToMatch, item, index))
+                    .AsSequential()
+                    .OrderBy(pair => pair.Index)
+                    .Select(pair => pair.Result)
                     .Where(r => r != null)
                     .ToList();
             }
 
             // TODO: cache result?
-            private ProvenanceResult ComputeProvenanceResult(string itemToMatch, ProjectItemElement itemElement)
+            private (ProvenanceResult Result, int Index) ComputeProvenanceResult(string itemToMatch, ProjectItemElement itemElement, int index)
             {
                 ProvenanceResult SingleItemSpecProvenance(string itemSpec, IElementLocation elementLocation, Operation operation)
                 {
@@ -2807,9 +2808,11 @@ ProvenanceResult SingleItemSpecProvenance(string itemSpec, IElementLocation elem
                 }
 
                 ProvenanceResult result = SingleItemSpecProvenance(itemElement.Include, itemElement.IncludeLocation, Operation.Include);
-                return result == null ?
+                result = result == null ?
                     SingleItemSpecProvenance(itemElement.Update, itemElement.UpdateLocation, Operation.Update) ?? SingleItemSpecProvenance(itemElement.Remove, itemElement.RemoveLocation, Operation.Remove) :
                     SingleItemSpecProvenance(itemElement.Exclude, itemElement.ExcludeLocation, Operation.Exclude) ?? result;
+
+                return (result, index);
             }
 
             /// <summary>
