diff --git a/build/SignToolData.json b/build/SignToolData.json
index 130cb49b16c..b322929f0d2 100644
--- a/build/SignToolData.json
+++ b/build/SignToolData.json
@@ -21,9 +21,6 @@
                 "bin/Microsoft.Build.Conversion/net46/Microsoft.Build.Engine.dll",
                 "bin/Microsoft.Build.Conversion/net46/Microsoft.Build.Conversion.Core.dll",
 
-                "bin/MSBuild/netcoreapp2.0/MSBuild.dll",
-                "bin/MSBuild/netcoreapp2.0/Microsoft.Build.dll",
-
                 "bin/MSBuild/netcoreapp2.1/MSBuild.dll",
                 "bin/MSBuild/netcoreapp2.1/Microsoft.Build.dll",
                 "bin/MSBuild/netcoreapp2.1/Microsoft.Build.Framework.dll",
diff --git a/src/Directory.Build.props b/src/Directory.Build.props
index 5eb9926f07a..11f05240d76 100644
--- a/src/Directory.Build.props
+++ b/src/Directory.Build.props
@@ -28,7 +28,7 @@
     <PlatformTarget>AnyCPU</PlatformTarget>
 
     <!-- Target frameworks for Exe and unit test projects (ie projects with runtime output) -->
-    <RuntimeOutputTargetFrameworks>netcoreapp2.0;netcoreapp2.1</RuntimeOutputTargetFrameworks>
+    <RuntimeOutputTargetFrameworks>netcoreapp2.1</RuntimeOutputTargetFrameworks>
     <RuntimeOutputTargetFrameworks Condition="'$(OsEnvironment)'=='windows'">net46;$(RuntimeOutputTargetFrameworks)</RuntimeOutputTargetFrameworks>
     <RuntimeOutputTargetFrameworks Condition="'$(MonoBuild)' == 'true'">net461</RuntimeOutputTargetFrameworks>
 
