diff --git a/src/Build.UnitTests/ProjectCache/ProjectCacheTests.cs b/src/Build.UnitTests/ProjectCache/ProjectCacheTests.cs
index b742e03d868..ec460ade72b 100644
--- a/src/Build.UnitTests/ProjectCache/ProjectCacheTests.cs
+++ b/src/Build.UnitTests/ProjectCache/ProjectCacheTests.cs
@@ -398,6 +398,15 @@ public void ProjectCacheByBuildParametersAndGraphBuildWorks(GraphCacheResponse t
 
             var graphResult = buildSession.BuildGraph(graph);
 
+            if (buildParameters.DisableInProcNode
+                && testData.NonCacheMissResults.Any(c => c.Value.ProxyTargets is not null))
+            {
+                // TODO: remove this branch when the DisableInProcNode failure is fixed by https://github.com/dotnet/msbuild/pull/6400
+                graphResult.OverallResult.ShouldBe(BuildResultCode.Failure);
+                buildSession.Logger.Errors.First().Code.ShouldBe("MSB4223");
+                return;
+            }
+
             graphResult.OverallResult.ShouldBe(BuildResultCode.Success);
 
             buildSession.Dispose();
@@ -425,6 +434,17 @@ public void ProjectCacheByBuildParametersAndBottomUpBuildWorks(GraphCacheRespons
             foreach (var node in graph.ProjectNodesTopologicallySorted)
             {
                 var buildResult = buildSession.BuildProjectFile(node.ProjectInstance.FullPath);
+
+                if (buildParameters.DisableInProcNode &&
+                    testData.NonCacheMissResults.TryGetValue(GetProjectNumber(node), out var cacheResult) &&
+                    cacheResult.ProxyTargets is not null)
+                {
+                    // TODO: remove this branch when the DisableInProcNode failure is fixed by https://github.com/dotnet/msbuild/pull/6400
+                    buildResult.OverallResult.ShouldBe(BuildResultCode.Failure);
+                    buildSession.Logger.Errors.First().Code.ShouldBe("MSB4223");
+                    return;
+                }
+
                 buildResult.OverallResult.ShouldBe(BuildResultCode.Success);
 
                 nodesToBuildResults[node] = buildResult;
@@ -525,6 +545,14 @@ public void RunningProxyBuildsOnOutOfProcNodesShouldIssueWarning(bool disableInp
 
             var graphResult = buildSession.BuildGraph(graph);
 
+            if (!disableInprocNodeViaEnvironmentVariable)
+            {
+                // TODO: remove this branch when the DisableInProcNode failure is fixed by https://github.com/dotnet/msbuild/pull/6400
+                graphResult.OverallResult.ShouldBe(BuildResultCode.Failure);
+                buildSession.Logger.Errors.First().Code.ShouldBe("MSB4223");
+                return;
+            }
+
             graphResult.OverallResult.ShouldBe(BuildResultCode.Success);
 
             buildSession.Dispose();
