diff --git a/src/Build/Evaluation/ExpressionShredder.cs b/src/Build/Evaluation/ExpressionShredder.cs
index ffa04158d9d..afb12090a4d 100644
--- a/src/Build/Evaluation/ExpressionShredder.cs
+++ b/src/Build/Evaluation/ExpressionShredder.cs
@@ -110,12 +110,14 @@ internal static List<ItemExpressionCapture> GetReferencedItemExpressions(string
         {
             List<ItemExpressionCapture> subExpressions = null;
 
-            if (expression.IndexOf('@') < 0)
+            int startInd = expression.IndexOf('@', start, end - start);
+
+            if (startInd < 0)
             {
                 return null;
             }
 
-            for (int i = start; i < end; i++)
+            for (int i = startInd; i < end; i++)
             {
                 int restartPoint;
                 int startPoint;
