diff --git a/src/MSBuild/CommandLineSwitches.cs b/src/MSBuild/CommandLineSwitches.cs
index 22206252993..8ab4afc6a2f 100644
--- a/src/MSBuild/CommandLineSwitches.cs
+++ b/src/MSBuild/CommandLineSwitches.cs
@@ -438,29 +438,6 @@ internal void SetParameterlessSwitch(ParameterlessSwitch parameterlessSwitch, st
         // list of recognized switch parameter separators -- for switches that take multiple parameters
         private static readonly char[] s_parameterSeparators = { ',', ';' };
 
-        /// <summary>
-        /// Called when a recognized switch that takes parameters is detected on the command line,
-        /// but an invalid switch of the same kind was detected first. Here we pretend the first
-        /// switch didn't exist and override it.
-        /// </summary>
-        /// <param name="parameterizedSwitch"></param>
-        /// <param name="switchParameters"></param>
-        /// <param name="multipleParametersAllowed"></param>
-        /// <param name="unquoteParameters"></param>
-        /// <returns>true, if the given parameters were successfully stored</returns>
-        internal bool OverrideParameterizedSwitch(
-            ParameterizedSwitch parameterizedSwitch,
-            string commandLineArg,
-            string switchParameters,
-            bool multipleParametersAllowed,
-            bool unquoteParameters,
-            bool emptyParametersAllowed
-            )
-        {
-            _parameterizedSwitches[(int)parameterizedSwitch].commandLineArg = null;
-            return SetParameterizedSwitch(parameterizedSwitch, commandLineArg, switchParameters, multipleParametersAllowed, unquoteParameters, emptyParametersAllowed);
-        }
-
         /// <summary>
         /// Called when a recognized switch that takes parameters is detected on the command line.
         /// </summary>
