diff --git a/src/Shared/NodeEndpointOutOfProcBase.cs b/src/Shared/NodeEndpointOutOfProcBase.cs
index 6f86db9c054..091e2be3be1 100644
--- a/src/Shared/NodeEndpointOutOfProcBase.cs
+++ b/src/Shared/NodeEndpointOutOfProcBase.cs
@@ -523,7 +523,7 @@ private void RunReadLoop(BufferedReadStream localReadPipe, NamedPipeServerStream
             byte[] headerByte = new byte[5];
 #if !TASKHOST
             // Use a separate reuseable wait handle to avoid allocating on Task.AsyncWaitHandle.
-            AutoResetEvent readTaskEvent = new(false);
+            using AutoResetEvent readTaskEvent = new(false);
             ValueTask<int> readTask = CommunicationsUtilities.ReadAsync(localReadPipe, headerByte, headerByte.Length, readTaskEvent);
 #else
             IAsyncResult result = localReadPipe.BeginRead(headerByte, 0, headerByte.Length, null, null);
