diff --git a/src/StringTools/InternableString.Simple.cs b/src/StringTools/InternableString.Simple.cs
index 296365717a8..39c867fea9f 100644
--- a/src/StringTools/InternableString.Simple.cs
+++ b/src/StringTools/InternableString.Simple.cs
@@ -51,7 +51,6 @@ public ref struct Enumerator
             /// </summary>
             private int _charIndex;
 
-            // https://github.com/dotnet/msbuild/pull/7790#issuecomment-1213603253
             public Enumerator(scoped ref InternableString spanBuilder)
             {
                 _string = spanBuilder;
diff --git a/src/StringTools/InternableString.cs b/src/StringTools/InternableString.cs
index 1076c26d24e..6a2bbcc3473 100644
--- a/src/StringTools/InternableString.cs
+++ b/src/StringTools/InternableString.cs
@@ -33,7 +33,6 @@ public ref struct Enumerator
             /// </summary>
             private int _charIndex;
 
-            // https://github.com/dotnet/msbuild/pull/7790#issuecomment-1213603253
             internal Enumerator(scoped ref InternableString str)
             {
                 _string = str;
