diff --git a/src/Framework/Telemetry/OpenTelemetryManager.cs b/src/Framework/Telemetry/OpenTelemetryManager.cs
index 472fbd89580..43e6f640b1a 100644
--- a/src/Framework/Telemetry/OpenTelemetryManager.cs
+++ b/src/Framework/Telemetry/OpenTelemetryManager.cs
@@ -16,9 +16,9 @@
 using OpenTelemetry;
 using OpenTelemetry.Trace;
 #endif
-#if DEBUG && NETFRAMEWORK
-using OpenTelemetry.Exporter;
-#endif
+// #if DEBUG && NETFRAMEWORK
+// using OpenTelemetry.Exporter;
+// #endif
 
 namespace Microsoft.Build.Framework.Telemetry
 {
@@ -63,10 +63,12 @@ public static void Initialize(bool isStandalone)
 
                 s_tracerProvider =
                     tracerProviderBuilder
+                        /*
 #if DEBUG
                         .AddOtlpExporter()
 #endif
-                       .Build();
+                        */
+                        .Build();
 
                 // create collector if not in vs
                 if (isStandalone)
@@ -207,7 +209,7 @@ public static Activity WithTag(this Activity activity, string name, object value
             activity.SetTag($"{TelemetryConstants.PropertyPrefix}{name}", hashed ? value.GetHashCode() : value);
             return activity;
         }
-        
+
         public static Activity WithStartTime(this Activity activity, DateTime? startTime)
         {
             if (startTime.HasValue)
