diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index facf5c948e8..ce330e93bdd 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -727,15 +727,16 @@ void InitializeCaches()
         private void InitializeTelemetry()
         {
             OpenTelemetryManager.Instance.Initialize(isStandalone: false);
+            string? failureMessage = OpenTelemetryManager.Instance.LoadFailureExceptionMessage;
             if (_deferredBuildMessages != null &&
-                OpenTelemetryManager.Instance.LoadFailureExceptionMessage != null &&
+                failureMessage != null &&
                 _deferredBuildMessages is ICollection<DeferredBuildMessage> deferredBuildMessagesCollection)
             {
                 deferredBuildMessagesCollection.Add(
                     new DeferredBuildMessage(
                         ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword(
                             "OpenTelemetryLoadFailed",
-                            OpenTelemetryManager.Instance.LoadFailureExceptionMessage),
+                            failureMessage),
                     MessageImportance.Low));
 
                 // clean up the message from OpenTelemetryManager to avoid double logging it
