diff --git a/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs b/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
index 19dbb3252d8..286d08a84a5 100644
--- a/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
@@ -924,7 +924,7 @@ public void ImportWithVSPathThrowsCorrectError()
                 // Does not matter that the file or folder does not exist, we are checking for the VS pathing here
                 importPath = "path\\that\\does\\not\\exist\\Microsoft\\VisualStudio\\FileName.txt";
 
-                string content = ObjectModelHelpers.CleanupFileContents(@"
+                var content = env.CreateTestProjectWithFiles(@"
                     <Project ToolsVersion=""msbuilddefaulttoolsversion"" xmlns='msbuildnamespace' >
                         <Import Project='" + importPath + @"'/>
                     </Project>
@@ -932,7 +932,7 @@ public void ImportWithVSPathThrowsCorrectError()
 
                 InvalidProjectFileException ex = Assert.Throws<InvalidProjectFileException>(() =>
                 {
-                        Project project = new Project(XmlReader.Create(new StringReader(content)));
+                        Project project = new Project(content.ProjectFile, null, null);
                 });
 
                 Assert.Contains("MSB4278", ex.ErrorCode);
