diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index 8e7702b9cb9..94af66f7e4a 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -456,7 +456,7 @@ private bool CopySingleThreaded(
                 }
                 else
                 {
-                    MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destPath);
                 }
 
                 if (copyComplete)
@@ -564,7 +564,7 @@ private bool CopyParallel(
                         }
                         else
                         {
-                            MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec)
+                            MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec);
                         }
 
                         if (copyComplete)
@@ -721,7 +721,7 @@ private bool DoCopyIfNecessary(FileState sourceFileState, FileState destinationF
                         "SkipUnchangedFiles",
                         "true"
                     );
-                    MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name);
                 }
                 // We only do the cheap check for identicalness here, we try the more expensive check
                 // of comparing the fullpaths of source and destination to see if they are identical,
@@ -731,12 +731,12 @@ private bool DoCopyIfNecessary(FileState sourceFileState, FileState destinationF
                              destinationFileState.Name,
                              StringComparison.OrdinalIgnoreCase))
                 {
-                    MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name);
                     success = DoCopyWithRetries(sourceFileState, destinationFileState, copyFile);
                 }
                 else
                 {
-                    MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name);
                 }
             }
             catch (OperationCanceledException)
