diff --git a/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs b/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs
index b2c82c88ab6..75b79e38a64 100644
--- a/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs
+++ b/src/Build/BackEnd/Client/MSBuildClientPacketPump.cs
@@ -68,6 +68,8 @@ internal sealed class MSBuildClientPacketPump : INodePacketHandler, INodePacketF
 
         public MSBuildClientPacketPump(Stream stream)
         {
+            ErrorUtilities.VerifyThrowArgumentNull(stream, nameof(stream));
+
             _stream = stream;
             _packetFactory = new NodePacketFactory();
 
