diff --git a/src/Build/BackEnd/Components/Logging/LoggingService.cs b/src/Build/BackEnd/Components/Logging/LoggingService.cs
index 1e8ab55c00e..27e27c29221 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -526,16 +526,16 @@ public bool HasBuildSubmissionLoggedErrors(int submissionId)
         /// <returns>A Hashset containing warning codes that should be treated as errors.</returns>
         public HashSet<string> GetWarningsToBeLoggedAsErrorsByProject(BuildEventContext context)
         {
-            if(_warningsAsErrorsByProject == null)
+            if (_warningsAsErrorsByProject == null)
             {
                 return null;
             }
 
             int key = GetWarningsAsErrorOrMessageKey(context);
 
-            HashSet<string> warningsAsErrorsExcludingMessages = new HashSet<string>(_warningsAsErrorsByProject?[key]);
+            HashSet<string> warningsAsErrorsExcludingMessages = new HashSet<string>(_warningsAsErrorsByProject[key]);
 
-            if(_warningsAsMessagesByProject != null)
+            if (_warningsAsMessagesByProject != null)
             {
                 warningsAsErrorsExcludingMessages.ExceptWith(_warningsAsMessagesByProject[key]);
 
diff --git a/src/Build/BackEnd/Components/Logging/TaskLoggingContext.cs b/src/Build/BackEnd/Components/Logging/TaskLoggingContext.cs
index 3f8208156dc..ad0af0ff9a3 100644
--- a/src/Build/BackEnd/Components/Logging/TaskLoggingContext.cs
+++ b/src/Build/BackEnd/Components/Logging/TaskLoggingContext.cs
@@ -1,10 +1,10 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System;
 using Microsoft.Build.Framework;
 using Microsoft.Build.Shared;
 using Microsoft.Build.Execution;
+using System;
 using System.Collections.Generic;
 
 namespace Microsoft.Build.BackEnd.Logging
