diff --git a/src/Build/Evaluation/Expander.cs b/src/Build/Evaluation/Expander.cs
index 70f2b3f6429..5327195d854 100644
--- a/src/Build/Evaluation/Expander.cs
+++ b/src/Build/Evaluation/Expander.cs
@@ -774,6 +774,10 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
             return expression.IndexOf(quoteChar, index);
         }
 
+        /// <summary>
+        /// Add the argument represented by the slices to the arguments list, handling nulls
+        /// appropriately.
+        /// </summary>
         private static void AddArgumentFromSlices(List<string> arguments, List<Tuple<int, int>> slices, string arg)
         {
             // This shouldn't happen
