diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 466df85aed3..ce4a68bbb05 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1055,6 +1055,7 @@ private static ExitType OutputPropertiesAfterEvaluation(string[] getProperty, st
             {
                 JsonOutputFormatter jsonOutputFormatter = new();
                 jsonOutputFormatter.AddPropertiesInJsonFormat(getProperty, property => project.GetPropertyValue(property));
+                jsonOutputFormatter.AddItemsInJsonFormat(getItem, project);
                 Console.WriteLine(jsonOutputFormatter.ToString());
             }
 
