diff --git a/src/MSBuild/LiveLogger/ProjectNode.cs b/src/MSBuild/LiveLogger/ProjectNode.cs
index 9561ead1647..998c8247a62 100644
--- a/src/MSBuild/LiveLogger/ProjectNode.cs
+++ b/src/MSBuild/LiveLogger/ProjectNode.cs
@@ -88,8 +88,10 @@ public void Log()
             }
 
             ShouldRerender = false;
+            // Summary (messages, warnings and errors)
+            string lineSummary = MessageCount + WarningCount + ErrorCount > 0 ? $"({MessageCount} ℹ️, {WarningCount} ⚠️, {ErrorCount} ❌)" : string.Empty;
             // Project details
-            string lineContents = ANSIBuilder.Alignment.SpaceBetween(ToANSIString(), $"({MessageCount} ℹ️, {WarningCount} ⚠️, {ErrorCount} ❌)", Console.BufferWidth - 1);
+            string lineContents = ANSIBuilder.Alignment.SpaceBetween(ToANSIString(), lineSummary, Console.BufferWidth - 1);
             // Create or update line
             if (Line is null)
             {
