diff --git a/src/Build/FileSystem/IDirectoryCache.cs b/src/Build/FileSystem/IDirectoryCache.cs
index e74d7c71a2b..a9612ec6f61 100644
--- a/src/Build/FileSystem/IDirectoryCache.cs
+++ b/src/Build/FileSystem/IDirectoryCache.cs
@@ -8,11 +8,13 @@
 using Microsoft.IO.Enumeration;
 #endif
 
+using Microsoft.Build.Evaluation;
+
 namespace Microsoft.Build.FileSystem
 {
     public interface IDirectoryCacheFactory
     {
-        IDirectoryCache GetDirectoryCacheForProject(string projectPath);
+        IDirectoryCache GetDirectoryCacheForProject(Project project);
     }
 
     public interface IDirectoryCache
