diff --git a/src/Framework/ExtendedBuildErrorEventArgs.cs b/src/Framework/ExtendedBuildErrorEventArgs.cs
index fa63d0e52f8..85d10939236 100644
--- a/src/Framework/ExtendedBuildErrorEventArgs.cs
+++ b/src/Framework/ExtendedBuildErrorEventArgs.cs
@@ -26,7 +26,7 @@ public sealed class ExtendedBuildErrorEventArgs : BuildErrorEventArgs, IExtended
     /// <summary>
     /// Default constructor. Used for deserialization.
     /// </summary>
-    public ExtendedBuildErrorEventArgs() : this("undefined") { }
+    internal ExtendedBuildErrorEventArgs() : this("undefined") { }
 
     /// <summary>
     /// This constructor specifies only type of extended data.
diff --git a/src/Framework/ExtendedBuildMessageEventArgs.cs b/src/Framework/ExtendedBuildMessageEventArgs.cs
index daa932c8bea..53ec510c8da 100644
--- a/src/Framework/ExtendedBuildMessageEventArgs.cs
+++ b/src/Framework/ExtendedBuildMessageEventArgs.cs
@@ -26,7 +26,7 @@ public sealed class ExtendedBuildMessageEventArgs : BuildMessageEventArgs, IExte
     /// <summary>
     /// Default constructor. Used for deserialization.
     /// </summary>
-    public ExtendedBuildMessageEventArgs() : this("undefined") { }
+    internal ExtendedBuildMessageEventArgs() : this("undefined") { }
 
     /// <summary>
     /// This constructor specifies only type of extended data.
diff --git a/src/Framework/ExtendedBuildWarningEventArgs.cs b/src/Framework/ExtendedBuildWarningEventArgs.cs
index d06fdf64121..9481d2bfbdc 100644
--- a/src/Framework/ExtendedBuildWarningEventArgs.cs
+++ b/src/Framework/ExtendedBuildWarningEventArgs.cs
@@ -26,7 +26,7 @@ public sealed class ExtendedBuildWarningEventArgs : BuildWarningEventArgs, IExte
     /// <summary>
     /// Default constructor. Used for deserialization.
     /// </summary>
-    public ExtendedBuildWarningEventArgs() : this("undefined") { }
+    internal ExtendedBuildWarningEventArgs() : this("undefined") { }
 
     /// <summary>
     /// This constructor specifies only type of extended data.
diff --git a/src/Framework/ExtendedCustomBuildEventArgs.cs b/src/Framework/ExtendedCustomBuildEventArgs.cs
index 3329ccaaac2..0669c025843 100644
--- a/src/Framework/ExtendedCustomBuildEventArgs.cs
+++ b/src/Framework/ExtendedCustomBuildEventArgs.cs
@@ -27,7 +27,7 @@ public sealed class ExtendedCustomBuildEventArgs : CustomBuildEventArgs, IExtend
     /// This constructor allows event data to be initialized.
     /// </summary>
     /// <seealso cref="IExtendedBuildEventArgs.ExtendedType"/>
-    public ExtendedCustomBuildEventArgs() : this("undefined") {}
+    internal ExtendedCustomBuildEventArgs() : this("undefined") {}
 
     /// <summary>
     /// This constructor allows event data to be initialized.
