diff --git a/src/Framework/Microsoft.Build.Framework.csproj b/src/Framework/Microsoft.Build.Framework.csproj
index 5e0d9a48a53..df75564473e 100644
--- a/src/Framework/Microsoft.Build.Framework.csproj
+++ b/src/Framework/Microsoft.Build.Framework.csproj
@@ -15,7 +15,7 @@
       an imported package. This suppression should be removed if/when the project is migrated to enable nullable
       reference types.
     -->
-    <NoWarn>$(NoWarn);CS8632</NoWarn>
+    <NoWarn>$(NoWarn),CS8632</NoWarn>
   </PropertyGroup>
 
   <ItemGroup>
