diff --git a/src/Build/BackEnd/Node/OutOfProcServerNode.cs b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
index 5d442fd32a4..468681779bf 100644
--- a/src/Build/BackEnd/Node/OutOfProcServerNode.cs
+++ b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
@@ -466,8 +466,6 @@ protected override void Dispose(bool disposing)
                 {
                     _timer.Dispose();
                     Flush();
-
-                    _syncWriter?.Dispose();
                 }
 
                 base.Dispose(disposing);
