diff --git a/src/Build/Logging/BinaryLogger/BuildEventArgsExtensions.cs b/src/Build/Logging/BuildEventArgsExtensions.cs
similarity index 96%
rename from src/Build/Logging/BinaryLogger/BuildEventArgsExtensions.cs
rename to src/Build/Logging/BuildEventArgsExtensions.cs
index 845500d6991..e65ffcd7902 100644
--- a/src/Build/Logging/BinaryLogger/BuildEventArgsExtensions.cs
+++ b/src/Build/Logging/BuildEventArgsExtensions.cs
@@ -1,14 +1,9 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
 using System.Collections;
 using System.Collections.Generic;
-using System.IO;
-using System.Linq;
-using Microsoft.Build.Evaluation;
 using Microsoft.Build.Framework;
-using static Microsoft.Build.Internal.Utilities;
 
 namespace Microsoft.Build.Logging;
 
diff --git a/src/Build/Microsoft.Build.csproj b/src/Build/Microsoft.Build.csproj
index 5697c631b25..75fbdb18d2e 100644
--- a/src/Build/Microsoft.Build.csproj
+++ b/src/Build/Microsoft.Build.csproj
@@ -170,7 +170,7 @@
     <Compile Include="FileAccess\ReportedFileOperation.cs" />
     <Compile Include="FileAccess\RequestedAccess.cs" />
     <Compile Include="Instance\IPropertyElementWithLocation.cs" />
-    <Compile Include="Logging\BinaryLogger\BuildEventArgsExtensions.cs" />
+    <Compile Include="Logging\BuildEventArgsExtensions.cs" />
     <Compile Include="Utilities\ReaderWriterLockSlimExtensions.cs" />
     <Compile Include="BackEnd\Node\ConsoleOutput.cs" />
     <Compile Include="BackEnd\Node\PartialBuildTelemetry.cs" />
