diff --git a/eng/Packages.props b/eng/Packages.props
index 36b51662364..1077c2f56df 100644
--- a/eng/Packages.props
+++ b/eng/Packages.props
@@ -2,7 +2,6 @@
   <ItemGroup>
     <PackageReference Update="jnm2.ReferenceAssemblies.net35" Version="1.0.1" />
     <PackageReference Update="LargeAddressAware" Version="1.0.3" />
-    <PackageReference Update="Microsoft.Bcl.HashCode" Version="1.1.0" />
     <PackageReference Update="Microsoft.Build.NuGetSdkResolver" Version="$(NuGetBuildTasksVersion)" />
     <PackageReference Update="Microsoft.CodeAnalysis.Build.Tasks" Version="$(MicrosoftNetCompilersToolsetVersion)" />
     <PackageReference Update="Microsoft.DotNet.BuildTools.GenAPI" Version="2.1.0-prerelease-02404-02" />
diff --git a/src/Build/BackEnd/Components/ProjectCache/ProjectCacheItem.cs b/src/Build/BackEnd/Components/ProjectCache/ProjectCacheItem.cs
index 964cb5294d6..21af5dd54ed 100644
--- a/src/Build/BackEnd/Components/ProjectCache/ProjectCacheItem.cs
+++ b/src/Build/BackEnd/Components/ProjectCache/ProjectCacheItem.cs
@@ -54,17 +54,17 @@ public override bool Equals(object obj)
 
         public override int GetHashCode()
         {
-            var hashCode = new HashCode();
+            int hashCode = -1043047289;
 
-            hashCode.Add(PluginPath);
+            hashCode = (hashCode * -1521134295) + PluginPath.GetHashCode();
 
             foreach (var pluginSetting in _pluginSettingsSorted)
             {
-                hashCode.Add(pluginSetting.Key);
-                hashCode.Add(pluginSetting.Value);
+                hashCode = (hashCode * -1521134295) + pluginSetting.Key.GetHashCode();
+                hashCode = (hashCode * -1521134295) + pluginSetting.Value.GetHashCode();
             }
 
-            return hashCode.ToHashCode();
+            return hashCode;
         }
     }
 }
diff --git a/src/Build/Microsoft.Build.csproj b/src/Build/Microsoft.Build.csproj
index 3f647ebf409..35b6e45e851 100644
--- a/src/Build/Microsoft.Build.csproj
+++ b/src/Build/Microsoft.Build.csproj
@@ -40,7 +40,6 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
-    <PackageReference Include="Microsoft.Bcl.HashCode" />
     <PackageReference Include="Microsoft.VisualStudio.Setup.Configuration.Interop" />
     <Reference Include="System.Configuration" />
     <Reference Include="System.IO.Compression" />
diff --git a/src/Package/MSBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec b/src/Package/MSBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec
index c092ccf2dc5..9482f3faf5d 100644
--- a/src/Package/MSBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec
+++ b/src/Package/MSBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec
@@ -35,7 +35,6 @@
     <file src="$X86BinPath$/System.Memory.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Text.Json.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/Microsoft.Bcl.AsyncInterfaces.dll" target="v15.0/bin" />
-    <file src="$X86BinPath$/Microsoft.Bcl.HashCode.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Buffers.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Text.Encodings.Web.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Threading.Tasks.Extensions.dll" target="v15.0/bin" />
@@ -90,7 +89,6 @@
     <file src="$X86BinPath$/System.Memory.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/System.Text.Json.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/Microsoft.Bcl.AsyncInterfaces.dll" target="v15.0/bin/amd64" />
-    <file src="$X86BinPath$/Microsoft.Bcl.HashCode.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/System.Buffers.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/System.Text.Encodings.Web.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/System.Threading.Tasks.Extensions.dll" target="v15.0/bin/amd64" />
diff --git a/src/Package/MSBuild.VSSetup/files.swr b/src/Package/MSBuild.VSSetup/files.swr
index 82e75e9b6c5..2f87ef174f2 100644
--- a/src/Package/MSBuild.VSSetup/files.swr
+++ b/src/Package/MSBuild.VSSetup/files.swr
@@ -41,7 +41,6 @@ folder InstallDir:\MSBuild\Current\Bin
   file source=$(X86BinPath)System.Memory.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Text.Json.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)Microsoft.Bcl.AsyncInterfaces.dll vs.file.ngenArchitecture=all
-  file source=$(X86BinPath)Microsoft.Bcl.HashCode.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Text.Encodings.Web.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Threading.Tasks.Extensions.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.ValueTuple.dll vs.file.ngenArchitecture=all
@@ -191,7 +190,6 @@ folder InstallDir:\MSBuild\Current\Bin\amd64
   file source=$(X86BinPath)System.Memory.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Text.Json.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)Microsoft.Bcl.AsyncInterfaces.dll vs.file.ngenArchitecture=all
-  file source=$(X86BinPath)Microsoft.Bcl.HashCode.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Text.Encodings.Web.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.Threading.Tasks.Extensions.dll vs.file.ngenArchitecture=all
   file source=$(X86BinPath)System.ValueTuple.dll vs.file.ngenArchitecture=all
