diff --git a/.editorconfig b/.editorconfig
index 2cde82d369f..d6d4b0dcc51 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -421,4 +421,7 @@ dotnet_diagnostic.SA1010.severity = none
 dotnet_diagnostic.IDE0100.severity = suggestion
 
 # File header should match the template, making it error since couple of files met in the code base without any header
-dotnet_diagnostic.IDE0073.severity = error
\ No newline at end of file
+dotnet_diagnostic.IDE0073.severity = error
+
+Use 'System.Threading.Lock'
+dotnet_diagnostic.IDE0330.severity = suggestion
\ No newline at end of file
diff --git a/Directory.Build.props b/Directory.Build.props
index b3524ccc277..f27847923f1 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -58,6 +58,9 @@
     -->
 
     <NoWarn>$(NoWarn);NU1507;NU1603;NU5105;1701;1702;SYSLIB0011;SYSLIB0037;SYSLIB0044;RS0016;RS0017;</NoWarn>
+    <!-- Loading certificate data through the constructor or Import is obsolete. Use X509CertificateLoader instead to load certific
+ates https://learn.microsoft.com/en-gb/dotnet/fundamentals/syslib-diagnostics/syslib0057 -->
+    <NoWarn>$(NoWarn);SYSLIB0057;</NoWarn>
   </PropertyGroup>
 
   <!-- Configuration MSBuild for portable (xcopy-install) toolsets: works on WinNT and linux/mac via Mono, isolates from machine environment:
diff --git a/eng/Version.Details.xml b/eng/Version.Details.xml
index f402448b04f..b04aff9230a 100644
--- a/eng/Version.Details.xml
+++ b/eng/Version.Details.xml
@@ -91,17 +91,17 @@
       <Sha>73f0850939d96131c28cf6ea6ee5aacb4da0083a</Sha>
       <SourceBuild RepoName="xliff-tasks" ManagedOnly="true" />
     </Dependency>
-    <Dependency Name="NuGet.Build.Tasks" Version="6.11.0-rc.122">
+    <Dependency Name="NuGet.Build.Tasks" Version="6.12.0-preview.1.83">
       <Uri>https://github.com/nuget/nuget.client</Uri>
-      <Sha>c0d441fc5e99e6765c9e0b2f77de162b9866b305</Sha>
+      <Sha>c8d14f3c28e3af3bdc3dcd38b23a212a91d13234</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.11.0-3.24378.3">
+    <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.12.0-2.24425.4">
       <Uri>https://github.com/dotnet/roslyn</Uri>
-      <Sha>5e3a11e2e7f952da93f9d35bd63a2fa181c0608b</Sha>
+      <Sha>13b65d3b5638457d94c83b545235f47f0d11e483</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.SourceBuild.Intermediate.roslyn" Version="4.11.0-3.24378.3">
+    <Dependency Name="Microsoft.SourceBuild.Intermediate.roslyn" Version="4.12.0-2.24425.4">
       <Uri>https://github.com/dotnet/roslyn</Uri>
-      <Sha>5e3a11e2e7f952da93f9d35bd63a2fa181c0608b</Sha>
+      <Sha>13b65d3b5638457d94c83b545235f47f0d11e483</Sha>
       <SourceBuild RepoName="roslyn" ManagedOnly="true" />
     </Dependency>
     <Dependency Name="Microsoft.DotNet.XUnitExtensions" Version="8.0.0-beta.24413.2">
diff --git a/eng/Versions.props b/eng/Versions.props
index ef00a7e39aa..83b1c851b77 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -52,8 +52,8 @@
     <MicrosoftDotNetXUnitExtensionsVersion>8.0.0-beta.24413.2</MicrosoftDotNetXUnitExtensionsVersion>
     <MicrosoftExtensionsDependencyModelVersion>7.0.0</MicrosoftExtensionsDependencyModelVersion>
     <MicrosoftIORedistVersion>6.0.1</MicrosoftIORedistVersion>
-    <MicrosoftNetCompilersToolsetVersion>4.11.0-3.24378.3</MicrosoftNetCompilersToolsetVersion>
-    <NuGetBuildTasksVersion>6.11.0-rc.122</NuGetBuildTasksVersion>
+    <MicrosoftNetCompilersToolsetVersion>4.12.0-2.24425.4</MicrosoftNetCompilersToolsetVersion>
+    <NuGetBuildTasksVersion>6.12.0-preview.1.83</NuGetBuildTasksVersion>
     <SystemRuntimeCompilerServicesUnsafeVersion>6.0.0</SystemRuntimeCompilerServicesUnsafeVersion>
   </PropertyGroup>
   <PropertyGroup Condition="!$(TargetFramework.StartsWith('net4'))">
@@ -75,7 +75,6 @@
       <FileVersion>$(VersionPrefix).$(FileVersion.Split('.')[3])</FileVersion>
     </PropertyGroup>
   </Target>
-
   <!--  Freeze the version for the current upgrade of the sdk. 
         TODO:Delete the lines below once the Retargeting merged. 
   -->
