diff --git a/src/Shared/UnitTests/MockEngine.cs b/src/Shared/UnitTests/MockEngine.cs
index c9b5ad61fe8..b51fb33a09a 100644
--- a/src/Shared/UnitTests/MockEngine.cs
+++ b/src/Shared/UnitTests/MockEngine.cs
@@ -33,6 +33,7 @@ namespace Microsoft.Build.UnitTests
      **************************************************************************/
     internal sealed class MockEngine : IBuildEngine8
     {
+        private readonly string ResourceSemaphoreName = $"MSBuildTestResourceSemaphore{Guid.NewGuid().ToString()}";
         private readonly object _lockObj = new object();  // Protects _log, _output
         private readonly ITestOutputHelper _output;
         private readonly StringBuilder _log = new StringBuilder();
@@ -491,7 +492,7 @@ public object UnregisterTaskObject(object key, RegisteredTaskObjectLifetime life
         Semaphore cpuCount;
         public int RequestCores(int requestedCores)
         {
-            cpuCount ??= Semaphore.OpenExisting("cpuCount");
+            cpuCount ??= Semaphore.OpenExisting(ResourceSemaphoreName);
 
             int coresAcquiredBeforeMoreCoresGetAcquired = runningTotal;
 
@@ -515,7 +516,7 @@ public int RequestCores(int requestedCores)
 
         public void ReleaseCores(int coresToRelease)
         {
-            cpuCount ??= Semaphore.OpenExisting("cpuCount");
+            cpuCount ??= Semaphore.OpenExisting(ResourceSemaphoreName);
 
             coresToRelease = Math.Min(runningTotal, coresToRelease);
 
@@ -528,7 +529,7 @@ public void ReleaseCores(int coresToRelease)
 
         public void BlockingWaitForCore()
         {
-            cpuCount ??= Semaphore.OpenExisting("cpuCount");
+            cpuCount ??= Semaphore.OpenExisting(ResourceSemaphoreName);
 
             cpuCount.WaitOne();
         }
