diff --git a/src/Framework/CriticalTaskException.cs b/src/Framework/CriticalTaskException.cs
index 04e842e0a63..e372866c74a 100644
--- a/src/Framework/CriticalTaskException.cs
+++ b/src/Framework/CriticalTaskException.cs
@@ -10,6 +10,9 @@
 
 namespace Microsoft.Build.Framework
 {
+    /// <summary>
+    /// A wrapper exception for exceptions thrown by tasks (in TaskBuilder) that are critical to the build process.
+    /// </summary>
     internal sealed class CriticalTaskException : BuildExceptionBase
     {
         public CriticalTaskException(
