diff --git a/src/Build/Evaluation/Conditionals/Token.cs b/src/Build/Evaluation/Conditionals/Token.cs
index da2e6a79abf..daa990e5cdd 100644
--- a/src/Build/Evaluation/Conditionals/Token.cs
+++ b/src/Build/Evaluation/Conditionals/Token.cs
@@ -35,21 +35,26 @@ internal enum TokenType
             Comma,
             LeftParenthesis,
             RightParenthesis,
+
             LessThan,
             GreaterThan,
             LessThanOrEqualTo,
             GreaterThanOrEqualTo,
+
             And,
             Or,
+
             EqualTo,
             NotEqualTo,
             Not,
+
             Property,
             String,
             Numeric,
             ItemList,
             ItemMetadata,
             Function,
+
             EndOfInput,
         }
 
