diff --git a/src/Framework/StringBuilderCache.cs b/src/Framework/StringBuilderCache.cs
index 9861b36eb06..6ca6563823f 100644
--- a/src/Framework/StringBuilderCache.cs
+++ b/src/Framework/StringBuilderCache.cs
@@ -31,6 +31,8 @@
 **
 ===========================================================*/
 
+//#define ASSERT_BALANCE
+
 using System;
 using System.Diagnostics;
 using System.Text;
@@ -80,9 +82,11 @@ public static void Release(StringBuilder sb)
         {
             if (sb.Capacity <= MAX_BUILDER_SIZE)
             {
+#if ASSERT_BALANCE
                 // Assert we are not replacing another string builder. That could happen when Acquire is reentered.
                 // User of StringBuilderCache has to make sure that calling method call stacks do not also use StringBuilderCache.
                 Debug.Assert(StringBuilderCache.t_cachedInstance == null, "Unexpected replacing of other StringBuilder.");
+#endif
 
                 StringBuilderCache.t_cachedInstance = sb;
             }
