diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 609d7058795..13a77520ee4 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -37,7 +37,6 @@
 using Microsoft.Build.Shared.Debugging;
 using Microsoft.Build.Experimental;
 using Microsoft.Build.Framework.Telemetry;
-using System.Runtime.CompilerServices;
 using Microsoft.Build.Internal;
 using Microsoft.Build.Logging.LiveLogger;
 using System.Runtime.InteropServices;
@@ -837,10 +836,6 @@ public static ExitType Execute(
                     // if there was no need to start the build e.g. because /help was triggered
                     // do nothing
                 }
-
-                // The encoding may be changed to support non-en characters for environment variables set by external tools. We don't want to impact other programs on the console.
-                //Console.OutputEncoding = originalOutputEncoding;
-                //Console.InputEncoding = originalInputEncoding;
             }
             /**********************************************************************************************************************
              * WARNING: Do NOT add any more catch blocks below! Exceptions should be caught as close to their point of origin as
