diff --git a/src/Utilities/ToolLocationHelper.cs b/src/Utilities/ToolLocationHelper.cs
index 58bca6ee69a..baa7385f829 100644
--- a/src/Utilities/ToolLocationHelper.cs
+++ b/src/Utilities/ToolLocationHelper.cs
@@ -186,11 +186,13 @@ public enum DotNetFrameworkArchitecture
         /// <summary>
         /// Indicates the 32-bit .NET Framework
         /// </summary>
+        [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "Bitness", Justification = "Bitness is a reasonable term")]
         Bitness32 = 1,
 
         /// <summary>
         /// Indicates the 64-bit .NET Framework
         /// </summary>
+        [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "Bitness", Justification = "Bitness is a reasonable term")]
         Bitness64 = 2
     }
 
@@ -363,6 +365,7 @@ public static IList<AssemblyFoldersFromConfigInfo> GetAssemblyFoldersFromConfigI
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>IDictionary of installed SDKS and their location. K:SDKName V:SDK installation location</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IDictionary<string, string> GetPlatformExtensionSDKLocations(string targetPlatformIdentifier, Version targetPlatformVersion)
             => GetPlatformExtensionSDKLocations(null, null, targetPlatformIdentifier, targetPlatformVersion);
 
@@ -374,6 +377,7 @@ public static IDictionary<string, string> GetPlatformExtensionSDKLocations(strin
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>IDictionary of installed SDKS and their location. K:SDKName V:SDK installation location</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IDictionary<string, string> GetPlatformExtensionSDKLocations(string[] diskRoots, string registryRoot, string targetPlatformIdentifier, Version targetPlatformVersion)
             => GetPlatformExtensionSDKLocations(diskRoots, null, registryRoot, targetPlatformIdentifier, targetPlatformVersion);
 
@@ -386,6 +390,7 @@ public static IDictionary<string, string> GetPlatformExtensionSDKLocations(strin
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>IDictionary of installed SDKS and their location. K:SDKName V:SDK installation location</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IDictionary<string, string> GetPlatformExtensionSDKLocations(string[] diskRoots, string[] extensionDiskRoots, string registryRoot, string targetPlatformIdentifier, Version targetPlatformVersion)
         {
             var extensionSDKs = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
@@ -406,6 +411,7 @@ public static IDictionary<string, string> GetPlatformExtensionSDKLocations(strin
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>IDictionary of installed SDKS and their tuples containing (location, platform version).</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Casing kept to maintain consistency with existing APIs")]
         public static IDictionary<string, Tuple<string, string>> GetPlatformExtensionSDKLocationsAndVersions(string targetPlatformIdentifier, Version targetPlatformVersion)
             => GetPlatformExtensionSDKLocationsAndVersions(null, null, targetPlatformIdentifier, targetPlatformVersion);
 
@@ -417,6 +423,7 @@ public static IDictionary<string, Tuple<string, string>> GetPlatformExtensionSDK
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>IDictionary of installed SDKS and their tuples containing (location, platform version).</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Casing kept to maintain consistency with existing APIs")]
         public static IDictionary<string, Tuple<string, string>> GetPlatformExtensionSDKLocationsAndVersions(string[] diskRoots, string registryRoot, string targetPlatformIdentifier, Version targetPlatformVersion)
             => GetPlatformExtensionSDKLocationsAndVersions(diskRoots, null, registryRoot, targetPlatformIdentifier, targetPlatformVersion);
 
@@ -429,6 +436,7 @@ public static IDictionary<string, Tuple<string, string>> GetPlatformExtensionSDK
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>IDictionary of installed SDKS and their tuples containing (location, platform version). Version may be null if the SDK targets multiple versions.</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Casing kept to maintain consistency with existing APIs")]
         public static IDictionary<string, Tuple<string, string>> GetPlatformExtensionSDKLocationsAndVersions(string[] diskRoots, string[] multiPlatformDiskRoots, string registryRoot, string targetPlatformIdentifier, Version targetPlatformVersion)
         {
             var extensionSDKsAndVersions = new Dictionary<string, Tuple<string, string>>(StringComparer.OrdinalIgnoreCase);
@@ -476,6 +484,7 @@ private static IEnumerable<TargetPlatformSDK> GetTargetPlatformMonikers(string[]
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string targetPlatformIdentifier, Version targetPlatformVersion)
             => GetPlatformExtensionSDKLocation(sdkMoniker, targetPlatformIdentifier, targetPlatformVersion, null, null);
 
@@ -489,6 +498,7 @@ public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string t
         /// <param name="diskRoots">List of disk roots to search for sdks within</param>
         /// <param name="registryRoot">Registry root to look for sdks within</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string targetPlatformIdentifier, Version targetPlatformVersion, string[] diskRoots, string registryRoot)
             => GetPlatformExtensionSDKLocation(sdkMoniker, targetPlatformIdentifier, targetPlatformVersion, diskRoots, null, registryRoot);
 
@@ -503,6 +513,7 @@ public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string t
         /// <param name="extensionDiskRoots">List of disk roots to look for manifest driven extension sdks</param>
         /// <param name="registryRoot">Registry root to look for sdks within</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string targetPlatformIdentifier, Version targetPlatformVersion, string[] diskRoots, string[] extensionDiskRoots, string registryRoot)
         {
             ErrorUtilities.VerifyThrowArgumentLength(targetPlatformIdentifier, nameof(targetPlatformIdentifier));
@@ -539,6 +550,7 @@ public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string t
         /// <param name="targetPlatformIdentifier">Identifier for the platform</param>
         /// <param name="targetPlatformVersion">Version of the platform</param>
         /// <returns>A full path to the sdk root if the sdk exists in the targeted platform or an empty string if it does not exist.</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string targetPlatformIdentifier, string targetPlatformVersion)
             => GetPlatformExtensionSDKLocation(sdkMoniker, targetPlatformIdentifier, targetPlatformVersion, null, null);
 
@@ -552,6 +564,7 @@ public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string t
         /// <param name="diskRoots">List of disk roots to search for sdks within</param>
         /// <param name="registryRoot">Registry root to look for sdks within</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string targetPlatformIdentifier, string targetPlatformVersion, string diskRoots, string registryRoot)
             => GetPlatformExtensionSDKLocation(sdkMoniker, targetPlatformIdentifier, targetPlatformVersion, diskRoots, null, registryRoot);
 
@@ -566,6 +579,7 @@ public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string t
         /// <param name="extensionDiskRoots">List of disk roots to look for manifest driven extension sdks</param>
         /// <param name="registryRoot">Registry root to look for sdks within</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string targetPlatformIdentifier, string targetPlatformVersion, string diskRoots, string extensionDiskRoots, string registryRoot)
         {
             ErrorUtilities.VerifyThrowArgumentNull(targetPlatformVersion, nameof(targetPlatformVersion));
@@ -598,6 +612,7 @@ public static string GetPlatformExtensionSDKLocation(string sdkMoniker, string t
         /// <param name="targetPlatformVersion"></param>
         /// <param name="extensionSdks"></param>
         /// <returns>A IDictionary collection of filtered extension SDKs</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Not worth breaking customers")]
         public static IDictionary<string, string> FilterPlatformExtensionSDKs(Version targetPlatformVersion, IDictionary<string, string> extensionSdks)
         {
             var filteredExtensionSdks = new Dictionary<string, string>();
@@ -620,6 +635,7 @@ public static IDictionary<string, string> FilterPlatformExtensionSDKs(Version ta
         /// </summary>
         /// <param name="sdkRoot">Root folder for the SDK</param>
         /// <returns>A list of folders in the order which they should be used when looking for references in the SDK</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IList<string> GetSDKReferenceFolders(string sdkRoot) => GetSDKReferenceFolders(sdkRoot, retailConfigurationName, neutralArchitectureName);
 
         /// <summary>
@@ -630,6 +646,7 @@ public static IDictionary<string, string> FilterPlatformExtensionSDKs(Version ta
         /// <param name="targetConfiguration">The configuration the SDK is targeting. This should be Debug or Retail</param>
         /// <param name="targetArchitecture">The architecture the SDK is targeting</param>
         /// <returns>A list of folders in the order which they should be used when looking for references in the SDK</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IList<string> GetSDKReferenceFolders(string sdkRoot, string targetConfiguration, string targetArchitecture)
         {
             ErrorUtilities.VerifyThrowArgumentLength(sdkRoot, nameof(sdkRoot));
@@ -682,6 +699,7 @@ private static void AddSDKPaths(string sdkRoot, string folderName, string target
         /// </summary>
         /// <param name="sdkRoot">Root folder for the SDK must contain a redist folder</param>
         /// <returns>A list of folders in the order which they should be used when looking for redist files in the SDK</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IList<string> GetSDKRedistFolders(string sdkRoot) => GetSDKRedistFolders(sdkRoot, retailConfigurationName, neutralArchitectureName);
 
         /// <summary>
@@ -692,6 +710,7 @@ private static void AddSDKPaths(string sdkRoot, string folderName, string target
         /// <param name="targetConfiguration">The configuration the SDK is targeting. This should be Debug or Retail</param>
         /// <param name="targetArchitecture">The architecture the SDK is targeting</param>
         /// <returns>A list of folders in the order which they should be used when looking for redist files in the SDK</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IList<string> GetSDKRedistFolders(string sdkRoot, string targetConfiguration, string targetArchitecture)
         {
             ErrorUtilities.VerifyThrowArgumentLength(sdkRoot, nameof(sdkRoot));
@@ -710,6 +729,7 @@ public static IList<string> GetSDKRedistFolders(string sdkRoot, string targetCon
         /// </summary>
         /// <param name="sdkRoot">Root folder for the SDK must contain a Designtime folder</param>
         /// <returns>A list of folders in the order which they should be used when looking for DesignTime files in the SDK</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IList<string> GetSDKDesignTimeFolders(string sdkRoot) => GetSDKDesignTimeFolders(sdkRoot, retailConfigurationName, neutralArchitectureName);
 
         /// <summary>
@@ -720,6 +740,7 @@ public static IList<string> GetSDKRedistFolders(string sdkRoot, string targetCon
         /// <param name="targetConfiguration">The configuration the SDK is targeting. This should be Debug or Retail</param>
         /// <param name="targetArchitecture">The architecture the SDK is targeting</param>
         /// <returns>A list of folders in the order which they should be used when looking for DesignTime files in the SDK</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static IList<string> GetSDKDesignTimeFolders(string sdkRoot, string targetConfiguration, string targetArchitecture)
         {
             ErrorUtilities.VerifyThrowArgumentLength(sdkRoot, nameof(sdkRoot));
@@ -1225,6 +1246,7 @@ public static string GetSDKContentFolderPath(
         /// <param name="targetPlatformIdentifier">Targeted platform to find SDKs for</param>
         /// <param name="targetPlatformVersion">Targeted platform version to find SDKs for</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformSDKLocation(string targetPlatformIdentifier, Version targetPlatformVersion)
             => GetPlatformSDKLocation(targetPlatformIdentifier, targetPlatformVersion, null, null);
 
@@ -1236,6 +1258,7 @@ public static string GetPlatformSDKLocation(string targetPlatformIdentifier, Ver
         /// <param name="diskRoots">List of disk roots to search for sdks within</param>
         /// <param name="registryRoot">Registry root to look for sdks within</param>
         /// <returns>Location of the SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformSDKLocation(string targetPlatformIdentifier, Version targetPlatformVersion, string[] diskRoots, string registryRoot)
         {
             var targetPlatform = GetMatchingPlatformSDK(targetPlatformIdentifier, targetPlatformVersion, diskRoots, null, registryRoot);
@@ -1248,6 +1271,7 @@ public static string GetPlatformSDKLocation(string targetPlatformIdentifier, Ver
         /// <param name="targetPlatformIdentifier">Identifier for the platform</param>
         /// <param name="targetPlatformVersion">Version of the platform</param>
         /// <returns>A full path to the sdk root if the sdk exists in the targeted platform or an empty string if it does not exist.</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformSDKLocation(string targetPlatformIdentifier, string targetPlatformVersion)
             => GetPlatformSDKLocation(targetPlatformIdentifier, targetPlatformVersion, null, null);
 
@@ -1259,6 +1283,7 @@ public static string GetPlatformSDKLocation(string targetPlatformIdentifier, str
         /// <param name="diskRoots">List of disk roots to search for sdks within</param>
         /// <param name="registryRoot">Registry root to look for sdks within</param>
         /// <returns>Location of the platform SDK if it is found, empty string if it could not be found</returns>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static string GetPlatformSDKLocation(string targetPlatformIdentifier, string targetPlatformVersion, string diskRoots, string registryRoot)
         {
             ErrorUtilities.VerifyThrowArgumentNull(targetPlatformVersion, nameof(targetPlatformVersion));
@@ -2347,6 +2372,7 @@ private static IList<string> GetPathAndChainReferenceAssemblyLocations(string ta
         /// <summary>
         /// Clear out the appdomain wide cache of Platform and Extension SDKs.
         /// </summary>
+        [SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SDK", Justification = "Shipped this way in Dev11 Beta (go-live)")]
         public static void ClearSDKStaticCache()
         {
             lock (s_locker)
