diff --git a/eng/Packages.props b/eng/Packages.props
index 9d922d46242..521f4d5be95 100644
--- a/eng/Packages.props
+++ b/eng/Packages.props
@@ -47,7 +47,7 @@
     <PackageReference Update="xunit.core" Version="$(XUnitVersion)" />
   </ItemGroup>
 
-  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'netcoreapp2.1'))">
+  <ItemGroup Condition="($([MSBuild]::IsOSPlatform('osx')) and '$(MonoBuild)' == 'true') or $([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'netcoreapp2.1'))">
     <PackageReference Update="Microsoft.Extensions.DependencyModel" Version="5.0.0"/>
   </ItemGroup>
 
