diff --git a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs
index 5fd373aec73..ba95e84d703 100644
--- a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs
@@ -511,7 +511,7 @@ void ValidateNetHostSdkVersion(string path)
                 }
 
                 if (int.TryParse(lastDirectoryName.Substring(0, dotIndex), out int majorVersion)
-                    && majorVersion < minimumSdkVersion)
+                    && (majorVersion < minimumSdkVersion) || lastDirectoryName.Contains("preview"))
                 {
                     ErrorUtilities.ThrowInternalError($"SDK version {majorVersion} is below the minimum required version. {errorMessage}");
                 }
