diff --git a/eng/Signing.props b/eng/Signing.props
index 4b36a5a843a..3dadae19399 100644
--- a/eng/Signing.props
+++ b/eng/Signing.props
@@ -8,9 +8,9 @@
 
     <FileSignInfo Include="RuntimeContracts.dll" CertificateName="3PartySHA2" />
 
-    <FileSignInfo Include="OpenTelemetry.dll" CertificateName="None" />
-    <FileSignInfo Include="OpenTelemetry.Api.dll" CertificateName="None" />
-    <FileSignInfo Include="OpenTelemetry.Api.ProviderBuilderExtensions.dll" CertificateName="None" />
+    <FileSignInfo Include="OpenTelemetry.dll" CertificateName="3PartySHA2" />
+    <FileSignInfo Include="OpenTelemetry.Api.dll" CertificateName="3PartySHA2" />
+    <FileSignInfo Include="OpenTelemetry.Api.ProviderBuilderExtensions.dll" CertificateName="3PartySHA2" />
   </ItemGroup>
 
   <PropertyGroup>
