diff --git a/src/Framework/NativeMethods.cs b/src/Framework/NativeMethods.cs
index a882115a631..85f940db852 100644
--- a/src/Framework/NativeMethods.cs
+++ b/src/Framework/NativeMethods.cs
@@ -1712,7 +1712,6 @@ out FILETIME lpLastWriteTime
     internal static extern bool CreateSymbolicLink(string symLinkFileName, string targetFileName, SymbolicLink dwFlags);
 
     [DllImport("libc", SetLastError = true)]
-    [SupportedOSPlatform("linux")]
     internal static extern int symlink(string oldpath, string newpath);
 
     #endregion
