diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 1b5ce0283b4..186d984470f 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1367,7 +1367,7 @@ private static bool PrintTargets(string projectFile, string toolsVersion, Dictio
                     MessageImportance.Low),
             };
 
-            if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("MSBuildDebugEngine")))
+            if (Traits.Instance.DebugEngine)
             {
                 messages.Add(
                     new BuildManager.DeferredBuildMessage(
