diff --git a/src/Tasks/Delete.cs b/src/Tasks/Delete.cs
index 8cf304d4729..55e935ee475 100644
--- a/src/Tasks/Delete.cs
+++ b/src/Tasks/Delete.cs
@@ -125,9 +125,11 @@ public override bool Execute()
                             {
                                 Log.LogWarningFromResources("Delete.DeletingFile", file.ItemSpec);
                             }
-
-                            // Do not log a fake command line as well, as it's superfluous, and also potentially expensive
-                            Log.LogMessageFromResources(MessageImportance.Normal, "Delete.DeletingFile", file.ItemSpec);
+                            else
+                            {
+                                // Do not log a fake command line as well, as it's superfluous, and also potentially expensive
+                                Log.LogMessageFromResources(MessageImportance.Normal, "Delete.DeletingFile", file.ItemSpec);
+                            }
 
                             File.Delete(file.ItemSpec);
                         }
