diff --git a/src/MSBuildTaskHost/Immutable/ImmutableDictionary.cs b/src/MSBuildTaskHost/Immutable/ImmutableDictionary.cs
index 9c16c9505b6..9fa606eedf9 100644
--- a/src/MSBuildTaskHost/Immutable/ImmutableDictionary.cs
+++ b/src/MSBuildTaskHost/Immutable/ImmutableDictionary.cs
@@ -145,6 +145,8 @@ internal ImmutableDictionary<K, V> SetItems(IEnumerable<KeyValuePair<K, V>> item
             {
                 clone[item.Key] = item.Value;
             }
+
+            return clone;
         }
 
         internal ImmutableDictionary<K, V> Remove(K key)
