diff --git a/src/MSBuild.UnitTests/MSBuildServer_Tests.cs b/src/MSBuild.UnitTests/MSBuildServer_Tests.cs
index 2914de50e73..8b7feae9a83 100644
--- a/src/MSBuild.UnitTests/MSBuildServer_Tests.cs
+++ b/src/MSBuild.UnitTests/MSBuildServer_Tests.cs
@@ -4,6 +4,7 @@
 using System;
 using System.Diagnostics;
 using System.Reflection;
+using System.Text.RegularExpressions;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Build.Framework;
@@ -251,9 +252,9 @@ public void MultiProcBuildOnServer()
 
         private int ParseNumber(string searchString, string toFind)
         {
-            int indexOfId = searchString.IndexOf(toFind) + toFind.Length;
-            int endOfId = searchString.IndexOfAny(new char[] { '\r', '\n' }, indexOfId);
-            return int.Parse(searchString.Substring(indexOfId, endOfId - indexOfId));
+            Regex regex = new(@$"{toFind}(\d+)");
+            var x = regex.Match(searchString);
+            return int.Parse(x.Groups[1].Value);
         }
     }
 }
