diff --git a/src/Tasks/IVbcHostObject5.cs b/src/Tasks/IVbcHostObject5.cs
index 75b483b17cb..bbf959e4f68 100644
--- a/src/Tasks/IVbcHostObject5.cs
+++ b/src/Tasks/IVbcHostObject5.cs
@@ -20,7 +20,7 @@ public interface IVbcHostObject5 : IVbcHostObject4
     {
         IVbcHostObjectFreeThreaded GetFreeThreadedHostObject();
         [PreserveSig]
-        public int CompileAsync(out IntPtr buildSucceededEvent, out IntPtr buildFailedEvent);
+        int CompileAsync(out IntPtr buildSucceededEvent, out IntPtr buildFailedEvent);
         [PreserveSig]
         int EndCompile(bool buildSuccess);
 
