diff --git a/eng/Versions.props b/eng/Versions.props
index 91576299ed7..e4f44a8a029 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -17,6 +17,7 @@
   <PropertyGroup>
     <VersionPrefix>16.4.0</VersionPrefix>
     <AssemblyVersion>15.1.0.0</AssemblyVersion>
+    <DotNetFinalVersionKind>release</DotNetFinalVersionKind>
     <PreReleaseVersionLabel>preview</PreReleaseVersionLabel>
     <DotNetUseShippingVersions>true</DotNetUseShippingVersions>
     <!-- Workaround for https://github.com/dotnet/roslyn/issues/35793 -->
