diff --git a/eng/Packages.props b/eng/Packages.props
index d2d7d223faa..d6917bd7207 100644
--- a/eng/Packages.props
+++ b/eng/Packages.props
@@ -1,6 +1,6 @@
 <Project>
   <!--
-    Make sure to update the binding redirects (in src\MSBuild\app.config) for any changes to
+    Make sure to update the binding redirects (in src\MSBuild\app.config and src\MSBuild\app.amd64.config) for any changes to
     the list of assemblies redistributed by MSBuild (non-MSBuild assemblies in the .vsix package).
      -->
   <ItemGroup>
