diff --git a/src/Tasks.UnitTests/XmlPeek_Tests.cs b/src/Tasks.UnitTests/XmlPeek_Tests.cs
index 6ce9601852d..d7435d0a714 100644
--- a/src/Tasks.UnitTests/XmlPeek_Tests.cs
+++ b/src/Tasks.UnitTests/XmlPeek_Tests.cs
@@ -4,6 +4,7 @@
 using System;
 using System.IO;
 
+using Microsoft.Build.Evaluation;
 using Microsoft.Build.Tasks;
 using Microsoft.Build.Utilities;
 
@@ -323,12 +324,12 @@ public void PeekWithoutUsingTask()
         [Fact]
         public void PeekWithNoParameters()
         {
-            MockEngine engine = new(true);
+            MockLogger log = new();
+            Project project = ObjectModelHelpers.CreateInMemoryProject(@"<Project><Target Name=""Test""><XmlPeek /></Target></Project>", log);
 
-            XmlPeek task = new() { BuildEngine = engine };
-
-            task.Execute().ShouldBeFalse();
-            engine.Log.ShouldContain("MSB4044");
+            project.Build().ShouldBeFalse();
+            log.AssertLogContains("MSB4044");
+            log.AssertLogContains("\"Query\"");
         }
 
         private void Prepare(string xmlFile, out string xmlInputPath)
diff --git a/src/Tasks.UnitTests/XmlPoke_Tests.cs b/src/Tasks.UnitTests/XmlPoke_Tests.cs
index 10459f2caa4..f0978f95c48 100644
--- a/src/Tasks.UnitTests/XmlPoke_Tests.cs
+++ b/src/Tasks.UnitTests/XmlPoke_Tests.cs
@@ -7,6 +7,7 @@
 using System.Linq;
 using System.Xml;
 
+using Microsoft.Build.Evaluation;
 using Microsoft.Build.Tasks;
 using Microsoft.Build.Utilities;
 
@@ -140,45 +141,37 @@ public void PokeAttributeWithCondition()
         [Fact]
         public void PokeWithNoParameters()
         {
-            MockEngine engine = new(true);
-
-            XmlPoke task = new() { BuildEngine = engine };
+            MockLogger log = new();
+            Project project = ObjectModelHelpers.CreateInMemoryProject(@"<Project><Target Name=""Test""><XmlPoke /></Target></Project>", log);
 
-            task.Execute().ShouldBeFalse();
-            engine.Log.ShouldContain("MSB4044");
+            project.Build().ShouldBeFalse();
+            log.AssertLogContains("MSB4044");
         }
 
         [Fact]
         public void PokeWithMissingRequiredQuery()
         {
-            MockEngine engine = new(true);
-            Prepare(_xmlFileNoNs, out string xmlInputPath);
+            const string projectContent = @"<Project><Target Name=""Test""><XmlPoke XmlInputPath=""nonesuch"" /></Target></Project>";
 
-            XmlPoke task = new()
-            {
-                BuildEngine = engine,
-                XmlInputPath = new TaskItem(xmlInputPath),
-            };
+            MockLogger log = new();
+            Project project = ObjectModelHelpers.CreateInMemoryProject(projectContent, log);
 
-            task.Execute().ShouldBeFalse();
-            engine.Log.ShouldContain("MSB4044");
-            engine.Log.ShouldContain("Query");
+            project.Build().ShouldBeFalse();
+            log.AssertLogContains("MSB4044");
+            log.AssertLogContains("\"Query\"");
         }
 
         [Fact]
         public void PokeWithMissingRequiredXmlInputPath()
         {
-            MockEngine engine = new(true);
+            const string projectContent = @"<Project><Target Name=""Test""><XmlPoke Query=""nonesuch"" /></Target></Project>";
 
-            XmlPoke task = new()
-            {
-                BuildEngine = engine,
-                Query = "//variable/@Name",
-            };
+            MockLogger log = new();
+            Project project = ObjectModelHelpers.CreateInMemoryProject(projectContent, log);
 
-            task.Execute().ShouldBeFalse();
-            engine.Log.ShouldContain("MSB4044");
-            engine.Log.ShouldContain("XmlInputPath");
+            project.Build().ShouldBeFalse();
+            log.AssertLogContains("MSB4044");
+            log.AssertLogContains("\"XmlInputPath\"");
         }
 
         [Fact]
