diff --git a/src/Framework/NativeMethods.cs b/src/Framework/NativeMethods.cs
index 78b5a184bd5..df945ce4a04 100644
--- a/src/Framework/NativeMethods.cs
+++ b/src/Framework/NativeMethods.cs
@@ -1101,6 +1101,7 @@ internal static bool IsSymLink(FileInfo fileInfo)
 #endif
     }
 
+    [SupportedOSPlatform("windows")]
     private static bool IsSymLinkFileInternal(string path)
     {
         using SafeFileHandle handle = CreateFile(path,
