diff --git a/src/Build/Construction/Solution/SolutionFile.cs b/src/Build/Construction/Solution/SolutionFile.cs
index 2fdef86e4c8..5b1f862c87a 100644
--- a/src/Build/Construction/Solution/SolutionFile.cs
+++ b/src/Build/Construction/Solution/SolutionFile.cs
@@ -164,20 +164,7 @@ internal SolutionFile()
         /// <summary>
         /// Returns Visual Studio major version
         /// </summary>
-        internal int VisualStudioVersion
-        {
-            get
-            {
-                if (_currentVisualStudioVersion != null)
-                {
-                    return _currentVisualStudioVersion.Major;
-                }
-                else
-                {
-                    return Version - 1;
-                }
-            }
-        }
+        internal int VisualStudioVersion => _currentVisualStudioVersion?.Major ?? Version - 1;
 
         /// <summary>
         /// Returns true if the solution contains any web projects
