diff --git a/src/Tasks/Microsoft.CSharp.CurrentVersion.targets b/src/Tasks/Microsoft.CSharp.CurrentVersion.targets
index b28ab078b60..4e348de04aa 100644
--- a/src/Tasks/Microsoft.CSharp.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.CSharp.CurrentVersion.targets
@@ -247,13 +247,13 @@ Copyright (C) Microsoft Corporation. All rights reserved.
          ChecksumAlgorithm="$(ChecksumAlgorithm)"
          CodeAnalysisRuleSet="$(ResolvedCodeAnalysisRuleSet)"
          CodePage="$(CodePage)"
-         DebugType="$(DebugType)"
+         DebugType="none"
          DefineConstants="$(DefineConstants)"
          DelaySign="$(DelaySign)"
          DisabledWarnings="$(NoWarn)"
          DisableSdkPath="$(DisableSdkPath)"
          DocumentationFile="@(DocFileItem)"
-         EmbedAllSources="$(EmbedAllSources)"
+         EmbedAllSources="false"
          EmbeddedFiles="@(EmbeddedFiles)"
          EmitDebugInformation="$(DebugSymbols)"
          EnvironmentVariables="$(CscEnvironment)"
@@ -281,23 +281,23 @@ Copyright (C) Microsoft Corporation. All rights reserved.
          Optimize="$(Optimize)"
          Deterministic="$(Deterministic)"
          PublicSign="$(PublicSign)"
-         OutputAssembly="@(IntermediateAssembly)"
-         OutputRefAssembly="@(IntermediateRefAssembly)"
-         PdbFile="$(PdbFile)"
+         OutputAssembly="@(XamlIntermediateAssembly)"
+         OutputRefAssembly=""
+         PdbFile=""
          Platform="$(PlatformTarget)"
          Prefer32Bit="$(Prefer32Bit)"
          PreferredUILang="$(PreferredUILang)"
          ProjectName="$(MSBuildProjectName)"
          ProvideCommandLineArgs="$(ProvideCommandLineArgs)"
-         References="@(ReferencePathWithRefAssemblies)"
-         RefOnly="$(ProduceOnlyReferenceAssembly)"
+         References="@(ReferencePath)"
+         RefOnly="false"
          ReportAnalyzer="$(ReportAnalyzer)"
          ReportIVTs="$(ReportIVTs)"
          Resources="@(_CoreCompileResourceInputs);@(CompiledLicenseFile)"
          ResponseFiles="$(CompilerResponseFile)"
          RuntimeMetadataVersion="$(RuntimeMetadataVersion)"
          SharedCompilationId="$(SharedCompilationId)"
-         SkipAnalyzers="$(_SkipAnalyzers)"
+         SkipAnalyzers="true"
          SkipCompilerExecution="$(SkipCompilerExecution)"
          Sources="@(Compile)"
          SubsystemVersion="$(SubsystemVersion)"
@@ -317,9 +317,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
          Win32Manifest="$(Win32Manifest)"
          Win32Resource="$(Win32Resource)"
          PathMap="$(PathMap)"
-         SourceLink="$(SourceLink)">
-      <Output TaskParameter="CommandLineArgs" ItemName="CscCommandLineArgs" />
-    </Csc>
+         SourceLink="$(SourceLink)" />
 
 <!-- Only Applicable to the regular CoreCompile:
               <ItemGroup>
diff --git a/src/Tasks/Microsoft.VisualBasic.CurrentVersion.targets b/src/Tasks/Microsoft.VisualBasic.CurrentVersion.targets
index fe8768fabec..1e64f51125c 100644
--- a/src/Tasks/Microsoft.VisualBasic.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.VisualBasic.CurrentVersion.targets
@@ -237,13 +237,13 @@ Copyright (C) Microsoft Corporation. All rights reserved.
          ChecksumAlgorithm="$(ChecksumAlgorithm)"
          CodeAnalysisRuleSet="$(ResolvedCodeAnalysisRuleSet)"
          CodePage="$(CodePage)"
-         DebugType="$(DebugType)"
+         DebugType="none"
          DefineConstants="$(FinalDefineConstants)"
          DelaySign="$(DelaySign)"
          DisableSdkPath="$(DisableSdkPath)"
          DisabledWarnings="$(NoWarn)"
          DocumentationFile="@(DocFileItem)"
-         EmbedAllSources="$(EmbedAllSources)"
+         EmbedAllSources="false"
          EmbeddedFiles="@(EmbeddedFiles)"
          EmitDebugInformation="$(DebugSymbols)"
          EnvironmentVariables="$(VbcEnvironment)"
@@ -275,16 +275,16 @@ Copyright (C) Microsoft Corporation. All rights reserved.
          OptionInfer="$(OptionInfer)"
          OptionStrict="$(OptionStrict)"
          OptionStrictType="$(OptionStrictType)"
-         OutputAssembly="@(IntermediateAssembly)"
-         OutputRefAssembly="@(IntermediateRefAssembly)"
-         PdbFile="$(PdbFile)"
+         OutputAssembly="@(XamlIntermediateAssembly)"
+         OutputRefAssembly=""
+         PdbFile=""
          Platform="$(PlatformTarget)"
          Prefer32Bit="$(Prefer32Bit)"
          PreferredUILang="$(PreferredUILang)"
          ProjectName="$(MSBuildProjectName)"
          ProvideCommandLineArgs="$(ProvideCommandLineArgs)"
-         References="@(ReferencePathWithRefAssemblies)"
-         RefOnly="$(ProduceOnlyReferenceAssembly)"
+         References="@(ReferencePath)"
+         RefOnly="false"
          RemoveIntegerChecks="$(RemoveIntegerChecks)"
          ReportAnalyzer="$(ReportAnalyzer)"
          ReportIVTs="$(ReportIVTs)"
@@ -317,9 +317,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
          Win32Resource="$(Win32Resource)"
          VBRuntime="$(VBRuntime)"
          PathMap="$(PathMap)"
-         SourceLink="$(SourceLink)">
-      <Output TaskParameter="CommandLineArgs" ItemName="VbcCommandLineArgs" />
-    </Vbc>
+         SourceLink="$(SourceLink)" />
 
  <!-- Only Applicable to the regular CoreCompile:
               <ItemGroup>
