diff --git a/src/Framework/MSBuildEventSource.cs b/src/Framework/MSBuildEventSource.cs
index 5ba9fa07883..cf0fd17fa59 100644
--- a/src/Framework/MSBuildEventSource.cs
+++ b/src/Framework/MSBuildEventSource.cs
@@ -220,15 +220,15 @@ public void GenerateResourceOverallStop()
         }
 
         [Event(27, Keywords = Keywords.All | Keywords.PerformanceLog)]
-        public void RarOverallStart()
+        public void RarOverallStart(int assembliesCount, int assemblyFilesCount, bool findDependencies)
         {
-            WriteEvent(27);
+            WriteEvent(27, assembliesCount, assemblyFilesCount, findDependencies);
         }
 
         [Event(28, Keywords = Keywords.All | Keywords.PerformanceLog)]
-        public void RarOverallStop()
+        public void RarOverallStop(int resolvedFilesCount, int resolvedDependencyFilesCount, int copyLocalFilesCount)
         {
-            WriteEvent(28);
+            WriteEvent(28, resolvedFilesCount, resolvedDependencyFilesCount, copyLocalFilesCount);
         }
 
         /// <summary>
diff --git a/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs b/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs
index 169eebec17a..dd96ef8f9d1 100644
--- a/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs
+++ b/src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs
@@ -2109,7 +2109,7 @@ ReadMachineTypeFromPEHeader readMachineTypeFromPEHeader
         )
         {
             bool success = true;
-            MSBuildEventSource.Log.RarOverallStart();
+            MSBuildEventSource.Log.RarOverallStart(_assemblyNames.Length, _assemblyFiles.Length, _findDependencies);
             {
                 try
                 {
@@ -2579,7 +2579,7 @@ out _copyLocalFiles
                             }
                         }
                     }
-                    MSBuildEventSource.Log.RarOverallStop();
+                    MSBuildEventSource.Log.RarOverallStop(_resolvedFiles.Length, _resolvedDependencyFiles.Length, _copyLocalFiles.Length);
                     return success && !Log.HasLoggedErrors;
                 }
                 catch (ArgumentException e)
@@ -2596,7 +2596,7 @@ out _copyLocalFiles
                 }
             }
 
-            MSBuildEventSource.Log.RarOverallStop();
+            MSBuildEventSource.Log.RarOverallStop(_resolvedFiles.Length, _resolvedDependencyFiles.Length, _copyLocalFiles.Length);
 
             return success && !Log.HasLoggedErrors;
         }
