diff --git a/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj b/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj
index 27bf2a1542b..7e26e8ebb45 100644
--- a/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj
+++ b/src/BuildCheck.UnitTests/Microsoft.Build.BuildCheck.UnitTests.csproj
@@ -15,6 +15,10 @@
     <ProjectReference Include="..\UnitTests.Shared\Microsoft.Build.UnitTests.Shared.csproj" />
     <ProjectReference Include="..\Xunit.NetCore.Extensions\Xunit.NetCore.Extensions.csproj" />
   </ItemGroup>
+
+  <PropertyGroup>
+    <ArtifactsNonShippingPackagesDir>$(ArtifactsBinDir)Microsoft.Build.BuildCheck.UnitTests\CustomChecks</ArtifactsNonShippingPackagesDir>
+  </PropertyGroup>
   
   <ItemGroup Label="TestAssests">
     <ProjectReference Include=".\TestAssets\CustomCheck\CustomCheck.csproj" />
@@ -46,4 +50,10 @@
     </None>
   </ItemGroup>
 
+  <Target Name="MSBuildPackagesForTests" AfterTargets="Build">
+    <Exec Command="dotnet pack &quot;..\Build\Microsoft.Build.csproj&quot; -o &quot;$(ArtifactsNonShippingPackagesDir)&quot;" />
+    <Exec Command="dotnet pack &quot;..\Framework\Microsoft.Build.Framework.csproj&quot; -o &quot;$(ArtifactsNonShippingPackagesDir)&quot;"/>
+    <Exec Command="dotnet pack &quot;..\StringTools\StringTools.csproj&quot; -o &quot;$(ArtifactsNonShippingPackagesDir)&quot;"/>
+  </Target>
+
 </Project>
diff --git a/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props b/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props
index d79475be5f0..989cfa9c609 100644
--- a/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props
+++ b/src/BuildCheck.UnitTests/TestAssets/Common/CommonTest.props
@@ -11,7 +11,7 @@
 
   <!-- In the real world scenario, the dependencies are added as Nuget PackageReference, modified for test purposes only. -->
   <ItemGroup>
-    <ProjectReference Include="..\..\..\Build\Microsoft.Build.csproj" IncludeInPackage="true" />
+    <ProjectReference Include="..\..\..\Build\Microsoft.Build.csproj" />
   </ItemGroup>
 
 </Project>
