diff --git a/.editorconfig b/.editorconfig
index 7b0f1419bb8..58a8ef2c257 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -409,3 +409,6 @@ dotnet_diagnostic.IDE0290.severity = suggestion
 dotnet_diagnostic.IDE0300.severity = suggestion
 dotnet_diagnostic.IDE0301.severity = suggestion
 dotnet_diagnostic.IDE0305.severity = suggestion
+
+# Temporarily disable SA1010 "Opening square brackets should not be preceded by a space" until https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3687 if fixed
+dotnet_diagnostic.SA1010.severity = none
