diff --git a/src/Build/BackEnd/Client/MSBuildClient.cs b/src/Build/BackEnd/Client/MSBuildClient.cs
index 7520bd02272..8f2d33a8d56 100644
--- a/src/Build/BackEnd/Client/MSBuildClient.cs
+++ b/src/Build/BackEnd/Client/MSBuildClient.cs
@@ -312,10 +312,7 @@ private ServerNodeBuildCommand GetServerNodeBuildCommand(string commandLine)
 
         private ServerNodeHandshake GetHandshake()
         {
-            return new ServerNodeHandshake(
-                CommunicationsUtilities.GetHandshakeOptions(taskHost: false, architectureFlagToSet: XMakeAttributes.GetCurrentMSBuildArchitecture()),
-                string.IsNullOrEmpty(_dllLocation) ? _exeLocation : _dllLocation
-            );
+            return new ServerNodeHandshake(CommunicationsUtilities.GetHandshakeOptions(taskHost: false, architectureFlagToSet: XMakeAttributes.GetCurrentMSBuildArchitecture()));
         }
 
         /// <summary>
@@ -400,7 +397,7 @@ private bool TryConnectToServer(int timeout)
                 }
 
                 // This indicates that we have finished all the parts of our handshake; hopefully the endpoint has as well.
-                _nodeStream.WriteIntForHandshake(ServerNodeHandshake.EndOfHandshakeSignal);
+                _nodeStream.WriteEndOfHandshakeSignal();
 
                 CommunicationsUtilities.Trace("Reading handshake from pipe {0}", _pipeName);
 
diff --git a/src/MSBuild/MSBuildClientApp.cs b/src/MSBuild/MSBuildClientApp.cs
index a029b19bc64..de83c3a10e3 100644
--- a/src/MSBuild/MSBuildClientApp.cs
+++ b/src/MSBuild/MSBuildClientApp.cs
@@ -51,7 +51,7 @@ CancellationToken cancellationToken
             if (!NativeMethodsShared.IsMono)
             {
                 // _exeFileLocation consists the msbuild dll instead.
-                dllLocation = BuildEnvironmentHelper.Instance.CurrentMSBuildExePath;;
+                dllLocation = BuildEnvironmentHelper.Instance.CurrentMSBuildExePath;
                 exeLocation = GetCurrentHost();
             }
             else
