diff --git a/src/Build/Construction/Solution/SolutionFile.cs b/src/Build/Construction/Solution/SolutionFile.cs
index 68892206db0..63ce5b9dcee 100644
--- a/src/Build/Construction/Solution/SolutionFile.cs
+++ b/src/Build/Construction/Solution/SolutionFile.cs
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
@@ -322,7 +322,7 @@ internal void ParseUsingNewParser()
                     ProjectFileErrorUtilities.ThrowInvalidProjectFile(
                         new BuildEventFileInfo(errorLocation),
                         "InvalidProjectFile",
-                        solutionEx.Message);
+                        solutionEx.ToString());
                 }
                 catch (Exception ex)
                 {
