diff --git a/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs b/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
index c5c90d7a549..9a0c8575efa 100644
--- a/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
@@ -1517,8 +1517,9 @@ public void CancelledBuildWithUnexecutedSubmission()
         /// A canceled build
         /// </summary>
         [Fact(Timeout = 20_000)]
-        public void CancelledBuild()
+        public async void CancelledBuild()
         {
+            
             Console.WriteLine("Starting CancelledBuild test that is known to hang.");
             string contents = CleanupFileContents(@"
 <Project xmlns='msbuildnamespace' ToolsVersion='msbuilddefaulttoolsversion'>
@@ -1528,7 +1529,6 @@ public void CancelledBuild()
  </Target>
 </Project>
 ");
-
             BuildParameters parameters = new()
             {
                 ShutdownInProcNodeOnBuildFinish = true,
@@ -1545,7 +1545,6 @@ public void CancelledBuild()
             BuildSubmission asyncResult = _buildManager.PendBuildRequest(data);
             Console.WriteLine("CancelledBuild: pend build returned");
 
-
             asyncResult.ExecuteAsync(null, null);
             Console.WriteLine("CancelledBuild: ExecuteAsync called");
             _buildManager.CancelAllSubmissions();
