diff --git a/src/MSBuild/TerminalLogger/TerminalLogger.cs b/src/MSBuild/TerminalLogger/TerminalLogger.cs
index 47e9b56e7ee..19fff3c69ed 100644
--- a/src/MSBuild/TerminalLogger/TerminalLogger.cs
+++ b/src/MSBuild/TerminalLogger/TerminalLogger.cs
@@ -1032,7 +1032,7 @@ private static void WrapText(StringBuilder sb, string text, int maxLength, strin
             int length = Math.Min(maxLength - indent.Length, text.Length - start);
             sb.AppendLine();
             sb.Append(indent);
-            sb.Append(text.Substring(start, length));
+            sb.Append(text.AsSpan().Slice(start, length));
 
             start += length;
         }
