diff --git a/src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs b/src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs
index 0be03427f20..59553953c79 100644
--- a/src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs
+++ b/src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs
@@ -20,7 +20,7 @@ public byte[] GetAllBytes()
         {
             if (_stream == null)
             {
-                return Array.Empty<byte>;
+                return Array.Empty<byte>();
             }
             else if (ArchiveFilePath == null)
             {
