diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 58d0e1901ce..db9b8bf844b 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2534,16 +2534,23 @@ private static bool ProcessTerminalLoggerConfiguration(CommandLineSwitches comma
             }
             else
             {
+                // Keep MSBUILDLIVELOGGER supporitng existing use. But MSBUILDTERMINALLOGGER takes precedence.
+                string liveLoggerArg = Environment.GetEnvironmentVariable("MSBUILDLIVELOGGER");
                 terminalloggerArg = Environment.GetEnvironmentVariable("MSBUILDTERMINALLOGGER");
-
-                if (string.IsNullOrWhiteSpace(terminalloggerArg))
+                if (!string.IsNullOrEmpty(terminalloggerArg))
                 {
-                    return false;
+                    s_globalMessagesToLogInBuildLoggers.Add(
+                        new BuildManager.DeferredBuildMessage($"The environment variable MSBUILDTERMINALLOGGER was set to {terminalloggerArg}.", MessageImportance.Low));
                 }
-                else
+                else if (!string.IsNullOrEmpty(liveLoggerArg))
                 {
+                    terminalloggerArg = liveLoggerArg;
                     s_globalMessagesToLogInBuildLoggers.Add(
-                        new BuildManager.DeferredBuildMessage($"The environment variable MSBUILDTERMINALLOGGER was set to {terminalloggerArg}.", MessageImportance.Low));
+                        new BuildManager.DeferredBuildMessage($"The environment variable MSBUILDLIVELOGGER was set to {liveLoggerArg}.", MessageImportance.Low));
+                }
+                else
+                {
+                    return false;
                 }
             }
 
