diff --git a/src/Tasks/Microsoft.Common.CrossTargeting.targets b/src/Tasks/Microsoft.Common.CrossTargeting.targets
index 6c16b460c23..1476da694bd 100644
--- a/src/Tasks/Microsoft.Common.CrossTargeting.targets
+++ b/src/Tasks/Microsoft.Common.CrossTargeting.targets
@@ -26,9 +26,9 @@ Copyright (C) Microsoft Corporation. All rights reserved.
           Returns="@(_ThisProjectPlatformMetadata)">
     <ItemGroup>
       <_ThisProjectPlatformMetadata Include="$(MSBuildProjectFullPath)">
-        <PlatformOptions>$(Platform),$(Platforms),$(PlatformTarget)</PlatformOptions>
+        <PlatformOptions>$(Platforms)</PlatformOptions>
         <CanMultiPlatform>false</CanMultiPlatform>
-        <CanMultiPlatform Condition="'$(Platforms)' != ''">true</CanMultiPlatform>
+        <CanMultiPlatform Condition="$(Platforms.Contains(',')) == 'true'">true</CanMultiPlatform>
       </_ThisProjectPlatformMetadata>
     </ItemGroup>
   </Target>
diff --git a/src/Tasks/Microsoft.Common.CurrentVersion.targets b/src/Tasks/Microsoft.Common.CurrentVersion.targets
index c381610b89d..be2596010c4 100644
--- a/src/Tasks/Microsoft.Common.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.Common.CurrentVersion.targets
@@ -1917,7 +1917,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
           Returns="@(_ThisProjectPlatformMetadata)">
     <ItemGroup>
       <_ThisProjectPlatformMetadata Include="$(MSBuildProjectFullPath)">
-        <PlatformOptions>$(Platform),$(Platforms),$(PlatformTarget)</PlatformOptions>
+        <PlatformOptions>$(Platforms)</PlatformOptions>
         <CanMultiPlatform>false</CanMultiPlatform>
         <CanMultiPlatform Condition="$(Platforms.Contains(',')) == 'true'">true</CanMultiPlatform>
       </_ThisProjectPlatformMetadata>
