diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index ba417e063f2..a6fa1148e68 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -999,7 +999,7 @@ dotnet_diagnostic.SA1508.severity = suggestion
 dotnet_diagnostic.SA1509.severity = suggestion
 
 # 'else' statement should not be preceded by a blank line
-dotnet_diagnostic.SA1510.severity = suggestion
+dotnet_diagnostic.SA1510.severity = warning
 
 # Single-line comments should not be followed by blank line
 dotnet_diagnostic.SA1512.severity = suggestion
