diff --git a/src/Build/Resources/Constants.cs b/src/Build/Resources/Constants.cs
index 7aa43ef0dd4..fc992aef817 100644
--- a/src/Build/Resources/Constants.cs
+++ b/src/Build/Resources/Constants.cs
@@ -325,6 +325,7 @@ private static void InitializeAvailableMethods()
                         availableStaticMethods.TryAdd("System.IO.Directory::GetLastWriteTime", directoryType);
                         availableStaticMethods.TryAdd("System.IO.Directory::GetParent", directoryType);
                         availableStaticMethods.TryAdd("System.IO.Directory::Exists", directoryType);
+
                         availableStaticMethods.TryAdd("System.IO.File::Exists", fileType);
                         availableStaticMethods.TryAdd("System.IO.File::GetCreationTime", fileType);
                         availableStaticMethods.TryAdd("System.IO.File::GetAttributes", fileType);
