diff --git a/src/BuildCheck.UnitTests/EndToEndTests.cs b/src/BuildCheck.UnitTests/EndToEndTests.cs
index b63084d16fd..7a77cf746a1 100644
--- a/src/BuildCheck.UnitTests/EndToEndTests.cs
+++ b/src/BuildCheck.UnitTests/EndToEndTests.cs
@@ -401,14 +401,6 @@ void ReplaceStringInFile(string filePath, string original, string replacement)
         }
     }
 
-    private static int GetWarningsCount(string output)
-    {
-        Regex regex = new Regex(@"(\d+) Warning\(s\)");
-        Match match = regex.Match(output);
-        match.Success.ShouldBeTrue("Expected Warnings section not found in the build output.");
-        return int.Parse(match.Groups[1].Value);
-    }
-
 
     [Fact]
     public void ConfigChangeReflectedOnReuse()
