diff --git a/src/Build/BackEnd/Components/Logging/LoggingService.cs b/src/Build/BackEnd/Components/Logging/LoggingService.cs
index 18abf8c13d3..bd2bba28490 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -1750,7 +1750,7 @@ private bool WarningAsErrorNotOverriden(BuildWarningEventArgs warningEvent)
         {
             int key = GetWarningsAsErrorOrMessageKey(warningEvent);
 
-            return WarningsNotAsErrors?.Contains(warningEvent.Code) != true && !(_warningsNotAsErrorsByProject.TryGetValue(key, out ISet<string> notToError) && notToError.Contains(warningEvent.Code));
+            return WarningsNotAsErrors?.Contains(warningEvent.Code) != true && !(_warningsNotAsErrorsByProject?.TryGetValue(key, out ISet<string> notToError) == true && notToError.Contains(warningEvent.Code));
         }
 
         /// <summary>
