diff --git a/src/Build.UnitTests/ChangeWaves_Tests.cs b/src/Build.UnitTests/ChangeWaves_Tests.cs
index a10893b1567..78fddbcce0f 100644
--- a/src/Build.UnitTests/ChangeWaves_Tests.cs
+++ b/src/Build.UnitTests/ChangeWaves_Tests.cs
@@ -3,10 +3,10 @@
 
 using Xunit;
 using Shouldly;
+using Microsoft.Build.Evaluation;
 using Microsoft.Build.Utilities;
 using Microsoft.Build.UnitTests;
 using Xunit.Abstractions;
-using Microsoft.Build.Evaluation;
 using System;
 
 namespace Microsoft.Build.Engine.UnitTests
@@ -40,7 +40,7 @@ private void buildSimpleProjectAndValidateChangeWave(TestEnvironment testEnviron
             TransientTestFile file = testEnvironment.CreateFile("proj.csproj", projectFile);
 
             ProjectCollection collection = new ProjectCollection();
-            MockLogger log = new MockLogger();
+            MockLogger log = new MockLogger(_output);
             collection.RegisterLogger(log);
 
             Project p = collection.LoadProject(file.Path);
diff --git a/src/Shared/ChangeWaves.cs b/src/Framework/ChangeWaves.cs
similarity index 100%
rename from src/Shared/ChangeWaves.cs
rename to src/Framework/ChangeWaves.cs
diff --git a/src/Framework/Microsoft.Build.Framework.csproj b/src/Framework/Microsoft.Build.Framework.csproj
index 62c1a97964b..5983457827a 100644
--- a/src/Framework/Microsoft.Build.Framework.csproj
+++ b/src/Framework/Microsoft.Build.Framework.csproj
@@ -18,9 +18,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Compile Include="..\Shared\ChangeWaves.cs">
-      <Link>SharedUtilities\ChangeWaves.cs</Link>
-    </Compile>
     <Compile Include="..\Shared\Constants.cs">
       <Link>Shared\Constants.cs</Link>
     </Compile>
