diff --git a/src/Build/Definition/ProjectItem.cs b/src/Build/Definition/ProjectItem.cs
index b1e2eae13bf..b88305e7aae 100644
--- a/src/Build/Definition/ProjectItem.cs
+++ b/src/Build/Definition/ProjectItem.cs
@@ -475,9 +475,7 @@ public bool HasMetadata(string name)
         /// </summary>
         string IItem.GetMetadataValueEscaped(string name)
         {
-            string value = GetMetadataValueEscaped(name, false);
-
-            return value ?? String.Empty;
+            return GetMetadataValueEscaped(name, false);
         }
 
         /// <summary>
diff --git a/src/Build/Instance/ProjectItemInstance.cs b/src/Build/Instance/ProjectItemInstance.cs
index aa3a34763ee..98a22673d3e 100644
--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -1280,8 +1280,7 @@ public string GetMetadata(string metadataName)
             /// </summary>
             public string GetMetadataEscaped(string metadataName)
             {
-                string value = GetMetadataEscaped(metadataName, false);
-                return value ?? String.Empty;
+                return GetMetadataEscaped(metadataName, false);
             }
 
             /// <summary>
