diff --git a/src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs b/src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs
index c6816fc14ad..7debf6e927d 100644
--- a/src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs
+++ b/src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs
@@ -111,8 +111,6 @@ public void InlineTaskWithAssembly()
                 _testOutputHelper.WriteLine($"AssemblyProj: {assemblyProj.Path}, {assemblyProj.ToString()}");
                 TransientTestFile csFile = env.CreateFile(folder, "Class1.cs", @"
 using System;
-using System.Text.Json;
-using System.Memory;
 
 namespace _5106 {
     public class Class1 {
diff --git a/src/Tasks/Microsoft.Build.Tasks.csproj b/src/Tasks/Microsoft.Build.Tasks.csproj
index 42423ab74d0..a6f26407daf 100644
--- a/src/Tasks/Microsoft.Build.Tasks.csproj
+++ b/src/Tasks/Microsoft.Build.Tasks.csproj
@@ -353,10 +353,6 @@
     <Compile Include="XamlTaskFactory\TaskParser.cs" />
     <Compile Include="XamlTaskFactory\XamlDataDrivenToolTask.cs" />
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETFramework'">
-    <PackageReference Include="System.Text.Json" />
-    <PackageReference Include="System.Memory" />
-  </ItemGroup>
   <ItemGroup Condition="$(TargetFrameworkIdentifier) == '.NETFramework'">
     <!-- Shim targets only work when the destination targets are installed. -->
     <Content Include="Microsoft.Data.Entity.targets">
diff --git a/src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactory.cs b/src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactory.cs
index 37d2eae79a7..de50ca4d4a6 100644
--- a/src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactory.cs
+++ b/src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactory.cs
@@ -58,7 +58,11 @@ public sealed class RoslynCodeTaskFactory : ITaskFactory
                     "Microsoft.Build.Framework",
                     "Microsoft.Build.Utilities.Core",
                     "mscorlib",
-                    "netstandard"
+                    "netstandard",
+                       #if RUNTIME_TYPE_NETCORE || NETCOREAPP
+            "System.Text.Json",
+            "System.Memory",
+            #endif
                 }
             },
             // CSharp specific assembly references
