diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 92d64fc8401..29054a820bd 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -3258,6 +3258,12 @@ private static ILogger[] ProcessLoggingSwitches(
             }
             else
             {
+                // If supported, advertise livelogger
+                if (DoesEnvironmentSupportLiveLogger())
+                {
+                    messagesToLogInBuildLoggers.Add(
+                        new BuildManager.DeferredBuildMessage("Try out the new LiveLogger using the switch -livelogger or -ll", MessageImportance.High));
+                }
                 ProcessConsoleLoggerSwitch(noConsoleLogger, consoleLoggerParameters, distributedLoggerRecords, verbosity, cpuCount, loggers);
             }
 
