diff --git a/src/Build/BackEnd/Client/MSBuildClient.cs b/src/Build/BackEnd/Client/MSBuildClient.cs
index b2329da56b5..f8aafa5bd5d 100644
--- a/src/Build/BackEnd/Client/MSBuildClient.cs
+++ b/src/Build/BackEnd/Client/MSBuildClient.cs
@@ -481,7 +481,8 @@ private void HandlePacket(INodePacket packet)
                 case NodePacketType.ServerNodeBuildResult:
                     HandleServerNodeBuildResult((ServerNodeBuildResult)packet);
                     break;
-                default: throw new InvalidOperationException($"Unexpected packet type {packet.GetType().Name}");
+                default:
+                    throw new InvalidOperationException($"Unexpected packet type {packet.GetType().Name}");
             }
         }
 
