diff --git a/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs b/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs
index 78f35bc8f4e..633554ab190 100644
--- a/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs
+++ b/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs
@@ -75,7 +75,7 @@ internal virtual IReadOnlyList<SdkResolver> LoadAllResolvers(ElementLocation loc
         internal virtual IReadOnlyList<SdkResolverManifest> GetResolversManifests(ElementLocation location)
         {
             MSBuildEventSource.Log.SdkResolverFindResolversManifestsStart();
-            IReadOnlyList <SdkResolverManifest> allResolversManifests = null;
+            IReadOnlyList<SdkResolverManifest> allResolversManifests = null;
             try
             {
                 allResolversManifests = FindPotentialSdkResolversManifests(
