diff --git a/src/MSBuild/MSBuild/Microsoft.Build.CommonTypes.xsd b/src/MSBuild/MSBuild/Microsoft.Build.CommonTypes.xsd
index d555d9a731a..870998c8d61 100644
--- a/src/MSBuild/MSBuild/Microsoft.Build.CommonTypes.xsd
+++ b/src/MSBuild/MSBuild/Microsoft.Build.CommonTypes.xsd
@@ -203,7 +203,7 @@ elementFormDefault="qualified">
               <xs:element name="NoWarn">
                 <xs:annotation>
                   <xs:documentation>
-                    <!-- _locID_text="PackageReference_NoWarn" _locComment="" -->Comma-separated list of warning codes to ignore (such as NU1605)
+                    <!-- _locID_text="PackageReference_NoWarn" _locComment="" -->Semicolon-separated list of warning codes to ignore (such as NU1605)
                   </xs:documentation>
                 </xs:annotation>
               </xs:element>
@@ -247,7 +247,7 @@ elementFormDefault="qualified">
           <xs:attribute name="NoWarn" type="xs:string" use="optional">
             <xs:annotation>
               <xs:documentation>
-                <!-- _locID_text="PackageReference_Attribute_NoWarn" _locComment="" -->Comma-separated list of warning codes to ignore (such as NU1605)
+                <!-- _locID_text="PackageReference_Attribute_NoWarn" _locComment="" -->Semicolon-separated list of warning codes to ignore (such as NU1605)
               </xs:documentation>
             </xs:annotation>
           </xs:attribute>
