diff --git a/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs b/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs
index 2239ac8d079..c802415a697 100644
--- a/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/MSBuild_Tests.cs
@@ -758,13 +758,13 @@ public void ItemsIncludeExcludePathsCombinations()
                 logger.AssertLogContains("iout3=[]");
                 logger.AssertLogContains("iout4=[]");
                 logger.AssertLogContains("iout5=[a/b.foo]");
-                logger.AssertLogContains("iout6=[a\\b.foo]");
+                logger.AssertLogContains($"iout6=[a{Path.DirectorySeparatorChar}b.foo]");
                 logger.AssertLogContains("iin1=[]");
                 logger.AssertLogContains("iin2=[]");
                 logger.AssertLogContains("iin3=[]");
                 logger.AssertLogContains("iin4=[]");
                 logger.AssertLogContains("iin5=[a/b.foo]");
-                logger.AssertLogContains("iin6=[a\\b.foo]");
+                logger.AssertLogContains($"iin6=[a{Path.DirectorySeparatorChar}b.foo]");
             }
             finally
             {
