diff --git a/.vsts-dotnet-ci.yml b/.vsts-dotnet-ci.yml
index fa35d9e2828..5463ea9fffc 100644
--- a/.vsts-dotnet-ci.yml
+++ b/.vsts-dotnet-ci.yml
@@ -350,7 +350,7 @@ jobs:
     continueOnError: true
     condition: eq(variables.onlyDocChanged, 0)
     inputs:
-      filePath: $(Build.SourcesDirectory)\eng\process-coverage.ps1
+      filePath: $(Build.SourcesDirectory)/eng/process-coverage.sh
       arguments: -repoRoot $(Build.SourcesDirectory) -deleteCoverageReportDir
       pwsh: true
   - task: PublishPipelineArtifact@1
@@ -436,7 +436,7 @@ jobs:
     continueOnError: true
     condition: eq(variables.onlyDocChanged, 0)
     inputs:
-      filePath: $(Build.SourcesDirectory)\eng\process-coverage.ps1
+      filePath: $(Build.SourcesDirectory)/eng/process-coverage.ps1
       arguments: -repoRoot $(Build.SourcesDirectory) -deleteCoverageReportDir
       pwsh: true
   - task: PublishPipelineArtifact@1
diff --git a/Coverage.runsettings b/Coverage.runsettings
index ef1e3e46821..1959bca1bce 100644
--- a/Coverage.runsettings
+++ b/Coverage.runsettings
@@ -1,22 +1,22 @@
 <?xml version="1.0" encoding="utf-8"?>
 <RunSettings>
-  <DataCollectionRunSettings>
-    <DataCollectors>
-      <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0" assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
-        <Configuration>
-          <IncludeTestAssembly>True</IncludeTestAssembly>
-          <CodeCoverage>
-            <UseVerifiableInstrumentation>False</UseVerifiableInstrumentation>
-            <EnableStaticNativeInstrumentation>False</EnableStaticNativeInstrumentation>
-            <EnableDynamicNativeInstrumentation>False</EnableDynamicNativeInstrumentation>
-            <EnableStaticManagedInstrumentation>False</EnableStaticManagedInstrumentation>
-            <EnableDynamicManagedInstrumentation>True</EnableDynamicManagedInstrumentation>
-          </CodeCoverage>
-        </Configuration>
-      </DataCollector>
-    </DataCollectors>
-  </DataCollectionRunSettings>
-  <MSTest>
-    <DeploymentEnabled>false</DeploymentEnabled>
-  </MSTest>
+    <DataCollectionRunSettings>
+        <DataCollectors>
+            <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0" assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
+                <Configuration>
+                    <DeterministicReport>True</DeterministicReport>
+                    <IncludeTestAssembly>true</IncludeTestAssembly>
+                    <CodeCoverage>
+                        <CommunicationTimeout>120000</CommunicationTimeout>
+                        <PipeClientsCount>1270</PipeClientsCount>
+                        <UseVerifiableInstrumentation>False</UseVerifiableInstrumentation>
+                        <EnableStaticNativeInstrumentation>True</EnableStaticNativeInstrumentation>
+                        <EnableDynamicNativeInstrumentation>False</EnableDynamicNativeInstrumentation>
+                        <EnableStaticManagedInstrumentation>False</EnableStaticManagedInstrumentation>
+                        <EnableDynamicManagedInstrumentation>True</EnableDynamicManagedInstrumentation>
+                    </CodeCoverage>
+                </Configuration>
+            </DataCollector>
+        </DataCollectors>
+    </DataCollectionRunSettings>
 </RunSettings>
\ No newline at end of file
diff --git a/src/Directory.Build.targets b/src/Directory.Build.targets
index b8e32da022c..15906e76a98 100644
--- a/src/Directory.Build.targets
+++ b/src/Directory.Build.targets
@@ -37,7 +37,7 @@
 
     <XunitOptions>$(XunitOptions)&amp;Traits=failing" /Diag:$(TargetName).log</XunitOptions>
     <!-- Enable code coverage -->
-    <XunitOptions>$(XunitOptions) --collect "Code Coverage;CoverageFileName=$(TargetDir)$(TargetName).coverage" --test-adapter-path $(NuGetPackageRoot)microsoft.codecoverage/17.14.0/build/netstandard2.0</XunitOptions>
+    <XunitOptions>$(XunitOptions) --settings $(RepoRoot)coverage.runsettings --test-adapter-path $(NuGetPackageRoot)microsoft.codecoverage/17.14.0/build/netstandard2.0</XunitOptions>
 
     <TestRunnerAdditionalArguments>$(XunitOptions)</TestRunnerAdditionalArguments>
 
