diff --git a/src/Build/Logging/FancyLogger/FancyLogger.cs b/src/Build/Logging/FancyLogger/FancyLogger.cs
index a915fec1fcb..8cbd1b668df 100644
--- a/src/Build/Logging/FancyLogger/FancyLogger.cs
+++ b/src/Build/Logging/FancyLogger/FancyLogger.cs
@@ -134,9 +134,9 @@ void eventSource_ErrorRaised(object sender, BuildErrorEventArgs e)
 
         public void Shutdown() {
             // Keep open if autoscroll disabled (the user is reading info)
-            while (true)
+            while (FancyLoggerBuffer.AutoScrollEnabled || !FancyLoggerBuffer.IsTerminated)
             {
-                if (FancyLoggerBuffer.AutoScrollEnabled) break;
+
             }
             FancyLoggerBuffer.Terminate();
             Console.WriteLine("Build status, warnings and errors will be shown here after the build has ended and the interactive logger has closed");
diff --git a/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs b/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
index 4989f70ed9d..cac4aafd7ef 100644
--- a/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
+++ b/src/Build/Logging/FancyLogger/FancyLoggerBuffer.cs
@@ -69,6 +69,9 @@ public static void Initialize()
                     if (IsTerminated) return;
                     switch (Console.ReadKey().Key)
                     {
+                        case ConsoleKey.Q:
+                            IsTerminated = true;
+                        break;
                         case ConsoleKey.UpArrow:
                             ScrollToLine(CurrentTopLineIndex - 1);
                             break;
