diff --git a/src/Build/Evaluation/Evaluator.cs b/src/Build/Evaluation/Evaluator.cs
index 4ba973eb301..9b3b8028ed7 100644
--- a/src/Build/Evaluation/Evaluator.cs
+++ b/src/Build/Evaluation/Evaluator.cs
@@ -1929,14 +1929,12 @@ private LoadImportsResult ExpandAndLoadImportsFromUnescapedImportExpression(stri
         {
             imports = new List<ProjectRootElement>();
 
-            bool ignoreInvalidImports = _loadSettings.HasFlag(ProjectLoadSettings.IgnoreInvalidImports);
-
             string importExpressionEscaped = _expander.ExpandIntoStringLeaveEscaped(unescapedExpression, ExpanderOptions.ExpandProperties, importElement.ProjectLocation);
             ElementLocation importLocationInProject = importElement.Location;
 
             if (String.IsNullOrWhiteSpace(importExpressionEscaped))
             {
-                if (ignoreInvalidImports)
+                if ((_loadSettings & ProjectLoadSettings.IgnoreInvalidImports) != 0)
                 {
                     // Log message for import skipped
                     ProjectImportedEventArgs eventArgs = new ProjectImportedEventArgs(
