diff --git a/src/Tasks/SystemState.cs b/src/Tasks/SystemState.cs
index 2ff64df7e92..face0482fc9 100644
--- a/src/Tasks/SystemState.cs
+++ b/src/Tasks/SystemState.cs
@@ -172,6 +172,7 @@ internal FileState(SerializationInfo info, StreamingContext context)
                     var frameworkProfile = info.GetString("fmProf");
                     frameworkName = new FrameworkName(frameworkIdentifier, frameworkNameVersion, frameworkProfile);
                 }
+                ModuleVersionID = (Guid)info.GetValue("mvid", typeof(Guid));
             }
 
             /// <summary>
@@ -195,6 +196,7 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
                     info.AddValue("fnId", frameworkName.Identifier);
                     info.AddValue("fmProf", frameworkName.Profile);
                 }
+                info.AddValue("mvid", ModuleVersionID);
             }
 
             /// <summary>
