diff --git a/src/Build/Graph/ParallelWorkSet.cs b/src/Build/Graph/ParallelWorkSet.cs
index dc61e30b0ef..fc2bd69ae63 100644
--- a/src/Build/Graph/ParallelWorkSet.cs
+++ b/src/Build/Graph/ParallelWorkSet.cs
@@ -188,8 +188,14 @@ private void ExecuteWorkItem()
         {
             if (_queue.TryDequeue(out Lazy<TResult> workItem))
             {
-                TResult _ = workItem.Value;
-                Interlocked.Decrement(ref _pendingCount);
+                try
+                {
+                    TResult _ = workItem.Value;
+                }
+                finally
+                {
+                    Interlocked.Decrement(ref _pendingCount);
+                }
             }
         }
     }
