diff --git a/src/Build/Evaluation/Evaluator.cs b/src/Build/Evaluation/Evaluator.cs
index d401418e501..35097b0aeda 100644
--- a/src/Build/Evaluation/Evaluator.cs
+++ b/src/Build/Evaluation/Evaluator.cs
@@ -1143,7 +1143,7 @@ private void AddBuiltInProperties()
             SetBuiltInProperty(ReservedPropertyNames.version, MSBuildAssemblyFileVersion.Instance.MajorMinorBuild);
             SetBuiltInProperty(ReservedPropertyNames.fileVersion, CachedFileVersion);
             SetBuiltInProperty(ReservedPropertyNames.semanticVersion, ProjectCollection.DisplayVersion);
-            SetBuiltInProperty(ReservedPropertyNames.windowsKitsPath, FrameworkLocationHelper.PathOfInstalledWindowsKits);
+
             ValidateChangeWaveState();
 
             SetBuiltInProperty(ReservedPropertyNames.msbuilddisablefeaturesfromversion, ChangeWaves.DisabledWave.ToString());
diff --git a/src/Build/Resources/Constants.cs b/src/Build/Resources/Constants.cs
index 500d4c185bb..2436979d94f 100644
--- a/src/Build/Resources/Constants.cs
+++ b/src/Build/Resources/Constants.cs
@@ -60,7 +60,6 @@ internal static class ReservedPropertyNames
         internal const string frameworkToolsRoot = "MSBuildFrameworkToolsRoot";
         internal const string interactive = "MSBuildInteractive";
         internal const string msbuilddisablefeaturesfromversion = "MSBuildDisableFeaturesFromVersion";
-        internal const string windowsKitsPath = "WindowsKitsDirectory";
 
         /// <summary>
         /// Lookup for reserved property names. Intentionally do not include MSBuildExtensionsPath* or MSBuildUserExtensionsPath in this list.  We need tasks to be able to override those.
diff --git a/src/Shared/FrameworkLocationHelper.cs b/src/Shared/FrameworkLocationHelper.cs
index f8bc93693d3..3ff6eb90662 100644
--- a/src/Shared/FrameworkLocationHelper.cs
+++ b/src/Shared/FrameworkLocationHelper.cs
@@ -101,7 +101,6 @@ internal static class FrameworkLocationHelper
 
 #if FEATURE_WIN32_REGISTRY
         private const string fallbackDotNetFrameworkSdkRegistryInstallPath = "SOFTWARE\\Microsoft\\Microsoft SDKs\\Windows";
-        private const string windowsKitsInstalledRootsRegistryPath = "SOFTWARE\\Microsoft\\Windows Kits\\Installed Roots";
 #endif // FEATURE_WIN32_REGISTRY
         internal const string fallbackDotNetFrameworkSdkInstallKeyValue = "CurrentInstallFolder";
 
@@ -623,25 +622,6 @@ private static string PathToV4ToolsInFallbackDotNetFrameworkSdk
             }
         }
 
-        private static string pathOfInstalledWindowsKits = null;
-
-        internal static string PathOfInstalledWindowsKits
-        {
-            get
-            {
-                #if FEATURE_WIN32_REGISTRY
-                if (FrameworkLocationHelper.pathOfInstalledWindowsKits == null)
-                {
-                    FrameworkLocationHelper.pathOfInstalledWindowsKits = FindRegistryValueUnderKey(
-                       windowsKitsInstalledRootsRegistryPath,
-                       "KitsRoot10");
-                }
-                #endif
-
-                return FrameworkLocationHelper.pathOfInstalledWindowsKits;
-            }
-        }
-
         #endregion // Static properties
 
         #region Internal methods
diff --git a/src/Tasks/Microsoft.Common.CurrentVersion.targets b/src/Tasks/Microsoft.Common.CurrentVersion.targets
index f2e186e2d84..3c88dc06d8b 100644
--- a/src/Tasks/Microsoft.Common.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.Common.CurrentVersion.targets
@@ -2518,7 +2518,7 @@ Copyright (C) Microsoft Corporation. All rights reserved.
     <SDKReferenceRegistryRoot Condition="'$(SDKReferenceRegistryRoot)' == ''">Software\Microsoft\Microsoft SDKs</SDKReferenceRegistryRoot>
     <SDKReferenceDirectoryRoot Condition="'$(SDKReferenceDirectoryRoot)' == ''">$(LocalAppData)\Microsoft SDKs;$(MSBuildProgramFiles32)\Microsoft SDKs</SDKReferenceDirectoryRoot>
     <!-- Manifest driven extension SDK locations -->
-    <SDKExtensionDirectoryRoot Condition="'$(SDKExtensionDirectoryRoot)' == '' and '$(SDKIdentifier)' != ''">$(MSBuildProgramFiles32)\Microsoft SDKs\Windows Kits\10;$(WindowsKitsDirectory)</SDKExtensionDirectoryRoot>
+    <SDKExtensionDirectoryRoot Condition="'$(SDKExtensionDirectoryRoot)' == '' and '$(SDKIdentifier)' != ''">$(MSBuildProgramFiles32)\Microsoft SDKs\Windows Kits\10;$(MSBuildProgramFiles32)\Windows Kits\10</SDKExtensionDirectoryRoot>
 
     <!-- UAP projects by default should support Windows 8.1 SDKs -->
     <SupportWindows81SDKs Condition="'$(SupportWindows81SDKs)' == '' and '$(Support81SDKs)' != 'false' and '$(TargetPlatformIdentifier)' == 'UAP'">true</SupportWindows81SDKs>
