diff --git a/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs b/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs
index 8d3484cf052..aa57103cf80 100644
--- a/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs
+++ b/src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs
@@ -230,7 +230,10 @@ async Task<CacheResult> ProcessCacheRequest(CacheRequest request)
                     await LateInitializePluginForVsWorkaround(request);
                 }
 
-                return await GetCacheResultAsync(cacheRequest.Submission.BuildRequestData);
+                return await GetCacheResultAsync(
+                    new BuildRequestData(
+                        request.Configuration.Project,
+                        request.Submission.BuildRequestData.TargetNames.ToArray()));
             }
 
             static bool IsDesignTimeBuild(ProjectInstance project)
