diff --git a/src/Shared/CopyOnWriteDictionary.cs b/src/Shared/CopyOnWriteDictionary.cs
index 6b1820bc5d5..2b4ab1917a2 100644
--- a/src/Shared/CopyOnWriteDictionary.cs
+++ b/src/Shared/CopyOnWriteDictionary.cs
@@ -227,6 +227,18 @@ public void Add(string key, V value)
             _backing = _backing.SetItem(key, value);
         }
 
+        /// <summary>
+        /// Adds several value to the dictionary.
+        /// </summary>
+        public void SetItems(IEnumerable<KeyValuePair<string, V>> items)
+        {
+            _backing = _backing.SetItems(items);
+        }
+
+        public IEnumerable<KeyValuePair<string, V>> Where(Func<KeyValuePair<string, V>, bool> predicate)
+        {
+            return _backing.Where(predicate);
+        }
         /// <summary>
         /// Returns true if the dictionary contains the specified key.
         /// </summary>
diff --git a/src/Utilities/TaskItem.cs b/src/Utilities/TaskItem.cs
index 154c053026c..a27113c29e7 100644
--- a/src/Utilities/TaskItem.cs
+++ b/src/Utilities/TaskItem.cs
@@ -298,12 +298,21 @@ public void CopyMetadataTo(ITaskItem destinationItem)
             if (_metadata != null)
             {
                 // Avoid a copy if we can
-                if (destinationItem is TaskItem destinationAsTaskItem && destinationAsTaskItem.Metadata?.Count < _metadata?.Count)
+                if (destinationItem is TaskItem destinationAsTaskItem)
                 {
-                    CopyOnWriteDictionary<string> copiedMetadata = _metadata.Clone(); // Copy on write!
-                    foreach (KeyValuePair<string, string> entry in destinationAsTaskItem.Metadata)
+                    CopyOnWriteDictionary<string> copiedMetadata;
+                    if (destinationAsTaskItem.Metadata == null || destinationAsTaskItem.Metadata.Count < _metadata.Count)
                     {
-                        copiedMetadata[entry.Key] = entry.Value;
+                        copiedMetadata = _metadata.Clone(); // Copy on write!
+                        if (destinationAsTaskItem.Metadata != null)
+                        {
+                            copiedMetadata.SetItems(destinationAsTaskItem.Metadata);
+                        }
+                    }
+                    else
+                    {
+                        copiedMetadata = destinationAsTaskItem.Metadata.Clone();
+                        copiedMetadata.SetItems(Metadata.Where(entry => !destinationAsTaskItem.Metadata.ContainsKey(entry.Key)));
                     }
                     destinationAsTaskItem.Metadata = copiedMetadata;
                 }
