diff --git a/src/Build/BackEnd/Components/Scheduler/Scheduler.cs b/src/Build/BackEnd/Components/Scheduler/Scheduler.cs
index 771292b5a40..9fa3752a6ef 100644
--- a/src/Build/BackEnd/Components/Scheduler/Scheduler.cs
+++ b/src/Build/BackEnd/Components/Scheduler/Scheduler.cs
@@ -2092,7 +2092,6 @@ static bool SkipNonexistentTargetsIfExistentTargetsHaveResults(BuildRequest buil
 
             static Action<ILoggingService> GetLoggingServiceAction(IConfigCache configCache, BuildRequest request, SchedulingData schedulingData)
             {
-                // retrieving the configs is not quite free, so avoid computing them eagerly
                 (BuildRequestConfiguration requestConfig, BuildRequestConfiguration parentConfig) = GetConfigurations(configCache, request, schedulingData);
 
                 Action<ILoggingService> emitNonErrorLogs = ls => ls.LogComment(
