diff --git a/src/Build.UnitTests/BackEnd/TaskHostConfiguration_Tests.cs b/src/Build.UnitTests/BackEnd/TaskHostConfiguration_Tests.cs
index e30b34ac072..144f30bea1a 100644
--- a/src/Build.UnitTests/BackEnd/TaskHostConfiguration_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/TaskHostConfiguration_Tests.cs
@@ -639,7 +639,6 @@ public void TestTranslationWithWarningsAsMessages()
 
             Assert.NotNull(deserializedConfig.WarningsAsMessages);
             config.WarningsAsMessages.SequenceEqual(deserializedConfig.WarningsAsMessages, StringComparer.Ordinal).ShouldBeTrue();
-
         }
 
         /// <summary>
diff --git a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
index cdc69691c8e..8fa1fe2919b 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
@@ -725,7 +725,7 @@ private ICollection<string> WarningsAsMessages
         public bool ShouldTreatWarningAsError(string warningCode)
         {
             // Warnings as messages overrides warnings as errors.
-            if (WarningsAsErrors == null || (WarningsAsMessages != null && WarningsAsMessages.Contains(warningCode)))
+            if (WarningsAsErrors == null || WarningsAsMessages?.Contains(warningCode) == true)
             {
                 return false;
             }
diff --git a/src/MSBuild/OutOfProcTaskHostNode.cs b/src/MSBuild/OutOfProcTaskHostNode.cs
index 13f895bb868..b4e8c4f0301 100644
--- a/src/MSBuild/OutOfProcTaskHostNode.cs
+++ b/src/MSBuild/OutOfProcTaskHostNode.cs
@@ -280,7 +280,7 @@ public bool IsRunningMultipleNodes
         public bool ShouldTreatWarningAsError(string warningCode)
         {
             // Warnings as messages overrides warnings as errors.
-            if (WarningsAsErrors == null || (WarningsAsMessages != null && WarningsAsMessages.Contains(warningCode)))
+            if (WarningsAsErrors == null || WarningsAsMessages?.Contains(warningCode) == true)
             {
                 return false;
             }
diff --git a/src/Shared/TaskHostConfiguration.cs b/src/Shared/TaskHostConfiguration.cs
index e3a0bbaec0e..1bcae1d7574 100644
--- a/src/Shared/TaskHostConfiguration.cs
+++ b/src/Shared/TaskHostConfiguration.cs
@@ -397,14 +397,14 @@ public void Translate(ITranslator translator)
             translator.Translate(collection: ref _warningsAsErrors,
                                  objectTranslator: (ITranslator t, ref string s) => t.Translate(ref s),
 #if CLR2COMPATIBILITY
-                                 collectionFactory: count => new HashSet<string>());
+                                 collectionFactory: count => new HashSet<string>(StringComparer.OrdinalIgnoreCase));
 #else
                                  collectionFactory: count => new HashSet<string>(count, StringComparer.OrdinalIgnoreCase));
 #endif
             translator.Translate(collection: ref _warningsAsMessages,
                                  objectTranslator: (ITranslator t, ref string s) => t.Translate(ref s),
 #if CLR2COMPATIBILITY
-                                 collectionFactory: count => new HashSet<string>());
+                                 collectionFactory: count => new HashSet<string>(StringComparer.OrdinalIgnoreCase));
 #else
                                  collectionFactory: count => new HashSet<string>(count, StringComparer.OrdinalIgnoreCase));
 #endif
