diff --git a/src/Build/BackEnd/BuildManager/BuildSubmission.cs b/src/Build/BackEnd/BuildManager/BuildSubmission.cs
index 44d08c99a37..ca2147dfae2 100644
--- a/src/Build/BackEnd/BuildManager/BuildSubmission.cs
+++ b/src/Build/BackEnd/BuildManager/BuildSubmission.cs
@@ -103,7 +103,7 @@ internal BuildSubmission(BuildManager buildManager, int submissionId, BuildReque
         /// <summary>
         /// The BuildRequestData being used for this submission.
         /// </summary>
-        public BuildRequestData BuildRequestData { get; }
+        internal BuildRequestData BuildRequestData { get; }
 
         /// <summary>
         /// The build request for execution.
