diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index f289efadd25..3f98ef728bf 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -3222,7 +3222,10 @@ out bool enableProfiler
 
             // Choose default console logger
             // TODO: Add conditions for terminals that do not support ANSI
-            if(shouldUseFancyLogger)
+            if(
+                shouldUseFancyLogger &&
+                !Console.IsOutputRedirected // Avoid using the FancyLogger when output is redirected to a file
+            )
             {
                 ProcessFancyLogger(noConsoleLogger, loggers);
             }
