diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index a461337a22d..906046bcc86 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1110,21 +1110,7 @@ string outputResultsCache
 
                 if (targetsOnly)
                 {
-                    try
-                    {
-                        Project project = projectCollection.LoadProject(projectFile, globalProperties, toolsVersion);
-
-                        targetsWriter.WriteLine(string.Join(Environment.NewLine, project.Targets.Keys));
-
-                        projectCollection.UnloadProject(project);
-                        success = true;
-                    }
-                    catch (Exception ex)
-                    {
-                        var message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword("TargetsCouldNotBePrinted", ex.Message);
-                        Console.Error.WriteLine(message);
-                        success = false;
-                    }
+                    success = PrintTargets(projectFile, toolsVersion, globalProperties, targetsWriter, projectCollection);
                 }
 
                 if (!preprocessOnly && !targetsOnly)
@@ -1317,6 +1303,25 @@ string outputResultsCache
             return success;
         }
 
+        private static bool PrintTargets(string projectFile, string toolsVersion, Dictionary<string, string> globalProperties, TextWriter targetsWriter, ProjectCollection projectCollection)
+        {
+            try
+            {
+                Project project = projectCollection.LoadProject(projectFile, globalProperties, toolsVersion);
+
+                targetsWriter.WriteLine(string.Join(Environment.NewLine, project.Targets.Keys));
+
+                projectCollection.UnloadProject(project);
+                return true;
+            }
+            catch (Exception ex)
+            {
+                var message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword("TargetsCouldNotBePrinted", ex.Message);
+                Console.Error.WriteLine(message);
+                return false;
+            }
+        }
+
         private static (BuildResultCode result, Exception exception) ExecuteBuild(BuildManager buildManager, BuildRequestData request)
         {
             BuildSubmission submission;
