diff --git a/src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs b/src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs
index d932ab16814..bd98358c4ad 100644
--- a/src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs
+++ b/src/Build/BackEnd/Components/Logging/ProjectLoggingContext.cs
@@ -23,11 +23,6 @@ internal class ProjectLoggingContext : BuildLoggingContext
         /// </summary>
         private string _projectFullPath;
 
-        /// <summary>
-        /// The parent node logging context this context was derived from.
-        /// </summary>
-        private NodeLoggingContext _nodeLoggingContext;
-
         /// <summary>
         /// Constructs a project logging context.
         /// </summary>
@@ -79,7 +74,6 @@ internal ProjectLoggingContext(
         private ProjectLoggingContext(NodeLoggingContext nodeLoggingContext, int submissionId, int configurationId, string projectFullPath, List<string> targets, string toolsVersion, PropertyDictionary<ProjectPropertyInstance> projectProperties, ItemDictionary<ProjectItemInstance> projectItems, BuildEventContext parentBuildEventContext, int evaluationId = BuildEventContext.InvalidEvaluationId)
             : base(nodeLoggingContext)
         {
-            _nodeLoggingContext = nodeLoggingContext;
             _projectFullPath = projectFullPath;
 
             ProjectPropertyInstanceEnumeratorProxy properties = null;
@@ -150,17 +144,6 @@ private ProjectLoggingContext(NodeLoggingContext nodeLoggingContext, int submiss
             this.IsValid = true;
         }
 
-        /// <summary>
-        /// Retrieves the node logging context.
-        /// </summary>
-        internal NodeLoggingContext NodeLoggingContext
-        {
-            get
-            {
-                return _nodeLoggingContext;
-            }
-        }
-
         /// <summary>
         /// Log that the project has finished
         /// </summary>
