diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 86da1102387..8411c5408ba 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -2721,13 +2721,6 @@ private static bool ProcessCommandLineSwitches(
 
                     isBuildCheckEnabled = IsBuildCheckEnabled(commandLineSwitches);
 
-                    // BuildCheck is not compatible with node reusing, see #10317.
-                    // Disable node reuse when build check is on.
-                    if (isBuildCheckEnabled)
-                    {
-                        enableNodeReuse = false;
-                    }
-
                     inputResultsCaches = ProcessInputResultsCaches(commandLineSwitches);
 
                     outputResultsCache = ProcessOutputResultsCache(commandLineSwitches);
