diff --git a/src/MSBuild/TerminalLogger/TerminalLogger.cs b/src/MSBuild/TerminalLogger/TerminalLogger.cs
index 64052300494..398583aa7df 100644
--- a/src/MSBuild/TerminalLogger/TerminalLogger.cs
+++ b/src/MSBuild/TerminalLogger/TerminalLogger.cs
@@ -635,13 +635,18 @@ private bool IsImmediateMessage(string message)
     /// </summary>
     private void ErrorRaised(object sender, BuildErrorEventArgs e)
     {
+        BuildEventContext? buildEventContext = e.BuildEventContext;
+        Project? project = null;
+        bool isTrackedProject = buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out project);
         string message = EventArgsFormatting.FormatEventMessage(
                 category: AnsiCodes.Colorize("error", TerminalColor.Red),
                 subcategory: e.Subcategory,
                 message: e.Message,
                 code: AnsiCodes.Colorize(e.Code, TerminalColor.Red),
                 file: HighlightFileName(e.File),
-                projectFile: null,
+
+                // for the tracked projects the project file name is included in the final output result.
+                projectFile: isTrackedProject ? null : e.ProjectFile ?? string.Empty,
                 lineNumber: e.LineNumber,
                 endLineNumber: e.EndLineNumber,
                 columnNumber: e.ColumnNumber,
@@ -649,10 +654,9 @@ private void ErrorRaised(object sender, BuildErrorEventArgs e)
                 threadId: e.ThreadId,
                 logOutputProperties: null);
 
-        BuildEventContext? buildEventContext = e.BuildEventContext;
-        if (buildEventContext is not null && _projects.TryGetValue(new ProjectContext(buildEventContext), out Project? project))
+        if (isTrackedProject)
         {
-            project.AddBuildMessage(MessageSeverity.Error, message);
+            project!.AddBuildMessage(MessageSeverity.Error, message);
         }
 
         // It is necessary to display error messages reported by MSBuild, even if it's not tracked in _projects collection.
@@ -662,7 +666,7 @@ private void ErrorRaised(object sender, BuildErrorEventArgs e)
         }
     }
 
-#endregion
+    #endregion
 
     #region Refresher thread implementation
 
