diff --git a/src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs b/src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs
index fe5a270f1a6..eb0824fac7a 100644
--- a/src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs
+++ b/src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs
@@ -70,7 +70,7 @@ protected override void ApplyImpl(OrderedItemDataCollection.Builder listBuilder,
                             items.Add(item.Item);
                         }
                     }
-                    if (items != null)
+                    if (items is not null)
                     {
                         listBuilder.RemoveAll(items);
                     }
