diff --git a/eng/common/tools.ps1 b/eng/common/tools.ps1
index 3e650534899..9014e062514 100644
--- a/eng/common/tools.ps1
+++ b/eng/common/tools.ps1
@@ -506,7 +506,7 @@ function InitializeBuildTool() {
       ExitWithExitCode 1
     }
     $dotnetPath = Join-Path $dotnetRoot (GetExecutableFileName 'dotnet')
-    $buildTool = @{ Path = $dotnetPath; Command = 'msbuild'; Tool = 'dotnet'; Framework = 'net5.0' }
+    $buildTool = @{ Path = $dotnetPath; Command = 'msbuild'; Tool = 'dotnet'; Framework = 'netcoreapp2.1' }
   } elseif ($msbuildEngine -eq "vs") {
     try {
       $msbuildPath = InitializeVisualStudioMSBuild -install:$restore
diff --git a/eng/common/tools.sh b/eng/common/tools.sh
index aeb05822cbd..b5d63cb1b7c 100755
--- a/eng/common/tools.sh
+++ b/eng/common/tools.sh
@@ -309,7 +309,7 @@ function InitializeBuildTool {
   # return values
   _InitializeBuildTool="$_InitializeDotNetCli/dotnet"
   _InitializeBuildToolCommand="msbuild"
-  _InitializeBuildToolFramework="net5.0"
+  _InitializeBuildToolFramework="netcoreapp2.1"
 }
 
 # Set RestoreNoCache as a workaround for https://github.com/NuGet/Home/issues/3116
