diff --git a/src/Tasks.UnitTests/GetCompatiblePlatform_Tests.cs b/src/Tasks.UnitTests/GetCompatiblePlatform_Tests.cs
index 48ea89bda94..47fb145aabc 100644
--- a/src/Tasks.UnitTests/GetCompatiblePlatform_Tests.cs
+++ b/src/Tasks.UnitTests/GetCompatiblePlatform_Tests.cs
@@ -233,7 +233,7 @@ public void PlatformIsChosenAsDefault(string referencedPlatforms, string referen
         public void ChosenPlatformMatchesDefault()
         {
             TaskItem projectReference = new TaskItem("foo.bar");
-            projectReference.SetMetadata("Platforms", "AnyCPU");
+            projectReference.SetMetadata("Platforms", "AnyCPU;x64");
             projectReference.SetMetadata("Platform", "AnyCPU");
 
             GetCompatiblePlatform task = new GetCompatiblePlatform()
