diff --git a/src/Shared/FileMatcher.cs b/src/Shared/FileMatcher.cs
index 0d5246f5d97..c90cc0b4498 100644
--- a/src/Shared/FileMatcher.cs
+++ b/src/Shared/FileMatcher.cs
@@ -857,6 +857,7 @@ private void GetFilesRecursive(
             // This fails in tests with the MockFileSystem when they don't have real paths.
             catch (IOException) { }
             catch (ArgumentException) { }
+            catch (UnauthorizedAccessException) { }
 #endif
 
             ErrorUtilities.VerifyThrow((recursionState.SearchData.Filespec == null) || (recursionState.SearchData.RegexFileMatch == null),
