diff --git a/src/Framework/Traits.cs b/src/Framework/Traits.cs
index b99b9d57d9e..16f54527c99 100644
--- a/src/Framework/Traits.cs
+++ b/src/Framework/Traits.cs
@@ -400,7 +400,7 @@ public bool EnableWarningOnCustomBuildEvent
             }
         }
 
-        public bool UnquoteSwitchParameterForTragetParametrizedSwitch
+        public bool UnquoteTragetSwitchParameters
         {
             get
             {
diff --git a/src/MSBuild/CommandLineSwitches.cs b/src/MSBuild/CommandLineSwitches.cs
index 5531321c307..226e4dc0f62 100644
--- a/src/MSBuild/CommandLineSwitches.cs
+++ b/src/MSBuild/CommandLineSwitches.cs
@@ -664,7 +664,7 @@ internal string[][] GetFileLoggerParameters()
         /// </summary>
         private bool IsMultipleAllowedSwithParameterDueToUnquote(bool unquoteParameter, ParameterizedSwitch parameterizedSwitch)
         {
-            if (!unquoteParameter || !Traits.Instance.EscapeHatches.UnquoteSwitchParameterForTragetParametrizedSwitch)
+            if (!unquoteParameter || !Traits.Instance.EscapeHatches.UnquoteTragetSwitchParameters)
             {
                 return false;
             }
