diff --git a/src/Tasks/Al.cs b/src/Tasks/Al.cs
index 6baa852ba02..eaa1b3bfb89 100644
--- a/src/Tasks/Al.cs
+++ b/src/Tasks/Al.cs
@@ -306,8 +306,9 @@ protected override string GenerateFullPathToTool()
             if (String.IsNullOrEmpty(pathToTool) || !FileSystems.Default.FileExists(pathToTool))
             {
                 // The bitness of al.exe should match the platform being built
-                string archToLookFor =  Platform.Equals("x86", StringComparison.OrdinalIgnoreCase) ? Platform :
-                                        Platform.Equals("x64", StringComparison.OrdinalIgnoreCase) ? ProcessorArchitecture.AMD64 : // x64 maps to AMD64 in GeneratePathToTool
+                // Yoda condition prevents null reference exception if Platform is null.
+                string archToLookFor =  "x86".Equals(Platform, StringComparison.OrdinalIgnoreCase) ? Platform :
+                                        "x64".Equals(Platform, StringComparison.OrdinalIgnoreCase) ? ProcessorArchitecture.AMD64 : // x64 maps to AMD64 in GeneratePathToTool
                                         ProcessorArchitecture.CurrentProcessArchitecture;
 
                 pathToTool = SdkToolsPathUtility.GeneratePathToTool(f => SdkToolsPathUtility.FileInfoExists(f), archToLookFor, SdkToolsPath, ToolExe, Log, true);
