diff --git a/src/Tasks.UnitTests/PopulateSupportedArchitectures_Tests.cs b/src/Tasks.UnitTests/PopulateSupportedArchitectures_Tests.cs
index 660d73af13b..ceca0294ef8 100644
--- a/src/Tasks.UnitTests/PopulateSupportedArchitectures_Tests.cs
+++ b/src/Tasks.UnitTests/PopulateSupportedArchitectures_Tests.cs
@@ -129,7 +129,6 @@ public void E2EScenarioTests(string manifestName, bool expectedResult)
                     {
                         actualDoc.Load(stream);
                     }
-                    actualDoc.Save(Path.Combine("C:\\Users\\ykovalova\\OneDrive - Microsoft\\Documents", "actualManifest.xml"));
 
                     expectedDoc.OuterXml.ShouldBe(actualDoc.OuterXml);
                     expectedDoc.InnerXml.ShouldBe(actualDoc.InnerXml);
