diff --git a/src/MSBuild.UnitTests/XMake_Tests.cs b/src/MSBuild.UnitTests/XMake_Tests.cs
index 62e26e60d84..b940d898c3b 100644
--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -2546,7 +2546,7 @@ public void LoggerThrowsMemberAccessExceptionWhenClassIsInvalid(string memberAcc
 
                 (string projectFilePath, string tempLoggerProjDir) = CopyTestAssetsToTestEnv(tempDir, projectContent, memberAccess);
 
-                string projectCheckBuildLog = RunnerUtilities.ExecBootstrapedMSBuild(
+                string loggerBuildLog = RunnerUtilities.ExecBootstrapedMSBuild(
                 $"{Path.Combine(tempLoggerProjDir, $"CustomLogger.csproj")} -restore -verbosity:n", out bool success);
 
                 var loggerDllPath = Path.Combine(tempLoggerProjDir, "artifacts", "bin", "netstandard2.0", expectedLoggerName);
