diff --git a/src/Tasks/DependencyFile.cs b/src/Tasks/DependencyFile.cs
index c6722b14d6c..9908267c4a8 100644
--- a/src/Tasks/DependencyFile.cs
+++ b/src/Tasks/DependencyFile.cs
@@ -14,7 +14,12 @@ namespace Microsoft.Build.Tasks
     /// <remarks>
     /// Represents a single input to a compilation-style task.
     /// Keeps track of timestamp for later comparison.
+    ///
+    /// Must be serializable because instances may be marshaled cross-AppDomain, see <see cref="ProcessResourceFiles.PortableLibraryCacheInfo"/>.
     /// </remarks>
+#if FEATURE_APPDOMAIN
+    [Serializable]
+#endif
     internal class DependencyFile
     {
         // Filename
