diff --git a/src/Shared/BuildEnvironmentHelper.cs b/src/Shared/BuildEnvironmentHelper.cs
index 6f305333738..f2a7c836944 100644
--- a/src/Shared/BuildEnvironmentHelper.cs
+++ b/src/Shared/BuildEnvironmentHelper.cs
@@ -508,7 +508,7 @@ public BuildEnvironment(BuildEnvironmentMode mode, string currentMSBuildExePath,
             VisualStudioInstallRootDirectory = visualStudioPath;
 
 #if !NO_FRAMEWORK_IVT
-            Framework.BuildEnvironmentState.s_runningTests = true;
+            Framework.BuildEnvironmentState.s_runningTests = runningTests;
             Framework.BuildEnvironmentState.s_runningInVisualStudio = runningInVisualStudio;
 #endif
 
