diff --git a/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs b/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
index 3c0ec06b3c2..1ebd4c20522 100644
--- a/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
+++ b/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
@@ -590,7 +590,10 @@ private void PropagateImport(int evaluationId, string originalProjectFile, strin
             {
                 lock (importedProjects)
                 {
-                    importedProjects.Add(newImportedProjectFile);
+                    if (importedProjects.Contains(originalProjectFile))
+                    {
+                        importedProjects.Add(newImportedProjectFile);
+                    }
                 }
             }
         }
