diff --git a/src/BuildCheck.UnitTests/EndToEndTests.cs b/src/BuildCheck.UnitTests/EndToEndTests.cs
index 22b2ce73979..c3fa31f48c0 100644
--- a/src/BuildCheck.UnitTests/EndToEndTests.cs
+++ b/src/BuildCheck.UnitTests/EndToEndTests.cs
@@ -462,7 +462,8 @@ public void CustomCheckTest_NoEditorConfig(string checkCandidate)
 
             string projectCheckBuildLog = RunnerUtilities.ExecBootstrapedMSBuild(
                 $"{Path.Combine(checkCandidatePath, $"{checkCandidate}.csproj")} /m:1 -nr:False -restore -check -verbosity:n",
-                out bool successBuild);
+                out bool successBuild,
+                outputHelper: env.Output);
 
             projectCheckBuildLog.ShouldNotBeEmpty();
             projectCheckBuildLog.ShouldContain("Build started");
@@ -487,7 +488,7 @@ public void CustomCheckTest_WithEditorConfig(string checkCandidate, string ruleI
                 checkCandidatePath));
 
             string projectCheckBuildLog = RunnerUtilities.ExecBootstrapedMSBuild(
-                $"{Path.Combine(checkCandidatePath, $"{checkCandidate}.csproj")} /m:1 -nr:False -restore -check -verbosity:n", out bool _);
+                $"{Path.Combine(checkCandidatePath, $"{checkCandidate}.csproj")} /m:1 -nr:False -restore -check -verbosity:n", out bool _, outputHelper: env.Output);
 
             projectCheckBuildLog.ShouldNotBeEmpty();
             projectCheckBuildLog.ShouldContain("Build started");
