diff --git a/src/Tasks/XmlPoke.cs b/src/Tasks/XmlPoke.cs
index 3e4c169f6cc..501d0bc9428 100644
--- a/src/Tasks/XmlPoke.cs
+++ b/src/Tasks/XmlPoke.cs
@@ -49,7 +49,7 @@ public class XmlPoke : TaskExtension
         /// <summary>
         /// Executes the XMLPoke task.
         /// </summary>
-        /// <returns>true if transformation succeeds.</returns>
+        /// <returns>true if task execution succeeds.</returns>
         public override bool Execute()
         {
             if (Value == null)
