diff --git a/src/Tasks.UnitTests/Copy_Tests.cs b/src/Tasks.UnitTests/Copy_Tests.cs
index 1792dad790f..18cff678cb5 100644
--- a/src/Tasks.UnitTests/Copy_Tests.cs
+++ b/src/Tasks.UnitTests/Copy_Tests.cs
@@ -1406,6 +1406,36 @@ public void CopyFileOnItself2(bool isUseHardLinks, bool isUseSymbolicLinks)
             }
         }
 
+        /// <summary>
+        /// Copying a file on top of itself should be a success (no-op) whether
+        /// or not skipUnchangedFiles is true or false.
+        /// </summary>
+        [Theory]
+        [MemberData(nameof(GetHardLinksSymLinks))]
+        public void CopyFileItselfNonExistentFile(bool isUseHardLinks, bool isUseSymbolicLinks)
+        {
+            string temp = Path.GetTempPath();
+            string file = Path.Combine(temp, "FileThatDoesNotExist");
+
+            var engine = new MockEngine(_testOutputHelper);
+            var t = new Copy
+            {
+                RetryDelayMilliseconds = 1, // speed up tests!
+                BuildEngine = engine,
+                SourceFiles = new ITaskItem[] { new TaskItem(file) },
+                DestinationFiles = new ITaskItem[] { new TaskItem(file) },
+                SkipUnchangedFiles = false,
+                UseHardlinksIfPossible = isUseHardLinks,
+                UseSymboliclinksIfPossible = isUseSymbolicLinks,
+            };
+
+            bool success = t.Execute();
+
+            Assert.False(success);
+
+            ((MockEngine)t.BuildEngine).AssertLogDoesntContain("MSB3026"); // Didn't do retries, nothing to do
+        }
+
         /// <summary>
         /// Copying a file on top of itself should be a success (no-op) whether
         /// or not skipUnchangedFiles is true or false. Variation with a second copy failure.
