diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index cb71d55758c..f8c199d9c6c 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -373,7 +373,7 @@ dotnet_diagnostic.CA2153.severity = none
 dotnet_diagnostic.CA2201.severity = none
 
 # Instantiate argument exceptions correctly
-dotnet_diagnostic.CA2208.severity = suggestion
+dotnet_diagnostic.CA2208.severity = warning
 
 # Non-constant fields should not be visible
 dotnet_diagnostic.CA2211.severity = none
@@ -1107,4 +1107,4 @@ dotnet_diagnostic.SA1649.severity = suggestion
 dotnet_diagnostic.SA1652.severity = none
 
 # Do not use equality check to check for collection size.
-dotnet_diagnostic.xUnit2013.severity = none
\ No newline at end of file
+dotnet_diagnostic.xUnit2013.severity = none
