diff --git a/src/Build/Evaluation/IItem.cs b/src/Build/Evaluation/IItem.cs
index dbeaaa67f08..ad6392828fc 100644
--- a/src/Build/Evaluation/IItem.cs
+++ b/src/Build/Evaluation/IItem.cs
@@ -53,6 +53,15 @@ string ProjectDirectory
         /// </summary>
         string GetMetadataValueEscaped(string name);
 
+        /// <summary>
+        /// Returns true if a particular piece of metadata is defined on this item (even if
+        /// its value is empty string) otherwise false.
+        /// This includes built-in metadata and metadata from item definitions.
+        /// </summary>
+        /// <remarks>
+        /// It has to include all of these because it's used for batching, which doesn't
+        /// care where the metadata originated.
+        /// </remarks>
         bool HasMetadata(string name);
     }
 }
