diff --git a/src/Tasks/WriteCodeFragment.cs b/src/Tasks/WriteCodeFragment.cs
index 6fc6270daf4..f6e07a804e5 100644
--- a/src/Tasks/WriteCodeFragment.cs
+++ b/src/Tasks/WriteCodeFragment.cs
@@ -449,7 +449,7 @@ private bool AddArguments(
                             value = ConvertParameterValueToInferredType(
                                 constructorParameterTypes[i],
                                 parameter.Value,
-                                parameter.PositionalParameterName); /* back to 1 based */
+                                parameter.PositionalParameterName); 
                         }
                         else
                         {
