diff --git a/src/Build/BackEnd/Node/OutOfProcServerNode.cs b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
index 624f6c0183a..2901298ed0a 100644
--- a/src/Build/BackEnd/Node/OutOfProcServerNode.cs
+++ b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
@@ -316,7 +316,7 @@ private void HandleServerShutdownCommand(NodeBuildComplete buildComplete)
             _shutdownEvent.Set();
         }
 
-        private static void HandleBuildCancel() 
+        private void HandleBuildCancel()
         {
             CommunicationsUtilities.Trace("Received request to cancel build running on MSBuild Server. MSBuild server will shutdown.}");
             _cancelRequested = true;
