diff --git a/src/Tasks/ManifestUtil/ManifestReader.cs b/src/Tasks/ManifestUtil/ManifestReader.cs
index d1a3676cac5..829bfec6961 100644
--- a/src/Tasks/ManifestUtil/ManifestReader.cs
+++ b/src/Tasks/ManifestUtil/ManifestReader.cs
@@ -104,7 +104,7 @@ private static Manifest ReadEmbeddedManifest(string path)
             return manifest;
         }
 
-        private static void ReadExist(FileStream stream, byte[] buffer, int offset, int count)
+        private static void ReadExist(Stream stream, byte[] buffer, int offset, int count)
         {
             while (count > 0)
             {
diff --git a/src/Tasks/StrongNameUtils.cs b/src/Tasks/StrongNameUtils.cs
index b24f5353b51..4c2ef5f2b3c 100644
--- a/src/Tasks/StrongNameUtils.cs
+++ b/src/Tasks/StrongNameUtils.cs
@@ -57,7 +57,7 @@ internal static void ReadKeyFile(TaskLoggingHelper log, string keyFile, out Stro
                     int offset = 0;
                     while (count > 0)
                     {
-                        int read = fs.Read(buffer, offset, count);
+                        int read = fs.Read(keyFileContents, offset, count);
                         if (read <= 0)
                         {
                             throw new EndOfStreamException();
