diff --git a/src/Tasks/GenerateDeploymentManifest.cs b/src/Tasks/GenerateDeploymentManifest.cs
index a19de20c07d..a9e8fae1d8b 100644
--- a/src/Tasks/GenerateDeploymentManifest.cs
+++ b/src/Tasks/GenerateDeploymentManifest.cs
@@ -141,11 +141,8 @@ private bool BuildResolvedSettings(DeployManifest manifest)
             }
             else if (String.IsNullOrEmpty(manifest.Publisher))
             {
-                string org = string.Empty;
-                if (NativeMethodsShared.IsWindows)
-                {
-                    org = Util.GetRegisteredOrganization();
-                }
+                string org = NativeMethodsShared.IsWindows ? Util.GetRegisteredOrganization() : string.Empty;
+
                 manifest.Publisher = !String.IsNullOrEmpty(org) ? org : manifest.Product;
             }
             Debug.Assert(!String.IsNullOrEmpty(manifest.Publisher));
