diff --git a/ref/Microsoft.Build.Tasks.Core/net/Microsoft.Build.Tasks.Core.cs b/ref/Microsoft.Build.Tasks.Core/net/Microsoft.Build.Tasks.Core.cs
index b269e54a881..c2152afe98a 100644
--- a/ref/Microsoft.Build.Tasks.Core/net/Microsoft.Build.Tasks.Core.cs
+++ b/ref/Microsoft.Build.Tasks.Core/net/Microsoft.Build.Tasks.Core.cs
@@ -226,7 +226,7 @@ protected CreateManifestResourceName() { }
         protected abstract string CreateManifestName(string fileName, string linkFileName, string rootNamespaceName, string dependentUponFileName, System.IO.Stream binaryStream);
         public override bool Execute() { throw null; }
         protected abstract bool IsSourceFile(string fileName);
-        public static string MakeValidEverettIdentifier(string name) { throw null; }
+        public static void MakeValidEverettIdentifier(System.Text.StringBuilder builder, string name) { }
     }
     public partial class CreateProperty : Microsoft.Build.Tasks.TaskExtension
     {
diff --git a/ref/Microsoft.Build.Tasks.Core/netstandard/Microsoft.Build.Tasks.Core.cs b/ref/Microsoft.Build.Tasks.Core/netstandard/Microsoft.Build.Tasks.Core.cs
index 1babf5ee2d4..4fbc33482af 100644
--- a/ref/Microsoft.Build.Tasks.Core/netstandard/Microsoft.Build.Tasks.Core.cs
+++ b/ref/Microsoft.Build.Tasks.Core/netstandard/Microsoft.Build.Tasks.Core.cs
@@ -156,7 +156,7 @@ protected CreateManifestResourceName() { }
         protected abstract string CreateManifestName(string fileName, string linkFileName, string rootNamespaceName, string dependentUponFileName, System.IO.Stream binaryStream);
         public override bool Execute() { throw null; }
         protected abstract bool IsSourceFile(string fileName);
-        public static string MakeValidEverettIdentifier(string name) { throw null; }
+        public static void MakeValidEverettIdentifier(System.Text.StringBuilder builder, string name) { }
     }
     public partial class CreateProperty : Microsoft.Build.Tasks.TaskExtension
     {
