diff --git a/src/Tasks.UnitTests/CodeTaskFactoryTests.cs b/src/Tasks.UnitTests/CodeTaskFactoryTests.cs
index 9ea47a8b181..3684ebf8bec 100644
--- a/src/Tasks.UnitTests/CodeTaskFactoryTests.cs
+++ b/src/Tasks.UnitTests/CodeTaskFactoryTests.cs
@@ -1047,9 +1047,8 @@ public override bool Execute()
         public void EmbedsSourceFileInBinlog()
         {
             string taskName = "HelloTask";
-            string targetName = "SayHello";
 
-            using TestEnvironment env = TestEnvironment.Create();
+            using var env = TestEnvironment.Create();
             var folder = env.CreateFolder(createFolder: true);
             var taskClass = env.CreateFile(folder, $"{taskName}.cs", $$"""
                 namespace InlineTask
@@ -1080,7 +1079,7 @@ public override bool Execute()
                     </Task>
                   </UsingTask>
 
-                    <Target Name="{targetName}">
+                    <Target Name="SayHello">
                         <{taskName} />
                     </Target>
 
@@ -1113,9 +1112,8 @@ public override bool Execute()
         public void EmbedsSourceFileInBinlogWhenFailsToCompile()
         {
             string taskName = "HelloTask";
-            string targetName = "SayHello";
 
-            using TestEnvironment env = TestEnvironment.Create();
+            using var env = TestEnvironment.Create();
             var folder = env.CreateFolder(createFolder: true);
             var taskClass = env.CreateFile(folder, $"{taskName}.cs", $$"""
                 namespace InlineTask
@@ -1139,7 +1137,7 @@ public class {{taskName}} : Task
                     </Task>
                   </UsingTask>
 
-                    <Target Name="{targetName}">
+                    <Target Name="SayHello">
                         <{taskName} />
                     </Target>
 
