diff --git a/src/Tasks/XamlTaskFactory/XamlTaskFactory.cs b/src/Tasks/XamlTaskFactory/XamlTaskFactory.cs
index a9195d3f89d..456d45273e2 100644
--- a/src/Tasks/XamlTaskFactory/XamlTaskFactory.cs
+++ b/src/Tasks/XamlTaskFactory/XamlTaskFactory.cs
@@ -113,7 +113,7 @@ public bool Initialize(string taskName, IDictionary<string, TaskPropertyInfo> ta
 
             CodeCompileUnit dom = generator.GenerateCode();
 
-            string pathToMSBuildBinaries = ToolLocationHelper.GetPathToBuildTools(ToolLocationHelper.CurrentToolsVersion);
+            string pathToMSBuildBinaries = ToolLocationHelper.GetPathToBuildDlls(ToolLocationHelper.CurrentToolsVersion);
 
             // create the code generator options    
             // Since we are running msbuild 12.0 these had better load.
diff --git a/src/Utilities/PublicAPI/net/PublicAPI.Unshipped.txt b/src/Utilities/PublicAPI/net/PublicAPI.Unshipped.txt
index e69de29bb2d..16c5137f456 100644
--- a/src/Utilities/PublicAPI/net/PublicAPI.Unshipped.txt
+++ b/src/Utilities/PublicAPI/net/PublicAPI.Unshipped.txt
@@ -0,0 +1 @@
+static Microsoft.Build.Utilities.ToolLocationHelper.GetPathToBuildDlls(string toolsVersion) -> string
\ No newline at end of file
diff --git a/src/Utilities/ToolLocationHelper.cs b/src/Utilities/ToolLocationHelper.cs
index baa7385f829..211dc94c255 100644
--- a/src/Utilities/ToolLocationHelper.cs
+++ b/src/Utilities/ToolLocationHelper.cs
@@ -3555,6 +3555,23 @@ private static string GetPathToWindowsSdkFile(string fileName, string pathToSdk,
             return filePath;
         }
 
+        /// <summary>
+        /// Given a ToolsVersion, return the path to the MSBuild dll's.
+        /// </summary>
+        /// <remarks>This was created for XamlToolTask, which only looks for the dll's surrounding msbuild.exe.
+        /// arm64 MSBuild specifically uses the dll's in the root folder.</remarks>
+        public static string GetPathToBuildDlls(string toolsVersion)
+        {
+            return toolsVersion switch
+            {
+                "2.0" => GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version20, UtilitiesDotNetFrameworkArchitecture.Current),
+                "3.5" => GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version35, UtilitiesDotNetFrameworkArchitecture.Current),
+                "4.0" => GetPathToDotNetFramework(TargetDotNetFrameworkVersion.Version40, UtilitiesDotNetFrameworkArchitecture.Current),
+                // When looking for MSBuild dll's under a VS installation, the root is the "canonical" location for all dll's.
+                _ => BuildEnvironmentHelper.Instance.MSBuildToolsDirectoryRoot,
+            };
+        }
+
         /// <summary>
         /// Given a ToolsVersion, return the path to the MSBuild tools for that ToolsVersion
         /// </summary>
