diff --git a/src/Framework/FileClassifier.cs b/src/Framework/FileClassifier.cs
index f6b8a3a3eb9..5d7512c0687 100644
--- a/src/Framework/FileClassifier.cs
+++ b/src/Framework/FileClassifier.cs
@@ -24,7 +24,7 @@ namespace Microsoft.Build.Framework
     ///         </list>
     ///     </para>
     /// </remarks>
-    internal sealed class FileClassifier
+    internal class FileClassifier     
     {
         private const StringComparison PathComparison = StringComparison.OrdinalIgnoreCase;
 
@@ -143,7 +143,7 @@ private static string EnsureTrailingSlash(string fileSpec)
     ///         Cache is add only. It does not updates already existing cached items.
     ///     </remarks>
     /// </summary>
-    internal sealed class ImmutableFilesTimestampCache
+    internal class ImmutableFilesTimestampCache
     {
         private readonly ConcurrentDictionary<string, DateTime> _cache = new(StringComparer.OrdinalIgnoreCase);
 
@@ -152,7 +152,6 @@ internal sealed class ImmutableFilesTimestampCache
         /// </summary>
         public static ImmutableFilesTimestampCache Shared { get; } = new();
 
-
         /// <summary>
         ///     Try get 'Last Write File Utc' time of particular file.
         /// </summary>
