diff --git a/src/Shared/Debugging/DebugUtils.cs b/src/Shared/Debugging/DebugUtils.cs
index 5fc8a67f991..aa83426e5af 100644
--- a/src/Shared/Debugging/DebugUtils.cs
+++ b/src/Shared/Debugging/DebugUtils.cs
@@ -26,7 +26,6 @@ static DebugUtils()
             if (Traits.Instance.DebugEngine)
             {
                 debugDirectory ??= Path.Combine(Directory.GetCurrentDirectory(), "MSBuild_Logs");
-                FileUtilities.EnsureDirectoryExists(debugDirectory);
 
                 // Out of proc nodes do not know the startup directory so set the environment variable for them.
                 if (string.IsNullOrWhiteSpace(environmentDebugPath))
@@ -35,6 +34,11 @@ static DebugUtils()
                 }
             }
 
+            if (debugDirectory is not null)
+            {
+                FileUtilities.EnsureDirectoryExists(debugDirectory);
+            }
+
             DebugPath = debugDirectory;
         }
 
