diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index 5c6f0be703e..12a3d83376e 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -2880,6 +2880,7 @@ private static void LogDeferredMessages(ILoggingService loggingService, IEnumera
             foreach (var message in deferredBuildMessages)
             {
                 loggingService.LogCommentFromText(BuildEventContext.Invalid, message.Importance, message.Text);
+
                 // If message includes a file path, include that file
                 if (message.FilePath is not null)
                 {
diff --git a/src/Build/BackEnd/Components/Logging/LoggingContext.cs b/src/Build/BackEnd/Components/Logging/LoggingContext.cs
index 3a24dcffcab..c8018767633 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingContext.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingContext.cs
@@ -269,6 +269,7 @@ internal void LogFatalBuildError(Exception exception, BuildEventFileInfo file)
             LoggingService.LogFatalBuildError(BuildEventContext, exception, file);
             _hasLoggedErrors = true;
         }
+
         /// <summary>
         /// Logs a file to be included in the binary logger
         /// </summary>
diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 07154a581da..2fa7bee90db 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1340,9 +1340,10 @@ string[] commandLine
                     }
 
                     // Log a message for every response file and include it in log
+                    List<BuildManager.DeferredBuildMessage> messagesToBuildInLoggersList = messagesToLogInBuildLoggers.ToList();
                     foreach (var responseFilePath in s_includedResponseFiles.ToList())
                     {
-                        messagesToLogInBuildLoggers = messagesToLogInBuildLoggers.Append(
+                        messagesToBuildInLoggersList.Add(
                             new BuildManager.DeferredBuildMessage(
                                 String.Format("Included response file: {0}", responseFilePath),
                                 MessageImportance.Normal,
@@ -1351,7 +1352,7 @@ string[] commandLine
                     }
 
 
-                    buildManager.BeginBuild(parameters, messagesToLogInBuildLoggers);
+                    buildManager.BeginBuild(parameters, messagesToBuildInLoggersList);
 
                     Exception exception = null;
                     try
