diff --git a/src/MSBuild/MSBuild.csproj b/src/MSBuild/MSBuild.csproj
index b09871fedb2..02bd03d97b6 100644
--- a/src/MSBuild/MSBuild.csproj
+++ b/src/MSBuild/MSBuild.csproj
@@ -188,7 +188,7 @@
     <Compile Include="NodeEndpointOutOfProcTaskHost.cs" />
     <Compile Include="LogMessagePacket.cs" />
     <Compile Include="NodeEndpointRarTaskHost.cs" />
-    <Compile Include="RarServiceTaskHostNode .cs" />
+    <Compile Include="RarServiceTaskHostNode.cs" />
     <Compile Include="ProjectSchemaValidationHandler.cs">
       <ExcludeFromStyleCop>true</ExcludeFromStyleCop>
     </Compile>
diff --git a/src/MSBuildTaskHost/MSBuildTaskHost.csproj b/src/MSBuildTaskHost/MSBuildTaskHost.csproj
index edf4ec79a30..25e4bab8f7b 100644
--- a/src/MSBuildTaskHost/MSBuildTaskHost.csproj
+++ b/src/MSBuildTaskHost/MSBuildTaskHost.csproj
@@ -46,7 +46,7 @@
       <Link>BuildEngineResult.cs</Link>
     </Compile>
     <Compile Include="..\Framework\TaskExecutionContext.cs">
-      <Link>BuildEngineResult.cs</Link>
+      <Link>TaskExecutionContext.cs</Link>
     </Compile>
     <Compile Include="..\Framework\IBuildEngine3.cs">
       <Link>IBuildEngine3.cs</Link>
