diff --git a/src/Shared/XMakeElements.cs b/src/Shared/XMakeElements.cs
index 327fb8c9aa5..dee836adeb3 100644
--- a/src/Shared/XMakeElements.cs
+++ b/src/Shared/XMakeElements.cs
@@ -38,7 +38,7 @@ internal static class XMakeElements
         internal static char[] InvalidTargetNameCharacters => [ '$', '@', '(', ')', '%', '*', '?', '.' ];
 
         // Names that cannot be used as property or item names because they are reserved
-        internal static readonly HashSet<string> ReservedItemNames =
+        internal static HashSet<string> ReservedItemNames =>
         [
             // project, "Project" is not reserved, because unfortunately ProjectReference items already use it as metadata name.
             visualStudioProject,
