diff --git a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
index bc54fc6227c..95b0a377366 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs
@@ -5,7 +5,7 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Globalization;
-using System.IO.Pipes;
+using System.IO;
 #if FEATURE_APPDOMAIN
 using System.Runtime.Remoting.Lifetime;
 using System.Runtime.Remoting;
@@ -1005,9 +1005,9 @@ string IRarBuildEngine.GetRarPipeName()
         }
 
         /// <summary>
-        /// Constructs <seealso cref="NamedPipeClientStream"/>
+        /// Constructs <seealso cref="Stream"/> for RAR client
         /// </summary>
-        NamedPipeClientStream IRarBuildEngine.GetRarClientStream(string pipeName, int timeout)
+        Stream IRarBuildEngine.GetRarClientStream(string pipeName, int timeout)
         {
             return NamedPipeUtil.TryConnectToProcess(pipeName, timeout, null);
         }
diff --git a/src/Framework/IRarBuildEngine.cs b/src/Framework/IRarBuildEngine.cs
index b5178887550..12861e2b296 100644
--- a/src/Framework/IRarBuildEngine.cs
+++ b/src/Framework/IRarBuildEngine.cs
@@ -1,7 +1,7 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System.IO.Pipes;
+using System.IO;
 
 namespace Microsoft.Build.Framework
 {
@@ -21,8 +21,8 @@ internal interface IRarBuildEngine
         internal string GetRarPipeName();
 
         /// <summary>
-        /// Constructs <seealso cref="NamedPipeClientStream"/>
+        /// Constructs <seealso cref="Stream"/>
         /// </summary>
-        internal NamedPipeClientStream GetRarClientStream(string pipeName, int timeout);
+        internal Stream GetRarClientStream(string pipeName, int timeout);
     }
 }
