diff --git a/.vsts-dotnet.yml b/.vsts-dotnet.yml
index 04c19b31687..273de1b8036 100644
--- a/.vsts-dotnet.yml
+++ b/.vsts-dotnet.yml
@@ -82,7 +82,7 @@ stages:
         command: restore
         feedsToUse: config
         restoreSolution: 'eng\common\internal\Tools.csproj'
-        nugetConfigPath: 'eng\common\internal\NuGet.config'
+        nugetConfigPath: 'NuGet.config'
         restoreDirectory: '$(Build.SourcesDirectory)\.packages'
 
     - task: ms-vseng.MicroBuildTasks.30666190-6959-11e5-9f96-f56098202fef.MicroBuildSigningPlugin@1
diff --git a/eng/Version.Details.xml b/eng/Version.Details.xml
index 17f1cac3020..ea69077a58b 100644
--- a/eng/Version.Details.xml
+++ b/eng/Version.Details.xml
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Dependencies>
   <ToolsetDependencies>
-    <Dependency Name="Microsoft.DotNet.Arcade.Sdk" Version="6.0.0-beta.22377.7">
+    <Dependency Name="Microsoft.DotNet.Arcade.Sdk" Version="6.0.0-beta.22378.6">
       <Uri>https://github.com/dotnet/arcade</Uri>
       <Sha>b9f4ee617d6976d0937dfbad71481c685052d256</Sha>
       <SourceBuild RepoName="arcade" ManagedOnly="true" />
@@ -14,7 +14,7 @@
       <Uri>https://github.com/dotnet/roslyn</Uri>
       <Sha>29e657c0582904529bae2a87c227220e03f509cf</Sha>
     </Dependency>
-    <Dependency Name="Microsoft.DotNet.XUnitExtensions" Version="6.0.0-beta.22377.7">
+    <Dependency Name="Microsoft.DotNet.XUnitExtensions" Version="6.0.0-beta.22378.6">
       <Uri>https://github.com/dotnet/arcade</Uri>
       <Sha>b9f4ee617d6976d0937dfbad71481c685052d256</Sha>
     </Dependency>
diff --git a/eng/Versions.props b/eng/Versions.props
index 9bfbcbd35a4..29c0e5a1cb2 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -42,7 +42,7 @@
          Otherwise, this version of dotnet will not be installed and the build will error out. -->
     <DotNetCliVersion>$([System.Text.RegularExpressions.Regex]::Match($([System.IO.File]::ReadAllText('$(MSBuildThisFileDirectory)..\global.json')), '"dotnet": "([^"]*)"').Groups.get_Item(1))</DotNetCliVersion>
     <MicrosoftCodeAnalysisCollectionsVersion>4.2.0-1.22102.8</MicrosoftCodeAnalysisCollectionsVersion>
-    <MicrosoftDotNetXUnitExtensionsVersion>6.0.0-beta.22377.7</MicrosoftDotNetXUnitExtensionsVersion>
+    <MicrosoftDotNetXUnitExtensionsVersion>6.0.0-beta.22378.6</MicrosoftDotNetXUnitExtensionsVersion>
     <MicrosoftExtensionsDependencyModelVersion>6.0.0-preview.2.21154.6</MicrosoftExtensionsDependencyModelVersion>
     <MicrosoftIORedistVersion>6.0.0</MicrosoftIORedistVersion>
     <MicrosoftNetCompilersToolsetVersion>4.3.0-3.22329.30</MicrosoftNetCompilersToolsetVersion>
diff --git a/global.json b/global.json
index 1667f6a5bab..80f82656174 100644
--- a/global.json
+++ b/global.json
@@ -10,6 +10,6 @@
   },
   "msbuild-sdks": {
     "Microsoft.Build.CentralPackageVersions": "2.0.1",
-    "Microsoft.DotNet.Arcade.Sdk": "6.0.0-beta.22377.7"
+    "Microsoft.DotNet.Arcade.Sdk": "6.0.0-beta.22378.6"
   }
 }
