diff --git a/src/Shared/ExceptionHandling.cs b/src/Shared/ExceptionHandling.cs
index 410bc5e0991..0d2d75ab274 100644
--- a/src/Shared/ExceptionHandling.cs
+++ b/src/Shared/ExceptionHandling.cs
@@ -372,9 +372,8 @@ internal static void DumpExceptionToFile(Exception ex)
                     }
                 }
             }
-            catch (OutOfMemoryException)
+            catch
             {
-                throw;
             }
         }
 
