diff --git a/src/Build.UnitTests/Construction/SolutionFile_Tests.cs b/src/Build.UnitTests/Construction/SolutionFile_Tests.cs
index bbe3952f98c..06189e92ff1 100644
--- a/src/Build.UnitTests/Construction/SolutionFile_Tests.cs
+++ b/src/Build.UnitTests/Construction/SolutionFile_Tests.cs
@@ -2508,6 +2508,21 @@ public void TryParseConfigurationPlatform(string input, bool isPlatformRequired,
             false,
             0,
             0)]
+        [InlineData(
+            //// Too many blank lines at the start
+            """
+
+
+
+
+            Microsoft Visual Studio Solution File, Format Version 12.00
+            # Visual Studio Version 17
+            VisualStudioVersion = 17.0.31903.59
+            MinimumVisualStudioVersion = 17.0.31903.59
+            """,
+            false,
+            0,
+            0)]
         public void GetSolutionFileAndVisualStudioMajorVersions(string content, bool expectedSuccess, int expectedSolutionVersion, int expectedVisualStudioMajorVersion)
         {
             string solutionPath = NativeMethodsShared.IsWindows ? "c:\\foo.sln" : "/foo.sln";
