diff --git a/src/Tasks.UnitTests/Microsoft.Build.Tasks.UnitTests.csproj b/src/Tasks.UnitTests/Microsoft.Build.Tasks.UnitTests.csproj
index 723c1932c9f..81b5048f0f7 100644
--- a/src/Tasks.UnitTests/Microsoft.Build.Tasks.UnitTests.csproj
+++ b/src/Tasks.UnitTests/Microsoft.Build.Tasks.UnitTests.csproj
@@ -158,8 +158,5 @@
     <None Update="TestResources\mycert.pfx">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
-    <None Update="TestResources\sample.zip">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
   </ItemGroup>
 </Project>
diff --git a/src/Tasks.UnitTests/TestResources/sample.zip b/src/Tasks.UnitTests/TestResources/sample.zip
deleted file mode 100644
index 95ba2b3d3ac..00000000000
Binary files a/src/Tasks.UnitTests/TestResources/sample.zip and /dev/null differ
