diff --git a/src/Shared/CoreCLRAssemblyLoader.cs b/src/Shared/CoreCLRAssemblyLoader.cs
index a7848f6f4fb..e8f89cf533d 100644
--- a/src/Shared/CoreCLRAssemblyLoader.cs
+++ b/src/Shared/CoreCLRAssemblyLoader.cs
@@ -26,9 +26,9 @@ internal sealed class CoreClrAssemblyLoader
         private static string _msbuildDirPath;
         private static readonly Version _currentAssemblyVersion = new Version(Microsoft.Build.Shared.MSBuildConstants.CurrentAssemblyVersion);
 
-        internal CoreClrAssemblyLoader()
+        static CoreClrAssemblyLoader()
         {
-            _msbuildDirPath = FileUtilities.NormalizePath(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath);
+            _msbuildDirPath = FileUtilities.NormalizePath(typeof(CoreClrAssemblyLoader).Assembly.Location);
             _msbuildDirPath = Path.GetDirectoryName(_msbuildDirPath);
         }
 
@@ -63,7 +63,8 @@ public Assembly LoadFromPath(string fullPath)
             // the load is part of the platform, and load it using the Default ALC.
             string assemblyDir = Path.GetDirectoryName(fullPath);
 
-            if (Traits.Instance.EscapeHatches.UseSingleLoadContext || string.Equals(assemblyDir, _msbuildDirPath))
+            if (Traits.Instance.EscapeHatches.UseSingleLoadContext ||
+                FileUtilities.ComparePathsNoThrow(assemblyDir, _msbuildDirPath, string.Empty))
             {
                 return LoadUsingLegacyDefaultContext(fullPath);
             }
