diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index d6eaa69285d..70a2d359a04 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -836,6 +836,8 @@ public GraphBuildSubmission PendBuildRequest(GraphBuildRequestData requestData)
 
                 if (KnownTelemetry.BuildTelemetry != null)
                 {
+                    // Project graph can have multiple entry points, for purposes of identifying event for same build project,
+                    // we believe that including only one entry point will provide enough precision.
                     KnownTelemetry.BuildTelemetry.Project ??= requestData.ProjectGraphEntryPoints?.FirstOrDefault().ProjectFile;
                     KnownTelemetry.BuildTelemetry.Target ??= string.Join(",", requestData.TargetNames);
                 }
