diff --git a/src/Build.UnitTests/BackEnd/SdkResolverService_Tests.cs b/src/Build.UnitTests/BackEnd/SdkResolverService_Tests.cs
index 65b7ddf7db5..65e21871880 100644
--- a/src/Build.UnitTests/BackEnd/SdkResolverService_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/SdkResolverService_Tests.cs
@@ -94,7 +94,7 @@ public void AssertResolutionWarnsIfResolvedVersionIsDifferentFromReferencedVersi
 
             result.Path.ShouldBe("path");
 
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
 
             _logger.WarningCount.ShouldBe(1);
             _logger.Warnings.First().Code.ShouldStartWith("MSB4241");
@@ -125,7 +125,7 @@ public void AssertSecondResolverWithPatternCanResolve()
             var result = SdkResolverService.Instance.ResolveSdk(BuildEventContext.InvalidSubmissionId, sdk, _loggingContext, new MockElementLocation("file"), "sln", "projectPath", interactive: false, isRunningInVisualStudio: false, failOnUnresolvedSdk: true);
 
             result.Path.ShouldBe("resolverpathwithresolvablesdkpattern2");
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain("MockSdkResolverWithResolvableSdkPattern2 running");
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldNotContain("MockSdkResolver1 running");
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldNotContain("MockSdkResolver2 running");
@@ -143,7 +143,7 @@ public void AssertFirstResolverCanResolve()
             var result = SdkResolverService.Instance.ResolveSdk(BuildEventContext.InvalidSubmissionId, sdk, _loggingContext, new MockElementLocation("file"), "sln", "projectPath", interactive: false, isRunningInVisualStudio: false, failOnUnresolvedSdk: true);
 
             result.Path.ShouldBe("resolverpath1");
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain("MockSdkResolver1 running");
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldNotContain("MockSdkResolverWithResolvableSdkPattern1 running");
         }
@@ -165,7 +165,7 @@ public void AssertFirstResolverWithPatternCantResolveChangeWave17_4()
                 var result = SdkResolverService.Instance.ResolveSdk(BuildEventContext.InvalidSubmissionId, sdk, _loggingContext, new MockElementLocation("file"), "sln", "projectPath", interactive: false, isRunningInVisualStudio: false, failOnUnresolvedSdk: true);
 
                 result.Path.ShouldBe("resolverpath1");
-                _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+                _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
                 _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain("MockSdkResolver1 running");
                 _logger.BuildMessageEvents.Select(i => i.Message).ShouldNotContain("MockSdkResolverWithResolvableSdkPattern1 running");
                 ChangeWaves.ResetStateForTests();
@@ -184,7 +184,7 @@ public void AssertFirstResolverWithPatternCanResolve()
             var result = SdkResolverService.Instance.ResolveSdk(BuildEventContext.InvalidSubmissionId, sdk, _loggingContext, new MockElementLocation("file"), "sln", "projectPath", interactive: false, isRunningInVisualStudio: false, failOnUnresolvedSdk: true);
 
             result.Path.ShouldBe("resolverpathwithresolvablesdkpattern1");
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain("MockSdkResolverWithResolvableSdkPattern1 running");
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldNotContain("MockSdkResolver1 running");
         }
@@ -201,7 +201,7 @@ public void AssertFirstResolverErrorsSupressedWhenResolved()
             var result = SdkResolverService.Instance.ResolveSdk(BuildEventContext.InvalidSubmissionId, sdk, _loggingContext, new MockElementLocation("file"), "sln", "projectPath", interactive: false, isRunningInVisualStudio: false, failOnUnresolvedSdk: true);
 
             result.Path.ShouldBe("resolverpath2");
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
 
             // Both resolvers should run, and no ERROR string.
             _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain("MockSdkResolver1 running");
@@ -372,7 +372,7 @@ public void SdkResolverCanReturnNoPaths(bool includePropertiesAndItems)
 
             ValidateExpectedPropertiesAndItems(includePropertiesAndItems, result);
 
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.WarningCount.ShouldBe(0);
         }
 
@@ -409,7 +409,7 @@ public void SdkResultCanReturnPropertiesAndItems()
 
             ValidateExpectedPropertiesAndItems(true, result);
 
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.WarningCount.ShouldBe(0);
         }
 
@@ -462,7 +462,7 @@ public void SdkResultCanReturnMultiplePaths(bool includePropertiesAndItems)
 
             ValidateExpectedPropertiesAndItems(includePropertiesAndItems, result);
 
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.WarningCount.ShouldBe(0);
         }
 
@@ -508,7 +508,7 @@ public void AssertResolutionWarnsIfResolvedVersionIsDifferentFromReferencedVersi
 
             ValidateExpectedPropertiesAndItems(true, result);
 
-            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SuccededToResolveSDK", sdk.ToString()));
+            _logger.BuildMessageEvents.Select(i => i.Message).ShouldContain(ResourceUtilities.FormatResourceStringStripCodeAndKeyword("SucceededToResolveSDK", sdk.ToString()));
             _logger.WarningCount.ShouldBe(1);
             _logger.Warnings.First().Code.ShouldStartWith("MSB4241");
         }
diff --git a/src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs b/src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs
index 0b61eff04f4..9d847022d74 100644
--- a/src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs
+++ b/src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs
@@ -355,7 +355,7 @@ private bool TryResolveSdkUsingSpecifiedResolvers(
 
                 if (result.Success)
                 {
-                    loggingContext.LogComment(MessageImportance.Low, "SuccededToResolveSDK", sdk.ToString());
+                    loggingContext.LogComment(MessageImportance.Low, "SucceededToResolveSDK", sdk.ToString());
 
                     LogWarnings(loggingContext, sdkReferenceLocation, result.Warnings);
 
diff --git a/src/Build/Resources/Strings.resx b/src/Build/Resources/Strings.resx
index cea0f831182..fe1bdce18e3 100644
--- a/src/Build/Resources/Strings.resx
+++ b/src/Build/Resources/Strings.resx
@@ -1324,7 +1324,7 @@
     <value>Could not resolve SDK "{0}". Exactly one of the probing messages below indicates why we could not resolve the SDK. Investigate and resolve that message to correctly specify the SDK.
   {1}</value>
   </data>
-  <data name="SuccededToResolveSDK" xml:space="preserve">
+  <data name="SucceededToResolveSDK" xml:space="preserve">
     <value>SDK "{0}" successfully resolved.</value>
   </data>
   <data name="CouldNotRunNuGetSdkResolver" xml:space="preserve">
diff --git a/src/Build/Resources/xlf/Strings.cs.xlf b/src/Build/Resources/xlf/Strings.cs.xlf
index 6b8e731127d..b565268b2f6 100644
--- a/src/Build/Resources/xlf/Strings.cs.xlf
+++ b/src/Build/Resources/xlf/Strings.cs.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.de.xlf b/src/Build/Resources/xlf/Strings.de.xlf
index bfc208782d4..ef3da8e53fe 100644
--- a/src/Build/Resources/xlf/Strings.de.xlf
+++ b/src/Build/Resources/xlf/Strings.de.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.es.xlf b/src/Build/Resources/xlf/Strings.es.xlf
index 1e2aa3990f2..73cf5de3092 100644
--- a/src/Build/Resources/xlf/Strings.es.xlf
+++ b/src/Build/Resources/xlf/Strings.es.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.fr.xlf b/src/Build/Resources/xlf/Strings.fr.xlf
index 5ad132a95a6..e7eb053a796 100644
--- a/src/Build/Resources/xlf/Strings.fr.xlf
+++ b/src/Build/Resources/xlf/Strings.fr.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.it.xlf b/src/Build/Resources/xlf/Strings.it.xlf
index 4034bc8bfe1..83f1d4dab5d 100644
--- a/src/Build/Resources/xlf/Strings.it.xlf
+++ b/src/Build/Resources/xlf/Strings.it.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.ja.xlf b/src/Build/Resources/xlf/Strings.ja.xlf
index 566f2fda55d..8ead006fd8f 100644
--- a/src/Build/Resources/xlf/Strings.ja.xlf
+++ b/src/Build/Resources/xlf/Strings.ja.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.ko.xlf b/src/Build/Resources/xlf/Strings.ko.xlf
index 63eb232629c..e02f39b5c48 100644
--- a/src/Build/Resources/xlf/Strings.ko.xlf
+++ b/src/Build/Resources/xlf/Strings.ko.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.pl.xlf b/src/Build/Resources/xlf/Strings.pl.xlf
index a6833fc6176..dd1b162b97b 100644
--- a/src/Build/Resources/xlf/Strings.pl.xlf
+++ b/src/Build/Resources/xlf/Strings.pl.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.pt-BR.xlf b/src/Build/Resources/xlf/Strings.pt-BR.xlf
index 6a488f3f418..e87e2f2185a 100644
--- a/src/Build/Resources/xlf/Strings.pt-BR.xlf
+++ b/src/Build/Resources/xlf/Strings.pt-BR.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.ru.xlf b/src/Build/Resources/xlf/Strings.ru.xlf
index 78114c910b7..9af788bc454 100644
--- a/src/Build/Resources/xlf/Strings.ru.xlf
+++ b/src/Build/Resources/xlf/Strings.ru.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.tr.xlf b/src/Build/Resources/xlf/Strings.tr.xlf
index d22708ed8cb..fc27ddc2af5 100644
--- a/src/Build/Resources/xlf/Strings.tr.xlf
+++ b/src/Build/Resources/xlf/Strings.tr.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.zh-Hans.xlf b/src/Build/Resources/xlf/Strings.zh-Hans.xlf
index 882b685f4d0..9e086bd1aa8 100644
--- a/src/Build/Resources/xlf/Strings.zh-Hans.xlf
+++ b/src/Build/Resources/xlf/Strings.zh-Hans.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
diff --git a/src/Build/Resources/xlf/Strings.zh-Hant.xlf b/src/Build/Resources/xlf/Strings.zh-Hant.xlf
index 42ff379309b..a74c5134842 100644
--- a/src/Build/Resources/xlf/Strings.zh-Hant.xlf
+++ b/src/Build/Resources/xlf/Strings.zh-Hant.xlf
@@ -451,7 +451,7 @@
       LOCALIZATION: {0} is a file, {1} and {2} are semicolon delimited lists of messages
     </note>
       </trans-unit>
-      <trans-unit id="SuccededToResolveSDK">
+      <trans-unit id="SucceededToResolveSDK">
         <source>SDK "{0}" successfully resolved.</source>
         <target state="new">SDK "{0}" successfully resolved.</target>
         <note />
