diff --git a/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs b/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs
index 792a7500088..d35996d1640 100644
--- a/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs
+++ b/src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs
@@ -7638,7 +7638,7 @@ public void DoNotAssumeFilesDescribedByRedistListExistOnDisk()
         }
 
         [Fact]
-        [Trait("Category", "netcore-linux-failing")]
+        [Trait("Category", "mono-osx-failing")]
         public void HandleFilesInSearchPathsWhichDiffersOnlyInCasing()
         {
             string redistListPath = CreateGenericRedistList();
