diff --git a/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs b/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs
index 54373593d2b..935d146d443 100644
--- a/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs
+++ b/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs
@@ -210,7 +210,6 @@ public partial interface IBuildEngine6 : Microsoft.Build.Framework.IBuildEngine,
     public partial interface IBuildEngine7 : Microsoft.Build.Framework.IBuildEngine, Microsoft.Build.Framework.IBuildEngine2, Microsoft.Build.Framework.IBuildEngine3, Microsoft.Build.Framework.IBuildEngine4, Microsoft.Build.Framework.IBuildEngine5, Microsoft.Build.Framework.IBuildEngine6
     {
         bool AllowFailureWithoutError { get; set; }
-        void BlockingWaitForCore();
         void ReleaseCores(int coresToRelease);
         int RequestCores(int requestedCores);
     }
diff --git a/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs b/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs
index 693b593d9c9..b77ffbb111d 100644
--- a/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs
+++ b/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs
@@ -210,7 +210,6 @@ public partial interface IBuildEngine6 : Microsoft.Build.Framework.IBuildEngine,
     public partial interface IBuildEngine7 : Microsoft.Build.Framework.IBuildEngine, Microsoft.Build.Framework.IBuildEngine2, Microsoft.Build.Framework.IBuildEngine3, Microsoft.Build.Framework.IBuildEngine4, Microsoft.Build.Framework.IBuildEngine5, Microsoft.Build.Framework.IBuildEngine6
     {
         bool AllowFailureWithoutError { get; set; }
-        void BlockingWaitForCore();
         void ReleaseCores(int coresToRelease);
         int RequestCores(int requestedCores);
     }
diff --git a/src/Framework/IBuildEngine7.cs b/src/Framework/IBuildEngine7.cs
index cb7bc8688d2..ca82448e707 100644
--- a/src/Framework/IBuildEngine7.cs
+++ b/src/Framework/IBuildEngine7.cs
@@ -24,11 +24,5 @@ public interface IBuildEngine7 : IBuildEngine6
         /// </summary>
         /// <param name="coresToRelease">Number of cores no longer in use.</param>
         void ReleaseCores(int coresToRelease);
-
-        /// <summary>
-        /// Block until a "core" from the managed pool for this build is available.
-        /// </summary>
-        void BlockingWaitForCore();
-
     }
 }
diff --git a/src/Tasks/SemaphoreCPUTask.cs b/src/Tasks/SemaphoreCPUTask.cs
index 7c8ca923dbd..9b2cd9cc63d 100644
--- a/src/Tasks/SemaphoreCPUTask.cs
+++ b/src/Tasks/SemaphoreCPUTask.cs
@@ -61,7 +61,7 @@ void LaunchAndComplete(int i, Action completionCallback)
         {
             Stopwatch s = new Stopwatch();
             s.Start();
-            BuildEngine7.BlockingWaitForCore();
+            BuildEngine7.RequestCores(1);
             Log.LogMessageFromText($"Action {i} started from {System.Diagnostics.Process.GetCurrentProcess().Id}, waited {s.Elapsed}", Framework.MessageImportance.High);
             Thread.Sleep(2_000);
             Log.LogMessageFromText($"Action {i} completed from {System.Diagnostics.Process.GetCurrentProcess().Id}, total {s.Elapsed}", Framework.MessageImportance.High);
