diff --git a/src/Build/Logging/TerminalLogger/TerminalLogger.cs b/src/Build/Logging/TerminalLogger/TerminalLogger.cs
index 5c6cc00e4a5..ecf6119332b 100644
--- a/src/Build/Logging/TerminalLogger/TerminalLogger.cs
+++ b/src/Build/Logging/TerminalLogger/TerminalLogger.cs
@@ -885,7 +885,6 @@ private void MessageRaised(object sender, BuildMessageEventArgs e)
 
             if (Verbosity > LoggerVerbosity.Quiet)
             {
-                // Show immediate messages to the user.
                 if (e.Code == "NETSDK1057" && !_loggedPreviewMessage)
                 {
                     // The SDK will log the high-pri "not-a-warning" message NETSDK1057
