diff --git a/eng/dependabot/Packages.props b/eng/dependabot/Packages.props
index 2c9a1ee8d64..87cf29fd2d2 100644
--- a/eng/dependabot/Packages.props
+++ b/eng/dependabot/Packages.props
@@ -19,6 +19,9 @@
     <PackageVersion Include="LargeAddressAware" Version="1.0.5" />
     <PackageVersion Update="LargeAddressAware" Condition="'$(LargeAddressAwareVersion)' != ''" Version="$(LargeAddressAwareVersion)" />
 
+    <PackageVersion Include="Microsoft.CodeCoverage" Version="17.12.0" />
+    <PackageVersion Update="Microsoft.CodeCoverage" Condition="'$(MicrosoftCodeCoverageVersion)' != ''" Version="$(MicrosoftCodeCoverageVersion)" />
+
     <PackageVersion Include="Microsoft.BuildXL.Processes" Version="0.1.0-20241105.3" />
     <PackageVersion Update="Microsoft.BuildXL.Processes" Condition="'$(BuildXLProcessesVersion)' != ''" Version="$(BuildXLProcessesVersion)" />
 
diff --git a/src/Directory.Build.targets b/src/Directory.Build.targets
index ea5bc853cbb..2321a7b03e5 100644
--- a/src/Directory.Build.targets
+++ b/src/Directory.Build.targets
@@ -52,6 +52,7 @@
     <PackageReference Include="Newtonsoft.Json" />
 
     <PackageReference Include="Microsoft.DotNet.XUnitExtensions" />
+    <PackageReference Include="Microsoft.CodeCoverage" />
 
     <!-- Don't localize unit test projects -->
     <PackageReference Remove="XliffTasks" />
