diff --git a/src/Build.UnitTests/Collections/CopyOnReadEnumerable_Tests.cs b/src/Build.UnitTests/Collections/CopyOnReadEnumerable_Tests.cs
index 3cc9f46d848..03c8e352ee7 100644
--- a/src/Build.UnitTests/Collections/CopyOnReadEnumerable_Tests.cs
+++ b/src/Build.UnitTests/Collections/CopyOnReadEnumerable_Tests.cs
@@ -28,6 +28,7 @@ public void EnumeratesBackingCollection()
                     enumerator.MoveNext();
                     enumerator.Current.ShouldBe(i);
                 }
+                enumerator.MoveNext().ShouldBeFalse();
             }
         }
 
@@ -38,22 +39,18 @@ public void CopiesBackingCollection()
 
             CopyOnReadEnumerable<string> enumerable = new CopyOnReadEnumerable<string>(values, values);
 
-            int count1 = 0;
-            using (IEnumerator<string> enumerator = values.GetEnumerator())
-            {
-                count1++;
-            }
-            count1.ShouldBe(values.Count);
+            int expectedCount = values.Count;
+            var enumerator = enumerable.GetEnumerator();
 
             // The list has been copied and adding to it has no effect on the enumerable.
             values.Add("d");
 
-            int count2 = 0;
-            using (IEnumerator<string> enumerator = values.GetEnumerator())
+            int actualCount = 0;
+            while (enumerator.MoveNext())
             {
-                count2++;
+                actualCount++;
             }
-            count2.ShouldBe(count1);
+            actualCount.ShouldBe(expectedCount);
         }
     }
 }
