diff --git a/src/MSBuild/LiveLogger/LiveLogger.cs b/src/MSBuild/LiveLogger/LiveLogger.cs
index 7edd3adc977..7772b29a753 100644
--- a/src/MSBuild/LiveLogger/LiveLogger.cs
+++ b/src/MSBuild/LiveLogger/LiveLogger.cs
@@ -377,16 +377,11 @@ private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
                             // Ignore any GetDirectoryName exceptions.
                         }
 
-                        string urlString;
-                        try
-                        {
-                            // This should generate file:// schema url string which is better handled by various Terminal clients than raw folder name.
-                            urlString = new Uri(url.ToString()).AbsoluteUri;
-                        }
-                        catch
+                        // Generates file:// schema url string which is better handled by various Terminal clients than raw folder name.
+                        string urlString = url.ToString();
+                        if (Uri.TryCreate(urlString, UriKind.Absolute, out Uri? uri))
                         {
-                            // If Uri constructor throws use raw folder name instead.
-                            urlString = url.ToString();
+                            urlString = uri.AbsoluteUri;
                         }
 
                         Terminal.WriteLine(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("ProjectFinished_OutputPath",
