diff --git a/src/Framework/MSBuildEventSource.cs b/src/Framework/MSBuildEventSource.cs
index c624cc58e25..c352ddde8f1 100644
--- a/src/Framework/MSBuildEventSource.cs
+++ b/src/Framework/MSBuildEventSource.cs
@@ -551,49 +551,49 @@ public void OutOfProcSdkResolverServiceRequestSdkPathFromMainNodeStop(int submis
             WriteEvent(75, submissionId, sdkName, solutionPath, projectPath, success, wasResultCached);
         }
 
-        [Event(76, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(76, Keywords = Keywords.All)]
         public void ProjectCacheCreatePluginInstanceStart(string pluginAssemblyPath)
         {
             WriteEvent(76, pluginAssemblyPath);
         }
 
-        [Event(77, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(77, Keywords = Keywords.All)]
         public void ProjectCacheCreatePluginInstanceStop(string pluginAssemblyPath, string pluginTypeName)
         {
             WriteEvent(77, pluginAssemblyPath, pluginTypeName);
         }
 
-        [Event(78, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(78, Keywords = Keywords.All)]
         public void ProjectCacheBeginBuildStart(string pluginTypeName)
         {
             WriteEvent(78, pluginTypeName);
         }
 
-        [Event(79, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(79, Keywords = Keywords.All)]
         public void ProjectCacheBeginBuildStop(string pluginTypeName)
         {
             WriteEvent(79, pluginTypeName);
         }
 
-        [Event(80, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(80, Keywords = Keywords.All)]
         public void ProjectCacheGetCacheResultStart(string pluginTypeName, string projectPath, string targets)
         {
             WriteEvent(80, pluginTypeName, projectPath, targets);
         }
 
-        [Event(81, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(81, Keywords = Keywords.All)]
         public void ProjectCacheGetCacheResultStop(string pluginTypeName, string projectPath, string targets, string cacheResultType)
         {
             WriteEvent(81, pluginTypeName, projectPath, targets, cacheResultType);
         }
 
-        [Event(82, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(82, Keywords = Keywords.All)]
         public void ProjectCacheEndBuildStart(string pluginTypeName)
         {
             WriteEvent(82, pluginTypeName);
         }
 
-        [Event(83, Keywords = Keywords.All | Keywords.PerformanceLog)]
+        [Event(83, Keywords = Keywords.All)]
         public void ProjectCacheEndBuildStop(string pluginTypeName)
         {
             WriteEvent(83, pluginTypeName);
