diff --git a/src/Tasks/Exec.cs b/src/Tasks/Exec.cs
index 5598252baaa..a8e885b10f5 100644
--- a/src/Tasks/Exec.cs
+++ b/src/Tasks/Exec.cs
@@ -590,6 +590,8 @@ protected internal override void AddCommandLineCommands(CommandLineBuilderExtens
             {
                 commandLine.AppendSwitch("-c");
                 commandLine.AppendTextUnquoted(" \"");
+                // Set the locale to 'Computer C' English UTF-8 output.
+                // This makes the task output the same text regardless of the configured locale.
                 commandLine.AppendTextUnquoted("export LANG=C.UTF-8; export LC_ALL=C.UTF-8; . ");
                 commandLine.AppendFileNameIfNotNull(batchFileForCommandLine);
                 commandLine.AppendTextUnquoted("\"");
