diff --git a/src/Build.UnitTests/EscapingInProjects_Tests.cs b/src/Build.UnitTests/EscapingInProjects_Tests.cs
index cd356698448..2b6c9891649 100644
--- a/src/Build.UnitTests/EscapingInProjects_Tests.cs
+++ b/src/Build.UnitTests/EscapingInProjects_Tests.cs
@@ -912,8 +912,6 @@ public class FullProjectsUsingMicrosoftCommonTargets
     {
         private readonly ITestOutputHelper _testOutput;
 
-        [Fact(Skip = "skipped for failing tests when reviving pipeline")]
-
         public FullProjectsUsingMicrosoftCommonTargets(ITestOutputHelper output)
         {
             _testOutput = output;
