diff --git a/src/Build/BackEnd/BuildManager/BuildManager.cs b/src/Build/BackEnd/BuildManager/BuildManager.cs
index 98009f123d8..ec093612169 100644
--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -2950,7 +2950,7 @@ private void OnProjectStarted(object sender, ProjectStartedEventArgs e)
             });
         }
 
-        public BuildCheckBinaryLogReplaySourcerWrapper GetBuildCheckBinaryLogReplayEventSourceWrapper(
+        public IBinaryLogReplaySource GetBuildCheckBinaryLogReplayEventSourceWrapper(
             BinaryLogReplayEventSource replayEventSource)
         {
             _buildParameters = new BuildParameters
diff --git a/src/Build/BuildCheck/Infrastructure/BuildCheckBinaryLogReplaySourcerWrapper.cs b/src/Build/BuildCheck/Infrastructure/BuildCheckBinaryLogReplaySourcerWrapper.cs
index ad9e61fafc4..2d675db0527 100644
--- a/src/Build/BuildCheck/Infrastructure/BuildCheckBinaryLogReplaySourcerWrapper.cs
+++ b/src/Build/BuildCheck/Infrastructure/BuildCheckBinaryLogReplaySourcerWrapper.cs
@@ -12,7 +12,7 @@
 
 namespace Microsoft.Build.Experimental.BuildCheck.Infrastructure;
 
-public class BuildCheckBinaryLogReplaySourcerWrapper : IBinaryLogReplaySource
+internal class BuildCheckBinaryLogReplaySourcerWrapper : IBinaryLogReplaySource
 {
     private readonly BinaryLogReplayEventSource _replayEventSource;
     private readonly IBuildEventHandler _buildCheckEventHandler;
