diff --git a/eng/TestAssets.targets b/eng/TestAssets.targets
index ed68a2571cd..87f945431c1 100644
--- a/eng/TestAssets.targets
+++ b/eng/TestAssets.targets
@@ -8,9 +8,7 @@
   <ItemGroup Label="TestAssets">
     <ProjectsToPack Include="..\Build\Microsoft.Build.csproj" />
     <ProjectsToPack Include="..\Framework\Microsoft.Build.Framework.csproj" />
-    <ProjectsToPack Include="..\StringTools\StringTools.csproj">
-      <TargetFrameworks Condition="'$(MSBuildRuntimeType)' != 'Core'">$(LibraryTargetFrameworks);net35</TargetFrameworks>
-    </ProjectsToPack>
+    <ProjectsToPack Include="..\StringTools\StringTools.csproj"/>
   </ItemGroup>
 
   <Target Name="GetSourceFilesForPacking">
@@ -24,6 +22,6 @@
           DependsOnTargets="GetSourceFilesForPacking"
           Inputs="@(ProjectsToPack);@(SourceFilesForPacking)"
           Outputs="$(ArtifactsNonShippingPackagesDir)\Microsoft.Build.$(Version).nupkg;$(ArtifactsNonShippingPackagesDir)\Microsoft.Build.Framework.$(Version).nupkg;$(ArtifactsNonShippingPackagesDir)\Microsoft.NET.StringTools.$(Version).nupkg">
-    <Exec Command="dotnet pack &quot;%(ProjectsToPack.Identity)&quot; -o &quot;$(ArtifactsNonShippingPackagesDir)&quot; -p:PackageVersion=$(PackageVersion)" />
+    <Exec Command="dotnet pack &quot;%(ProjectsToPack.Identity)&quot; -o &quot;$(ArtifactsNonShippingPackagesDir)&quot; -p:PackageVersion=$(PackageVersion) --no-restore" />
   </Target>
-</Project>
\ No newline at end of file
+</Project>
