diff --git a/src/Build/Graph/ProjectInterpretation.cs b/src/Build/Graph/ProjectInterpretation.cs
index 8cee3b8f19e..b62715052f9 100644
--- a/src/Build/Graph/ProjectInterpretation.cs
+++ b/src/Build/Graph/ProjectInterpretation.cs
@@ -42,7 +42,7 @@ internal enum ProjectType
         {
             OuterBuild,
             InnerBuild,
-            NonMultitargeting
+            NonMultitargeting,
         }
 
         internal readonly struct ReferenceInfo
diff --git a/src/Tasks.UnitTests/ResolveComReference_Tests.cs b/src/Tasks.UnitTests/ResolveComReference_Tests.cs
index 23afd470bdf..0e15658be92 100644
--- a/src/Tasks.UnitTests/ResolveComReference_Tests.cs
+++ b/src/Tasks.UnitTests/ResolveComReference_Tests.cs
@@ -753,7 +753,8 @@ public void AxReferenceKnowsItsRCWPiaExists_IncludeVersion()
             CheckAxReferenceRCWTlbExists(RcwStyle.PreexistingPia /* pass in the PIA reference */, true /* include version in the interop name */);
         }
 
-        private enum RcwStyle {
+        private enum RcwStyle
+        {
             GenerateTlb,
             PreexistingTlb,
             PreexistingPia,
diff --git a/src/Tasks/BootstrapperUtil/Product.cs b/src/Tasks/BootstrapperUtil/Product.cs
index 73ba30e58e2..fd0ca0c24e3 100644
--- a/src/Tasks/BootstrapperUtil/Product.cs
+++ b/src/Tasks/BootstrapperUtil/Product.cs
@@ -14,7 +14,7 @@ internal enum CopyAllFilesType
     {
         CopyAllFilesFalse,
         CopyAllFilesTrue,
-        CopyAllFilesIfNotHomeSite
+        CopyAllFilesIfNotHomeSite,
     };
 
     /// <summary>
diff --git a/src/Tasks/FormatVersion.cs b/src/Tasks/FormatVersion.cs
index b0f40cec9f6..d89e00eb8fa 100644
--- a/src/Tasks/FormatVersion.cs
+++ b/src/Tasks/FormatVersion.cs
@@ -17,7 +17,8 @@ namespace Microsoft.Build.Tasks
     /// </comment>
     public sealed class FormatVersion : TaskExtension
     {
-        private enum _FormatType {
+        private enum _FormatType
+        {
             Version,
             Path,
         }
diff --git a/src/Tasks/GenerateApplicationManifest.cs b/src/Tasks/GenerateApplicationManifest.cs
index f4ea09d1a9d..984bcc847f9 100644
--- a/src/Tasks/GenerateApplicationManifest.cs
+++ b/src/Tasks/GenerateApplicationManifest.cs
@@ -17,7 +17,8 @@ namespace Microsoft.Build.Tasks
     /// </summary>
     public sealed class GenerateApplicationManifest : GenerateManifestBase
     {
-        private enum _ManifestType {
+        private enum _ManifestType
+        {
             Native,
             ClickOnce,
         }
diff --git a/src/Tasks/GenerateManifestBase.cs b/src/Tasks/GenerateManifestBase.cs
index ac5f5d74291..14b82c05669 100644
--- a/src/Tasks/GenerateManifestBase.cs
+++ b/src/Tasks/GenerateManifestBase.cs
@@ -16,14 +16,16 @@ namespace Microsoft.Build.Tasks
     /// </summary>
     public abstract class GenerateManifestBase : Task
     {
-        private enum AssemblyType {
+        private enum AssemblyType
+        {
             Unspecified,
             Managed,
             Native,
             Satellite,
         }
 
-        private enum DependencyType {
+        private enum DependencyType
+        {
             Install,
             Prerequisite,
         }
diff --git a/src/Tasks/StrongNameUtils.cs b/src/Tasks/StrongNameUtils.cs
index f03c3ecdc7d..d527661e3c6 100644
--- a/src/Tasks/StrongNameUtils.cs
+++ b/src/Tasks/StrongNameUtils.cs
@@ -21,7 +21,7 @@ internal enum StrongNameLevel
         None,
         DelaySigned,
         FullySigned,
-        Unknown
+        Unknown,
     };
 
     /// <summary>
