diff --git a/src/Build/Logging/FancyLogger/FancyLoggerNode.cs b/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
index 823f9385c93..519dcae6761 100644
--- a/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
+++ b/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
@@ -56,6 +56,8 @@ public void WriteEnd()
                     Console.WindowWidth
                 )
             );
+            if (CurrentTargetLine == null) return;
+            FancyLoggerBuffer.DeleteLine(CurrentTargetLine!.Id);
         }
 
         public void WriteTarget(TargetStartedEventArgs args)
@@ -63,7 +65,8 @@ public void WriteTarget(TargetStartedEventArgs args)
             if (Line == null) return;
             // Update spinner
             UpdateLine();
-            // Create target node
+            // Update target line
+            FancyLoggerBuffer.UpdateLine(CurrentTargetLine!.Id, $"   `- {ANSIBuilder.Formatting.Bold(args.TargetName)}");
 
             /* FancyLoggerTargetNode targetNode = new FancyLoggerTargetNode(args);
             FancyLoggerBuffer.WriteNewLineAfter(
