diff --git a/src/Build/Definition/ProjectCollection.cs b/src/Build/Definition/ProjectCollection.cs
index 247c986423f..9973e5d3d80 100644
--- a/src/Build/Definition/ProjectCollection.cs
+++ b/src/Build/Definition/ProjectCollection.cs
@@ -341,9 +341,6 @@ public ProjectCollection(IDictionary<string, string> globalProperties, IEnumerab
                 {
                     _globalProperties = new PropertyDictionary<ProjectPropertyInstance>(globalProperties.Count);
 
-                    // at this stage globalProperties collection contains entries passed from command line (e.g. /p:foo=bar).
-                    PropertiesFromCommandLine = [.. globalProperties.Keys];
-
                     foreach (KeyValuePair<string, string> pair in globalProperties)
                     {
                         try
@@ -503,7 +500,7 @@ public static string DisplayVersion
         /// <summary>
         /// Properties passed from the command line (e.g. by using /p:).
         /// </summary>
-        public ICollection<string> PropertiesFromCommandLine { get; }
+        public ICollection<string> PropertiesFromCommandLine { get; set; }
 
         /// <summary>
         /// The default tools version of this project collection. Projects use this tools version if they
diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index aeddef7aba4..9afa243f2bc 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -835,6 +835,8 @@ public static ExitType Execute(
                         {
                             using (ProjectCollection collection = new(globalProperties, loggers, ToolsetDefinitionLocations.Default))
                             {
+                                collection.PropertiesFromCommandLine = [.. globalProperties.Keys];
+
                                 Project project = collection.LoadProject(projectFile, globalProperties, toolsVersion);
 
                                 if (getResultOutputFile.Length == 0)
