diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index bbb2bf74c82..27a39a22b6d 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -1189,7 +1189,7 @@ internal static bool BuildProject(
                 // This is a hack for now to make sure the perf hit only happens
                 // on diagnostic. This should be changed to pipe it through properly,
                 // perhaps as part of a fuller tracing feature.
-                bool logTaskInputs = verbosity == LoggerVerbosity.Diagnostic;
+                bool logTaskInputs = verbosity == LoggerVerbosity.Diagnostic || loggers.Any(l => l is BinaryLogger);
 
                 if (!logTaskInputs)
                 {
