diff --git a/src/MSBuild/LiveLogger/LiveLogger.cs b/src/MSBuild/LiveLogger/LiveLogger.cs
index 05d00913bce..e2039ec6a15 100644
--- a/src/MSBuild/LiveLogger/LiveLogger.cs
+++ b/src/MSBuild/LiveLogger/LiveLogger.cs
@@ -31,6 +31,8 @@ internal sealed class LiveLogger : INodeLogger
 
     private ProjectContext _restoreContext;
 
+    private Thread _refresher;
+
     public LoggerVerbosity Verbosity { get => LoggerVerbosity.Minimal; set { } }
     public string Parameters { get => ""; set { } }
 
@@ -77,8 +79,8 @@ public void Initialize(IEventSource eventSource)
         eventSource.WarningRaised += new BuildWarningEventHandler(WarningRaised);
         eventSource.ErrorRaised += new BuildErrorEventHandler(ErrorRaised);
 
-        Thread refresher = new(ThreadProc);
-        refresher.Start();
+        _refresher = new(ThreadProc);
+        _refresher.Start();
     }
 
     private void ThreadProc()
@@ -261,6 +263,7 @@ private void ErrorRaised(object sender, BuildErrorEventArgs e)
     public void Shutdown()
     {
         _cts.Cancel();
+        _refresher?.Join();
     }
 }
 
