diff --git a/src/Tasks/GetSDKReferenceFiles.cs b/src/Tasks/GetSDKReferenceFiles.cs
index bef6ec4b078..284999b5a52 100644
--- a/src/Tasks/GetSDKReferenceFiles.cs
+++ b/src/Tasks/GetSDKReferenceFiles.cs
@@ -1271,16 +1271,16 @@ public void Translate(ITranslator translator)
                     info.ImageRuntime = imageRuntime;
                     info.IsManagedWinmd = isManagedWinmd;
                     info.IsWinMD = isWinmd;
-                }, count => new Dictionary<string, SdkReferenceInfo>(count));
+                }, count => new Dictionary<string, SdkReferenceInfo>(count, StringComparer.OrdinalIgnoreCase));
 
                 translator.TranslateDictionary(ref _directoryToFileList, (ITranslator t, ref string s) => t.Translate(ref s), (ITranslator t, ref List<string> fileList) =>
                 {
                     t.Translate(ref fileList, (ITranslator t, ref string str) => { t.Translate(ref str); });
-                }, count => new Dictionary<string, List<string>>(count));
+                }, count => new Dictionary<string, List<string>>(count, StringComparer.OrdinalIgnoreCase));
                 translator.TranslateDictionary(ref _directoryToFileList, (ITranslator t, ref string s) => t.Translate(ref s), (ITranslator t, ref List<string> fileList) =>
                 {
                     t.Translate(ref fileList, (ITranslator t, ref string str) => { t.Translate(ref str); });
-                }, count => new Dictionary<string, List<string>>(count));
+                }, count => new Dictionary<string, List<string>>(count, StringComparer.OrdinalIgnoreCase));
 
                 translator.Translate(ref _hash);
             }
