diff --git a/src/Framework/Features.cs b/src/Framework/Features.cs
index a7b1ec4b16c..4e7a81d677f 100644
--- a/src/Framework/Features.cs
+++ b/src/Framework/Features.cs
@@ -30,7 +30,7 @@ public enum FeatureStatus
         NotAvailable,
 
         /// <summary>
-        /// The feature is in preview.
+        /// The feature is in preview, subject to change API or behavior between releases.
         /// </summary>
         Preview,
     }
