diff --git a/src/Shared/TranslatorHelpers.cs b/src/Shared/TranslatorHelpers.cs
index 9cab3485c97..c3e7d12b6ad 100644
--- a/src/Shared/TranslatorHelpers.cs
+++ b/src/Shared/TranslatorHelpers.cs
@@ -261,9 +261,6 @@ public static void Translate(this ITranslator translator, ref AssemblyName assem
                     HashAlgorithm = hashAlgorithm,
                     VersionCompatibility = versionCompatibility,
                     CodeBase = codeBase,
-                    // AssemblyName.KeyPair is not used anywhere, additionally StrongNameKeyPair is not supported in .net core 5-
-                    // and throws platform not supported exception when serialized or deserialized
-                    KeyPair = null,
                 };
 
                 assemblyName.SetPublicKey(publicKey);
