diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index bff68d1bade..15faca0d420 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -272,7 +272,7 @@ private void LogDiagnostic(string message, params object[] messageArgs)
             // If the destination file is a hard or symbolic link, File.Copy would overwrite the source.
             // To prevent this, we need to delete the existing entry before we Copy or create a link.
             // We could try to figure out if the file is a link, but I can't think of a reason to not simply delete it always.
-            if (destinationFileState.FileExists && !destinationFileState.IsReadOnly)
+            if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_6) && destinationFileState.FileExists && !destinationFileState.IsReadOnly)
             {
                 FileUtilities.DeleteNoThrow(destinationFileState.Name);
             }
