diff --git a/src/Build/Evaluation/Evaluator.cs b/src/Build/Evaluation/Evaluator.cs
index 6aed10ece8b..bc7ec077b66 100644
--- a/src/Build/Evaluation/Evaluator.cs
+++ b/src/Build/Evaluation/Evaluator.cs
@@ -821,14 +821,17 @@ private void Evaluate()
 
         private IEnumerable FilterOutEnvironmentDerivedProperties(PropertyDictionary<P> dictionary)
         {
+            List<P> list = new();
             foreach (P p in dictionary)
             {
                 if (!((p is ProjectProperty pp && pp.IsEnvironmentProperty) ||
                     (p is EnvironmentDerivedProjectPropertyInstance)))
                 {
-                    yield return p;
+                    list.Add(p);
                 }
             }
+
+            return list;
         }
 
         private void CollectProjectCachePlugins()
