diff --git a/src/Build/Evaluation/Evaluator.cs b/src/Build/Evaluation/Evaluator.cs
index 1ece76227ce..8b1935050d7 100644
--- a/src/Build/Evaluation/Evaluator.cs
+++ b/src/Build/Evaluation/Evaluator.cs
@@ -2304,7 +2304,9 @@ private LoadImportsResult ExpandAndLoadImportsFromUnescapedImportExpression(stri
                                 continue;
                             }
 
+#if !FEATURE_VISUALSTUDIOSETUP
                             VerifyVSDistributionPath(importElement.Project, importLocationInProject);
+#endif
 
                             ProjectErrorUtilities.ThrowInvalidProject(importLocationInProject, "ImportedProjectNotFound",
                                                                       importFileUnescaped, importExpressionEscaped);
@@ -2579,7 +2581,9 @@ private void ThrowForImportedProjectWithSearchPathsNotFound(ProjectImportPathMat
 
             string stringifiedListOfSearchPaths = StringifyList(onlyFallbackSearchPaths);
 
+#if !FEATURE_VISUALSTUDIOSETUP
             VerifyVSDistributionPath(importElement.Project, importElement.ProjectLocation);
+#endif
 
 #if FEATURE_SYSTEM_CONFIGURATION
             string configLocation = AppDomain.CurrentDomain.SetupInformation.ConfigurationFile;
@@ -2647,7 +2651,8 @@ private void SetAllProjectsProperty()
 
         private void VerifyVSDistributionPath(string path, ElementLocation importLocationInProject)
         {
-            if (path.IndexOf($"Microsoft{Path.DirectorySeparatorChar}VisualStudio", StringComparison.OrdinalIgnoreCase) >= 0)
+            if (path.IndexOf("Microsoft\\VisualStudio", StringComparison.OrdinalIgnoreCase) >= 0
+                || path.IndexOf("Microsoft/VisualStudio", StringComparison.OrdinalIgnoreCase) >= 0)
             {
                 ProjectErrorUtilities.ThrowInvalidProject(importLocationInProject, "ImportedProjectFromVSDistribution", path);
             }
diff --git a/src/Build/Resources/Strings.resx b/src/Build/Resources/Strings.resx
index ae9e7e663ef..f0f939e9abc 100644
--- a/src/Build/Resources/Strings.resx
+++ b/src/Build/Resources/Strings.resx
@@ -528,7 +528,7 @@
     <comment>{StrBegin="MSB4019: "}LOCALIZATION: &lt;Import&gt; should not be localized.</comment>
   </data>
   <data name="ImportedProjectFromVSDistribution" xml:space="preserve">
-    <value>MSB4278: The imported project {0} is from the Visual Studio distribution of MSBuild. Build with MSBuild.exe in a Developer Command Prompt. </value>
+    <value>MSB4278: The imported project {0} does not exist and appears to be part of Visual Studio. This project may require MSBuild.exe and fail to build in the dotnet CLI. </value>
     <comment>{StrBegin="MSB4278: "}LOCALIZATION: &lt;Import&gt; should not be localized.</comment>
   </data>
   <data name="ImportedProjectFromExtensionsPathNotFoundFromAppConfig" xml:space="preserve">
