diff --git a/src/Build/BackEnd/Shared/EventsCreatorHelper.cs b/src/Build/BackEnd/Shared/EventsCreatorHelper.cs
index d7f869d73c7..0b9503569b0 100644
--- a/src/Build/BackEnd/Shared/EventsCreatorHelper.cs
+++ b/src/Build/BackEnd/Shared/EventsCreatorHelper.cs
@@ -32,9 +32,9 @@ public static BuildMessageEventArgs CreateMessageEventFromText(BuildEventContext
 
     public static BuildErrorEventArgs CreateErrorEventFromText(BuildEventContext buildEventContext, string? subcategoryResourceName, string? errorCode, string? helpKeyword, BuildEventFileInfo file, string message)
     {
-        ErrorUtilities.VerifyThrow(buildEventContext != null, "Must specify the buildEventContext");
-        ErrorUtilities.VerifyThrow(file != null, "Must specify the associated file.");
-        ErrorUtilities.VerifyThrow(message != null, "Need error message.");
+        ErrorUtilities.VerifyThrowArgumentNull(buildEventContext);
+        ErrorUtilities.VerifyThrowArgumentNull(file);
+        ErrorUtilities.VerifyThrowArgumentNull(message);
 
         string? subcategory = null;
 
