diff --git a/Directory.Build.props b/Directory.Build.props
index 2cc17a1372f..6f80aaa5077 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -62,6 +62,11 @@
     <MachineIndependentBuild>true</MachineIndependentBuild>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <AssemblyInformationCachePaths Condition="Exists('$(MSBuildBinPath)\SdkPrecomputedAssemblyReferences.cache')">$(MSBuildBinPath)\SDKPrecomputedAssemblyReferences.cache</AssemblyInformationCachePaths>
+    <AssemblyInformationCachePaths Condition="Exists('$(NetCoreRoot)$(NetCoreSdkVersion)\SdkPrecomputedAssemblyReferences.cache')">$(NetCoreRoot)$(NetCoreSdkVersion)\SDKPrecomputedAssemblyReferences.cache</AssemblyInformationCachePaths>
+  </PropertyGroup>
+
   <PropertyGroup>
     <DefaultItemExcludes>$(DefaultItemExcludes);*.log</DefaultItemExcludes>
     <DefaultItemExcludes>$(DefaultItemExcludes);*.binlog</DefaultItemExcludes>
