diff --git a/src/StringTools/StringTools.csproj b/src/StringTools/StringTools.csproj
index e7713cb359a..e56f8e0a4d4 100644
--- a/src/StringTools/StringTools.csproj
+++ b/src/StringTools/StringTools.csproj
@@ -10,7 +10,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <Nullable>enable</Nullable>
 
-    <VersionPrefix>1.0.0</VersionPrefix>
+    <VersionPrefix>1.0.1</VersionPrefix>
     <AssemblyVersion></AssemblyVersion>
     <SemanticVersioningV1>true</SemanticVersioningV1>
 
