diff --git a/src/Build/BackEnd/Components/Logging/LoggingContext.cs b/src/Build/BackEnd/Components/Logging/LoggingContext.cs
index 4f0713a6e7e..92036ec0363 100644
--- a/src/Build/BackEnd/Components/Logging/LoggingContext.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingContext.cs
@@ -214,7 +214,7 @@ internal void LogWarning(string subcategoryResourceName, BuildEventFileInfo file
 
             // Log an error if the warning we were about to log is listed as a WarningAsError
             // https://github.com/dotnet/msbuild/issues/5511
-            if(_loggingService.WarningsAsErrors != null)
+            if(_loggingService.WarningsAsErrors != null && _loggingService.WarningsAsErrors.Count > 0)
             {
                 string warningCode;
                 string message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword(out warningCode, out _, messageResourceName, messageArgs);
@@ -243,9 +243,13 @@ internal void LogWarningFromText(string subcategoryResourceName, string warningC
 
             // Log an error if the warning we were about to log is listed as a WarningAsError
             // https://github.com/dotnet/msbuild/issues/5511
-            if (_loggingService.WarningsAsErrors.Contains(warningCode))
+            if (_loggingService.WarningsAsErrors != null && _loggingService.WarningsAsErrors.Count > 0)
             {
-                LogErrorFromText(subcategoryResourceName, warningCode, helpKeyword, file, message);
+                if (_loggingService.WarningsAsErrors.Contains(warningCode))
+                {
+                    LogErrorFromText(subcategoryResourceName, warningCode, helpKeyword, file, message);
+                    return;
+                }
             }
 
             _loggingService.LogWarningFromText(_eventContext, subcategoryResourceName, warningCode, helpKeyword, file, message);
