diff --git a/src/Build/BackEnd/Components/Communications/ServerNodeEndpointOutOfProc.cs b/src/Build/BackEnd/Components/Communications/ServerNodeEndpointOutOfProc.cs
index 5cb65f898d3..528d27056da 100644
--- a/src/Build/BackEnd/Components/Communications/ServerNodeEndpointOutOfProc.cs
+++ b/src/Build/BackEnd/Components/Communications/ServerNodeEndpointOutOfProc.cs
@@ -7,7 +7,7 @@
 namespace Microsoft.Build.BackEnd
 {
     /// <summary>
-    /// This is an implementation of INodeEndpoint for the out-of-proc nodes.  It acts only as a client.
+    /// This is an implementation of out-of-proc server node endpoint.
     /// </summary>
     internal sealed class ServerNodeEndpointOutOfProc : NodeEndpointOutOfProcBase
     {
diff --git a/src/Build/BackEnd/Node/OutOfProcServerNode.cs b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
index b61ec44399f..3a5a4fe2043 100644
--- a/src/Build/BackEnd/Node/OutOfProcServerNode.cs
+++ b/src/Build/BackEnd/Node/OutOfProcServerNode.cs
@@ -92,8 +92,8 @@ public NodeEngineShutdownReason Run(out Exception? shutdownException)
 
             string pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuildServer-" + handshake.ComputeHash());
 
-            string serverRunningMutexName = $@"Global\server-running-{pipeName}";
-            _serverBusyMutexName = $@"Global\server-busy-{pipeName}";
+            string serverRunningMutexName = $@"{ServerNamedMutex.RunningServerMutexNamePrefix}{pipeName}";
+            _serverBusyMutexName = $@"{ServerNamedMutex.BusyServerMutexNamePrefix}{pipeName}";
 
             // TODO: shall we address possible race condition. It is harmless as it, with acceptable probability, just cause unnecessary process spawning
             // and of two processes will become victim and fails, build will not be affected
@@ -367,7 +367,10 @@ public override void Flush()
 
         internal sealed class ServerNamedMutex : IDisposable
         {
-            public readonly Mutex _serverMutex;
+            public const string RunningServerMutexNamePrefix = @"Global\server-running-";
+            public const string BusyServerMutexNamePrefix = @"Global\server-busy-";
+
+            private readonly Mutex _serverMutex;
 
             public bool IsDisposed { get; private set; }
 
