diff --git a/src/Framework/Sdk/SdkResultItem.cs b/src/Framework/Sdk/SdkResultItem.cs
index ca9502d0f45..5d12fbf3241 100644
--- a/src/Framework/Sdk/SdkResultItem.cs
+++ b/src/Framework/Sdk/SdkResultItem.cs
@@ -40,21 +40,6 @@ public override bool Equals(object obj)
                    Metadata?.Count == item.Metadata?.Count)
             {
                 return Metadata.All(m => item.Metadata.TryGetValue(m.Key, out var itemValue) && itemValue == m.Value);
-                {
-                    if (!item.Metadata.TryGetValue(kvp.Key, out var itemValue))
-                    {
-                        return false;
-                    }
-                    else
-                    {
-                        if (kvp.Value != itemValue)
-                        {
-                            return false;
-                        }
-                    }
-                }
-
-                return true;
             }
             return false;
         }
