diff --git a/src/Build/Logging/FancyLogger/ANSIBuilder.cs b/src/Build/Logging/FancyLogger/ANSIBuilder.cs
index e0a244aa62e..afb8a0f7124 100644
--- a/src/Build/Logging/FancyLogger/ANSIBuilder.cs
+++ b/src/Build/Logging/FancyLogger/ANSIBuilder.cs
@@ -157,8 +157,9 @@ public static string Overlined(string text)
                 return String.Format("\x1b[53m{0}\x1b[55m", text);
             }
 
-            public static string Hyperlink(string text, string url)
+            public static string Hyperlink(string text, string rawUrl)
             {
+                string url = rawUrl.Length > 0 ? new System.Uri(rawUrl).AbsoluteUri : rawUrl;
                 return $"\x1b]8;;{url}\x1b\\{text}\x1b]8;;\x1b\\";
             }
 
