diff --git a/src/MSBuild/LiveLogger/FancyLogger.cs b/src/MSBuild/LiveLogger/LiveLogger.cs
similarity index 99%
rename from src/MSBuild/LiveLogger/FancyLogger.cs
rename to src/MSBuild/LiveLogger/LiveLogger.cs
index 4e9029d02e5..a3727cbfcb6 100644
--- a/src/MSBuild/LiveLogger/FancyLogger.cs
+++ b/src/MSBuild/LiveLogger/LiveLogger.cs
@@ -8,7 +8,7 @@
 
 namespace Microsoft.Build.Logging.LiveLogger
 {
-    internal class FancyLogger : ILogger
+    internal class LiveLogger : ILogger
     {
         private Dictionary<int, ProjectNode> projects = new Dictionary<int, ProjectNode>();
 
@@ -21,7 +21,7 @@ internal class FancyLogger : ILogger
 
         public LoggerVerbosity Verbosity { get; set; }
 
-        public FancyLogger()
+        public LiveLogger()
         {
             Parameters = "";
         }
diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 5f15c8c2d9e..ff326891411 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -3428,7 +3428,7 @@ private static void ProcessLiveLogger(
             // Check for flags and env variables
             if (!noConsoleLogger)
             {
-                FancyLogger l = new FancyLogger();
+                LiveLogger l = new LiveLogger();
                 loggers.Add(l);
             }
         }
