diff --git a/src/Tasks/system.design/stronglytypedresourcebuilder.cs b/src/Tasks/system.design/stronglytypedresourcebuilder.cs
index 4ca7d651073..b7806d19b8b 100644
--- a/src/Tasks/system.design/stronglytypedresourcebuilder.cs
+++ b/src/Tasks/system.design/stronglytypedresourcebuilder.cs
@@ -244,7 +244,7 @@ private static CodeCompileUnit InternalCreate(Dictionary<String, ResourceData> r
             {
                 String propertyName = entry.Key;
                 // The resourceName will be the original value, before fixups, if any.
-                if (reverseFixupTable.TryGetValue(propertyName, out string resourceName))
+                if (!reverseFixupTable.TryGetValue(propertyName, out string resourceName))
                 {
                     resourceName = propertyName;
                 }
