diff --git a/eng/Common.globalconfig b/eng/Common.globalconfig
index ba417e063f2..c47852e8692 100644
--- a/eng/Common.globalconfig
+++ b/eng/Common.globalconfig
@@ -843,7 +843,7 @@ dotnet_diagnostic.SA1132.severity = warning
 dotnet_diagnostic.SA1133.severity = suggestion
 
 # Each attribute should be placed on its own line of code
-dotnet_diagnostic.SA1134.severity = suggestion
+dotnet_diagnostic.SA1134.severity = warning
 
 # Using directive should be qualified
 dotnet_diagnostic.SA1135.severity = suggestion
diff --git a/src/MSBuildTaskHost/Concurrent/ConcurrentQueue.cs b/src/MSBuildTaskHost/Concurrent/ConcurrentQueue.cs
index fefb43b48bb..f6b0b468289 100644
--- a/src/MSBuildTaskHost/Concurrent/ConcurrentQueue.cs
+++ b/src/MSBuildTaskHost/Concurrent/ConcurrentQueue.cs
@@ -554,7 +554,9 @@ internal struct Slot
     [StructLayout(LayoutKind.Explicit, Size = 192)] // padding before/between/after fields based on typical cache line size of 64
     internal struct PaddedHeadAndTail
     {
-        [FieldOffset(64)] public int Head;
-        [FieldOffset(128)] public int Tail;
+        [FieldOffset(64)]
+        public int Head;
+        [FieldOffset(128)]
+        public int Tail;
     }
 }
