diff --git a/src/Build/BackEnd/Components/Communications/NodeManager.cs b/src/Build/BackEnd/Components/Communications/NodeManager.cs
index 2bb110790c8..6d9f73590eb 100644
--- a/src/Build/BackEnd/Components/Communications/NodeManager.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeManager.cs
@@ -349,7 +349,7 @@ NodeConfiguration AcquiredNodeConfigurationFactory(NodeInfo nodeInfo)
 
         public IEnumerable<Process> GetProcesses()
         {
-            return _outOfProcNodeProvider?.GetProcesses() ?? IEnumerable.Empty();
+            return _outOfProcNodeProvider?.GetProcesses()!;
         }
     }
 }
