diff --git a/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs b/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs
index 0a3651d317a..0b6c7769f80 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs
@@ -124,9 +124,6 @@ public async Task<BuildResult> BuildTargets(ProjectLoggingContext loggingContext
             // Get the actual target objects from the names
             BuildRequestConfiguration configuration = _requestEntry.RequestConfiguration;
 
-            var initialTargets = configuration.ProjectInitialTargets;
-            var defaultTargets = configuration.ProjectDefaultTargets;
-
             bool previousCacheableStatus = configuration.IsCacheable;
             configuration.IsCacheable = false;
             configuration.RetrieveFromCache();
@@ -161,11 +158,11 @@ public async Task<BuildResult> BuildTargets(ProjectLoggingContext loggingContext
                     {
                         buildReason = TargetBuiltReason.EntryTarget;
                     }
-                    else if (initialTargets.Contains(targetName))
+                    else if (configuration.ProjectInitialTargets.Contains(targetName))
                     {
                         buildReason = TargetBuiltReason.InitialTarget;
                     } 
-                    else if (defaultTargets.Contains(targetName))
+                    else if (configuration.ProjectDefaultTargets.Contains(targetName))
                     {
                         buildReason = TargetBuiltReason.DefaultTarget;
                     }
