diff --git a/src/Build/BackEnd/BuildManager/BuildParameters.cs b/src/Build/BackEnd/BuildManager/BuildParameters.cs
index 2e417b9108a..9649c0815dc 100644
--- a/src/Build/BackEnd/BuildManager/BuildParameters.cs
+++ b/src/Build/BackEnd/BuildManager/BuildParameters.cs
@@ -830,7 +830,7 @@ public BuildParameters Clone()
 
         internal bool UsesInputCaches() => InputResultsCacheFiles != null;
 
-        internal bool SkippedResultsDoNotCauseCacheMiss() => IsolateProjects != IsolateProjects.False;
+        internal bool SkippedResultsDoNotCauseCacheMiss() => IsolateProjects == IsolateProjects.True;
 
         /// <summary>
         /// Implementation of the serialization mechanism.
