diff --git a/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs b/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs
index 823592db259..f05986f7532 100644
--- a/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs
+++ b/src/MSBuild/ValidateMSBuildPackageDependencyVersions.cs
@@ -1,6 +1,5 @@
 using Microsoft.Build.Framework;
 using Microsoft.Build.Utilities;
-using System;
 using System.IO;
 using System.Reflection;
 using System.Xml;
@@ -35,7 +34,6 @@ public override bool Execute()
                                         {
                                             string name = string.Empty;
                                             string version = string.Empty;
-                                            bool check = true;
                                             foreach (var dependentAssembly in assemblyBindingXmlElement.ChildNodes)
                                             {
                                                 if (dependentAssembly is XmlElement dependentAssemblyXmlElement)
@@ -54,19 +52,15 @@ public override bool Execute()
                                                     {
                                                         foreach (var bindingRedirectAttribute in dependentAssemblyXmlElement.Attributes)
                                                         {
-                                                            if (bindingRedirectAttribute is XmlAttribute bindingRedirectVersion && bindingRedirectVersion.Name.Equals("newVersion", System.StringComparison.OrdinalIgnoreCase))
+                                                            if (bindingRedirectAttribute is XmlAttribute bindingRedirectAttributeXmlElement && bindingRedirectAttributeXmlElement.Name.Equals("newVersion", System.StringComparison.OrdinalIgnoreCase))
                                                             {
-                                                                version = bindingRedirectVersion.Value;
-                                                            }
-                                                            else if (bindingRedirectAttribute is XmlAttribute notToCheck && notToCheck.Name.Equals("notToBeChecked", System.StringComparison.OrdinalIgnoreCase))
-                                                            {
-                                                                check = !notToCheck.Value.Equals("true", System.StringComparison.OrdinalIgnoreCase);
+                                                                version = bindingRedirectAttributeXmlElement.Value;
                                                             }
                                                         }
                                                     }
                                                 }
                                             }
-                                            if (check && !string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(version))
+                                            if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(version))
                                             {
                                                 string path = Path.Combine(AssemblyPath, name + ".dll");
                                                 if (File.Exists(path) && !version.Equals(Assembly.LoadFile(path).GetName().Version.ToString()))
diff --git a/src/MSBuild/app.amd64.config b/src/MSBuild/app.amd64.config
index 95a07e4c238..517c74df63f 100644
--- a/src/MSBuild/app.amd64.config
+++ b/src/MSBuild/app.amd64.config
@@ -111,7 +111,7 @@
           <!-- It is unusual to redirect down, but in this case it's ok: 4.0.3.0 forwards
                to 4.0.0.0 in the GAC, so this just removes the need to redistribute a file
                and makes that resolution faster. -->
-          <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.0.0" notToBeChecked="true" />
+          <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.0.0" />
         </dependentAssembly>
 
         <!-- Redirects for components dropped by Visual Studio -->
diff --git a/src/MSBuild/app.config b/src/MSBuild/app.config
index 83f40b73d94..f22bacc449e 100644
--- a/src/MSBuild/app.config
+++ b/src/MSBuild/app.config
@@ -101,7 +101,7 @@
           <!-- It is unusual to redirect down, but in this case it's ok: 4.0.3.0 forwards
                to 4.0.0.0 in the GAC, so this just removes the need to redistribute a file
                and makes that resolution faster. -->
-          <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.0.0" notToBeChecked="true" />
+          <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.0.0" />
         </dependentAssembly>
 
         <!-- Redirects for components dropped by Visual Studio -->
