diff --git a/eng/Version.Details.xml b/eng/Version.Details.xml
index 37c8b7904fd..e2ac29d690a 100644
--- a/eng/Version.Details.xml
+++ b/eng/Version.Details.xml
@@ -58,9 +58,9 @@
       <Sha>3aa0b2b84cab7d94b9136547563d027fd78e82a6</Sha>
       <SourceBuild RepoName="xliff-tasks" ManagedOnly="true" />
     </Dependency>
-    <Dependency Name="NuGet.Build.Tasks" Version="6.8.0-preview.1.41">
+    <Dependency Name="NuGet.Build.Tasks" Version="6.8.0-preview.1.44">
       <Uri>https://github.com/nuget/nuget.client</Uri>
-      <Sha>972557a7258d77dbaed3b7a649d2f13a4083f94c</Sha>
+      <Sha>c7035e0564fc33c43bf3f17b612a052e0a01c95b</Sha>
     </Dependency>
     <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.8.0-1.23378.8">
       <Uri>https://github.com/dotnet/roslyn</Uri>
diff --git a/eng/Versions.props b/eng/Versions.props
index 9417f7eb508..a36df633e67 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -52,7 +52,7 @@
     <MicrosoftExtensionsDependencyModelVersion>7.0.0</MicrosoftExtensionsDependencyModelVersion>
     <MicrosoftIORedistVersion>6.0.0</MicrosoftIORedistVersion>
     <MicrosoftNetCompilersToolsetVersion>4.8.0-1.23378.8</MicrosoftNetCompilersToolsetVersion>
-    <NuGetBuildTasksVersion>6.8.0-preview.1.41</NuGetBuildTasksVersion>
+    <NuGetBuildTasksVersion>6.8.0-preview.1.44</NuGetBuildTasksVersion>
     <SystemRuntimeCompilerServicesUnsafeVersion>6.0.0</SystemRuntimeCompilerServicesUnsafeVersion>
     <SystemTextJsonVersion>7.0.3</SystemTextJsonVersion>
     <SystemThreadingTasksDataflowVersion>7.0.0</SystemThreadingTasksDataflowVersion>
