diff --git a/src/Shared/TaskLoggingHelper.cs b/src/Shared/TaskLoggingHelper.cs
index 432882d1f33..1abe0144d1b 100644
--- a/src/Shared/TaskLoggingHelper.cs
+++ b/src/Shared/TaskLoggingHelper.cs
@@ -519,7 +519,8 @@ public void LogIncludeGeneratedFile(string filePath, string content)
                 {
                     { "FILEPATH", filePath },
                     { "CONTENT", content },
-                }
+                },
+                Importance = MessageImportance.High,
             };
 
             BuildEngine.LogMessageEvent(e);
