diff --git a/eng/Packages.props b/eng/Packages.props
index 1c74031c544..fd01e882795 100644
--- a/eng/Packages.props
+++ b/eng/Packages.props
@@ -28,7 +28,7 @@
     <PackageReference Update="System.Resources.Extensions" Version="$(SystemResourcesExtensionsPackageVersion)" />
     <PackageReference Update="System.Runtime" Version="4.3.1" />
     <PackageReference Update="System.Runtime.CompilerServices.Unsafe" Version="6.0.0" />
-    <PackageReference Update="System.Security.Cryptography.X509Certificates" Version="4.1.2" />
+    <PackageReference Update="System.Security.Cryptography.X509Certificates" Version="4.3.2" />
     <PackageReference Update="System.Security.Cryptography.Pkcs" Version="4.7.0" />
     <PackageReference Update="System.Security.Cryptography.Xml" Version="4.7.0" />
     <PackageReference Update="System.Security.Permissions" Version="4.7.0" />
