diff --git a/src/Build/Utilities/Utilities.cs b/src/Build/Utilities/Utilities.cs
index 467b7cecd3c..65b4dc12728 100644
--- a/src/Build/Utilities/Utilities.cs
+++ b/src/Build/Utilities/Utilities.cs
@@ -700,10 +700,8 @@ public static void EnumerateProperties<TArg>(IEnumerable properties, TArg arg, A
             //  * <see cref="IItem"/>
             //  * <see cref="ITaskItem"/>
             //  * possibly others
-
             // That's why we here wrap with ItemAccessor if needed
 
-
             if (items == null)
             {
                 return [];
