diff --git a/src/Analyzers.UnitTests/ConfigurationProvider_Tests.cs b/src/Analyzers.UnitTests/ConfigurationProvider_Tests.cs
index a1af515a394..bd923db1ed2 100644
--- a/src/Analyzers.UnitTests/ConfigurationProvider_Tests.cs
+++ b/src/Analyzers.UnitTests/ConfigurationProvider_Tests.cs
@@ -9,13 +9,13 @@
 using System.Reflection;
 using System.Text;
 using System.Threading.Tasks;
-using Microsoft.Build.BuildCop.Infrastructure;
-using Microsoft.Build.BuildCop.Infrastructure.EditorConfig;
+using Microsoft.Build.BuildCheck.Infrastructure;
+using Microsoft.Build.BuildCheck.Infrastructure.EditorConfig;
 using Microsoft.Build.Experimental.BuildCheck;
 using Microsoft.Build.UnitTests;
 using Shouldly;
 using Xunit;
-using static Microsoft.Build.BuildCop.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
+using static Microsoft.Build.BuildCheck.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
 
 namespace Microsoft.Build.Analyzers.UnitTests
 {
diff --git a/src/Analyzers.UnitTests/EditorConfigParser_Tests.cs b/src/Analyzers.UnitTests/EditorConfigParser_Tests.cs
index e37d912f773..e39dfb6681f 100644
--- a/src/Analyzers.UnitTests/EditorConfigParser_Tests.cs
+++ b/src/Analyzers.UnitTests/EditorConfigParser_Tests.cs
@@ -9,11 +9,11 @@
 using System.Reflection;
 using System.Text;
 using System.Threading.Tasks;
-using Microsoft.Build.BuildCop.Infrastructure.EditorConfig;
+using Microsoft.Build.BuildCheck.Infrastructure.EditorConfig;
 using Microsoft.Build.UnitTests;
 using Shouldly;
 using Xunit;
-using static Microsoft.Build.BuildCop.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
+using static Microsoft.Build.BuildCheck.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
 
 
 namespace Microsoft.Build.Analyzers.UnitTests
diff --git a/src/Analyzers.UnitTests/EditorConfig_Tests.cs b/src/Analyzers.UnitTests/EditorConfig_Tests.cs
index 64f4e9d2832..71b367cc7a0 100644
--- a/src/Analyzers.UnitTests/EditorConfig_Tests.cs
+++ b/src/Analyzers.UnitTests/EditorConfig_Tests.cs
@@ -8,10 +8,10 @@
 using System.Reflection;
 using System.Text;
 using System.Threading.Tasks;
-using Microsoft.Build.BuildCop.Infrastructure.EditorConfig;
+using Microsoft.Build.BuildCheck.Infrastructure.EditorConfig;
 using Microsoft.Build.UnitTests;
 using Xunit;
-using static Microsoft.Build.BuildCop.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
+using static Microsoft.Build.BuildCheck.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
 
 #nullable disable
 
diff --git a/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs b/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
index 96103663f2e..a371064d4d2 100644
--- a/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
+++ b/src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
@@ -122,6 +122,7 @@ public void ProcessAnalyzerAcquisition(AnalyzerAcquisitionData acquisitionData)
         internal BuildCheckManager(ILoggingService loggingService)
         {
             _loggingService = loggingService;
+            _buildCheckCentralContext = new(_configurationProvider);
             _buildEventsProcessor = new(_buildCheckCentralContext);
         }
 
