diff --git a/src/Tasks/Microsoft.Common.CurrentVersion.targets b/src/Tasks/Microsoft.Common.CurrentVersion.targets
index 66505eba420..38488f81508 100644
--- a/src/Tasks/Microsoft.Common.CurrentVersion.targets
+++ b/src/Tasks/Microsoft.Common.CurrentVersion.targets
@@ -1693,9 +1693,14 @@ Copyright (C) Microsoft Corporation. All rights reserved.
       <_MSBuildProjectReferenceExistent Remove="@(_MSBuildProjectReferenceExistent)" Condition="'%(_MSBuildProjectReferenceExistent.SkipGetTargetFrameworkProperties)' != 'true'" />
 
       <!-- If we skipped querying the other project, ensure we don't pollute it with the TargetFramework/RuntimeIdentifier
-           that may be set for this project as a global property (if we're in an inner build). -->
+           that may be set for this project as a global property (if we're in an inner build).
+
+           But allow setting TF explicitly. -->
+      <_MSBuildProjectReferenceExistent Condition="'%(_MSBuildProjectReferenceExistent.SetTargetFramework)' == ''">
+          <UndefineProperties>%(_MSBuildProjectReferenceExistent.UndefineProperties);TargetFramework</UndefineProperties>
+      </_MSBuildProjectReferenceExistent>
       <_MSBuildProjectReferenceExistent>
-          <UndefineProperties>%(_MSBuildProjectReferenceExistent.UndefineProperties);TargetFramework;RuntimeIdentifier</UndefineProperties>
+          <UndefineProperties>%(_MSBuildProjectReferenceExistent.UndefineProperties);RuntimeIdentifier</UndefineProperties>
       </_MSBuildProjectReferenceExistent>
 
       <!-- Then add all AnnotatedProjects back. -->
