diff --git a/src/Build/Logging/FancyLogger/FancyLogger.cs b/src/Build/Logging/FancyLogger/FancyLogger.cs
index 99daefc3dd4..578b553a2b4 100644
--- a/src/Build/Logging/FancyLogger/FancyLogger.cs
+++ b/src/Build/Logging/FancyLogger/FancyLogger.cs
@@ -158,9 +158,10 @@ void eventSource_ErrorRaised(object sender, BuildErrorEventArgs e)
 
         public void Shutdown() {
             // Keep open if autoscroll disabled (the user is reading info)
-            /*while (FancyLoggerBuffer.AutoScrollEnabled || !FancyLoggerBuffer.IsTerminated)
+            // TODO: Find more efficient way of keeping logger open? 
+            while (FancyLoggerBuffer.AutoScrollEnabled || !FancyLoggerBuffer.IsTerminated)
             {
-            }*/
+            }
             FancyLoggerBuffer.Terminate();
             Console.Clear();
             Console.WriteLine("Build status, warnings and errors will be shown here after the build has ended and the interactive logger has closed");
diff --git a/src/Build/Logging/FancyLogger/FancyLoggerNode.cs b/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
index ad797455947..a1ff7c29971 100644
--- a/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
+++ b/src/Build/Logging/FancyLogger/FancyLoggerNode.cs
@@ -144,6 +144,7 @@ public class FancyLoggerMessageNode
         public FancyLoggerBufferLine? Line;
         public FancyLoggerMessageNode(LazyFormattedBuildEventArgs args)
         {
+            // TODO: Replace
             if (args.Message == null)
             {
                 Message = "Message was undefined";
