diff --git a/src/Build/BuildCheck/Infrastructure/EditorConfig/EditorConfigParser.cs b/src/Build/BuildCheck/Infrastructure/EditorConfig/EditorConfigParser.cs
index 34c4cd7bcb9..017929f73d6 100644
--- a/src/Build/BuildCheck/Infrastructure/EditorConfig/EditorConfigParser.cs
+++ b/src/Build/BuildCheck/Infrastructure/EditorConfig/EditorConfigParser.cs
@@ -12,7 +12,7 @@ namespace Microsoft.Build.Experimental.BuildCheck.Infrastructure.EditorConfig;
 
 internal sealed class EditorConfigParser
 {
-    // Define a static property to hold the editorConfigFilePath
+    // static property for embedding resolved `.editorconfig`s in binlog
     private static ConcurrentBag<string> editorConfigFilePaths = new ConcurrentBag<string>();
     public static IEnumerable<string> EditorConfigFilePaths => editorConfigFilePaths;
 
diff --git a/src/Build/Logging/BinaryLogger/BinaryLogger.cs b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
index fc73dcc58df..b4a775c608e 100644
--- a/src/Build/Logging/BinaryLogger/BinaryLogger.cs
+++ b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
@@ -4,7 +4,6 @@
 using System;
 using System.IO;
 using System.IO.Compression;
-using System.Linq;
 using Microsoft.Build.Experimental.BuildCheck.Infrastructure.EditorConfig;
 using Microsoft.Build.Framework;
 using Microsoft.Build.Framework.Telemetry;
@@ -324,12 +323,9 @@ public void Shutdown()
             if (projectImportsCollector != null)
             {
                 // Write the build check editorconfig file paths to the log
-                if (EditorConfigParser.EditorConfigFilePaths.Any())
+                foreach (var filePath in EditorConfigParser.EditorConfigFilePaths)
                 {
-                    foreach (var filePath in EditorConfigParser.EditorConfigFilePaths)
-                    {
-                        projectImportsCollector.AddFile(filePath);
-                    }
+                    projectImportsCollector.AddFile(filePath);
                 }
                 projectImportsCollector.Close();
 
