diff --git a/src/Build.UnitTests/BinaryLogger_Tests.cs b/src/Build.UnitTests/BinaryLogger_Tests.cs
index 9ee2339b895..0ec915e2710 100644
--- a/src/Build.UnitTests/BinaryLogger_Tests.cs
+++ b/src/Build.UnitTests/BinaryLogger_Tests.cs
@@ -74,6 +74,8 @@ public BinaryLoggerTests(ITestOutputHelper output)
 
             // this is needed to ensure the binary logger does not pollute the environment
             _env.WithEnvironmentInvariant();
+
+            _logFile = _env.ExpectFile(".binlog").Path;
         }
 
         [Theory]
@@ -165,7 +167,7 @@ public void UnusedEnvironmentVariablesDoNotAppearInBinaryLog()
                 TransientTestFolder logFolder = env.CreateFolder(createFolder: true);
                 TransientTestFile projectFile = env.CreateFile(logFolder, "myProj.proj", contents);
                 BinaryLogger logger = new();
-                logger.Parameters = Path.Combine(logFolder.Path, "binlog.binlog");
+                logger.Parameters = _logFile;
                 RunnerUtilities.ExecMSBuild($"{projectFile.Path} -bl:{logger.Parameters}", out bool success);
                 success.ShouldBeTrue();
                 RunnerUtilities.ExecMSBuild($"{logger.Parameters} -flp:logfile={Path.Combine(logFolder.Path, "logFile.log")};verbosity=diagnostic", out success);
