diff --git a/src/Utilities/ToolLocationHelper.cs b/src/Utilities/ToolLocationHelper.cs
index 232bab41622..a6095129d51 100644
--- a/src/Utilities/ToolLocationHelper.cs
+++ b/src/Utilities/ToolLocationHelper.cs
@@ -160,7 +160,7 @@ public enum VisualStudioVersion
         Version160,
 
         /// <summary>
-        /// Visual Studio (Dev17)
+        /// Visual Studio 2022 (Dev17)
         /// </summary>
         Version170,
 
