diff --git a/src/MSBuild/TerminalLogger/TerminalLogger.cs b/src/MSBuild/TerminalLogger/TerminalLogger.cs
index 3a0d990241b..fc33271d50f 100644
--- a/src/MSBuild/TerminalLogger/TerminalLogger.cs
+++ b/src/MSBuild/TerminalLogger/TerminalLogger.cs
@@ -847,7 +847,7 @@ private string FormatEventMessage(
                     if (endLineNumber == 0)
                     {
                         builder.Append(endColumnNumber == 0 ?
-                            $"({lineNumber},{endColumnNumber}): " :
+                            $"({lineNumber},{columnNumber}): " :
                             $"({lineNumber},{columnNumber}-{endColumnNumber}): ");
                     }
                     else
