diff --git a/eng/AfterSigning.targets b/eng/AfterSigning.targets
index e699c81e7dc..0bc6b1db9cd 100644
--- a/eng/AfterSigning.targets
+++ b/eng/AfterSigning.targets
@@ -1,5 +1,5 @@
 <Project>
-    <Target Name="CopyAMD64Symbols" Condition="'$(OfficialBuild)' == 'true'" BeforeTargets="Build">
+    <Target Name="CopyAMD64Symbols" Condition="'$(OfficialBuild)' == 'true' AND '$(ArcadeBuildFromSource)' != 'true'" BeforeTargets="Build">
         <Copy 
             SourceFiles="$(ArtifactsBinDir)MSBuild\x64\Release\net472\MSBuild.pdb"
             DestinationFolder="$(ArtifactsSymStoreDirectory)\MSBuild\net472\amd64"
diff --git a/eng/Versions.props b/eng/Versions.props
index eae06d1dd7a..5a445f6014b 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -48,16 +48,4 @@
       <FileVersion>$(VersionPrefix).$(FileVersion.Split('.')[3])</FileVersion>
     </PropertyGroup>
   </Target>
-    <!-- Override with the target version from newer source build where '$(ArcadeBuildFromSource)' != 'true' was added.  -->
-
-      <Target Name="CopyAMD64Symbols" Condition="'$(OfficialBuild)' == 'true' AND '$(ArcadeBuildFromSource)' != 'true'" BeforeTargets="Build">
-        <Copy 
-            SourceFiles="$(ArtifactsBinDir)MSBuild\x64\Release\net472\MSBuild.pdb"
-            DestinationFolder="$(ArtifactsSymStoreDirectory)\MSBuild\net472\amd64"
-            />
-        <Copy 
-            SourceFiles="$(ArtifactsBinDir)MSBuildTaskHost\x64\Release\net35\MSBuildTaskHost.pdb"
-            DestinationFolder="$(ArtifactsSymStoreDirectory)\MSBuildTaskHost\net35\amd64"
-            />
-    </Target>
 </Project>
