diff --git a/src/Shared/FileUtilities.cs b/src/Shared/FileUtilities.cs
index 365071788b4..619dc0a36b9 100644
--- a/src/Shared/FileUtilities.cs
+++ b/src/Shared/FileUtilities.cs
@@ -246,8 +246,8 @@ internal static string EnsureNoTrailingSlash(string path)
         /// <summary>
         /// Ensures the path is enclosed within single quotes.
         /// </summary>
-        /// <param name="path"></param>
-        /// <returns></returns>
+        /// <param name="path">The path to check.</param>
+        /// <returns>The path enclosed by quotes.</returns>
         internal static string EnsureSingleQuotes(string path)
         {
             return EnsureQuotes(path);
@@ -256,8 +256,8 @@ internal static string EnsureSingleQuotes(string path)
         /// <summary>
         /// Ensures the path is enclosed within double quotes.
         /// </summary>
-        /// <param name="path"></param>
-        /// <returns></returns>
+        /// <param name="path">The path to check.</param>
+        /// <returns>The path enclosed by quotes.</returns>
         internal static string EnsureDoubleQuotes(string path)
         {
             return EnsureQuotes(path, isSingleQuote: false);
@@ -266,9 +266,9 @@ internal static string EnsureDoubleQuotes(string path)
         /// <summary>
         /// Ensures the path is enclosed within quotes.
         /// </summary>
-        /// <param name="path"></param>
-        /// <param name="isSingleQuote">indicates if single or double quotes should be used</param>
-        /// <returns></returns>
+        /// <param name="path">The path to check.</param>
+        /// <param name="isSingleQuote">Indicates if single or double quotes should be used</param>
+        /// <returns>The path enclosed by quotes.</returns>
         internal static string EnsureQuotes(string path, bool isSingleQuote = true)
         {
             path = FixFilePath(path);
