diff --git a/src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs b/src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs
index 23733c80155..b19506be953 100644
--- a/src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs
+++ b/src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs
@@ -946,15 +946,11 @@ private void ReadDiagnosticFields(BuildEventArgsFields fields)
 
         private ExtendedDataFields? ReadExtendedDataFields()
         {
-            ExtendedDataFields? fields = null;
+            string extendedType = ReadOptionalString()!;
+            IDictionary<string, string>? extendedMetadata = ReadStringDictionary();
+            string? extendedData = ReadOptionalString();
 
-            fields = new ExtendedDataFields();
-
-            fields.ExtendedType = ReadOptionalString();
-            fields.ExtendedMetadata = ReadStringDictionary();
-            fields.ExtendedData = ReadOptionalString();
-
-            return fields;
+            return new ExtendedDataFields(extendedType, extendedMetadata, extendedData);
         }
 
         private BuildEventArgsFields ReadBuildEventArgsFields(bool readImportance = false)
diff --git a/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs b/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs
index 76c805f4bae..0a21182e83c 100644
--- a/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs
+++ b/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs
@@ -484,7 +484,6 @@ private void Write(BuildMessageEventArgs e)
             }
         }
 
-
         private void Write(ProjectImportedEventArgs e)
         {
             Write(BinaryLogRecordKind.ProjectImported);
@@ -697,7 +696,7 @@ private void WriteMessageFields(BuildMessageEventArgs e, bool writeMessage = tru
             }
         }
 
-            private void WriteArguments(object[] arguments)
+        private void WriteArguments(object[] arguments)
         {
             if (arguments == null || arguments.Length == 0)
             {
