diff --git a/src/Tasks.UnitTests/Exec_Tests.cs b/src/Tasks.UnitTests/Exec_Tests.cs
index 7592f3aa949..50732b70703 100644
--- a/src/Tasks.UnitTests/Exec_Tests.cs
+++ b/src/Tasks.UnitTests/Exec_Tests.cs
@@ -158,7 +158,7 @@ public void ExitCodeCausesFailure()
         }
 
         [Fact]
-        public void Timeout1()
+        public void Timeout()
         {
             // On non-Windows the exit code of a killed process is SIGKILL (137)
             int expectedExitCode = NativeMethodsShared.IsWindows ? -1 : 137;
