diff --git a/src/Build/ElementLocation/ElementLocation.cs b/src/Build/ElementLocation/ElementLocation.cs
index 18e34b17440..ced2ed9f41c 100644
--- a/src/Build/ElementLocation/ElementLocation.cs
+++ b/src/Build/ElementLocation/ElementLocation.cs
@@ -185,7 +185,6 @@ internal static ElementLocation Create(string file)
         /// </remarks>
         internal static ElementLocation Create(string file, int line, int column)
         {
-            // return EmptyLocation;
             if (line == 0 && column == 0 && string.IsNullOrEmpty(file))
             {
                 return EmptyLocation;
