diff --git a/src/Directory.Build.props b/src/Directory.Build.props
index 1636c5b8c95..2263f0eef57 100644
--- a/src/Directory.Build.props
+++ b/src/Directory.Build.props
@@ -92,12 +92,6 @@
     <PublishWindowsPdb>false</PublishWindowsPdb>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(IsTestProject)' == 'true' and $(TargetFramework.StartsWith('net4'))">
-    <!-- OpenTelemetry in tests explicitly included because in framework it has PrivateAssets="all" -->
-    <PackageReference Include="Microsoft.VisualStudio.OpenTelemetry.ClientExtensions"/>
-    <PackageReference Include="Microsoft.VisualStudio.OpenTelemetry.Collector"/>
-  </ItemGroup>
-
   <!-- Ensure ref assemblies are placed under `ref/$(TargetFramework)` in the NuGet package -->
   <Target Name="ShipRefAssembliesToNuGetPackage" BeforeTargets="Pack" Condition="$(IsInnerBuild) == true">
     <ItemGroup>
