diff --git a/src/Tasks/GetAssembliesMetadata.cs b/src/Tasks/GetAssembliesMetadata.cs
index d5f67c62f36..520b979be03 100644
--- a/src/Tasks/GetAssembliesMetadata.cs
+++ b/src/Tasks/GetAssembliesMetadata.cs
@@ -50,12 +50,14 @@ public override bool Execute()
                 // During DTB the referenced project may not has been built yet, so we need to check if the assembly already exists.
                 if (File.Exists(assemblyPath))
                 {
-                    AssemblyInformation assemblyInformation = new(assemblyPath);
-                    AssemblyAttributes attributes = assemblyInformation.GetAssemblyMetadata();
-
-                    if (attributes != null)
+                    using (AssemblyInformation assemblyInformation = new(assemblyPath))
                     {
-                        assembliesMetadata.Add(CreateItemWithMetadata(attributes));
+                        AssemblyAttributes attributes = assemblyInformation.GetAssemblyMetadata();
+
+                        if (attributes != null)
+                        {
+                            assembliesMetadata.Add(CreateItemWithMetadata(attributes));
+                        }
                     }
                 }
             }
