diff --git a/src/MSBuild/TerminalLogger/TerminalLogger.cs b/src/MSBuild/TerminalLogger/TerminalLogger.cs
index fdc5170dbbc..e9699e87b31 100644
--- a/src/MSBuild/TerminalLogger/TerminalLogger.cs
+++ b/src/MSBuild/TerminalLogger/TerminalLogger.cs
@@ -1063,8 +1063,6 @@ private void EraseNodes()
     /// </summary>
     public static ILogger CreateTerminalOrConsoleLogger(LoggerVerbosity verbosity, string[]? args)
     {
-        Debugger.Launch();
-
         string tlArg = args?.FirstOrDefault(a => a.StartsWith("--tl:", StringComparison.InvariantCultureIgnoreCase)) ?? string.Empty;
 
         bool isDisabled =
