diff --git a/src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj b/src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj
index 57cba86c10c..fe4fa483058 100644
--- a/src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj
+++ b/src/Build.OM.UnitTests/Microsoft.Build.Engine.OM.UnitTests.csproj
@@ -16,7 +16,7 @@
 
   <PropertyGroup>
     <!-- Managed manually since PackageDownload is not supported by dependabot https://github.com/dependabot/dependabot-core/issues/2920 -->
-    <NuGetCommandLinePackageVersion>4.9.6</NuGetCommandLinePackageVersion>
+    <NuGetCommandLinePackageVersion>6.6.1</NuGetCommandLinePackageVersion>
   </PropertyGroup>
 
   <ItemGroup>
