diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index 2343c4845a5..45a3a11ae8d 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -596,7 +596,6 @@ private bool CopyParallel(
                     // Retail assert...
                     ErrorUtilities.ThrowInternalError("Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
                 }
-
             }
 
             partitionCopyActionBlock.Complete();
