diff --git a/setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec b/setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec
index ffef7075c05..c3de56bb5e5 100644
--- a/setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec
+++ b/setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec
@@ -32,7 +32,6 @@
     <file src="$X86BinPath$/System.Collections.Immutable.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.IO.Compression.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Runtime.InteropServices.RuntimeInformation.dll" target="v15.0/bin" />
-    <file src="$X86BinPath$/System.Security.Cryptography.Algorithms.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.Threading.Tasks.Dataflow.dll" target="v15.0/bin" />
 
     <file src="$X86BinPath$/Microsoft.Build.Core.xsd" target="v15.0/bin/MSBuild" />
@@ -75,7 +74,6 @@
     <file src="$X86BinPath$/System.Collections.Immutable.dll" target="v15.0/bin" />
     <file src="$X86BinPath$/System.IO.Compression.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/System.Runtime.InteropServices.RuntimeInformation.dll" target="v15.0/bin/amd64" />
-    <file src="$X86BinPath$/System.Security.Cryptography.Algorithms.dll" target="v15.0/bin/amd64" />
     <file src="$X86BinPath$/System.Threading.Tasks.Dataflow.dll" target="v15.0/bin/amd64" />
 
     <file src="$X86BinPath$/Microsoft.Build.Core.xsd" target="v15.0/bin/amd64/MSBuild" />
