diff --git a/ref/Microsoft.Build/net/Microsoft.Build.cs b/ref/Microsoft.Build/net/Microsoft.Build.cs
index 8daf09c374a..c0871fc11ab 100644
--- a/ref/Microsoft.Build/net/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/net/Microsoft.Build.cs
@@ -1,5 +1,12 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
+namespace Microsoft.Build.BackEnd
+{
+    public partial class CircularDependencyException : System.Exception
+    {
+        public CircularDependencyException(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context) { }
+    }
+}
 namespace Microsoft.Build.Construction
 {
     public abstract partial class ElementLocation
diff --git a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
index 23825be7aba..5951d38e2bc 100644
--- a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
@@ -1,5 +1,12 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
+namespace Microsoft.Build.BackEnd
+{
+    public partial class CircularDependencyException : System.Exception
+    {
+        public CircularDependencyException(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context) { }
+    }
+}
 namespace Microsoft.Build.Construction
 {
     public abstract partial class ElementLocation
diff --git a/src/Build/BackEnd/Shared/CircularDependencyException.cs b/src/Build/BackEnd/Shared/CircularDependencyException.cs
index 1f47f92edbe..a1281601b40 100644
--- a/src/Build/BackEnd/Shared/CircularDependencyException.cs
+++ b/src/Build/BackEnd/Shared/CircularDependencyException.cs
@@ -31,7 +31,7 @@ internal CircularDependencyException(string message)
         /// <summary>
         /// Constructor for deserialization.
         /// </summary>
-        protected CircularDependencyException(SerializationInfo info, StreamingContext context)
+        public CircularDependencyException(SerializationInfo info, StreamingContext context)
             : base(info, context)
         {
         }
