diff --git a/src/Tasks.UnitTests/ResolveComReference_Tests.cs b/src/Tasks.UnitTests/ResolveComReference_Tests.cs
index 81773360220..23afd470bdf 100644
--- a/src/Tasks.UnitTests/ResolveComReference_Tests.cs
+++ b/src/Tasks.UnitTests/ResolveComReference_Tests.cs
@@ -756,8 +756,8 @@ public void AxReferenceKnowsItsRCWPiaExists_IncludeVersion()
         private enum RcwStyle {
             GenerateTlb,
             PreexistingTlb,
-            PreexistingPia
-        };
+            PreexistingPia,
+        }
 
         /// <summary>
         /// Helper method that will new up an AX and matching TLB reference, and verify that the AX reference
diff --git a/src/Tasks/FormatVersion.cs b/src/Tasks/FormatVersion.cs
index 9a236cdb95c..b0f40cec9f6 100644
--- a/src/Tasks/FormatVersion.cs
+++ b/src/Tasks/FormatVersion.cs
@@ -19,8 +19,8 @@ public sealed class FormatVersion : TaskExtension
     {
         private enum _FormatType {
             Version,
-            Path
-        };
+            Path,
+        }
 
         private _FormatType _formatType = _FormatType.Version;
 
diff --git a/src/Tasks/GenerateApplicationManifest.cs b/src/Tasks/GenerateApplicationManifest.cs
index d9531a65c04..f4ea09d1a9d 100644
--- a/src/Tasks/GenerateApplicationManifest.cs
+++ b/src/Tasks/GenerateApplicationManifest.cs
@@ -19,8 +19,8 @@ public sealed class GenerateApplicationManifest : GenerateManifestBase
     {
         private enum _ManifestType {
             Native,
-            ClickOnce
-        };
+            ClickOnce,
+        }
 
         private ITaskItem[] _dependencies;
         private ITaskItem[] _files;
diff --git a/src/Tasks/GenerateManifestBase.cs b/src/Tasks/GenerateManifestBase.cs
index 5d96a55bd65..ac5f5d74291 100644
--- a/src/Tasks/GenerateManifestBase.cs
+++ b/src/Tasks/GenerateManifestBase.cs
@@ -20,13 +20,13 @@ private enum AssemblyType {
             Unspecified,
             Managed,
             Native,
-            Satellite
-        };
+            Satellite,
+        }
 
         private enum DependencyType {
             Install,
-            Prerequisite
-        };
+            Prerequisite,
+        }
 
         private string _processorArchitecture;
         private int _startTime;
