diff --git a/src/Build/Definition/ProjectCollection.cs b/src/Build/Definition/ProjectCollection.cs
index 49ffd7a431a..a228be8f4e2 100644
--- a/src/Build/Definition/ProjectCollection.cs
+++ b/src/Build/Definition/ProjectCollection.cs
@@ -1175,6 +1175,9 @@ public Project LoadProjectWithOptions(string fileName, ProjectOptions projectOpt
 
             using (_locker.EnterDisposableWriteLock())
             {
+                // We can only work on this Collection.
+                projectOptions.ProjectCollection = this;
+
                 if (projectOptions.GlobalProperties == null)
                 {
                     projectOptions.GlobalProperties = GlobalProperties;
