diff --git a/src/Build/BackEnd/Components/RequestBuilder/Lookup.cs b/src/Build/BackEnd/Components/RequestBuilder/Lookup.cs
index 4d84c13196c..93c744c5870 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/Lookup.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/Lookup.cs
@@ -482,6 +482,7 @@ public ICollection<ProjectItemInstance> GetItems(string itemType)
                     {
                         if (allAdds == null)
                         {
+                            // Use the List<T>(IEnumerable<T>) constructor to avoid an intermediate array allocation.
                             allAdds = new List<ProjectItemInstance>(adds);
                         }
                         else
