diff --git a/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs b/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs
index ad081c972c9..68961562f77 100644
--- a/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs
@@ -136,7 +136,7 @@ public void Build_WithCustomBuildArgs_NetCore(string envVariableValue, bool isWa
             }
         }
 
-        [WindowsOnlyTheory]
+        [WindowsFullFrameworkOnlyTheory]
         [InlineData("1", true)]
         [InlineData("0", false)]
         [InlineData("", false)]
