diff --git a/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs b/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs
index cbac77134a7..888db2e14cc 100644
--- a/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs
+++ b/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs
@@ -1062,14 +1062,10 @@ private List<string> GetUndeclaredProjects(MSBuild msbuildTask)
                     !(declaredProjects.Contains(normalizedMSBuildProject)
                       || _buildRequestEntry.RequestConfiguration.ShouldSkipIsolationConstraintsForReference(normalizedMSBuildProject)))
                 {
-                    if (undeclaredProjects == null)
-                    {
-                        undeclaredProjects = new List<string>(projectReferenceItems.Count);
-                    }
-
-                            undeclaredProjects.Add(normalizedMSBuildProject);
-                        }
+                    undeclaredProjects ??= new List<string>(projectReferenceItems.Count);
+                    undeclaredProjects.Add(normalizedMSBuildProject);
                 }
+            }
 
             return undeclaredProjects;
         }
