diff --git a/src/Shared/ExceptionHandling.cs b/src/Shared/ExceptionHandling.cs
index 3e3cae26f19..eca284ecb64 100644
--- a/src/Shared/ExceptionHandling.cs
+++ b/src/Shared/ExceptionHandling.cs
@@ -283,8 +283,8 @@ e is SerializationException ||
         /// </summary>
         internal static bool NotExpectedRegistryException(Exception e)
         {
-            if (e is UnauthorizedAccessException
-             || e is SecurityException
+            if (e is SecurityException
+             || e is UnauthorizedAccessException
              || e is IOException
              || e is ObjectDisposedException
              || e is ArgumentException)
