diff --git a/src/Deprecated/Engine/Engine/EngineLoggingServices.cs b/src/Deprecated/Engine/Engine/EngineLoggingServices.cs
index 15d7de1e2ce..018b94a2ae6 100644
--- a/src/Deprecated/Engine/Engine/EngineLoggingServices.cs
+++ b/src/Deprecated/Engine/Engine/EngineLoggingServices.cs
@@ -26,7 +26,7 @@ namespace Microsoft.Build.BuildEngine
     /// <remarks>
     /// We have made this an abstract class and not an interface to avoid forcing the "public" access
     /// modifier on the implementation of the internal logging services.
-    /// </remarks> 
+    /// </remarks>
     internal abstract class EngineLoggingServices
     {
         #region Initialization/Shutdown methods
