diff --git a/src/Shared/UnitTests/TestEnvironment.cs b/src/Shared/UnitTests/TestEnvironment.cs
index 3ea0ab9807b..b5f73d78f3a 100644
--- a/src/Shared/UnitTests/TestEnvironment.cs
+++ b/src/Shared/UnitTests/TestEnvironment.cs
@@ -257,7 +257,7 @@ public override void AssertInvariant(ITestOutputHelper output)
 
                     // Ignore clean shutdown trace logs.
                     if (Regex.IsMatch(file.Name, @"MSBuild_NodeShutdown_\d+\.txt") &&
-                        Regex.IsMatch(contents, @"Node shutting down with reason BuildComplete and exception:\s*"))
+                        Regex.IsMatch(contents, @"Node shutting down with reason BuildComplete(Reuse)* and exception:\s*"))
                     {
                         newFilesCount--;
                         continue;
