diff --git a/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs b/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
index 69836554daf..ea8a30e778e 100644
--- a/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/BuildManager_Tests.cs
@@ -3805,7 +3805,7 @@ public void OutOfProcFileBasedP2PBuildSucceeds()
         /// not reload it from disk but instead fully utilize the entire translate project instance state
         /// to do the build.
         /// </summary>
-        /// <param name="shouldSerializeEntireState">Serializes the entire project instance state if true.</param>
+        /// <param name="shouldSerializeEntireState"><see langword="true"/> to serialize the entire project instance state; otherwise, <see langword="false"/>.</param>
         [Theory]
         [InlineData(false)]
         [InlineData(true)]
