diff --git a/src/MSBuild.UnitTests/XMake_Tests.cs b/src/MSBuild.UnitTests/XMake_Tests.cs
index 18ddf7e4a94..e56d1355aae 100644
--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -37,8 +37,7 @@ public class XMakeAppTests : IDisposable
 #endif
 
         private readonly ITestOutputHelper _output;
-        private readonly TestEnvironment _env;
-        
+        private readonly TestEnvironment _env;        
         public XMakeAppTests(ITestOutputHelper output)
         {
             _output = output;
