diff --git a/src/Deprecated/Engine/Conditionals/Token.cs b/src/Deprecated/Engine/Conditionals/Token.cs
index cc23126a162..60d93cfaeee 100644
--- a/src/Deprecated/Engine/Conditionals/Token.cs
+++ b/src/Deprecated/Engine/Conditionals/Token.cs
@@ -16,24 +16,11 @@ internal sealed class Token
         /// </summary>
         internal enum TokenType 
         {
-            Comma,
-            LeftParenthesis,
-            RightParenthesis,
-            LessThan,
-            GreaterThan,
-            LessThanOrEqualTo,
-            GreaterThanOrEqualTo,
-            And,
-            Or,
-            EqualTo,
-            NotEqualTo,
-            Not,
-            Property,
-            String,
-            Numeric,
-            ItemList,
-            ItemMetadata,
-            Function,
+            Comma, LeftParenthesis, RightParenthesis,
+            LessThan, GreaterThan, LessThanOrEqualTo, GreaterThanOrEqualTo,
+            And, Or,
+            EqualTo, NotEqualTo, Not,
+            Property, String, Numeric, ItemList, ItemMetadata, Function,
             EndOfInput
         };
 
