diff --git a/.vsts-dotnet-ci.yml b/.vsts-dotnet-ci.yml
index 112f6725c36..d92440ece7a 100644
--- a/.vsts-dotnet-ci.yml
+++ b/.vsts-dotnet-ci.yml
@@ -22,7 +22,7 @@ phases:
     inputs:
       testRunTitle: 'Windows-on-full Full Framework'
       testRunner: XUnit
-      testResultsFiles: 'artifacts/**/*UnitTests_net46*.xml'
+      testResultsFiles: 'artifacts/**/*UnitTests_net472*.xml'
       publishRunAttachments: true
       mergeTestResults: true
     condition: always()
@@ -55,7 +55,7 @@ phases:
     inputs:
       testRunTitle: 'Windows-on-core Full Framework'
       testRunner: XUnit
-      testResultsFiles: 'artifacts/**/*UnitTests_net46*.xml'
+      testResultsFiles: 'artifacts/**/*UnitTests_net472*.xml'
       publishRunAttachments: true
       mergeTestResults: true
     condition: always()
diff --git a/Directory.Build.props b/Directory.Build.props
index 7838284b519..2de47b5b0e9 100644
--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -5,6 +5,10 @@
     <Configurations>Debug;Release;Debug-MONO;Release-MONO</Configurations>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <FullFrameworkTFM>net472</FullFrameworkTFM>
+  </PropertyGroup>
+
   <PropertyGroup>
     <OsEnvironment Condition="$([MSBuild]::IsOSPlatform('windows'))">windows</OsEnvironment>
     <OsEnvironment Condition="$([MSBuild]::IsOSPlatform('linux'))">linux</OsEnvironment>
diff --git a/build/SignToolData.json b/build/SignToolData.json
index b322929f0d2..51c5a5269f1 100644
--- a/build/SignToolData.json
+++ b/build/SignToolData.json
@@ -4,22 +4,22 @@
             "certificate": "Microsoft",
             "strongName": "StrongName",
             "values": [
-                "bin/MSBuild/net46/MSBuild.exe",
-                "bin/MSBuild/net46/Microsoft.Build.dll",
-                "bin/MSBuild/net46/Microsoft.Build.Framework.dll",
-                "bin/MSBuild/net46/Microsoft.Build.Tasks.Core.dll",
-                "bin/MSBuild/net46/Microsoft.Build.Utilities.Core.dll",
+                "bin/MSBuild/net472/MSBuild.exe",
+                "bin/MSBuild/net472/Microsoft.Build.dll",
+                "bin/MSBuild/net472/Microsoft.Build.Framework.dll",
+                "bin/MSBuild/net472/Microsoft.Build.Tasks.Core.dll",
+                "bin/MSBuild/net472/Microsoft.Build.Utilities.Core.dll",
 
-                "bin/MSBuild/net46/*/*.resources.dll",
+                "bin/MSBuild/net472/*/*.resources.dll",
 
                 "bin/MSBuildTaskHost/net35/MSBuildTaskHost.exe",
                 "bin/MSBuildTaskHost/net35/*/MSBuildTaskHost.resources.dll",
 
-                "bin/MSBuild/x64/net46/MSBuild.exe",
+                "bin/MSBuild/x64/net472/MSBuild.exe",
 
                 "bin/MSBuildTaskHost/x64/net35/MSBuildTaskHost.exe",
-                "bin/Microsoft.Build.Conversion/net46/Microsoft.Build.Engine.dll",
-                "bin/Microsoft.Build.Conversion/net46/Microsoft.Build.Conversion.Core.dll",
+                "bin/Microsoft.Build.Conversion/net472/Microsoft.Build.Engine.dll",
+                "bin/Microsoft.Build.Conversion/net472/Microsoft.Build.Conversion.Core.dll",
 
                 "bin/MSBuild/netcoreapp2.1/MSBuild.dll",
                 "bin/MSBuild/netcoreapp2.1/Microsoft.Build.dll",
diff --git a/build/build.ps1 b/build/build.ps1
index 5232e09bb38..c848786d1f0 100644
--- a/build/build.ps1
+++ b/build/build.ps1
@@ -275,12 +275,12 @@ function Build {
 
       if ($hostType -eq 'full')
       {
-        $msbuildToUse = Join-Path $bootstrapRoot "net46\MSBuild\15.0\Bin\MSBuild.exe"
+        $msbuildToUse = Join-Path $bootstrapRoot "net472\MSBuild\15.0\Bin\MSBuild.exe"
 
         if ($configuration -eq "Debug-MONO" -or $configuration -eq "Release-MONO")
         {
           # Copy MSBuild.dll to MSBuild.exe so we can run it without a host
-          $sourceDll = Join-Path $bootstrapRoot "net46\MSBuild\15.0\Bin\MSBuild.dll"
+          $sourceDll = Join-Path $bootstrapRoot "net472\MSBuild\15.0\Bin\MSBuild.dll"
           Copy-Item -Path $sourceDll -Destination $msbuildToUse
         }
       }
diff --git a/src/Build.UnitTests/Microsoft.Build.Engine.UnitTests.csproj b/src/Build.UnitTests/Microsoft.Build.Engine.UnitTests.csproj
index 356967d7047..61323b6e6cc 100644
--- a/src/Build.UnitTests/Microsoft.Build.Engine.UnitTests.csproj
+++ b/src/Build.UnitTests/Microsoft.Build.Engine.UnitTests.csproj
@@ -21,8 +21,8 @@
     <ProjectReference Include="..\Xunit.NetCore.Extensions\Xunit.NetCore.Extensions.csproj" />
 
     <ProjectReference Include="..\Samples\TaskWithDependency\TaskWithDependency.csproj" Private="false" ReferenceOutputAssembly="false" OutputItemType="TaskWithDependencyResolvedProjectReferencePath">
-      <SetTargetFramework Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">TargetFramework=net46</SetTargetFramework>
-      <SetTargetFramework Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(MonoBuild)' == 'true'">TargetFramework=net461</SetTargetFramework>
+      <SetTargetFramework Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">TargetFramework=$(FullFrameworkTFM)</SetTargetFramework>
+      <SetTargetFramework Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(MonoBuild)' == 'true'">TargetFramework=$(FullFrameworkTFM)</SetTargetFramework>
       <SetTargetFramework Condition="'$(TargetFrameworkIdentifier)' != '.NETFramework'">TargetFramework=netstandard2.0</SetTargetFramework>
     </ProjectReference>
     <ProjectReference Include="..\Samples\PortableTask\PortableTask.csproj" Private="false" ReferenceOutputAssembly="false" OutputItemType="PortableTaskResolvedProjectReferencePath" SetTargetFramework="TargetFramework=netstandard1.3" />
diff --git a/src/Build/Microsoft.Build.csproj b/src/Build/Microsoft.Build.csproj
index 2218d4074db..e98e8525db5 100644
--- a/src/Build/Microsoft.Build.csproj
+++ b/src/Build/Microsoft.Build.csproj
@@ -6,7 +6,7 @@
     <!-- Node reuse requires an API new to .NET Core 2.1 not yet available
          in .NETStandard. -->
     <TargetFrameworks>netcoreapp2.1</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OsEnvironment)'=='windows'">net46;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OsEnvironment)'=='windows'">$(FullFrameworkTFM);netcoreapp2.1</TargetFrameworks>
     <TargetFrameworks Condition="'$(MonoBuild)'=='true'">$(RuntimeOutputTargetFrameworks)</TargetFrameworks>
     <RootNamespace>Microsoft.Build</RootNamespace>
     <AssemblyName>Microsoft.Build</AssemblyName>
diff --git a/src/Deprecated/Conversion/Microsoft.Build.Conversion.csproj b/src/Deprecated/Conversion/Microsoft.Build.Conversion.csproj
index 774d31cc555..879976ebd73 100644
--- a/src/Deprecated/Conversion/Microsoft.Build.Conversion.csproj
+++ b/src/Deprecated/Conversion/Microsoft.Build.Conversion.csproj
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="RoslynTools.RepoToolset">
   <PropertyGroup>
     <AssemblyName>Microsoft.Build.Conversion.Core</AssemblyName>
-    <TargetFramework>net46</TargetFramework>
+    <TargetFramework>$(FullFrameworkTFM)</TargetFramework>
     <GenerateReferenceAssemblySources>true</GenerateReferenceAssemblySources>
     <EnableDefaultItems>false</EnableDefaultItems>
     <IsPackable>true</IsPackable>
diff --git a/src/Deprecated/Engine/Microsoft.Build.Engine.csproj b/src/Deprecated/Engine/Microsoft.Build.Engine.csproj
index 070ac7f4d48..7acb8bf589c 100644
--- a/src/Deprecated/Engine/Microsoft.Build.Engine.csproj
+++ b/src/Deprecated/Engine/Microsoft.Build.Engine.csproj
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="RoslynTools.RepoToolset">
   <PropertyGroup>
-    <TargetFramework>net46</TargetFramework>
+    <TargetFramework>$(FullFrameworkTFM)</TargetFramework>
     <NoWarn>$(NoWarn);618</NoWarn>
     <GenerateReferenceAssemblySources>true</GenerateReferenceAssemblySources>
     <EnableDefaultItems>false</EnableDefaultItems>
diff --git a/src/Directory.Build.props b/src/Directory.Build.props
index b1ffb2f3acd..80eaea9a32e 100644
--- a/src/Directory.Build.props
+++ b/src/Directory.Build.props
@@ -29,13 +29,13 @@
     <Platforms>AnyCPU;x64</Platforms>
     
     <!-- Defaults for target frameworks and architecture -->
-    <LibraryTargetFrameworks>net46;netstandard2.0</LibraryTargetFrameworks>
+    <LibraryTargetFrameworks>$(FullFrameworkTFM);netstandard2.0</LibraryTargetFrameworks>
     <LibraryTargetFrameworks Condition="'$(MonoBuild)'=='true'">net461</LibraryTargetFrameworks>
     <PlatformTarget>AnyCPU</PlatformTarget>
 
     <!-- Target frameworks for Exe and unit test projects (ie projects with runtime output) -->
     <RuntimeOutputTargetFrameworks>netcoreapp2.1</RuntimeOutputTargetFrameworks>
-    <RuntimeOutputTargetFrameworks Condition="'$(OsEnvironment)'=='windows'">net46;$(RuntimeOutputTargetFrameworks)</RuntimeOutputTargetFrameworks>
+    <RuntimeOutputTargetFrameworks Condition="'$(OsEnvironment)'=='windows'">$(FullFrameworkTFM);$(RuntimeOutputTargetFrameworks)</RuntimeOutputTargetFrameworks>
     <RuntimeOutputTargetFrameworks Condition="'$(MonoBuild)' == 'true'">net461</RuntimeOutputTargetFrameworks>
 
     <!-- Don't automatically append target framework to output path, since we want to put the Platform Target beforehand, if it's not AnyCPU -->
diff --git a/src/MSBuild/MSBuild.csproj b/src/MSBuild/MSBuild.csproj
index 9cc9fa09da3..b9b80f3f60f 100644
--- a/src/MSBuild/MSBuild.csproj
+++ b/src/MSBuild/MSBuild.csproj
@@ -249,14 +249,14 @@
   <Target Name="GetCustomPackageFiles" BeforeTargets="_GetPackageFiles" DependsOnTargets="RemoveSatelliteDllsFromBuildOutputInPackage">
 
     <ItemGroup>
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\MSBuild.exe.config" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.Data.Entity.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.ServiceModel.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.WinFx.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.WorkflowBuildExtensions.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.Xaml.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Workflow.Targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
-      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Workflow.VisualBasic.Targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == 'net46' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\MSBuild.exe.config" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.Data.Entity.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.ServiceModel.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.WinFx.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.WorkflowBuildExtensions.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Microsoft.Xaml.targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Workflow.Targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
+      <_OurFiles Include="$(OutputPath)%(_TargetFrameworks.Identity)\Workflow.VisualBasic.Targets" TargetFramework="%(_TargetFrameworks.Identity)" Condition=" '%(_TargetFrameworks.Identity)' == '$(FullFrameworkTFM)' " />
     </ItemGroup>
 
     <ItemGroup>
diff --git a/src/Package/DevDivPackage/DevDivPackage.csproj b/src/Package/DevDivPackage/DevDivPackage.csproj
index 5fd687244fe..1efb9c059b2 100644
--- a/src/Package/DevDivPackage/DevDivPackage.csproj
+++ b/src/Package/DevDivPackage/DevDivPackage.csproj
@@ -1,7 +1,7 @@
 ﻿<Project>
   <Import Project="Sdk.props" Sdk="RoslynTools.RepoToolset" />
   <PropertyGroup>
-    <TargetFramework>net46</TargetFramework>
+    <TargetFramework>$(FullFrameworkTFM)</TargetFramework>
     <NuspecFile>VS.ExternalAPIs.MSBuild.nuspec</NuspecFile>
     <PlatformTarget>x86</PlatformTarget>
   </PropertyGroup>
diff --git a/src/Package/DevDivPackage/VS.ExternalAPIs.MSBuild.nuspec b/src/Package/DevDivPackage/VS.ExternalAPIs.MSBuild.nuspec
index bd36cf0034a..00b1c66054f 100644
--- a/src/Package/DevDivPackage/VS.ExternalAPIs.MSBuild.nuspec
+++ b/src/Package/DevDivPackage/VS.ExternalAPIs.MSBuild.nuspec
@@ -8,17 +8,17 @@
     <version>0.0</version>
   </metadata>
   <files>
-    <file src="Microsoft.Build.dll" target="lib\net46" />
-    <file src="Microsoft.Build.Framework.dll" target="lib\net46" />
-    <file src="Microsoft.Build.Tasks.Core.dll" target="lib\net46" />
-    <file src="Microsoft.Build.Utilities.Core.dll" target="lib\net46" />
+    <file src="Microsoft.Build.dll" target="lib\net472" />
+    <file src="Microsoft.Build.Framework.dll" target="lib\net472" />
+    <file src="Microsoft.Build.Tasks.Core.dll" target="lib\net472" />
+    <file src="Microsoft.Build.Utilities.Core.dll" target="lib\net472" />
 
     <!-- Obsolete but still referenced in the VS repo -->
-    <file src="Microsoft.Build.Engine.dll" target="lib\net46" />
-    <file src="Microsoft.Build.Conversion.Core.dll" target="lib\net46" />
+    <file src="Microsoft.Build.Engine.dll" target="lib\net472" />
+    <file src="Microsoft.Build.Conversion.Core.dll" target="lib\net472" />
 
-    <file src="Microsoft.Build.tlb" target="lib\net46" />
-    <file src="Microsoft.Build.Framework.tlb" target="lib\net46" />
-    <file src="Microsoft.Build.Tasks.Core.tlb" target="lib\net46" />
+    <file src="Microsoft.Build.tlb" target="lib\net472" />
+    <file src="Microsoft.Build.Framework.tlb" target="lib\net472" />
+    <file src="Microsoft.Build.Tasks.Core.tlb" target="lib\net472" />
   </files>
 </package>
\ No newline at end of file
diff --git a/src/Package/GetBinPaths.targets b/src/Package/GetBinPaths.targets
index 2e338ca11ce..30eb690cf10 100644
--- a/src/Package/GetBinPaths.targets
+++ b/src/Package/GetBinPaths.targets
@@ -12,7 +12,7 @@
          We have to treat these separately from normal project references, as the AssignProjectConfiguration task would overwrite
          the SetPlatform item metadata if they were ProjectReferences.
     -->
-    <X64ProjectReference SetPlatform="Platform=x64" SetTargetFramework="TargetFramework=net46" Include="..\..\MSBuild\MSBuild.csproj" OutputItemType="MSBuildX64ResolvedProjectReferencePath" />
+    <X64ProjectReference SetPlatform="Platform=x64" SetTargetFramework="TargetFramework=$(FullFrameworkTFM)" Include="..\..\MSBuild\MSBuild.csproj" OutputItemType="MSBuildX64ResolvedProjectReferencePath" />
     <X64ProjectReference SetPlatform="Platform=x64" Include="..\..\MSBuildTaskHost\MSBuildTaskHost.csproj" OutputItemType="MSBuildTaskHostX64ResolvedProjectReferencePath" />
 
   </ItemGroup>
diff --git a/src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj b/src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj
index 283538260cf..2ecbce1ad36 100644
--- a/src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj
+++ b/src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="RoslynTools.RepoToolset">
   <PropertyGroup>
-    <TargetFramework>net46</TargetFramework>
+    <TargetFramework>$(FullFrameworkTFM)</TargetFramework>
     <NuspecFile>MSBuild.Engine.Corext.nuspec</NuspecFile>
     <PlatformTarget>x86</PlatformTarget>
 
diff --git a/src/Package/MSBuild.VSSetup/MSBuild.VSSetup.csproj b/src/Package/MSBuild.VSSetup/MSBuild.VSSetup.csproj
index 469d1c7383f..e2d25d9c949 100644
--- a/src/Package/MSBuild.VSSetup/MSBuild.VSSetup.csproj
+++ b/src/Package/MSBuild.VSSetup/MSBuild.VSSetup.csproj
@@ -9,7 +9,7 @@
    <Import Sdk="RoslynTools.RepoToolset" Project="Sdk.props" />
 
    <PropertyGroup>
-    <TargetFramework>net46</TargetFramework>
+    <TargetFramework>$(FullFrameworkTFM)</TargetFramework>
 
     <!-- The IBCMerge process tries to treat this project as
          the real Microsoft.Build.dll, producing thousands
diff --git a/src/nuget/Microsoft.Build.Conversion.Core.nuspec b/src/nuget/Microsoft.Build.Conversion.Core.nuspec
index f0a3b67e954..1bf12ec7c5b 100644
--- a/src/nuget/Microsoft.Build.Conversion.Core.nuspec
+++ b/src/nuget/Microsoft.Build.Conversion.Core.nuspec
@@ -16,13 +16,13 @@
     </description>
     <copyright>© Microsoft Corporation. All rights reserved.</copyright>
     <dependencies>
-      <group targetFramework="net46">
+      <group targetFramework="net472">
         <dependency id="Microsoft.Build" version="$version$" />
         <dependency id="Microsoft.Build.Engine" version="$version$" />
       </group>
     </dependencies>
   </metadata>
   <files>
-    <file src="Microsoft.Build.Conversion.Core.dll" target="lib\net46" />
+    <file src="Microsoft.Build.Conversion.Core.dll" target="lib\net472" />
   </files>
 </package>
\ No newline at end of file
diff --git a/src/nuget/Microsoft.Build.Engine.nuspec b/src/nuget/Microsoft.Build.Engine.nuspec
index 0c553c791e8..c67e3b03853 100644
--- a/src/nuget/Microsoft.Build.Engine.nuspec
+++ b/src/nuget/Microsoft.Build.Engine.nuspec
@@ -16,12 +16,12 @@
     </description>
     <copyright>© Microsoft Corporation. All rights reserved.</copyright>
     <dependencies>
-      <group targetFramework="net46">
+      <group targetFramework="net472">
         <dependency id="Microsoft.Build.Framework" version="$version$" />
       </group>
     </dependencies>
   </metadata>
   <files>
-    <file src="Microsoft.Build.Engine.dll" target="lib\net46" />
+    <file src="Microsoft.Build.Engine.dll" target="lib\net472" />
   </files>
 </package>
\ No newline at end of file
