diff --git a/src/Shared/NodeEndpointOutOfProcBase.cs b/src/Shared/NodeEndpointOutOfProcBase.cs
index e9f4b404cb3..7bb77dda88f 100644
--- a/src/Shared/NodeEndpointOutOfProcBase.cs
+++ b/src/Shared/NodeEndpointOutOfProcBase.cs
@@ -472,9 +472,11 @@ private void PacketPumpProc()
             {
                 if (localPipeServer.IsConnected)
                 {
+#if NETCOREAPP // OperatingSystem.IsWindows() is new in .NET 5.0
                     if (OperatingSystem.IsWindows())
+#endif
                     {
-                    localPipeServer.WaitForPipeDrain();
+                        localPipeServer.WaitForPipeDrain();
                     }
 
                     localPipeServer.Disconnect();
