diff --git a/src/StringTools/WeakStringCache.Concurrent.cs b/src/StringTools/WeakStringCache.Concurrent.cs
index fa6c0c31f2e..fc9078ba738 100644
--- a/src/StringTools/WeakStringCache.Concurrent.cs
+++ b/src/StringTools/WeakStringCache.Concurrent.cs
@@ -65,7 +65,7 @@ public string GetOrCreateEntry(ref InternableString internable, out bool cacheHi
             handle.SetString(result);
             if (_stringsByHashCode.TryAdd(hashCode, handle))
             {
-                Interlocked.Add(ref _count, 1);
+                Interlocked.Increment(ref _count);
             }
 
 
