diff --git a/src/Tasks/Exec.cs b/src/Tasks/Exec.cs
index b0696e0069c..f140c9dc73d 100644
--- a/src/Tasks/Exec.cs
+++ b/src/Tasks/Exec.cs
@@ -628,7 +628,7 @@ protected internal override void AddCommandLineCommands(CommandLineBuilderExtens
                     {
                         char c = batchFileForCommandLine[i];
 
-                        if (c == '&' || (ShouldEscapeCharacter(c)) && (i == 0 || batchFileForCommandLine[i - 1] != '^'))
+                        if ((c == '&' || ShouldEscapeCharacter(c)) && (i == 0 || batchFileForCommandLine[i - 1] != '^'))
                         {
                             fileName.Append('^');
                         }
