diff --git a/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs b/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs
index b00fdaf839a..1e88174045e 100644
--- a/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/TaskRegistry_Tests.cs
@@ -676,7 +676,7 @@ public void OverriddenTask_AlwaysWins(string firstArch, string firstOverride, st
         }
 
         [Fact]
-        public void OverriddenTask_MultipleOverridesCauseMSB4276()
+        public void OverriddenTask_MultipleOverridesCauseMSB4275()
         {
             string proj =
                 $"<Project>" +
@@ -708,8 +708,8 @@ public void OverriddenTask_MultipleOverridesCauseMSB4276()
                         parameters,
                         request);
 
-                    // We should see MSB4276: Multiple usingtask overrides with the same name
-                    logger.AssertLogContains("MSB4276");
+                    // We should see MSB4275: Multiple usingtask overrides with the same name
+                    logger.AssertLogContains("MSB4275");
                 }
             }
         }
