diff --git a/src/Utilities/TaskItem.cs b/src/Utilities/TaskItem.cs
index 6391d45e19a..d43175b77aa 100644
--- a/src/Utilities/TaskItem.cs
+++ b/src/Utilities/TaskItem.cs
@@ -75,7 +75,7 @@ public TaskItem()
         /// <summary>
         /// This constructor creates a new task item, given the item spec.
         /// </summary>
-        /// <comments>Assumes the itemspec passed in is escaped.</comments>
+        /// <comments>Assumes the itemspec passed in is escaped and represents a file path. </comments>
         /// <param name="itemSpec">The item-spec string.</param>
         public TaskItem(string itemSpec)
             : this(itemSpec, treatAsFilePath: true) { }
