diff --git a/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs b/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
index ec8b2b2dc43..44a21a2321f 100644
--- a/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
@@ -4616,7 +4616,8 @@ public void VerifyPropertyTrackingLoggingDefault()
                     logger
                         .AllBuildEvents
                         .OfType<PropertyReassignmentEventArgs>()
-                        .ShouldHaveSingleItem();
+                        .Count()
+                        .ShouldBe(2);
 
                     logger
                         .AllBuildEvents
@@ -4683,7 +4684,8 @@ public void VerifyPropertyTrackingLoggingNone()
                     logger
                        .AllBuildEvents
                        .OfType<PropertyReassignmentEventArgs>()
-                       .ShouldHaveSingleItem();
+                       .Count()
+                       .ShouldBe(2);
                 });
         }
 
