diff --git a/src/Framework/Logging/AnsiCodes.cs b/src/Framework/Logging/AnsiCodes.cs
index e3a8c3aabbf..399d7817217 100644
--- a/src/Framework/Logging/AnsiCodes.cs
+++ b/src/Framework/Logging/AnsiCodes.cs
@@ -124,6 +124,11 @@ internal static class AnsiCodes
     /// </remarks>
     public const string RemoveProgress = "\x1b]9;4;0;\x1b\\";
 
+    /// <summary>
+    /// Remove all ANSI escape codes from a string.
+    /// </summary>
+    private static readonly Regex RemoveAnsiRegex = new Regex(@"\x1B\[[^@-~]*[@-~]", RegexOptions.Compiled);
+
     public static string Colorize(string? s, TerminalColor color)
     {
         if (string.IsNullOrWhiteSpace(s))
@@ -160,6 +165,6 @@ public static string MakeBold(string? s)
     /// <returns></returns>
     public static string RemoveAnsiCodes(string text)
     {
-        return new Regex(@"\x1B\[[^@-~]*[@-~]").Replace(text, "");
+        return RemoveAnsiRegex.Replace(text, "");
     }
 }
