diff --git a/src/Shared/BinaryWriterExtensions.cs b/src/Shared/BinaryWriterExtensions.cs
index daaf7ebb7c1..44cffcddf63 100644
--- a/src/Shared/BinaryWriterExtensions.cs
+++ b/src/Shared/BinaryWriterExtensions.cs
@@ -87,7 +87,6 @@ public static void WriteGuid(this BinaryWriter writer, Guid value)
             }
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void WriteExtendedBuildEventData(this BinaryWriter writer, IExtendedBuildEventArgs data)
         {
             writer.Write(data.ExtendedType);
