diff --git a/src/Build/BackEnd/Node/OutOfProcNode.cs b/src/Build/BackEnd/Node/OutOfProcNode.cs
index 852e5e900bb..69f7990dd51 100644
--- a/src/Build/BackEnd/Node/OutOfProcNode.cs
+++ b/src/Build/BackEnd/Node/OutOfProcNode.cs
@@ -702,8 +702,21 @@ private void HandleNodeConfiguration(NodeConfiguration configuration)
                 NativeMethodsShared.SetCurrentDirectory(BuildEnvironmentHelper.Instance.CurrentMSBuildToolsDirectory);
             }
 
-            // Replicate the environment.
-            CommunicationsUtilities.SetEnvironment(_buildParameters.BuildProcessEnvironment);
+            // Replicate the environment.  First, unset any environment variables set by the previous configuration.
+            if (_currentConfiguration != null)
+            {
+                foreach (string key in _currentConfiguration.BuildParameters.BuildProcessEnvironment.Keys)
+                {
+                    Environment.SetEnvironmentVariable(key, null);
+                }
+            }
+
+            // Now set the new environment and update Traits class accordingly
+            foreach (KeyValuePair<string, string> environmentPair in _buildParameters.BuildProcessEnvironment)
+            {
+                Environment.SetEnvironmentVariable(environmentPair.Key, environmentPair.Value);
+            }
+
             Traits.UpdateFromEnvironment();
 
             // We want to make sure the global project collection has the toolsets which were defined on the parent
