diff --git a/eng/Packages.props b/eng/Packages.props
index 390a1f9eea6..489ea061c95 100644
--- a/eng/Packages.props
+++ b/eng/Packages.props
@@ -30,5 +30,6 @@
     <PackageVersion Include="System.Text.Encoding.CodePages" Version="$(SystemTextEncodingCodePagesVersion)" />
     <PackageVersion Include="System.Text.Json" Version="$(SystemTextJsonVersion)" />
     <PackageVersion Include="System.Threading.Tasks.Dataflow" Version="$(SystemThreadingTasksDataflowVersion)" />
+    <PackageVersion Include="xunit.console" Version="$(XUnitVersion)" />
   </ItemGroup>
 </Project>
diff --git a/src/Directory.Build.targets b/src/Directory.Build.targets
index f6abd29f8b0..68295e63b40 100644
--- a/src/Directory.Build.targets
+++ b/src/Directory.Build.targets
@@ -86,7 +86,6 @@
   </PropertyGroup>
 
   <ItemGroup Condition="'$(IsUnitTestProject)' == 'true' And '$(TargetFrameworkIdentifier)' != '.NETFramework' ">
-    <PackageVersion Include="xunit.console" Version="$(XUnitVersion)" />
     <PackageReference Include="xunit.console" />
   </ItemGroup>
 
diff --git a/src/Xunit.NetCore.Extensions/Xunit.NetCore.Extensions.csproj b/src/Xunit.NetCore.Extensions/Xunit.NetCore.Extensions.csproj
index 3c9533fd8e2..cc0e7d35a90 100644
--- a/src/Xunit.NetCore.Extensions/Xunit.NetCore.Extensions.csproj
+++ b/src/Xunit.NetCore.Extensions/Xunit.NetCore.Extensions.csproj
@@ -7,6 +7,7 @@
     <IsTestProject>false</IsTestProject><!-- while this project references xunit it isn't itself a test -->
   </PropertyGroup>
   <ItemGroup>
+    <!-- Implicit reference to xunit.core and xunit.assert in the test project by .NET SDK causes it's not able to put PackageVersion in .props file. -->
     <PackageVersion Include="xunit.core" Version="$(XUnitVersion)" />
     <PackageReference Include="xunit.core" />
   </ItemGroup>
