diff --git a/src/Shared/FileMatcher.cs b/src/Shared/FileMatcher.cs
index eee91652d72..17528d2b90b 100644
--- a/src/Shared/FileMatcher.cs
+++ b/src/Shared/FileMatcher.cs
@@ -178,14 +178,6 @@ internal static bool HasWildcards(string filespec)
             return -1 != filespec.LastIndexOfAny(s_wildcardCharacters);
         }
 
-        /// <summary>
-        /// Determines whether the given path has any wild card characters or semicolons.
-        /// </summary>
-        internal static bool HasWildcardsOrSemicolon(string filespec)
-        {
-            return -1 != filespec.LastIndexOfAny(s_wildcardAndSemicolonCharacters);
-        }
-
         /// <summary>
         /// Determines whether the given path has any wild card characters, any semicolons or any property references.
         /// </summary>
