diff --git a/src/Build/BackEnd/BuildManager/RequestedProjectState.cs b/src/Build/BackEnd/BuildManager/RequestedProjectState.cs
index c25377390c7..1706eb7c6db 100644
--- a/src/Build/BackEnd/BuildManager/RequestedProjectState.cs
+++ b/src/Build/BackEnd/BuildManager/RequestedProjectState.cs
@@ -94,8 +94,7 @@ internal RequestedProjectState Merge(RequestedProjectState other)
 
                 if (mergedProperties.Count > 0)
                 {
-                    result.PropertyFilters = mergedProperties?.ToList();
-                }
+                    result.PropertyFilters = mergedProperties.ToList();
             }
 
             // Merge item filters
