diff --git a/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs b/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
index 8c7ab120ec4..316303574c8 100644
--- a/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/Evaluator_Tests.cs
@@ -4597,7 +4597,7 @@ public void VerifyMSBuildLogsAMessageWhenLocalPropertyCannotOverrideValueOfGloba
         public void VerifyPropertyTrackingLoggingDefault()
         {
             // Having just environment variables defined should default to nothing being logged except one environment variable read.
-            this.VerifyPropertyTrackingLoggingScenario(
+            VerifyPropertyTrackingLoggingScenario(
                 null,
                 logger =>
                 {
@@ -4613,11 +4613,6 @@ public void VerifyPropertyTrackingLoggingDefault()
                         .EnvironmentVariableName
                         .ShouldBe("DEFINED_ENVIRONMENT_VARIABLE2");
 
-                    logger
-                        .AllBuildEvents
-                        .OfType<PropertyReassignmentEventArgs>()
-                        .ShouldBeEmpty();
-
                     logger
                         .AllBuildEvents
                         .OfType<PropertyInitialValueSetEventArgs>()
@@ -4628,7 +4623,7 @@ public void VerifyPropertyTrackingLoggingDefault()
         [Fact]
         public void VerifyPropertyTrackingLoggingPropertyReassignment()
         {
-            this.VerifyPropertyTrackingLoggingScenario(
+            VerifyPropertyTrackingLoggingScenario(
                 "1",
                 logger =>
                 {
@@ -4675,11 +4670,6 @@ public void VerifyPropertyTrackingLoggingNone()
                         .EnvironmentVariableName
                         .ShouldBe("DEFINED_ENVIRONMENT_VARIABLE2");
 
-                    logger
-                        .AllBuildEvents
-                        .OfType<PropertyReassignmentEventArgs>()
-                        .ShouldBeEmpty();
-
                     logger
                         .AllBuildEvents
                         .OfType<PropertyInitialValueSetEventArgs>()
