diff --git a/src/BuildCheck.UnitTests/EndToEndTests.cs b/src/BuildCheck.UnitTests/EndToEndTests.cs
index 3158343c281..07e37f4ca96 100644
--- a/src/BuildCheck.UnitTests/EndToEndTests.cs
+++ b/src/BuildCheck.UnitTests/EndToEndTests.cs
@@ -74,6 +74,7 @@ public void WarningsCountExceedsLimitTest(bool buildInOutOfProcessNode, bool lim
         PrepareSampleProjectsAndConfig(
             buildInOutOfProcessNode,
             out TransientTestFile projectFile,
+            out _,
             "PropsCheckTestWithLimit.csproj");
 
         if (limitReportsCount)
