diff --git a/src/Build/BackEnd/Components/Communications/NodeManager.cs b/src/Build/BackEnd/Components/Communications/NodeManager.cs
index 98fa267e20a..8fddd390d29 100644
--- a/src/Build/BackEnd/Components/Communications/NodeManager.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeManager.cs
@@ -93,7 +93,7 @@ public IList<NodeInfo> CreateNodes(NodeConfiguration configuration, NodeAffinity
             // We will prefer to make nodes on the "closest" providers first; in-proc, then
             // out-of-proc, then remote.
             // When we support distributed build, we will also consider the remote provider.
-            List<NodeInfo> nodes = new();
+            List<NodeInfo> nodes = new(numberOfNodesToCreate);
             if ((nodeAffinity == NodeAffinity.Any || nodeAffinity == NodeAffinity.InProc) && !_componentHost!.BuildParameters.DisableInProcNode)
             {
                 nodes.AddRange(AttemptCreateNode(_inProcNodeProvider!, configuration, numberOfNodesToCreate));
diff --git a/src/Build/BackEnd/Components/Communications/NodeProviderInProc.cs b/src/Build/BackEnd/Components/Communications/NodeProviderInProc.cs
index 7d169e96108..c9bfbaa7cb3 100644
--- a/src/Build/BackEnd/Components/Communications/NodeProviderInProc.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeProviderInProc.cs
@@ -196,7 +196,7 @@ public IList<NodeInfo> CreateNodes(int nextNodeId, INodePacketFactory factory, F
                 NodeInfo nodeInfo = new(nodeId, ProviderType);
                 if (!CreateNode(nodeId, factory, configurationFactory(nodeInfo)))
                 {
-                    // If it fails let it return what we have crated so far to so caller can somehow acquire missing nodes.
+                    // If it fails let it return what we have created so far so caller can somehow acquire missing nodes.
                     break;
                 }
 
