diff --git a/src/Build/BackEnd/Client/MSBuildClient.cs b/src/Build/BackEnd/Client/MSBuildClient.cs
index 93b79288a2f..d09100eda14 100644
--- a/src/Build/BackEnd/Client/MSBuildClient.cs
+++ b/src/Build/BackEnd/Client/MSBuildClient.cs
@@ -185,7 +185,7 @@ public MSBuildClientExitResult Execute(CancellationToken cancellationToken)
             {
                 CommunicationsUtilities.Trace("Failure to connect to a server.");
                 // Overwrite the client exit type from unexpected to connection error, since that would trigger the fallback to old build behavior.
-                _exitResult.MSBuildClientExitType = MSBuildClientExitType.ConnectionError;
+                _exitResult.MSBuildClientExitType = MSBuildClientExitType.CanNotConnect;
                 return _exitResult;
             }
 
@@ -533,7 +533,7 @@ private bool TryConnectToServer(int timeout)
             catch (Exception ex)
             {
                 CommunicationsUtilities.Trace("Failed to connect to server: {0}", ex);
-                _exitResult.MSBuildClientExitType = MSBuildClientExitType.ConnectionError;
+                _exitResult.MSBuildClientExitType = MSBuildClientExitType.CanNotConnect;
                 return false;
             }
 
diff --git a/src/Build/BackEnd/Client/MSBuildClientExitType.cs b/src/Build/BackEnd/Client/MSBuildClientExitType.cs
index b8061b3295e..53ee5283c0f 100644
--- a/src/Build/BackEnd/Client/MSBuildClientExitType.cs
+++ b/src/Build/BackEnd/Client/MSBuildClientExitType.cs
@@ -9,15 +9,15 @@ public enum MSBuildClientExitType
         /// </summary>
         Success,
         /// <summary>
-        /// Server is busy.
+        /// Server is busy. This would invoke a fallback behavior.
         /// </summary>
         ServerBusy,
         /// <summary>
-        /// Client was unable to connect to the server.
+        /// Client was unable to connect to the server. This would invoke a fallback behavior.
         /// </summary>
-        ConnectionError,
+        CanNotConnect,
         /// <summary>
-        /// Client was unable to launch the server.
+        /// Client was unable to launch the server. This would invoke a fallback behavior.
         /// </summary>
         LaunchError,
         /// <summary>
diff --git a/src/Build/PublicAPI/net/PublicAPI.Unshipped.txt b/src/Build/PublicAPI/net/PublicAPI.Unshipped.txt
index 7ce7bc26ef8..b82406837c7 100644
--- a/src/Build/PublicAPI/net/PublicAPI.Unshipped.txt
+++ b/src/Build/PublicAPI/net/PublicAPI.Unshipped.txt
@@ -9,7 +9,7 @@ Microsoft.Build.Experimental.MSBuildClientExitResult.MSBuildClientExitResult() -
 Microsoft.Build.Experimental.MSBuildClientExitResult.MSBuildClientExitType.get -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitResult.MSBuildClientExitType.set -> void
 Microsoft.Build.Experimental.MSBuildClientExitType
-Microsoft.Build.Experimental.MSBuildClientExitType.ConnectionError = 2 -> Microsoft.Build.Experimental.MSBuildClientExitType
+Microsoft.Build.Experimental.MSBuildClientExitType.CanNotConnect = 2 -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitType.LaunchError = 3 -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitType.ServerBusy = 1 -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitType.Success = 0 -> Microsoft.Build.Experimental.MSBuildClientExitType
diff --git a/src/Build/PublicAPI/netstandard/PublicAPI.Unshipped.txt b/src/Build/PublicAPI/netstandard/PublicAPI.Unshipped.txt
index 485d920eb6a..52364d2580b 100644
--- a/src/Build/PublicAPI/netstandard/PublicAPI.Unshipped.txt
+++ b/src/Build/PublicAPI/netstandard/PublicAPI.Unshipped.txt
@@ -9,7 +9,7 @@ Microsoft.Build.Experimental.MSBuildClientExitResult.MSBuildClientExitResult() -
 Microsoft.Build.Experimental.MSBuildClientExitResult.MSBuildClientExitType.get -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitResult.MSBuildClientExitType.set -> void
 Microsoft.Build.Experimental.MSBuildClientExitType
-Microsoft.Build.Experimental.MSBuildClientExitType.ConnectionError = 2 -> Microsoft.Build.Experimental.MSBuildClientExitType
+Microsoft.Build.Experimental.MSBuildClientExitType.CanNotConnect = 2 -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitType.LaunchError = 3 -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitType.ServerBusy = 1 -> Microsoft.Build.Experimental.MSBuildClientExitType
 Microsoft.Build.Experimental.MSBuildClientExitType.Success = 0 -> Microsoft.Build.Experimental.MSBuildClientExitType
diff --git a/src/MSBuild/MSBuildClientApp.cs b/src/MSBuild/MSBuildClientApp.cs
index b5187d198a1..cffa083e7ba 100644
--- a/src/MSBuild/MSBuildClientApp.cs
+++ b/src/MSBuild/MSBuildClientApp.cs
@@ -74,7 +74,8 @@ public static MSBuildApp.ExitType Execute(
             MSBuildClientExitResult exitResult = msbuildClient.Execute(cancellationToken);
 
             if (exitResult.MSBuildClientExitType == MSBuildClientExitType.ServerBusy ||
-                exitResult.MSBuildClientExitType == MSBuildClientExitType.ConnectionError)
+                exitResult.MSBuildClientExitType == MSBuildClientExitType.CanNotConnect ||
+                exitResult.MSBuildClientExitType == MSBuildClientExitType.LaunchError)
             {
                 // Server is busy, fallback to old behavior.
                 return MSBuildApp.Execute(commandLine);
