diff --git a/src/Utilities/ToolTask.cs b/src/Utilities/ToolTask.cs
index 50311c7a074..508f92c6fcf 100644
--- a/src/Utilities/ToolTask.cs
+++ b/src/Utilities/ToolTask.cs
@@ -862,7 +862,7 @@ protected void DeleteTempFile(string fileName)
 
                 // Warn only -- occasionally temp files fail to delete because of virus checkers; we
                 // don't want the build to fail in such cases
-                LogShared.LogWarningWithCodeFromResources("Shared.FailedDeletingTempFile", fileName, e.Message + " Processes: " + lockedFileMessage);
+                LogShared.LogWarningWithCodeFromResources("Shared.FailedDeletingTempFile", fileName, e.Message, lockedFileMessage);
             }
         }
 
