diff --git a/src/Framework.UnitTests/SystemSetup_Tests.cs b/src/Framework.UnitTests/SystemSetup_Tests.cs
index 82f2b77710e..132097e243e 100644
--- a/src/Framework.UnitTests/SystemSetup_Tests.cs
+++ b/src/Framework.UnitTests/SystemSetup_Tests.cs
@@ -14,7 +14,7 @@ public sealed class SystemSetup_Tests
         [Fact]
         public void VerifyLongPaths()
         {
-            NativeMethodsShared.MaxPath.ShouldBeGreaterThan(10000);
+            NativeMethodsShared.MaxPath.ShouldBeGreaterThan(10000, "Long paths are not enabled. Enable long paths via the registry.");
         }
 
 #if NETCOREAPP
@@ -28,7 +28,7 @@ public void VerifySymLinksEnabled()
             {
                 string symLink = File.CreateSymbolicLink(path, file.Path).FullName;
                 string contents = File.ReadAllText(path);
-                contents.ShouldBe("fileContents");
+                contents.ShouldBe("fileContents", "You do not have permissions to create symbolic links.");
             }
             finally
             {
