diff --git a/src/MSBuild/XMake.cs b/src/MSBuild/XMake.cs
index 6790805eaee..6376e0d3b8c 100644
--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -41,7 +41,6 @@
 using LoggerDescription = Microsoft.Build.Logging.LoggerDescription;
 using ForwardingLoggerRecord = Microsoft.Build.Logging.ForwardingLoggerRecord;
 using BinaryLogger = Microsoft.Build.Logging.BinaryLogger;
-using System.CodeDom;
 
 namespace Microsoft.Build.CommandLine
 {
@@ -1198,7 +1197,6 @@ string outputResultsCache
 #if MSBUILDENABLEVSPROFILING
                     DataCollection.CommentMarkProfile(8800, "Pending Build Request from MSBuild.exe");
 #endif
-
                     BuildResultCode? result = null;
 
                     var messagesToLogInBuildLoggers = Traits.Instance.EscapeHatches.DoNotSendDeferredMessagesToBuildManager
@@ -1231,7 +1229,6 @@ string outputResultsCache
                                 else
                                 {
                                     buildRequest = new BuildRequestData(projectFile, globalProperties, toolsVersion, targets, null);
-                                    
                                 }
                             }
 
