diff --git a/src/Tasks/GetSDKReferenceFiles.cs b/src/Tasks/GetSDKReferenceFiles.cs
index 284999b5a52..4468c610426 100644
--- a/src/Tasks/GetSDKReferenceFiles.cs
+++ b/src/Tasks/GetSDKReferenceFiles.cs
@@ -1273,10 +1273,6 @@ public void Translate(ITranslator translator)
                     info.IsWinMD = isWinmd;
                 }, count => new Dictionary<string, SdkReferenceInfo>(count, StringComparer.OrdinalIgnoreCase));
 
-                translator.TranslateDictionary(ref _directoryToFileList, (ITranslator t, ref string s) => t.Translate(ref s), (ITranslator t, ref List<string> fileList) =>
-                {
-                    t.Translate(ref fileList, (ITranslator t, ref string str) => { t.Translate(ref str); });
-                }, count => new Dictionary<string, List<string>>(count, StringComparer.OrdinalIgnoreCase));
                 translator.TranslateDictionary(ref _directoryToFileList, (ITranslator t, ref string s) => t.Translate(ref s), (ITranslator t, ref List<string> fileList) =>
                 {
                     t.Translate(ref fileList, (ITranslator t, ref string str) => { t.Translate(ref str); });
