diff --git a/src/Build.UnitTests/Instance/TaskItem_Tests.cs b/src/Build.UnitTests/Instance/TaskItem_Tests.cs
index f1286310c85..115752d7e04 100644
--- a/src/Build.UnitTests/Instance/TaskItem_Tests.cs
+++ b/src/Build.UnitTests/Instance/TaskItem_Tests.cs
@@ -178,7 +178,7 @@ public void TestDeepClone()
         }
 
         /// <summary>
-        /// Validate the presentation of metadata on a TaskItem, but of direct values and those inherited from
+        /// Validate the presentation of metadata on a TaskItem, both of direct values and those inherited from
         /// item definitions.
         /// </summary>
         [Fact]
