diff --git a/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs b/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs
index 7e5fbe3db1b..0aaa25088a2 100644
--- a/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs
+++ b/ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs
@@ -155,7 +155,6 @@ protected CustomBuildEventArgs(string message, string helpKeyword, string sender
     public abstract partial class EngineServices
     {
         public const int Version1 = 1;
-        public const int Version2 = 2;
         protected EngineServices() { }
         public virtual bool IsTaskInputLoggingEnabled { get { throw null; } }
         public virtual int Version { get { throw null; } }
diff --git a/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs b/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs
index 06d25ff198a..f4dc90e663c 100644
--- a/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs
+++ b/ref/Microsoft.Build.Framework/netstandard/Microsoft.Build.Framework.cs
@@ -155,7 +155,6 @@ protected CustomBuildEventArgs(string message, string helpKeyword, string sender
     public abstract partial class EngineServices
     {
         public const int Version1 = 1;
-        public const int Version2 = 2;
         protected EngineServices() { }
         public virtual bool IsTaskInputLoggingEnabled { get { throw null; } }
         public virtual int Version { get { throw null; } }
diff --git a/src/Framework/EngineServices.cs b/src/Framework/EngineServices.cs
index 2b78ae0db69..43b8d10f7c6 100644
--- a/src/Framework/EngineServices.cs
+++ b/src/Framework/EngineServices.cs
@@ -16,20 +16,15 @@ namespace Microsoft.Build.Framework
     public abstract class EngineServices
     {
         /// <summary>
-        /// Initial version with LogsMessagesOfImportance() as the only exposed member.
+        /// Initial version with LogsMessagesOfImportance() and IsTaskInputLoggingEnabled as the only exposed members.
         /// </summary>
         public const int Version1 = 1;
 
-        /// <summary>
-        /// This version added the IsTaskInputLoggingEnabled property.
-        /// </summary>
-        public const int Version2 = 2;
-
         /// <summary>
         /// An explicit version of this class. Must be incremented whenever new members are added. Derived classes should override
         /// the property to return the version actually being implemented.
         /// </summary>
-        public virtual int Version => Version2;
+        public virtual int Version => Version1; // Not updated since we have not shipped 17.0 yet
 
         /// <summary>
         /// Returns <see langword="true"/> if the given message importance is not guaranteed to be ignored by registered loggers.
