diff --git a/src/Build/Logging/BinaryLogger/BinaryLogger.cs b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
index cdc0c927a09..e2ae1c1e0f8 100644
--- a/src/Build/Logging/BinaryLogger/BinaryLogger.cs
+++ b/src/Build/Logging/BinaryLogger/BinaryLogger.cs
@@ -160,7 +160,11 @@ public void Initialize(IEventSource eventSource)
             stream = new BufferedStream(stream, bufferSize: 32768);
             binaryWriter = new BinaryWriter(stream);
             eventArgsWriter = new BuildEventArgsWriter(binaryWriter);
-            eventArgsWriter.EmbedFile += EventArgsWriter_EmbedFile;
+
+            if (projectImportsCollector != null)
+            {
+                eventArgsWriter.EmbedFile += EventArgsWriter_EmbedFile;
+            }
 
             binaryWriter.Write(FileFormatVersion);
 
diff --git a/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs b/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs
index b2b6c559787..4079c2c2ce0 100644
--- a/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs
+++ b/src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs
@@ -834,7 +834,9 @@ private void WriteProjectItems(IEnumerable items)
 
         private void CheckForFilesToEmbed(string itemType, object itemList)
         {
-            if (!string.Equals(itemType, ItemTypeNames.EmbedInBinlog, StringComparison.OrdinalIgnoreCase) || itemList is not IEnumerable list)
+            if (EmbedFile == null ||
+                !string.Equals(itemType, ItemTypeNames.EmbedInBinlog, StringComparison.OrdinalIgnoreCase) ||
+                itemList is not IEnumerable list)
             {
                 return;
             }
@@ -843,11 +845,11 @@ private void CheckForFilesToEmbed(string itemType, object itemList)
             {
                 if (item is ITaskItem taskItem && !string.IsNullOrEmpty(taskItem.ItemSpec))
                 {
-                    EmbedFile?.Invoke(taskItem.ItemSpec);
+                    EmbedFile.Invoke(taskItem.ItemSpec);
                 }
                 else if (item is string itemSpec && !string.IsNullOrEmpty(itemSpec))
                 {
-                    EmbedFile?.Invoke(itemSpec);
+                    EmbedFile.Invoke(itemSpec);
                 }
             }
         }
