diff --git a/eng/Version.Details.xml b/eng/Version.Details.xml
index 97360cbde1a..eff330da538 100644
--- a/eng/Version.Details.xml
+++ b/eng/Version.Details.xml
@@ -10,9 +10,9 @@
       <Sha>7324320f814152b72295946847ca72413507705a</Sha>
       <SourceBuild RepoName="arcade" ManagedOnly="true" />
     </Dependency>
-    <Dependency Name="NuGet.Build.Tasks" Version="6.0.0-rc.244">
+    <Dependency Name="NuGet.Build.Tasks" Version="6.0.0-rc.258">
       <Uri>https://github.com/nuget/nuget.client</Uri>
-      <Sha>f4ae349ec18d1ac5d5e7d1b0ed753e1300e9e157</Sha>
+      <Sha>e685bdb6d7b4d1f3664f6542f50b84376a19341c</Sha>
     </Dependency>
     <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.0.0-5.21469.2">
       <Uri>https://github.com/dotnet/roslyn</Uri>
diff --git a/eng/Versions.props b/eng/Versions.props
index a61d0ee2825..8e8249e1dfb 100644
--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -36,7 +36,7 @@
     <MicrosoftCodeAnalysisBannedApiAnalyzersVersion>3.3.2</MicrosoftCodeAnalysisBannedApiAnalyzersVersion>
     <MicrosoftExtensionsDependencyModelVersion>6.0.0-preview.2.21154.6</MicrosoftExtensionsDependencyModelVersion>
     <MicrosoftNetCompilersToolsetVersion>4.0.0-5.21469.2</MicrosoftNetCompilersToolsetVersion>
-    <NuGetBuildTasksVersion>6.0.0-rc.244</NuGetBuildTasksVersion>
+    <NuGetBuildTasksVersion>6.0.0-rc.258</NuGetBuildTasksVersion>
   </PropertyGroup>
   <Target Name="OverrideArcadeFileVersion" AfterTargets="_InitializeAssemblyVersion">
     <!-- See https://github.com/dotnet/arcade/issues/3386
