diff --git a/src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs b/src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs
index f3312b09973..8e654e34d0d 100644
--- a/src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs
+++ b/src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs
@@ -2355,7 +2355,6 @@ public void TestTargetFrameworkVersionGreaterThan4(bool useNewParser)
 #if !FEATURE_ASPNET_COMPILER
                     });
 #endif
-                }
 #pragma warning restore format
 
 #if FEATURE_ASPNET_COMPILER
@@ -2363,6 +2362,7 @@ public void TestTargetFrameworkVersionGreaterThan4(bool useNewParser)
                 string message = ResourceUtilities.FormatResourceStringStripCodeAndKeyword("AspNetCompiler.TargetingHigherFrameworksDefaultsTo40", solution.ProjectsInOrder[0].ProjectName, ver.ToString());
                 logger.AssertLogContains(message);
 #endif
+                }
             }
             finally
             {
