diff --git a/src/Framework/NativeMethods.cs b/src/Framework/NativeMethods.cs
index 608005241d2..a838e0dc9a6 100644
--- a/src/Framework/NativeMethods.cs
+++ b/src/Framework/NativeMethods.cs
@@ -9,9 +9,7 @@
 using System.IO;
 using System.Reflection;
 using System.Runtime.InteropServices;
-#if RUNTIME_TYPE_NETCORE
 using System.Runtime.Versioning;
-#endif
 using System.Text;
 using System.Threading;
 
@@ -25,23 +23,6 @@
 
 namespace Microsoft.Build.Framework;
 
-#if NETFRAMEWORK || NETSTANDARD2_0
-[AttributeUsage(AttributeTargets.Method | AttributeTargets.Property)]
-internal class SupportedOSPlatformGuard : Attribute
-{
-    internal SupportedOSPlatformGuard(string platformName)
-    {
-    }
-}
-[AttributeUsage(AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Class)]
-internal class SupportedOSPlatform : Attribute
-{
-    internal SupportedOSPlatform(string platformName)
-    {
-    }
-}
-#endif
-
 internal static class NativeMethods
 {
 #region Constants
diff --git a/src/Framework/SupportedOSPlatform.cs b/src/Framework/SupportedOSPlatform.cs
new file mode 100644
index 00000000000..71deb58019b
--- /dev/null
+++ b/src/Framework/SupportedOSPlatform.cs
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if !NET5_0_OR_GREATER
+namespace System.Runtime.Versioning
+{
+    /// <summary>
+    /// SupportedOSPlatform is a net5.0+ Attribute.
+    /// Create the same type only in full-framework and netstandard2.0 builds
+    /// to prevent many #if RUNTIME_TYPE_NETCORE checks.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Property)]
+    internal class SupportedOSPlatformGuard : Attribute
+    {
+        internal SupportedOSPlatformGuard(string platformName)
+        {
+        }
+    }
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Class)]
+    internal class SupportedOSPlatform : Attribute
+    {
+        internal SupportedOSPlatform(string platformName)
+        {
+        }
+    }
+}
+#endif
diff --git a/src/MSBuildTaskHost/MSBuildTaskHost.csproj b/src/MSBuildTaskHost/MSBuildTaskHost.csproj
index f56435ec284..becff23674d 100644
--- a/src/MSBuildTaskHost/MSBuildTaskHost.csproj
+++ b/src/MSBuildTaskHost/MSBuildTaskHost.csproj
@@ -142,6 +142,9 @@
     <Compile Include="..\Framework\StringBuilderCache.cs">
       <Link>StringBuilderCache.cs</Link>
     </Compile>
+    <Compile Include="..\Framework\SupportedOSPlatform.cs">
+      <Link>SupportedOSAttribute.cs</Link>
+    </Compile>
     <Compile Include="..\Shared\TaskEngineAssemblyResolver.cs">
       <Link>TaskEngineAssemblyResolver.cs</Link>
     </Compile>
diff --git a/src/Tasks/BootstrapperUtil/BootstrapperBuilder.cs b/src/Tasks/BootstrapperUtil/BootstrapperBuilder.cs
index a4ad3eb94c2..8b0ffffed17 100644
--- a/src/Tasks/BootstrapperUtil/BootstrapperBuilder.cs
+++ b/src/Tasks/BootstrapperUtil/BootstrapperBuilder.cs
@@ -17,11 +17,7 @@
 using System.Xml.XPath;
 using System.Xml.Xsl;
 using Microsoft.Build.Shared.FileSystem;
-#if RUNTIME_TYPE_NETCORE
 using System.Runtime.Versioning;
-#else
-using Microsoft.Build.Framework;
-#endif
 
 #nullable disable
 
diff --git a/src/Tasks/BootstrapperUtil/Util.cs b/src/Tasks/BootstrapperUtil/Util.cs
index bd46c5ce5cc..d62c8247070 100644
--- a/src/Tasks/BootstrapperUtil/Util.cs
+++ b/src/Tasks/BootstrapperUtil/Util.cs
@@ -5,14 +5,9 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.IO;
-#if RUNTIME_TYPE_NETCORE
 using System.Runtime.Versioning;
-#endif
 using Microsoft.Build.Shared;
 using Microsoft.Win32;
-#if NETFRAMEWORK
-using Microsoft.Build.Framework;
-#endif
 
 #nullable disable
 
diff --git a/src/Tasks/ManifestUtil/ComImporter.cs b/src/Tasks/ManifestUtil/ComImporter.cs
index be9ab0c1b42..30b34cc10fe 100644
--- a/src/Tasks/ManifestUtil/ComImporter.cs
+++ b/src/Tasks/ManifestUtil/ComImporter.cs
@@ -7,13 +7,9 @@
 using System.Globalization;
 using System.Resources;
 using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
 #if RUNTIME_TYPE_NETCORE
 using System.Runtime.InteropServices.ComTypes;
-using System.Runtime.Versioning;
-#endif
-
-#if NETFRAMEWORK
-using Microsoft.Build.Framework;
 #endif
 
 #nullable disable
diff --git a/src/Tasks/ManifestUtil/FileReference.cs b/src/Tasks/ManifestUtil/FileReference.cs
index dcd3e09f74e..e13dad0af4f 100644
--- a/src/Tasks/ManifestUtil/FileReference.cs
+++ b/src/Tasks/ManifestUtil/FileReference.cs
@@ -5,16 +5,10 @@
 using System.Collections.Generic;
 using System.ComponentModel;
 using System.Runtime.InteropServices;
-#if RUNTIME_TYPE_NETCORE
 using System.Runtime.Versioning;
-#endif
 using System.Text;
 using System.Xml.Serialization;
 
-#if NETFRAMEWORK
-using Microsoft.Build.Framework;
-#endif
-
 #nullable disable
 
 namespace Microsoft.Build.Tasks.Deployment.ManifestUtilities
diff --git a/src/Tasks/ManifestUtil/SecurityUtil.cs b/src/Tasks/ManifestUtil/SecurityUtil.cs
index 3bcadaa14fd..2ab1d221d09 100644
--- a/src/Tasks/ManifestUtil/SecurityUtil.cs
+++ b/src/Tasks/ManifestUtil/SecurityUtil.cs
@@ -1,9 +1,6 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-#if !RUNTIME_TYPE_NETCORE
-using Microsoft.Build.Framework;
-#endif
 using Microsoft.Build.Utilities;
 using Microsoft.Win32;
 using System;
@@ -33,6 +30,9 @@
 using System.Text;
 using System.Xml;
 using Microsoft.Build.Shared.FileSystem;
+#if !NET5_0_OR_GREATER
+using Microsoft.Build.Framework;
+#endif
 
 #nullable disable
 
diff --git a/src/Tasks/ManifestUtil/Util.cs b/src/Tasks/ManifestUtil/Util.cs
index ed7a00c9229..56ba265c495 100644
--- a/src/Tasks/ManifestUtil/Util.cs
+++ b/src/Tasks/ManifestUtil/Util.cs
@@ -13,9 +13,7 @@
 using System.IO;
 using System.Linq;
 using System.Reflection;
-#if RUNTIME_TYPE_NETCORE
 using System.Runtime.Versioning;
-#endif
 using System.Security;
 using System.Security.Cryptography;
 using System.Text;
diff --git a/src/Tasks/ManifestUtil/mansign2.cs b/src/Tasks/ManifestUtil/mansign2.cs
index 43d4683aee9..8d0d9fbf267 100644
--- a/src/Tasks/ManifestUtil/mansign2.cs
+++ b/src/Tasks/ManifestUtil/mansign2.cs
@@ -12,12 +12,7 @@
 using System.Text;
 using System.Xml;
 using System.Runtime.InteropServices;
-
-#if RUNTIME_TYPE_NETCORE
 using System.Runtime.Versioning;
-#else
-using Microsoft.Build.Framework;
-#endif
 
 using _FILETIME = System.Runtime.InteropServices.ComTypes.FILETIME;
 
diff --git a/src/Tasks/NativeMethods.cs b/src/Tasks/NativeMethods.cs
index 45bc7d9faad..94ec0181eb5 100644
--- a/src/Tasks/NativeMethods.cs
+++ b/src/Tasks/NativeMethods.cs
@@ -3,9 +3,12 @@
 
 using System;
 using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Build.Shared.FileSystem;
+
+#if FEATURE_COM_INTEROP
 using System.Text;
 using System.Reflection;
-using System.Runtime.InteropServices;
 using Microsoft.Build.Shared;
 using System.Collections.Generic;
 using System.Collections;
@@ -13,7 +16,7 @@
 using System.Linq;
 using System.Runtime.ExceptionServices;
 using System.Text.RegularExpressions;
-using Microsoft.Build.Shared.FileSystem;
+#endif
 
 #nullable disable
 
@@ -522,7 +525,7 @@ internal enum SymbolicLink
     /// </summary>
     internal static class NativeMethods
     {
-        #region Constants
+#region Constants
 
         internal static readonly IntPtr NullPtr = IntPtr.Zero;
         internal static readonly IntPtr InvalidIntPtr = new IntPtr(-1);
@@ -627,9 +630,9 @@ internal enum MoveFileFlags
             MOVEFILE_FAIL_IF_NOT_TRACKABLE = 0x00000020
         }
 
-        #endregion
+#endregion
 
-        #region NT header stuff
+#region NT header stuff
 
         internal const uint IMAGE_NT_OPTIONAL_HDR32_MAGIC = 0x10b;
         internal const uint IMAGE_NT_OPTIONAL_HDR64_MAGIC = 0x20b;
@@ -780,9 +783,9 @@ internal struct CRYPTOAPI_BLOB
             internal IntPtr pbData;
         }
 
-        #endregion
+#endregion
 
-        #region PInvoke
+#region PInvoke
         private const string Crypt32DLL = "crypt32.dll";
         private const string Advapi32DLL = "advapi32.dll";
 #if !RUNTIME_TYPE_NETCORE
