diff --git a/src/MSBuildTaskHost/OutOfProcTaskHost.cs b/src/MSBuildTaskHost/OutOfProcTaskHost.cs
index cff38e66b70..4f4b2ab02da 100644
--- a/src/MSBuildTaskHost/OutOfProcTaskHost.cs
+++ b/src/MSBuildTaskHost/OutOfProcTaskHost.cs
@@ -68,11 +68,6 @@ internal enum ExitType
         [MTAThread]
         public static int Main()
         {
-            while (true)
-            {
-                Microsoft.NET.StringTools.Strings.WeakIntern("hello");
-            }
-
             int exitCode = Execute() == ExitType.Success ? 0 : 1;
             return exitCode;
         }
