diff --git a/src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs b/src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs
index 823508be852..52360d65ff4 100644
--- a/src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs
+++ b/src/Build/BuildCheck/Checks/NoEnvironmentVariablePropertyCheck.cs
@@ -68,7 +68,7 @@ private void ProcessEnvironmentVariableReadAction(BuildCheckDataContext<Environm
             {
                 context.ReportResult(BuildCheckResult.Create(SupportedRule, context.Data.EvaluatedEnvironmentVariable.Location, buildCheckResultMessageArgs));
             }
-            else if (_scope == EvaluationCheckScope.WorkTreeImports)
+            else if (_scope != EvaluationCheckScope.ProjectFileOnly)
             {
                 _dataContext ??= context;
                 _buildCheckResults.Push(BuildCheckResult.Create(SupportedRule, context.Data.EvaluatedEnvironmentVariable.Location, buildCheckResultMessageArgs));
