diff --git a/src/Framework/MSBuildEventSource.cs b/src/Framework/MSBuildEventSource.cs
index 13bf34480b7..6e4db6d18bf 100644
--- a/src/Framework/MSBuildEventSource.cs
+++ b/src/Framework/MSBuildEventSource.cs
@@ -437,9 +437,9 @@ public void CopyUpToDateStart(string path)
         }
 
         [Event(59, Keywords = Keywords.All)]
-        public void CopyUpToDateStop(string path)
+        public void CopyUpToDateStop(string path, bool wasUpToDate)
         {
-            WriteEvent(59, path);
+            WriteEvent(59, path, wasUpToDate);
         }
         #endregion
     }
diff --git a/src/Tasks/Copy.cs b/src/Tasks/Copy.cs
index 869718fea00..6c91abeaab3 100644
--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -462,7 +462,7 @@ private bool CopySingleThreaded(
                 }
                 else
                 {
-                    MSBuildEventSource.Log.CopyUpToDateStop(destPath);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destPath, true);
                 }
 
                 if (copyComplete)
@@ -570,7 +570,7 @@ private bool CopyParallel(
                         }
                         else
                         {
-                            MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec);
+                            MSBuildEventSource.Log.CopyUpToDateStop(destItem.ItemSpec, true);
                         }
 
                         if (copyComplete)
@@ -727,7 +727,7 @@ private bool DoCopyIfNecessary(FileState sourceFileState, FileState destinationF
                         "SkipUnchangedFiles",
                         "true"
                     );
-                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name, true);
                 }
                 // We only do the cheap check for identicalness here, we try the more expensive check
                 // of comparing the fullpaths of source and destination to see if they are identical,
@@ -737,12 +737,12 @@ private bool DoCopyIfNecessary(FileState sourceFileState, FileState destinationF
                              destinationFileState.Name,
                              StringComparison.OrdinalIgnoreCase))
                 {
-                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name, false);
                     success = DoCopyWithRetries(sourceFileState, destinationFileState, copyFile);
                 }
                 else
                 {
-                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name);
+                    MSBuildEventSource.Log.CopyUpToDateStop(destinationFileState.Name, true);
                 }
             }
             catch (OperationCanceledException)
