diff --git a/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs b/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs
index 7e6fb2b278b..73516c3b2d6 100644
--- a/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs
+++ b/src/Build.UnitTests/BackEnd/BuildManager_Logging_Tests.cs
@@ -89,7 +89,7 @@ public void Build_WithCustomBuildArgs_NetCore(string envVariableValue, bool isWa
         [InlineData("1", true)]
         [InlineData("0", false)]
         [InlineData(null, false)]
-        public void Build_WithCustomBuildArgs_Framework(string? envVariableValue, bool isWarningExpected) =>
+        public void Build_WithCustomBuildArgs_Framework(string envVariableValue, bool isWarningExpected) =>
             TestCustomEventWarning(envVariableValue, isWarningExpected);
 
         private void TestCustomEventWarning(string envVariableValue, bool isWarningExpected)
