diff --git a/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs b/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs
index cd8ee90b582..d4322c9410f 100644
--- a/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs
+++ b/src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs
@@ -20,7 +20,7 @@ namespace Microsoft.Build.UnitTests.OM.Definition
     /// <summary>
     /// Tests for ProjectItem
     /// </summary>
-    public class ProjectItem_Tests
+    public class ProjectItem_Tests : IDisposable
     {
         internal const string ItemWithIncludeAndExclude = @"
                     <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
@@ -43,6 +43,18 @@ public class ProjectItem_Tests
                     </Project>
                 ";
 
+        protected TestEnvironment _env;
+
+        public ProjectItem_Tests()
+        {
+            _env = TestEnvironment.Create();
+        }
+
+        public void Dispose()
+        {
+            _env.Dispose();
+        }
+
         /// <summary>
         /// Project getter
         /// </summary>
@@ -3542,4 +3554,13 @@ public void FileNameMetadataEvaluationShouldNotDependsFromPlatformSpecificSlashe
             }
         }
     }
+
+    public class ProjectItemWithOptimizations_Tests : ProjectItem_Tests
+    {
+        public ProjectItemWithOptimizations_Tests()
+        {
+            // Make sure we always use the dictionary-based Remove logic.
+            _env.SetEnvironmentVariable("MSBUILDDICTIONARYBASEDITEMREMOVETHRESHOLD", "0");
+        }
+    }
 }
