diff --git a/src/Build/Utilities/Utilities.cs b/src/Build/Utilities/Utilities.cs
index 8614970efd4..d46d761618e 100644
--- a/src/Build/Utilities/Utilities.cs
+++ b/src/Build/Utilities/Utilities.cs
@@ -626,6 +626,9 @@ public static T[] ToArray<T>(this IEnumerator<T> enumerator)
             return enumerator.ToEnumerable().ToArray();
         }
 
+        /// <summary>
+        /// Iterates through the nongeneric enumeration and provides generic strong-typed enumeration of properties.
+        /// </summary>
         public static IEnumerable<PropertyData> EnumerateProperties(IEnumerable properties)
         {
             if (properties == null)
@@ -681,6 +684,9 @@ IEnumerable<PropertyData> CastOneByOne(IEnumerable props)
             }
         }
 
+        /// <summary>
+        /// Iterates through the nongeneric enumeration and provides generic strong-typed callback to handle the properties.
+        /// </summary>
         public static void EnumerateProperties<TArg>(IEnumerable properties, TArg arg, Action<TArg, KeyValuePair<string, string>> callback)
         {
             foreach (var tuple in EnumerateProperties(properties))
@@ -690,7 +696,7 @@ public static void EnumerateProperties<TArg>(IEnumerable properties, TArg arg, A
         }
 
         /// <summary>
-        /// Enumerates the given nongeneric enumeration and tries to match or wrap appropriate item types
+        /// Enumerates the given nongeneric enumeration and tries to match or wrap appropriate item types.
         /// </summary>
         public static IEnumerable<ItemData> EnumerateItems(IEnumerable items)
         {
@@ -779,6 +785,9 @@ IEnumerable<ItemData> CastOneByOne(IEnumerable itms)
             }
         }
 
+        /// <summary>
+        /// Iterates through the nongeneric enumeration of items and provides generic strong-typed callback to handle the items.
+        /// </summary>
         public static void EnumerateItems(IEnumerable items, Action<DictionaryEntry> callback)
         {
             foreach (var tuple in EnumerateItems(items))
