diff --git a/src/Framework/BuildSubmissionStartedEventArgs.cs b/src/Framework/BuildSubmissionStartedEventArgs.cs
index 17d2a8d92e4..e706f36d49e 100644
--- a/src/Framework/BuildSubmissionStartedEventArgs.cs
+++ b/src/Framework/BuildSubmissionStartedEventArgs.cs
@@ -26,7 +26,7 @@ public BuildSubmissionStartedEventArgs()
         }
 
         public BuildSubmissionStartedEventArgs(
-            IReadOnlyDictionary<string, string?>? globalProperties,
+            IReadOnlyDictionary<string, string?> globalProperties,
             IEnumerable<string> entryProjectsFullPath,
             IEnumerable<string> targetNames,
             BuildRequestDataFlags flags,
@@ -40,7 +40,7 @@ public BuildSubmissionStartedEventArgs(
             SubmissionId = submissionId;
         }
 
-        public IReadOnlyDictionary<string, string?>? GlobalProperties { get; set; }
+        public IReadOnlyDictionary<string, string?> GlobalProperties { get; set; }
 
         public IEnumerable<string> EntryProjectsFullPath { get; set; }
 
