diff --git a/ref/Microsoft.Build/net/Microsoft.Build.cs b/ref/Microsoft.Build/net/Microsoft.Build.cs
index b7463405425..9f42d2be544 100644
--- a/ref/Microsoft.Build/net/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/net/Microsoft.Build.cs
@@ -17,7 +17,6 @@ protected ElementLocation() { }
         public abstract int Column { get; }
         public abstract string File { get; }
         public abstract int Line { get; }
-        public Microsoft.Build.Construction.ElementLocation Location { get { throw null; } }
         public string LocationString { get { throw null; } }
         public override bool Equals(object obj) { throw null; }
         public override int GetHashCode() { throw null; }
diff --git a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
index 4df7931e636..c12fdafce68 100644
--- a/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
+++ b/ref/Microsoft.Build/netstandard/Microsoft.Build.cs
@@ -17,7 +17,6 @@ protected ElementLocation() { }
         public abstract int Column { get; }
         public abstract string File { get; }
         public abstract int Line { get; }
-        public Microsoft.Build.Construction.ElementLocation Location { get { throw null; } }
         public string LocationString { get { throw null; } }
         public override bool Equals(object obj) { throw null; }
         public override int GetHashCode() { throw null; }
diff --git a/src/Build/ElementLocation/ElementLocation.cs b/src/Build/ElementLocation/ElementLocation.cs
index 0d1275c7e57..ccd02a7c515 100644
--- a/src/Build/ElementLocation/ElementLocation.cs
+++ b/src/Build/ElementLocation/ElementLocation.cs
@@ -83,8 +83,6 @@ internal static ElementLocation EmptyLocation
             get { return s_emptyElementLocation; }
         }
 
-        public ElementLocation Location => this;
-
         IElementLocation IInternalLocation.Location => this;
 
         /// <summary>
