diff --git a/src/Utilities.UnitTests/TrackedDependencies/FileTrackerTests.cs b/src/Utilities.UnitTests/TrackedDependencies/FileTrackerTests.cs
index 7cfe45d5fc0..c5561753836 100644
--- a/src/Utilities.UnitTests/TrackedDependencies/FileTrackerTests.cs
+++ b/src/Utilities.UnitTests/TrackedDependencies/FileTrackerTests.cs
@@ -2326,7 +2326,7 @@ private static void InProcTrackingSpawnsToolWithTracker(bool useTrackerResponseF
             const string tlogRootName = "foo_inline";
             const string sourceFile = "inlinetrackingtest.txt";
             const string trackerResponseFile = "test-tracker.rsp";
-            const string fileTrackerParameters = "/d FileTracker.dll /r \"" + rootingMarker + "\"";
+            const string fileTrackerParameters = "/d FileTracker32.dll /r \"" + rootingMarker + "\"";
 
             File.Delete(toolReadTlog);
             File.Delete(sourceFile);
