diff --git a/src/Tasks/Exec.cs b/src/Tasks/Exec.cs
index ba75c6da31f..95f43a488bb 100644
--- a/src/Tasks/Exec.cs
+++ b/src/Tasks/Exec.cs
@@ -590,9 +590,7 @@ protected internal override void AddCommandLineCommands(CommandLineBuilderExtens
             {
                 commandLine.AppendSwitch("-c");
                 commandLine.AppendTextUnquoted(" \"");
-                // Set the locale to the POSIX 'Computer C' English locale.
-                // This makes the task command output the same text independent of the system locale.
-                commandLine.AppendTextUnquoted("export LANG=C; export LC_ALL=C; . ");
+                commandLine.AppendTextUnquoted(". ");
                 commandLine.AppendFileNameIfNotNull(batchFileForCommandLine);
                 commandLine.AppendTextUnquoted("\"");
             }
