diff --git a/src/Build/Logging/DistributedLoggers/ConfigurableForwardingLogger.cs b/src/Build/Logging/DistributedLoggers/ConfigurableForwardingLogger.cs
index 2adc797d495..5504f448377 100644
--- a/src/Build/Logging/DistributedLoggers/ConfigurableForwardingLogger.cs
+++ b/src/Build/Logging/DistributedLoggers/ConfigurableForwardingLogger.cs
@@ -202,8 +202,6 @@ public virtual void Initialize(IEventSource eventSource)
 
             ParseParameters(eventSource);
 
-            ResetLoggerState();
-
             if (!_forwardingSetFromParameters)
             {
                 SetForwardingBasedOnVerbosity(eventSource);
@@ -314,15 +312,6 @@ internal MessageImportance GetMinimumMessageImportance()
             return MessageImportance.High - 1;
         }
 
-        /// <summary>
-        /// Reset the states of per-build member variables.
-        /// Used when a build is finished, but the logger might be needed for the next build.
-        /// </summary>
-        private void ResetLoggerState()
-        {
-            // No state needs resetting
-        }
-
         /// <summary>
         /// Called when Engine is done with this logger
         /// </summary>
