diff --git a/src/MSBuild/LiveLogger/LiveLogger.cs b/src/MSBuild/LiveLogger/LiveLogger.cs
index e32d4483009..278926df9ba 100644
--- a/src/MSBuild/LiveLogger/LiveLogger.cs
+++ b/src/MSBuild/LiveLogger/LiveLogger.cs
@@ -150,7 +150,6 @@ private void eventSource_BuildStarted(object sender, BuildStartedEventArgs e)
 
         private void eventSource_BuildFinished(object sender, BuildFinishedEventArgs e)
         {
-            endTime = DateTime.Now;
             succeeded = e.Succeeded;
         }
 
@@ -323,7 +322,7 @@ public void Shutdown()
             Console.WriteLine();
 
             Debug.Assert(_stopwatch is not null, $"Expected {nameof(_stopwatch)} to be initialized long before Shutdown()");
-            TimeSpan buildDuration = _stopwatch.Elapsed;
+            TimeSpan buildDuration = _stopwatch!.Elapsed;
 
             string prettyDuration = buildDuration.TotalHours > 1.0 ?
                 buildDuration.ToString(@"h\:mm\:ss") :
